---
title: "R Notebook"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
```

``` {r library, include = FALSE}
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
```

``` {r functions}
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.salmon <- readRDS("data/txi.salmon.floride.RDS")

#The samples file generated for tximport
samples <- read.table("FL_samples.txt", header = TRUE)

```

The package `DESeq2` was used for the expression analysis outlined below

# Single factor analysis - Comparing Males v Females
To analyze your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions.
```{r DESeqDataSet}
#Make sure the conditions are in the samples file as a factor
samples$Sex <- as.factor(samples$Sex)
samples$Organ <- as.factor(samples$Organ)

#Create the DESeq dataset
dds_FL <- DESeqDataSetFromTximport(txi.salmon, 
                                   colData = samples,
                                   design = ~ Sex)

```
The data is then pre-filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

```{r pre-filtering}
#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_FL)) >= 10
dds_FL <- dds_FL[keep, ]

```


After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.
```{r diff-exp}
#Generate the expression values
dds_FL_exp <- DESeq(dds_FL)

#Compile the results
res <- results(dds_FL_exp)
res
```

Once that has finished we can now start exploring some of the results
```{r investigate-DESeq-results}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#How many ADJUSTED p-values were less than 0.1?
sum(res$padj < 0.1, na.rm = TRUE)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_FL_exp, alpha = 0.05)
summary(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```

## Visualizing the results {.tabset}
### MA-plot - MvF Diff. {-}
Generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be colored if the adjusted p-value is less that 0.1.
```{r MA-plot}
par(mfrow=c(1,2))
plotMA(res, ylim = c(-2,2))
plotMA(res)

```

### Heatmap - overall expression {-}
We can also generate a heat map to look ar overall expression levels across our samples. Note, this is not differentially expressed genes.
```{r heatmap}
select <- order(rowMeans(counts(dds_FL_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]
df <- as.data.frame(colData(dds_FL)[,c("Sex", "Organ")])

#Transform the data
vsd <- vst(dds_FL_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df)

#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))
write.table(heatmap_TG,
            'heatmap_trinitygenes.txt',
            sep = "",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)
```

#### Using BLAST in command line{-}
```{bash blastn, eval = FALSE}
#Create a BLAST database - Used Syngnathus scovelli for our BLAST database
makeblastdb -in ../ncbi_dataset/data/GCF_024217435.1/GCF_024217435.1_RoL_Ssco_1.1_genomic.fna -out s_scov_genome -dbtype nucl

#Run blastn on any set of samples
blastn -db ../genomes/s_scov/s_scov_genome -query XXXXX.fasta -out XXX.txt -evalue 0.001 -num_threads 12 outfmt "6 qseqid qstart qed stitle sstart send evalue bitscore length pident gaps"

```

### Sample-dist Heatmap {-}
We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples
```{r sample-dist}
#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$Sex, vsd$Organ, sep = "-")
colnames(sampleDistsMatrix) <- paste(vsd$Sex, vsd$Organ, sep = "-")

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)
```

### PCA plots {-}
```{r pca-plot}
colData <- as.data.frame(samples)
rownames(colData) <- samples$ID

#Prepare the PCA using the transformed data from above
pcaFL <- pca(assay(vsd), metadata = colData)

#Plot the PCA
biplot(pcaFL,
       showLoadings = FALSE,
       colby = 'Organ', colkey = c('Liver'='forestgreen', 'Gill'='purple', 'Gonad'='gold'),
       colLegendTitle = "Organ Type",
       hline = 0, vline = 0,
       vlineType = 'dashed',
       pointSize = 3,
       legendPosition = 'left', legendLabSize = 10, legendIconSize = 6.0, legendTitleSize = 10,
       shape = 'Sex', shapekey = c('M'=17, 'F'=8),
       shapeLegendTitle = 'Sex',
       drawConnectors = FALSE,
       title = 'PCA bi-plot for Syngnathus floridae')

pairsplot(pcaFL,
          components = getComponents(pcaFL, c(1, 2, 3)),
          triangle = FALSE,
          hline = 0, vline = 0,
          pointSize = 3,
          gridlines.major = FALSE, gridlines.minor = FALSE,
          colby = 'Organ', colkey = c('Liver'='forestgreen', 'Gill'='purple', 'Gonad'='gold'),
          shape = 'Sex', shapekey = c('M'=17, 'F'=8),
          title = 'Pairs plot', titleLabSize = 22,
          legendPosition = 'left',
          axisLabSize = 10, plotaxes = TRUE,
          margingaps = unit(c(0.1, 0.1, 0.1, 0.1), 'cm'))
```


# Multifactor design - Comparing M v F across the diff tissue types
If we investigate the column data of our DESeq dataset we can see that each sample has both a sex and organ type attached to it, we will be using these two factors in our multi-factor analysis


```{r multi-factor-analysis}
##Create a copy of the DESeq dataset
ddsMF_FL <- dds_FL

#Create an additional column the has the two conditions combined(sex and organ type)
ddsMF_FL$group <- factor(paste0(ddsMF_FL$Sex, ddsMF_FL$Organ))

#Reapply that new column as the design for this analysis
design(ddsMF_FL) <- ~ group

#Run the differential expression analysis
ddsMF_FL_exp <- DESeq(ddsMF_FL)
resultsNames(ddsMF_FL_exp)

```

## Invesitgate the results {.tabset}
### M-F Liver Comparisson {-}
```{r m-f-liver}
##Pulling out the liver M-F results
liver_con_res <- results(ddsMF_FL_exp, contrast = c("group", "MLiver", "FLiver"))
head(liver_con_res)

```

There are many criteria that have been employed previously to denote sex bias. For this study we are classifying sex-biased genes as **genes with a p-value < 0.05 AND a logFC >= |2|**. With that criteria in the liver there are `r sum(liver_con_res$padj < 0.05 & liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(liver_con_res$padj < 0.05 & liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes.

```{r top50-liver}
#Pulling the top 50 diff expressed genes in Males
top50_MLiver <- cbind(rownames(head(liver_con_res[order(liver_con_res$log2FoldChange, decreasing = TRUE),], n=50)))
write.table(top50_MLiver,
            'FL_maleL_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_Fliver <- cbind(rownames(head(liver_con_res[order(liver_con_res$log2FoldChange, decreasing = FALSE),], n=50)))
write.table(top50_Fliver,
            'FL_femL_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```

```{r MA-plot-liver}
liver_con_res05 <- results(ddsMF_FL_exp, contrast = c("group", "MLiver", "FLiver"), alpha = 0.05)

par(mfrow=c(1,1))
plotMA(liver_con_res05)

```

### M-F Gill Comparisson {-}
```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res <- results(ddsMF_FL_exp, contrast = c("group", "MGill", "FGill"))
head(gill_con_res)

```

In the gills there were `r sum(gill_con_res$padj < 0.05 & gill_con_res$log2FoldChange > 0, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res$padj < 0.05 & gill_con_res$log2FoldChange < 0, na.rm = TRUE)` female-biased genes based on p-values of < 0.05. However, because of the number of comparisons that have to be made in order to generate these p-values we have less confidence in them. Instead we can use a **log-fold-change cut off of 2** to denote male versus female biased. With this cutoff, there are `r sum(gill_con_res$log2FoldChange > 2)` male-biased genes in the liver and `r sum(gill_con_res$log2FoldChange < -2)` female-biased genes.

```{r top50-gill}
#Pulling the top 50 diff expressed genes in Males
top50_MGill <- cbind(rownames(head(gill_con_res[order(gill_con_res$log2FoldChange, decreasing = TRUE),], n=50)))
write.table(top50_MGill,
            'FL_maleG_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FGill <- cbind(rownames(head(gill_con_res[order(gill_con_res$log2FoldChange, decreasing = FALSE),], n=50)))
write.table(top50_FGill,
            'FL_femG_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```


### M-F Gonad Comparisson {-}
```{r m-f-gonads}
##Pulling out the gonad M-F results
gonad_con_res <- results(ddsMF_FL_exp, contrast = c("group", "MGonad", "FGonad"))
head(gonad_con_res)

```

Between the ovaries and testis (gonads) there were `r sum(gonad_con_res$padj < 0.05 & gonad_con_res$log2FoldChange > 0, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gonad_con_res$padj < 0.05 & gonad_con_res$log2FoldChange < 0, na.rm = TRUE)` female-biased genes based on p-values of < 0.05. However, because of the number of comparisons that have to be made in order to generate these p-values we have less confidence in them. Instead we can use a **log-fold-change cut off of 2** to denote male versus female biased. With this cutoff, there are `r sum(gonad_con_res$log2FoldChange > 2)` male-biased genes in the liver and `r sum(gonad_con_res$log2FoldChange < -2)` female-biased genes.

```{r top50-gonad}
#Pulling the top 50 diff expressed genes in Males
top50_MGonad <- cbind(rownames(head(gonad_con_res[order(gonad_con_res$log2FoldChange, decreasing = TRUE),], n=50)))
write.table(top50_MGonad,
            'FL_maleGon_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FGonad <- cbind(rownames(head(gonad_con_res[order(gonad_con_res$log2FoldChange, decreasing = FALSE),], n=50)))
write.table(top50_FGonad,
            'FL_femGon_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```

# Categorizing sex-specific genes
## Liver

## Gills
```{r create gill-dataset}
#Putting the counts into an object
fl_cnts <- counts(dds_FL)

#Pulling out the Gill samples from the counts
fl_gill<-fl_cnts[,colnames(fl_cnts)%in%samples$ID[ samples$Organ=="Gill"]]
head(fl_gill)

#Grabbing the IDs that correspond to the females and males
fem_gills_ids<-samples$ID[samples$Sex=="F" & samples$Organ=="Gill"]
fem_gills_ids

mal_gills_ids<-samples$ID[samples$Sex=="M" & samples$Organ=="Gill"]
mal_gills_ids

male_spp_gill<-fl_gill[rowSums(fl_gill[,fem_gills_ids]) <1 & rowSums(fl_gill[,mal_gills_ids])>=10,]

fem_spp_gill<-fl_gill[rowSums(fl_gill[,fem_gills_ids]) >=10 & rowSums(fl_gill[,mal_gills_ids])<1,]
head(fem_spp_gill)

```

## Gonads
