---
title: "R Notebook"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
```

``` {r library, include = FALSE}
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
library(ggpubr)
```

``` {r functions}
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.salmon <- readRDS("data/txi.salmon.floride.RDS")

#The samples file generated for tximport
samples <- read.table("FL_samples.txt", header = TRUE)

```

The package `DESeq2` was used for the expression analysis outlined below

# Single factor analysis - Comparing Males v Females
To analyze your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions.
```{r DESeqDataSet}
#Make sure the conditions are in the samples file as a factor
samples$Sex <- as.factor(samples$Sex)
samples$Organ <- as.factor(samples$Organ)

#Create the DESeq dataset
dds_FL <- DESeqDataSetFromTximport(txi.salmon, 
                                   colData = samples,
                                   design = ~ Sex)

```
The data is then pre-filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

```{r pre-filtering}
#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_FL)) >= 10
dds_FL <- dds_FL[keep, ]

```


After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.
```{r diff-exp}
#Generate the expression values
dds_FL_exp <- DESeq(dds_FL)

#Compile the results
res <- results(dds_FL_exp)
res
```

Once that has finished we can now start exploring some of the results
```{r investigate-DESeq-results}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#How many ADJUSTED p-values were less than 0.1?
sum(res$padj < 0.1, na.rm = TRUE)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_FL_exp, alpha = 0.05)
summary(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```

## Visualizing the results {.tabset}
### MA-plot - MvF Diff. {-}
Generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be colored if the adjusted p-value is less that 0.1.
```{r MA-plot}
par(mfrow=c(1,2))
plotMA(res, ylim = c(-2,2))
plotMA(res)

```

### Heatmap - overall expression {-}
We can also generate a heat map to look ar overall expression levels across our samples. Note, this is not differentially expressed genes.
```{r heatmap}
select <- order(rowMeans(counts(dds_FL_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]
df <- as.data.frame(colData(dds_FL)[,c("Sex", "Organ")])

#Transform the data
vsd <- vst(dds_FL_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df)

#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))
write.table(heatmap_TG,
            'heatmap_trinitygenes.txt',
            sep = "",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)
```

### Sample-dist Heatmap {-}
We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples
```{r sample-dist}
#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$Sex, vsd$Organ, sep = "-")
colnames(sampleDistsMatrix) <- paste(vsd$Sex, vsd$Organ, sep = "-")

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)
```

### PCA plots {-}
```{r pca-plot}
colData <- as.data.frame(samples)
rownames(colData) <- samples$ID

#Prepare the PCA using the transformed data from above
pcaFL <- pca(assay(vsd), metadata = colData)

#Plot the PCA
biplot(pcaFL,
       showLoadings = FALSE,
       colby = 'Organ', colkey = c('Liver'='forestgreen', 'Gill'='purple', 'Gonad'='gold'),
       colLegendTitle = "Organ Type",
       hline = 0, vline = 0,
       vlineType = 'dashed',
       pointSize = 3,
       legendPosition = 'left', legendLabSize = 10, legendIconSize = 6.0, legendTitleSize = 10,
       shape = 'Sex', shapekey = c('M'=17, 'F'=8),
       shapeLegendTitle = 'Sex',
       drawConnectors = FALSE,
       title = 'PCA bi-plot for Syngnathus floridae')

pairsplot(pcaFL,
          components = getComponents(pcaFL, c(1, 2, 3)),
          triangle = FALSE,
          hline = 0, vline = 0,
          pointSize = 3,
          gridlines.major = FALSE, gridlines.minor = FALSE,
          colby = 'Organ', colkey = c('Liver'='forestgreen', 'Gill'='purple', 'Gonad'='gold'),
          shape = 'Sex', shapekey = c('M'=17, 'F'=8),
          title = 'Pairs plot', titleLabSize = 22,
          legendPosition = 'left',
          axisLabSize = 10, plotaxes = TRUE,
          margingaps = unit(c(0.1, 0.1, 0.1, 0.1), 'cm'))
```


# Multifactor design - Comparing M v F across the diff tissue types
If we investigate the column data of our DESeq dataset we can see that each sample has both a sex and organ type attached to it, we will be using these two factors in our multi-factor analysis


```{r multi-factor-analysis}
##Create a copy of the DESeq dataset
ddsMF_FL <- dds_FL

#Create an additional column the has the two conditions combined(sex and organ type)
ddsMF_FL$group <- factor(paste0(ddsMF_FL$Sex, ddsMF_FL$Organ))

#Reapply that new column as the design for this analysis
design(ddsMF_FL) <- ~ group

#Run the differential expression analysis
ddsMF_FL_exp <- DESeq(ddsMF_FL)
resultsNames(ddsMF_FL_exp)

```

## Invesitgate the results {.tabset}
### M-F Liver Comparisson {-}
```{r m-f-liver}
##Pulling out the liver M-F results
liver_con_res <- results(ddsMF_FL_exp, contrast = c("group", "MLiver", "FLiver"), alpha = 0.05)
head(liver_con_res)

```

There are many criteria that have been employed previously to denote sex bias. For this study we are classifying sex-biased genes as **genes with a p-value < 0.05 AND a logFC $\ge$ |2|**. With that criteria in the liver there are `r sum(liver_con_res$padj < 0.05 & liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(liver_con_res$padj < 0.05 & liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes.

```{r top50-liver}
#Pulling out rows where padj is not NA and is less than out alpha of 0.05
liver_con_res_p05 <- liver_con_res[!is.na(liver_con_res$padj) & liver_con_res$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_MLiver <- cbind(rownames(head(liver_con_res_p05[order(liver_con_res_p05$log2FoldChange, decreasing = TRUE),], n=50)))
write.table(top50_MLiver,
            'FL_maleL_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_Fliver <- cbind(rownames(head(liver_con_res_p05[order(liver_con_res_p05$log2FoldChange, decreasing = FALSE),], n=50)))
write.table(top50_Fliver,
            'FL_femL_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```

### M-F Gill Comparisson {-}
```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res <- results(ddsMF_FL_exp, contrast = c("group", "MGill", "FGill"), alpha = 0.5)
head(gill_con_res)

```

In the gills there were `r sum(gill_con_res$padj < 0.05 & gill_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res$padj < 0.05 & gill_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

```{r top50-gill}
#Pulling out rows where padj is not NA and is less than out alpha of 0.05
gill_con_res_p05 <- gill_con_res[!is.na(gill_con_res$padj) & gill_con_res$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_MGill <- cbind(rownames(head(gill_con_res_p05[order(gill_con_res_p05$log2FoldChange, decreasing = TRUE),], n=50)))
write.table(top50_MGill,
            'FL_maleG_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

#Pulling the top 21 diff expressed genes in Females - Only 21 female-biased genes in the gills
top21_FGill <- cbind(rownames(head(gill_con_res_p05[order(gill_con_res_p05$log2FoldChange, decreasing = FALSE),], n=21)))
write.table(top21_FGill,
            'FL_femG_top21TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```

### M-F Gonad Comparisson {-}
```{r m-f-gonads}
##Pulling out the gonad M-F results
gonad_con_res <- results(ddsMF_FL_exp, contrast = c("group", "MGonad", "FGonad"), alpha = 0.05)
head(gonad_con_res)

```

Between the ovaries and testis (gonads) there were `r sum(gonad_con_res$padj < 0.05 & gonad_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gonad_con_res$padj < 0.05 & gonad_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

```{r top50-gonad}
#Pulling out rows where padj is not NA and is less than out alpha of 0.05
gonad_con_res_p05 <- gonad_con_res[!is.na(gonad_con_res$padj) & gonad_con_res$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_MGonad <- cbind(rownames(head(gonad_con_res_p05[order(gonad_con_res_p05$log2FoldChange, decreasing = TRUE),], n=50)))
write.table(top50_MGonad,
            'FL_maleGon_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FGonad <- cbind(rownames(head(gonad_con_res_p05[order(gonad_con_res_p05$log2FoldChange, decreasing = FALSE),], n=50)))
write.table(top50_FGonad,
            'FL_femGon_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```

### MA Plots {-}
```{r MA-plot-MF}
liver_MA <- ggmaplot(liver_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 20,
         main = expression("Male Liver" %->% "Female Liver"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("orange2", "yellowgreen", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gill_MA <- ggmaplot(gill_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 20,
         main = expression("Male Gill" %->% "Female Gill"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("orange2", "yellowgreen", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gonad_MA <- ggmaplot(gonad_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 20,
         main = expression("Testis" %->% "Ovaries"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("orange2", "yellowgreen", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

#Merge all the plots into one
merged_MA <- plot_grid(liver_MA, gill_MA, gonad_MA,
          ncol = 1,
          nrow = 3,
          labels = c('A', 'B', 'C'),
          label_fontfamily = 'sans',
          label_fontface = 'bold',
          label_size = 18)

plot_grid(merged_MA)
```

## Categorizing sex-specific genes

```{r sex-specific-forloop}
#Create a vector with the different organ types
organs <- levels(colData(dds_FL)$Organ)

#Create an empty list to store my datasets in
FL_sex_specific_genes <- list()

#Generate the for loop to identify MSpecific and FSpecific genes in each organ
for(organ in organs){
  
  #Male-Specific Genes
  ##Pull out all of the rows where fem count = 0
  fem0_organ <- assay(dds_FL)[rowSums(assay(dds_FL)[, dds_FL$Sex == "F"]) < 1, dds_FL$Organ == organ]
  
  ##Pull out rows where median of male count >=10 for that organ
  mal10_organ <- apply(assay(dds_FL)[, dds_FL$Sex == "M" & dds_FL$Organ == organ],
                       1,
                       function(x) median(x) >=10)
  
  ##Keep only the rows where fem=0 and the Male median>=10
  fem0_mal10_organ <- fem0_organ[rownames(fem0_organ) %in% names(mal10_organ[mal10_organ == TRUE]),]
  
  ##Create a new object with a name based on the organ type
  organ_malsp <- sub("$", "_male_specific", organ)
  FL_sex_specific_genes[[organ_malsp]] <- fem0_mal10_organ
  
  
  
  #Female-Specific Genes
  ##Pull out all of the rows where male count = 0
  mal0_organ <- assay(dds_FL)[rowSums(assay(dds_FL)[, dds_FL$Sex == "M"]) < 1, dds_FL$Organ == organ]
  
  ##Pull out rows where median of female count >=10 for that organ
  fem10_organ <- apply(assay(dds_FL)[, dds_FL$Sex == "F" & dds_FL$Organ == organ],
                       1,
                       function(x) median(x) >=10)
  
  #Keep only the rows where male=0 and the fem median>=10
  mal0_fem10_organ <- mal0_organ[rownames(mal0_organ) %in% names(fem10_organ[fem10_organ == TRUE]),]
  
  # Create a new object with a name based on the organ type
  organ_femsp <- sub("$", "_female_specific", organ)
  FL_sex_specific_genes[[organ_femsp]] <- mal0_fem10_organ
} 
```


```{r create gill-dataset-sarah}
# the one to try - for individual thresholds
fem0_gill<-assay(dds_FL)[rowSums(assay(dds_FL)[,dds_FL$Sex == "F"]) < 1 ,dds_FL$Organ == "Gill"]
malLots_gill<-which(rowSums(t(apply(assay(dds_FL)[,dds_FL$Sex == "M" & dds_FL$Organ == "Gill"],1,function(x) x>=3)))==ncol(assay(dds_FL)[,dds_FL$Sex == "M" & dds_FL$Organ == "Gill"]))
fem0_malLots_gill<-fem0_gill[rownames(fem0_gill) %in% names(malLots_gill),]

# based on median
fem0_gill<-assay(dds_FL)[rowSums(assay(dds_FL)[,dds_FL$Sex == "F"]) < 1 ,dds_FL$Organ == "Gill"]
malLots_gill<-apply(assay(dds_FL)[,dds_FL$Sex == "M" & dds_FL$Organ == "Gill"],1,function(x) median(x) >=3)
fem0_malLots_gill<-fem0_gill[rownames(fem0_gill) %in% names(malLots_gill[malLots_gill==TRUE]),]

```


## Gene Ontology Analysis
We wanted to figure out two things: 1) What are the genes that are sex-biased within each tissue type and 2) What are the GO terms for those genes. 

To accomplish this I first had to pull out the sequences that corresponded to the Trinity gene IDS pulled out above. Once those sequences were fetched they were blasted against the _Syngnathus scovelli_ genome. This was completed with the following bash script on the RCC. 

### Using BLAST in command line
```{bash blastn, eval = FALSE}
#Create a BLAST database - Used Syngnathus scovelli for our BLAST database
makeblastdb -in ../ncbi_dataset/data/GCF_024217435.1/GCF_024217435.1_RoL_Ssco_1.1_genomic.fna -out s_scov_genome -dbtype nucl

```

```{bash blast_pipeline, file = 'bash/blast_pipeline.sh', eval = FALSE}


```

This script was run as `bash bash_scripts/blast_pipeline.sh floridae_high_exp_genes/high_expTG_names other_scripts/subset_fasta_file trinity_supertran_floridae.fasta ../genomes/s_scov/s_scov_genome floridae_high_exp_genes/high_exp_results/`


Once the BLAST results were obtained in the RCC I read them into R for further filtering (keeping only one hit per gene)
```{r readin_blast_results}
#Specify the directory where BLAST results are located
blast_path <- "data/floridae_BLAST/"

#Create a list of the files I want
FL_blast_files <- list.files(blast_path)

#Create an empty list to store my datasets in
FL_blast_list <- list()

#Create a loop to read in all of the blast results
for(file_name in FL_blast_files){
  # Read the file and store it in a new object
  file_data <- read.delim(file.path(blast_path, file_name), stringsAsFactors = FALSE, quote = "", sep = "\t", header = FALSE)
  
  #Add column names to the dataset
  colnames(file_data) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "scov_gene_info", "scov_gene_start", "scov_gene_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  # Create a new object with a name based on the file name
  blast_name <- sub(".txt$", "", file_name) #Removes the file extension
  FL_blast_list[[blast_name]] <- file_data
}

```

Because of the way BLAST was run in the RCC, we can have multiple hits for one gene, I want to filter that out so that there is only one hit kept for each gene.
```{r filter-BLAST}
#Use lapply to apply the function to each dataset stored in the list created above
blast_output_filtered <- lapply(FL_blast_list, function(data_frame){
 
  #Pull out the Unique Trinity gene IDs
  uniqueID <- unique(data_frame[1])
  
  #Create an empty dataframe to store intermediate results in
  output <- data.frame(matrix(data = NA, ncol = ncol(data_frame), nrow = 0))
  colnames(output) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "scov_gene_info", "scov_gene_start", "scov_gene_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  #Generate a for loop that pulls out the lowest e-value + highest % identity for each gene
  for(gene in uniqueID$trin_geneid){
    
    #Subset the dataset for each Trinity gene
    this_trin_gene<- subset(data_frame, trin_geneid==gene)
    #Pull out gene with smallest e-value
    uni_gene_subset_lowe <- this_trin_gene[which(this_trin_gene$evalue == min(this_trin_gene$evalue)),]
    #In case mult. genes have same e-value, pull out gene with highest % identity
    uni_gene_subset_lowe_highpid <- uni_gene_subset_lowe[which(uni_gene_subset_lowe$pident == max(uni_gene_subset_lowe$pident)),]
    
    # keep only one of multiple identical rows
    uni_gene_subset_lowe_highpid <- unique(uni_gene_subset_lowe_highpid)
   
    # check that there is a single gene ID in scovelli, if not, only first row is kept
    if(length(gsub("^.*(GeneID:\\d+)\\].*$","\\1",uni_gene_subset_lowe_highpid$scov_gene_info))>1){
    #   browser()
    # } else{
      uni_gene_subset_lowe_highpid<-uni_gene_subset_lowe_highpid[1,]
    }
    
    #Add the final gene into the empty dataframe from above
    output <- rbind(output, uni_gene_subset_lowe_highpid)
  }
  
  return(output)
})

```