---
title: "Tissue Specificity in _Syngnathus floridae_"
author: "Coley Tosto"
date: "`r Sys.Date()`"
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: no
    toc: false
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',
                     fig_path="../imgs/")
```

``` {r library, message = FALSE, warning = FALSE}
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(ggplot2)
```

``` {r functions, include = FALSE}
```

``` {r read-data, message = FALSE, warning = FALSE}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.salmon <- readRDS("data/txi.salmon.floride.RDS")

#The samples file generated for tximport
samples <- read.table("FL_samples.txt", header = TRUE)

samples$group <- factor(paste0(samples$Sex, samples$Organ))

#Changing "Gonad" to be more specific to testis or ovaries
samples$Organ <- ifelse(samples$Sex == "F" & samples$Organ =="Gonad",
                        paste0("Ovaries"),
                        ifelse(samples$Sex == "M" & samples$Organ == "Gonad",
                               paste0("Testis"),
                               paste0(samples$Organ))
                        )

#Make sure the conditions are in the samples file as a factor
samples$Sex <- as.factor(samples$Sex)
samples$Organ <- as.factor(samples$Organ)

#Format colData to be used in the tau function
colData <- as.data.frame(samples)
rownames(colData) <- samples$ID

```

# Differential Expression Analysis
One of the goals for this analysis is to compare tissue specificity to sex-biased gene expression. In order to do that I began the differential expression analysis here using the package DESeq2. This was done using the abundance matrix generated from salmon and the model was run as counts ~ group, where group included both the sex and the organ type (i.e. MLiver, FLiver, etc.).

Additionally, because $\tau$ is calculated based on expression levels, I wanted to apply the same filtering used by DESeq2 for the sex-biased expression analysis to the dataset that will be used to calculate $\tau$.

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_FL <- DESeqDataSetFromTximport(txi.salmon, 
                                   colData = samples,
                                   design = ~ group)

#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_FL)) >= 10
dds_FL <- dds_FL[keep, ]

#Generate the expression values
dds_FL_exp <- DESeq(dds_FL)
```

# Calculating Tissue Specificity
To estimate tissue specificity, the TPM estimates are needed which is the number of transcripts from a particular gene normalized first by gene length, and then by sequencing depth (in millions) in the sample. The output quant.sf files from salmon have the following format:
```
Name | Length | EffectiveLength | TPM | NumReads

```

## Pulling out TPM values
From the salmon outputs I pulled out the TPM values for each sample.

```{r getTPMs}
#Get the list of file names/paths for all of the quant.sf files
files <- list.files(pattern = ".sf", path = "data/floridae_expression_files", 
                    full.names = TRUE)

#For each quant.sf file pull out the TPM column
tpms <- do.call(cbind, lapply(files, function(file){

  dat <- read.delim(file, row.names = "Name")
  tpm <- dat["TPM"]
  colnames(tpm) <- gsub("data/floridae_expression_files/(.*)_quant.sf","\\1",file)
  
  return(tpm)
}))
```

## Filtering the TPM dataset
Once all the TPMs were gathered for the different samples I filtered some of the rows out based on results of the differential expression analysis done by DESeq2. The two filtering steps applied to the TPM dataset included:

  1. Keeping only the rows that weren't filtered out in the DESeq2 dataset due to low counts (rowSum $\le$ 10).
  
  2. Removing the rows that corresponded to genes that were "outliers" in the DESeq2 analysis.

It should be noted that the DESeq2 filtering was done based off of the count data and not the TPMs, but as TPMs are just normalized counts they should be correlated and something that was had low gene counts should also have a low TPM and anything that was considered an outlier in the counts could also be an outlier in terms of TPM, but it may not be perfect.

```{r filterTPMs}
#Only keeping the rows that weren't filtered out due to low counts
tpms <- tpms[rownames(tpms) %in% rownames(dds_FL), ]

#Pulling out the geneIDs for genes that were categorized as "outliers" by DESeq2
#Calculating the Cooks threshold that would have been used
np <- length(resultsNames(dds_FL_exp))
nsamp <- ncol(dds_FL_exp)
cooks_thresh <- qf(0.99, df1 = np, df2 = 29-np)

out_ids <- names(mcols(dds_FL_exp)$maxCooks[mcols(dds_FL_exp)$maxCooks > cooks_thresh])

#Removing the rows in the tpm dataset that were deemed "outliers" by DESeq2
tpms <- tpms[!(rownames(tpms) %in% out_ids), ]

```

## Generating $\tau$
These filtered TPM estimates can then be used to estimate $\tau$, a tissue specificity estimator that can range from 0 to 1. $\tau$ is calculated for each tissue, $i$, as follows:

$$
\tau=\frac{\sum_i{[1-ln(TPM_i)/ln(TPM_{max})]}}{N-1}
$$

where $TPM_{max}$ is the maximum **average** TPM for a given tissue type, and $TPM_i$ is the **average** TPM for tissue $i$. If $\tau = 0$, that gene is evenly expressed across tissues; if $\tau=1$, the gene is expressed in an entirely tissue-specific fashion. Because TPM values approach 0 are impacted by sampling stochasticity, any genes that had an expression approaching 0 were set to $TPM=2$. 

```{r est_tauFXN}
#Function for estimating tau given the TPM matrix and metadata file
est_tau<-function(geneDat,colDat){
  
  #For each row in the TPM matrix cbind it with the metadata file,
  #this attaches organ type information to the TPM values
  tissue_dat<-data.frame(cbind(colDat,
                               geneDat))
  
  #For the TPM values approaching 0, set them to 2
  tissue_dat$geneDat[tissue_dat$geneDat < 1] <- 2
  
  
  #Get the average TPM for each tissue type (TPMi)
  tissue_avgs<-tapply(tissue_dat$geneDat,tissue_dat$Organ,mean)
  
  #Get the maximum value from the average TPMS (TPMmax)
  tpmMax <- max(tissue_avgs, na.rm=TRUE)
    
  #IF running tau on JUST males of JUST females, this accounts for the
  #fact that ovary or testis will return an NA in the averaging
  if(length(unique(tissue_dat$Organ)) == 3){
    tau <- sum(1-(log(tissue_avgs[unique(tissue_dat$Organ)])/log(tpmMax)))/
      (length(unique(tissue_dat$Organ))-1)
    
    return(tau)
  }
  
  #IF using the WHOLE dataset, calculate tau
  tau<-sum(1-(log(tissue_avgs)/log(tpmMax)))/(length(unique(tissue_dat$Organ))-1)
  
  return(tau)
}

```

I then applied that function across each row in the TPMs matrix with my metadata stored in the object `colData`. The metadata file includes the sample ID, Sex, and Organ type for every column present in the TPM matrix.

I ran the $\tau$ function with the whole dataset, and then for just males and just females.

```{r calcTau}
tau <- apply(tpms, 1, est_tau, colDat=colData)


tau_fem<-apply(tpms[,which(colData$Sex=="F")], 1, est_tau,
               colData[which(colData$Sex=="F"),])

tau_mal<-apply(tpms[,which(colData$Sex=="M")], 1, est_tau,
               colData[which(colData$Sex=="M"),])

```

```{r tauHist,fig.cap="Distribution of tissue specificity estimates for only the male samples (left) versus only the female samples (middle) versus the male samples and the female samples (right).", echo=FALSE}
par(mfrow=c(1,3))
hist(tau_mal,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="Male-only")

hist(tau_fem,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="Female-only")

hist(tau,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="All samples")
```

We can see that the distribution of $\tau$ doesn't vary drastically when looking at only male/female samples versus when we included all of the samples in the calculation (Fig. \@ref(fig:tauHist)). Because of this, I can be confident in using the $\tau$ with all the samples moving forward.

## Validating the $\tau$ calculations
To make sure $\tau$ is being calculated in a way that makes sense, I checked some of the TPM values and plotted the counts for the genes with a high tissue specificity index and a low tissue specificity index.

```{r ts-plot-counts-high, echo=FALSE, fig.height=20, fig.width=15, fig.cap="Plots of the counts from the DESeq2 Dataset for genes with the highest tau", message=FALSE}
#Getting list of genes with the greatest Tau
top_tau <- head(sort(tau, decreasing = TRUE), n=9)

#Looking at the TPMS
tpms[rownames(tpms)== names(top_tau[1]),]
tpms[rownames(tpms)== names(top_tau[2]),]
tpms[rownames(tpms)== names(top_tau[3]),]
tpms[rownames(tpms)== names(top_tau[4]),]
tpms[rownames(tpms)== names(top_tau[5]),]
tpms[rownames(tpms)== names(top_tau[6]),]
tpms[rownames(tpms)== names(top_tau[7]),]
tpms[rownames(tpms)== names(top_tau[8]),]
tpms[rownames(tpms)== names(top_tau[9]),]

#Plotting the gene counts from DESeq2 for the toptau genes
par(mfrow=c(3, 3))
plotCounts(dds_FL, gene= names(top_tau[1]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[2]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[3]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[4]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[5]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[6]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[7]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[8]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[9]), intgroup="Organ")
```

```{r ts-plot-counts-low, echo=FALSE, fig.height=20, fig.width=15, fig.cap="Plots of the counts from the DESeq2 Dataset for genes with the lowest tau", message=FALSE}
#Getting list of genes with the lowest Tau
bot_tau <- head(sort(tau, decreasing = FALSE), n=9)

#Looking at the TPMS
tpms[rownames(tpms)== names(bot_tau[1]),]
tpms[rownames(tpms)== names(bot_tau[2]),]
tpms[rownames(tpms)== names(bot_tau[3]),]
tpms[rownames(tpms)== names(bot_tau[4]),]
tpms[rownames(tpms)== names(bot_tau[5]),]
tpms[rownames(tpms)== names(bot_tau[6]),]
tpms[rownames(tpms)== names(bot_tau[7]),]
tpms[rownames(tpms)== names(bot_tau[8]),]
tpms[rownames(tpms)== names(bot_tau[9]),]

#Plotting the gene counts from DESeq2 for the toptau genes
par(mfrow=c(3, 3))
plotCounts(dds_FL, gene= names(bot_tau[1]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[2]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[3]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[4]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[5]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[6]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[7]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[8]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[9]), intgroup="Organ")

```


# Sex Bias and Tissue Specificity
From the DESeq2 analysis I started earlier I can pull out the FC information for all of the sex biased genes and plot that against the $\tau$ calculations.

## $\tau$ and logFC in sex-biased genes for each organ
A gene was determined to be sex-biased if it had a log-fold change $\ge |2|$ AND an adjusted p-value of $< 0.05$. I first investigated the relationship between $\tau$ and sex bias (fold-change) for only the genes that were deemed as sex-biased.

```{r tau-v-bias, echo=FALSE, fig.cap="Sex bias (in terms of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are sex biased in the gonads, liver, and gills."}
##Pull out the results of the MF contrasts for each organ
liver_con_res <- results(dds_FL_exp, contrast = c("group", "MLiver", "FLiver"), alpha = 0.05)
gonad_con_res <- results(dds_FL_exp, contrast = c("group", "MGonad", "FGonad"), alpha = 0.05)
gill_con_res <- results(dds_FL_exp, contrast = c("group", "MGill", "FGill"), alpha = 0.05)

#Remove rows where the p-value is NA (low counts and outliers) for each organ
liver_con_res_noNA <- liver_con_res[!is.na(liver_con_res$padj),]
gonad_con_res_noNA <- gonad_con_res[!is.na(gonad_con_res$padj),]
gill_con_res_noNA <- gill_con_res[!is.na(gill_con_res$padj),]

#Pull out the sex-biased genes (based on FC of >= |2| and padj of < 0.05) for each organ
liver_biased <- liver_con_res_noNA[which(abs(liver_con_res_noNA$log2FoldChange) >= 2 & liver_con_res_noNA$padj <= 0.05),]
gonad_biased <- gonad_con_res_noNA[which(abs(gonad_con_res_noNA$log2FoldChange) >= 2 & gonad_con_res_noNA$padj <= 0.05),]
gill_biased <- gill_con_res_noNA[which(abs(gill_con_res_noNA$log2FoldChange) >= 2 & gill_con_res_noNA$padj <= 0.05),]

#Merge the sex bias information with the tau values based on row names for each organ
liver_bias_tau <- merge(as.data.frame(liver_biased), as.data.frame(tau), by = "row.names")
gonad_bias_tau <- merge(as.data.frame(gonad_biased), as.data.frame(tau), by = "row.names")
gill_bias_tau <- merge(as.data.frame(gill_biased), as.data.frame(tau), by = "row.names")

#Plot the relationship between FC (bias) and tau for each organ
plot(sqrt(abs(gonad_bias_tau$log2FoldChange)),
     gonad_bias_tau$tau,
     pch=19,
     col = "#EEB42275",
     xlab="Sex bias (|log2 fold change|)",
     ylab=expression(tau["TPM"]))
points(sqrt(abs(liver_bias_tau$log2FoldChange)),
     liver_bias_tau$tau,
     pch=19,
     col="#EE826275")
points(sqrt(abs(gill_bias_tau$log2FoldChange)),
     gill_bias_tau$tau,
     pch=19,
     col = "#6E8B3D75")

#Add the best fit lines
abline(lm(gonad_bias_tau$tau ~ sqrt(abs(gonad_bias_tau$log2FoldChange))), col="#EEB422", lwd = 3)
abline(lm(liver_bias_tau$tau ~ sqrt(abs(liver_bias_tau$log2FoldChange))), col = "#EE8262", lwd = 3, lty = 2)
abline(lm(gill_bias_tau$tau ~ sqrt(abs(gill_bias_tau$log2FoldChange))), col = "#6E8B3D", lwd = 3, lty = 3)

#Add a legend
legend("bottomright", c("Gonads","Liver",
                    "Gills"), 
       lty=c(1,2,3), 
       col=c("#EEB422","#EE8262","#6E8B3D"), 
       bty='o',
       lwd=3)
legend('bottomright', c("","",""), 
       pch=c(21,21,21),
       pt.bg = c("#EEB42275","#EE826275","#6E8B3D75"),
       pt.lwd = 1,
       pt.cex = 1.5,
       col=c("#EEB422","#EE8262","#6E8B3D"), bty='n',
       text.width = 0.55,
       cex=1.0
       )
```

The absolute value of the log2 fold change was square root transformed. There appears to only be a relationship between sex bias and tissue specificity in the gonads and not as much in the gills and the liver (Fig. \@ref(fig:tau-v-bias)). We can look further into this relationship with a Spearman's rank correlation test. This was done between $\tau$ and the $|log2FoldChange|$ for each organ separately.

```{r tau-v-bias-cor, echo=FALSE, message=FALSE, warning=FALSE}
cor.test(gonad_bias_tau$tau, abs(gonad_bias_tau$log2FoldChange), method = "spearman")
cor.test(liver_bias_tau$tau, abs(liver_bias_tau$log2FoldChange), method = "spearman")
cor.test(gill_bias_tau$tau, abs(gill_bias_tau$log2FoldChange), method = "spearman")
```

Both the liver and the gonads show a significant correlation between sex bias and tissue specificity with rho values similar to what has been found in other studies. The lack of a correlation in the gills may be due to the low samples size.

## $\tau$ and logFC for all the signficantly expressed genes in each organ
Only looking at the logFC for the sex-biased genes may be affecting the relationship that I am seeing between sex bias and $\tau$, especially in the gill where the samples size is very small if I am only including sex-biased genes. 

I used the same steps as above to see how the relationship changes when I add in ALL of the genes that are significantly expressed for each organ. In this datasets I have removed all rows where the adjusted p-value is NA (outliers and low-counts).

```{r tau-v-FC, echo=FALSE, fig.cap="Sex bias (in terms of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are signifcantly expressed in the gonads, liver, and gills."}
#Merge the noNA datasets that include all of the sex-biased and non-sex-biased genes with the tau datasets
liver_bias_tau_all <- merge(as.data.frame(liver_con_res_noNA), as.data.frame(tau), by = "row.names")
gonad_bias_tau_all <- merge(as.data.frame(gonad_con_res_noNA), as.data.frame(tau), by = "row.names")
gill_bias_tau_all <- merge(as.data.frame(gill_con_res_noNA), as.data.frame(tau), by = "row.names")

#Plot logFC against Tau for all of the organs
plot(sqrt(abs(gonad_bias_tau_all$log2FoldChange)),
     gonad_bias_tau_all$tau,
     pch=19,
     col = "#EEB42275",
     xlab="sex bias (|log2 fold change|)",
     ylab=expression(tau["TPM"]))
points(sqrt(abs(liver_bias_tau_all$log2FoldChange)),
     liver_bias_tau_all$tau,
     pch=19,
     col="#EE826275")
points(sqrt(abs(gill_bias_tau_all$log2FoldChange)),
     gill_bias_tau_all$tau,
     pch=19,
     col = "#6E8B3D75")

#Add the best fit lines
abline(lm(gonad_bias_tau_all$tau ~ sqrt(abs(gonad_bias_tau_all$log2FoldChange))), col="#EEB422", lwd = 3)
abline(lm(liver_bias_tau_all$tau ~ sqrt(abs(liver_bias_tau_all$log2FoldChange))), col = "#EE8262", lwd = 3, lty = 2)
abline(lm(gill_bias_tau_all$tau ~ sqrt(abs(gill_bias_tau_all$log2FoldChange))), col = "#6E8B3D", lwd = 3, lty = 3)

#Add a legend
legend("bottomright", c("Gonads","Liver",
                    "Gills"), 
       lty=c(1,2,3), 
       col=c("#EEB422","#EE8262","#6E8B3D"), 
       bty='o',
       lwd=3)

legend('bottomright', c("","",""), 
       pch=c(21,21,21),
       pt.bg = c("#EEB42275","#EE826275","#6E8B3D75"),
       pt.lwd = 1,
       pt.cex = 1.5,
       col=c("#EEB422","#EE8262","#6E8B3D"), bty='n',
       text.width = 0.55,
       cex=1.0
       )
```

The absolute value of the log2 fold change was square root transformed. We can see similar patterns as when we only plotted the sex biased genes (Fig. \@ref(fig:tau-v-bias)) with the strongest relationship showing up in the gonads. When we add in all of the points, however, it seems that the best fit line for the gills now slightly points downward (Fig. \@ref(fig:tau-v-FC)). Let's see how the Spearman's rank correlation test have changed. This was done between $\tau$ and the $|log2FoldChange|$ for each organ separately.

```{r tau-v-FC-cor, echo=FALSE, warning=FALSE}
cor.test(gonad_bias_tau_all$tau, abs(gonad_bias_tau_all$log2FoldChange), method = "spearman")
cor.test(liver_bias_tau_all$tau, abs(liver_bias_tau_all$log2FoldChange), method = "spearman")
cor.test(gill_bias_tau_all$tau, abs(gill_bias_tau_all$log2FoldChange), method = "spearman")
```

We can see the rho values going down for both the gonads and the liver, but the correlations are still significant. We can also see that in the gills we now have a significant negative correlation between sex bias and $\tau$.

## Categories of sex bias vs $\tau$

To further investigate the relationship between $\tau$ and sex biased and possibly get a cleaner idea of what may be going on I categorized the sex biased genes based on a series of logFC thresholds. After that I can plot $\tau$ against the bias level.

```{r bias-bin-cats, echo=FALSE}
#Create a long format dataset that has the tissue information, logFC, tau, and geneID 
logFC_long <- data.frame(
  tissue=c(rep("Gill", nrow(gill_bias_tau)),
           rep("Gonad", nrow(gonad_bias_tau)),
           rep("Liver", nrow(liver_bias_tau))
         ),
  logFC=c(gill_bias_tau$log2FoldChange,
          gonad_bias_tau$log2FoldChange,
          liver_bias_tau$log2FoldChange
          ),
  geneID=c(gill_bias_tau$Row.names,
           gonad_bias_tau$Row.names,
           liver_bias_tau$Row.names
           ),
  tau=c(gill_bias_tau$tau,
        gonad_bias_tau$tau,
        liver_bias_tau$tau
        )
  
)

#Add a column to denote female-biased or male biased
logFC_long$bias <- ifelse(logFC_long$logFC < 0, paste0("FB"), paste0("MB"))

#Categorize the degree of sex bias in each row
#Make a vector that contains all of the groupings
biased_bins <- c("low", "med", "high", "extreme", "sex-specific")

#Create a new column in the dataset and use ifelse statements to set the category limits
#abs(logFC) was used to account for the fem-biased genes possessing negative values
logFC_long$bias_cat <- ifelse(abs(logFC_long$logFC) >= 2 & abs(logFC_long$logFC) < 3, biased_bins[1],
                                   ifelse(abs(logFC_long$logFC) >= 3 & abs(logFC_long$logFC) < 5, biased_bins[2],
                                          ifelse(abs(logFC_long$logFC) >= 5 & abs(logFC_long$logFC) < 10,
                                                 biased_bins[3], 
                                                 biased_bins[4])
                                          )
                                   )

#Making sure the ordering stays correct
logFC_long$bias_cat <- factor(logFC_long$bias_cat,
    levels = biased_bins, ordered = TRUE)
```

```{r tau-v-biascat-plot, echo=FALSE, fig.width=15, fig.height=8, fig.cap="Tissue specifcity (tau) across the different bias levels. Color denotes female-biased versus male biased and jitters were added to show all of the raw tau values."}
my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4")

plotdat<-ggplot(logFC_long,
       aes(x = bias_cat, y = tau, fill = bias)) +
  geom_boxplot() +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme(axis.line = element_blank(),
    plot.background = element_blank(),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    panel.border = element_blank()) +
  #theme_minimal() +
  #geom_point(data=logFC_long, aes(x=bias_cat, y=tau),
   #          size=0.4, 
    #         position = position_jitter(width = 0.15)) + 
  labs(x="Bias Category", y=expression(tau["TPM"])) +
  guides(fill = guide_legend(title = "Bias Level", order = 3))

plotdat + 
geom_point(data = logFC_long[logFC_long$bias == "MB",], aes(x=as.numeric(bias_cat)+0.185, y=tau),
          size=0.4, 
          position = position_jitter(width = 0.15),
          bg=my_colors["MB"],
          col="black",pch=21
          )+ 
geom_point(data = logFC_long[logFC_long$bias == "FB",], aes(x=as.numeric(bias_cat)-0.185, y=tau),
          size=0.4, 
          position = position_jitter(width = 0.15),
          bg=my_colors["FB"],
          col="black",pch=21
          )

```

I want to see how $\tau$ in unbiased genes compares to these categories as well. To do that I recreated the long-style dataset to include all of the sig. expressed genes rather than just the sex-biased genes.

```{r bias-bin-cats-all, echo=FALSE}
#Create a long format dataset that has the tissue information, logFC, tau, and geneID 
logFC_long_all <- data.frame(
  tissue=c(rep("Gill", nrow(gill_bias_tau_all)),
           rep("Gonad", nrow(gonad_bias_tau_all)),
           rep("Liver", nrow(liver_bias_tau_all))
         ),
  logFC=c(gill_bias_tau_all$log2FoldChange,
          gonad_bias_tau_all$log2FoldChange,
          liver_bias_tau_all$log2FoldChange
          ),
  geneID=c(gill_bias_tau_all$Row.names,
           gonad_bias_tau_all$Row.names,
           liver_bias_tau_all$Row.names
           ),
  tau=c(gill_bias_tau_all$tau,
        gonad_bias_tau_all$tau,
        liver_bias_tau_all$tau
        ),
  padj=c(gill_bias_tau_all$padj,
         gonad_bias_tau_all$padj,
         liver_bias_tau_all$padj)
  
)

#Add a column to denote female-biased or male biased
logFC_long_all$bias <- ifelse(logFC_long_all$logFC <= -2 & logFC_long_all$padj <= 0.05,
                              paste0("FB"),
                              ifelse(logFC_long_all$logFC >= 2 & logFC_long_all$padj <= 0.05,
                                     paste0("MB"),
                                     paste0("UB")
                                     )
                            )
  
#Categorize the degree of sex bias in each row
#Make a vector that contains all of the groupings
biased_bins <- c("Unbiased", "Low", "Med", "High", "Extreme")

#Create a new column in the dataset and use ifelse statements to set the category limits
#abs(logFC) was used to account for the fem-biased genes possessing negative values
logFC_long_all$bias_cat <- ifelse(logFC_long_all$bias == "UB", 
                              biased_bins[1],
                              ifelse(abs(logFC_long_all$logFC) >= 2 & abs(logFC_long_all$logFC) < 3,
                                     biased_bins[2],
                                     ifelse(abs(logFC_long_all$logFC) >= 3 & abs(logFC_long_all$logFC) < 5,
                                            biased_bins[3],
                                            ifelse(abs(logFC_long_all$logFC) >= 5 & abs(logFC_long_all$logFC) < 10,
                                                   biased_bins[4],
                                                   biased_bins[5])
                                            )
                                     )
                              )
  

#Making sure the ordering stays correct
logFC_long_all$bias_cat <- factor(logFC_long_all$bias_cat,
    levels = biased_bins, ordered = TRUE)
```

```{r tau-v-biascat-plot-all-bp, echo=FALSE, fig.width=15, fig.height=8, fig.cap="Tissue specifcity (tau) across the different bias levels. Color denotes female-biased versus male biased versus unbiased and jitters were added to show all of the raw tau values."}
my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "darkgray" )

plotdat <- ggplot(logFC_long_all,
       aes(x = bias_cat, y = tau, fill = bias)) +
  geom_boxplot() +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  #geom_point(data=logFC_long_all, aes(x=bias_cat, y=tau),
    #         size=0.4, 
    #         position = position_jitter(width = 0.15)) + 
  labs(x="Bias Category", y=expression(tau["TPM"])) +
  guides(fill = guide_legend(title = "Bias Level", order = 3))

plotdat + 
geom_point(data = logFC_long_all[logFC_long_all$bias == "MB",], aes(x=as.numeric(bias_cat)+0.185, y=tau),
          size=0.6, 
          position = position_jitter(width = 0.15),
          bg=my_colors["MB"],
          col="black",pch=21
          ) + 
geom_point(data = logFC_long_all[logFC_long_all$bias == "FB",], aes(x=as.numeric(bias_cat)-0.185, y=tau),
          size=0.6, 
          position = position_jitter(width = 0.15),
          bg=my_colors["FB"],
          col="black",pch=21
          ) + 
geom_point(data = logFC_long_all[logFC_long_all$bias == "UB",], aes(x=bias_cat, y=tau),
          size=0.6, 
          position = position_jitter(width = 0.2),
          bg=my_colors["UB"],
          col="black",pch=21)
```


```{r tau-v-biascat-plot-all-vp, echo=FALSE, fig.height=8, fig.cap="Violin plot showing tissue specifcity (tau) across the different bias levels. Color denotes female-biased versus male biased versus unbiased and jitters were added to show all of the raw tau values."}
my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "darkgray" )

plotdat <- ggplot(logFC_long_all,
       aes(x = bias_cat, y = tau, fill = bias)) +
  geom_violin(position = position_dodge(), draw_quantiles = c(0.5)) +
  geom_boxplot(width = 0.1, color = "black", position = position_dodge(width = 0.9)) +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +

  labs(x="Bias Category", y=expression(tau["TPM"])) +
  guides(fill = guide_legend(title = "Bias Level", order = 3))

plotdat + 
geom_point(data = logFC_long_all[logFC_long_all$bias == "MB",], aes(x=as.numeric(bias_cat)+0.185, y=tau),
          size=0.6, 
          position = position_jitter(width = 0.15),
          bg=my_colors["MB"],
          col="black",pch=21
          ) + 
geom_point(data = logFC_long_all[logFC_long_all$bias == "FB",], aes(x=as.numeric(bias_cat)-0.185, y=tau),
          size=0.6, 
          position = position_jitter(width = 0.15),
          bg=my_colors["FB"],
          col="black",pch=21
          ) + 
geom_point(data = logFC_long_all[logFC_long_all$bias == "UB",], aes(x=bias_cat, y=tau),
          size=0.6, 
          position = position_jitter(width = 0.15),
          bg=my_colors["UB"],
          col="black",pch=21)
```

```{r tau-v-biascat-plot-all-vp, echo=FALSE, fig.height=8, fig.cap="Violin plot showing tissue specifcity (tau) across the different bias levels. Color denotes female-biased versus male biased versus unbiased."}
my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "darkgray" )

sample_size = logFC_long_all %>% group_by(bias_cat, bias, tissue) %>% summarize(num=n())
logFC_long_all$group <- paste0(logFC_long$bias, logFC_long_all$bias_cat)

logFC_long_all %>%
  #left_join(sample_size) %>%
  #mutate(myaxis = paste0(bias_cat, "\n", "n=", num)) %>%
  ggplot(aes(x = bias_cat, y = tau, fill = bias)) +
    geom_violin(position = position_dodge(), draw_quantiles = c(0.5)) +
    geom_boxplot(width = 0.1, color = "black", position = position_dodge(width = 0.9)) +
    scale_fill_manual(values = my_colors) +
    facet_grid(. ~ tissue) +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(), 
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black")) +

    labs(x="Bias Category", y=expression(tau["TPM"]))  +
    stat_n_text(data = logFC_long_all[logFC_long_all$bias == "FB",], aes(x=bias_cat, y=tau),
                hjust = 1.2,
                #vjust = -2,
                y.pos = 0.95) +
    stat_n_text(data = logFC_long_all[logFC_long_all$bias == "MB",], aes(x=bias_cat, y=tau),
                hjust = -0.2,
                y.pos = 0,
                vjust = 2) +
    stat_n_text(data = logFC_long_all[logFC_long_all$bias == "UB",], aes(x=bias_cat, y=tau),
                vjust = -4 ) +
    guides(fill = guide_legend(title = "Bias Level", order = 3))

```

## Exploring the relationship between $\tau$ and sex bias in a single-factor analysis
Other papers have looked at the relationship between $\tau$ and sex bias across all tissue types together, without partitioning it out. To do this we need to get $log_2$ male/female expression levels across ALL of the tissues. I will be doing this by running a single-factor model with DESeq2 so instead of counts ~ group it will be counts ~ Sex.

Once we have the results from the new DESeq2 model, $\tau$ can once again be plotted against sex bias ($log_2$ fold change).

```{r tau-v-FC-MFbias, echo=FALSE, fig.cap="Sex bias (in terms of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are signifcantly expressed. The fold change in this plot  is a difference bewteen males and females regardless of tissue type.", message=FALSE}
##Re-running the DESeq2 model to look at M-F diff expression across ALL tissues
dds_FL_SF <- dds_FL
design(dds_FL_SF) <- ~ Sex
dds_FL_SF_exp <- DESeq(dds_FL_SF)

#Pulling out the results and removing NAs
sex_con_res <- results(dds_FL_SF_exp, alpha = 0.05)
sex_con_res_noNA <- sex_con_res[!is.na(sex_con_res$padj),]

#Merge the noNA dataset with the tau datasets
MFbias_tau_all <- merge(as.data.frame(sex_con_res_noNA), as.data.frame(tau), by = "row.names")

#Plot logFC against Tau for all of the organs
plot(sqrt(abs(MFbias_tau_all$log2FoldChange)),
     MFbias_tau_all$tau,
     pch=19,
     col = "#20B2AA75",
     xlab="sex bias (log2 fold change)",
     ylab=expression(tau["TPM"]))

#Add the best fit line
abline(lm(MFbias_tau_all$tau ~ sqrt(abs(MFbias_tau_all$log2FoldChange))), col="black", lwd = 3, lty = 2)
```

After plotting the relationship (Fig. \@ref(fig:tau-v-FC-MFbias)) we can once again look at the Spearman's rank correlation.

```{r tau-v-FC-MFbias-cor, echo=FALSE, message=FALSE, warning=FALSE}
cor.test(MFbias_tau_all$tau, abs(MFbias_tau_all$log2FoldChange), method = "spearman")
```

There is a significant correlation between $\tau$ and abs(log2FoldChange). From the previous plots (Fig. \@ref(fig:tau-v-FC) and Fig. \@ref(fig:tau-v-bias)) we can presume that this significant correlation and rho is likely driven by the gonads.

To get a better idea of what is going on I first grouped each gene into female-biased, male-biased, or unbiased based on the logFC and adjusted p-value. After creating the groups I can plot $\tau$ against the different biases.

```{r MF-long, echo=FALSE, fig.cap="Tau across male-biased, female-biased, and unbiased genes."}
#Create a long format dataset that has the tissue information, logFC, tau, and geneID 
MFlogFC_long <- data.frame(
  logFC=c(MFbias_tau_all$log2FoldChange),
  geneID=c(MFbias_tau_all$Row.names),
  tau=c(MFbias_tau_all$tau),
  padj=c(MFbias_tau_all$padj)
  
)

#Add a column to denote female-biased or male biased
MFlogFC_long$bias <- ifelse(MFlogFC_long$logFC <= -2 & MFlogFC_long$padj <= 0.05, paste0("FB"), 
                            ifelse(MFlogFC_long$logFC >= 2 & MFlogFC_long$padj <= 0.05, paste0("MB"),
                                   paste0("UB"))
                            )

#Plot tau versus the Bias group
my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "darkgray")

ggplot(MFlogFC_long,
       aes(x = bias, y = tau, fill = bias)) +
  geom_boxplot() +
  scale_fill_manual(values = my_colors) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_point(data=MFlogFC_long, aes(x=bias, y=tau),
             size=0.4, 
             position = position_jitter(width = 0.15)) + 
  labs(x="Bias Level", y=expression(tau["TPM"])) +
  guides(fill = guide_legend(title = "Bias Level", order = 3))

```

The jitter points added on top represent the raw $\tau$ value. There is a large spread of $\tau$ for all of the bias groups, but from looking at the boxplot it does appear that the unbiased genes have lower $\tau$ than both the male-biased and the female-biased (Fig. \@ref(fig:MF-long))
