---
title: "Differential Expression Analysis in _Syngnathus floridae_"
output:
  html_document:
    df_print: paged
    code_folding: show
    toc: yes
    toc_float: yes
    number_sections: yes
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
```

``` {r library, include = FALSE}
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
library(ggpubr)
library(cowplot)
library(UpSetR)
#library(ComplexUpset)
library(ggplot2)
library(kableExtra)
library(tidyverse)
library(knitr)
```

``` {r functions}
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.salmon <- readRDS("data/txi.salmon.floride.RDS")

#The samples file generated for tximport
samples <- read.table("FL_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$Sex <- as.factor(samples$Sex)
samples$Organ <- as.factor(samples$Organ)

#Adding in Mating success and Rep fitness(total number of eggs transferred)
samples$mate_success <- c("1", "1", "3", "1", "2", "2", "1", "2", "0",
                          "1", "1", "3", "3", "1", "2", "2", "1", "2", "0",
                          "1", "1", "3", "3", "0",
                          "1", "1", "2", "1", "1")

samples$rep_fittness <- c(169, 88, 762, 285, 467, 500, 427, 156, 0, 
                          169, 88, 762, 916, 285, 467, 500, 427, 156, 0, 
                          169, 88, 762, 916, 0,
                          159, 285, 156, 372, 166)
```

The package `DESeq2` was used for the differential expression analysis outlined below.

# Single factor analysis - Comparing Males v Females
To analyze your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions.
```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_FL <- DESeqDataSetFromTximport(txi.salmon, 
                                   colData = samples,
                                   design = ~ Sex)

```


The data is then pre-filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.
```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_FL)) >= 10
dds_FL <- dds_FL[keep, ]

```


After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.
```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_FL_exp <- DESeq(dds_FL)

#Compile the results
res <- results(dds_FL_exp)
res
```

Once that has finished we can now start exploring some of the single-factor analysis results
```{r investigate-DESeq-results-SF}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#How many ADJUSTED p-values were less than 0.1?
sum(res$padj < 0.1, na.rm = TRUE)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_FL_exp, alpha = 0.05)
summary(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```

## Visualizing the results {.tabset}
### MA-plot - MvF Diff. {-}
Generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be colored if the adjusted p-value is less that 0.1.
```{r MA-plot, echo=FALSE}
par(mfrow=c(1,2))
plotMA(res, ylim = c(-2,2))
plotMA(res)

```

### Heatmap - overall expression {-}
We can also generate a heat map to look ar overall expression levels across our samples. Note, this is not differentially expressed genes.
```{r heatmap, echo=FALSE}
#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(dds_FL_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]
df <- as.data.frame(colData(dds_FL)[,c("Sex", "Organ")])

#Transform the data
vsd <- vst(dds_FL_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df)
```

From the heatmap we can pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.
```{r heatmap_genes}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))
#Run once to create the file and then commented out
#write.table(heatmap_TG,
  #          'heatmap_trinitygenes.txt',
  #          sep = "",
  #          quote=FALSE,
  #          row.names = FALSE,
  #          col.names = FALSE)
```

### Sample-dist Heatmap {-}
We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples
```{r sample-dist, echo=FALSE}
#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$Sex, vsd$Organ, sep = "-")
colnames(sampleDistsMatrix) <- vsd$ID

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)
```

### PCA plots {-}
```{r pca-plot}
colData <- as.data.frame(samples)
rownames(colData) <- samples$ID

#Prepare the PCA using the transformed data from above
pcaFL <- pca(assay(vsd), metadata = colData)

#Plot the PCA
biplot(pcaFL,
       showLoadings = FALSE,
       colby = 'Organ', colkey = c('Liver'='forestgreen', 'Gill'='purple', 'Gonad'='gold'),
       colLegendTitle = "Organ Type",
       hline = 0, vline = 0,
       vlineType = 'dashed',
       pointSize = 3,
       legendPosition = 'left', legendLabSize = 10, legendIconSize = 6.0, legendTitleSize = 10,
       shape = 'Sex', shapekey = c('M'=17, 'F'=8),
       shapeLegendTitle = 'Sex',
       drawConnectors = FALSE,
       title = 'PCA bi-plot for Syngnathus floridae')

pairsplot(pcaFL,
          components = getComponents(pcaFL, c(1, 2, 3)),
          triangle = FALSE,
          hline = 0, vline = 0,
          pointSize = 3,
          gridlines.major = FALSE, gridlines.minor = FALSE,
          colby = 'Organ', colkey = c('Liver'='forestgreen', 'Gill'='purple', 'Gonad'='gold'),
          shape = 'Sex', shapekey = c('M'=17, 'F'=8),
          title = 'Pairs plot', titleLabSize = 22,
          legendPosition = 'left',
          axisLabSize = 10, plotaxes = TRUE,
          margingaps = unit(c(0.1, 0.1, 0.1, 0.1), 'cm'))
```

```{r pca_scaled, echo=FALSE}

pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)
pca_plotting_data$mate_success <- samples$mate_success
pca_plotting_data$repo_fit <- samples$rep_fittness
pca_plotting_data$Organ <- samples$Organ
pca_plotting_data$Sex <- samples$Sex

pcaData <- plotPCA(vsd, intgroup = c("Organ", "Sex"), returnData = TRUE, ntop = nrow(dds_FL))
pcaData$mate_success <- samples$mate_success
pcaData$repo_fit <- samples$rep_fittness

percentVar <- round(100 * attr(pcaData, "percentVar")) # adjust this too

my_colors <- c("Gill" = "#6E8B3D75", "Gonad" = "#EEB42275", "Liver" = "#EE826275")
my_shapes <- c(16, 17)

par(mfrow=c(1,2))
ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = Organ, shape = Sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(aes(size = mate_success)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"))

my_colors2 <- c("F" = "#7fc97f75", "M" = "#beaed475" )
my_shapes2 <- c(16, 17, 15)
ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = Sex, shape = Organ)) +
  scale_color_manual(values = my_colors2) +
  scale_shape_manual(values = my_shapes2) +
  geom_point(aes(size = mate_success)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"))

ggplot(pca_plotting_data, aes(x=PC1, y=PC3, color = Sex, shape = Organ)) +
  scale_color_manual(values = my_colors2) +
  scale_shape_manual(values = my_shapes2) +
  geom_point(aes(size = mate_success)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC3: 10% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"))

#ggplot(pcaData, aes(x=PC1, y=PC2, color = Organ, shape = Sex)) +
#  scale_color_manual(values = my_colors) +
#  scale_shape_manual(values = my_shapes) +
#  geom_point(aes(size = repo_fit)) +
#  xlab(paste0("PC1: ",percentVar[1],"% #variance")) +
#  ylab(paste0("PC2: ",percentVar[2],"% #variance")) + 
#  coord_fixed()
```


# Multifactor design - Comparing M v F across the diff tissue types
If we investigate the column data of our DESeq dataset we can see that each sample has both a sex and organ type attached to it, we will be using these two factors in our multi-factor analysis

```{r multi-factor-analysis, message=FALSE, warning=FALSE}
##Create a copy of the DESeq dataset
ddsMF_FL <- dds_FL

#Create an additional column the has the two conditions combined(sex and organ type)
ddsMF_FL$group <- factor(paste0(ddsMF_FL$Sex, ddsMF_FL$Organ))

#Reapply that new column as the design for this analysis
design(ddsMF_FL) <- ~ group

#Run the differential expression analysis
ddsMF_FL_exp <- DESeq(ddsMF_FL)
resultsNames(ddsMF_FL_exp)

```

## Invesitgate the results of the differential expression {.tabset}
Thanks to the multi-factor analysis we can now explore differential expression between all of the different combinations: \  

  - Male Liver v. Female Liver \  
  - Male Gill v. Female Gill \  
  - Male Gonad v. Female Gonad \  
  - All of the within sex tissue comparisons (e.g. Male Liver v. Male Gill, etc.) \  

### M-F Liver Comparisson {-}
```{r m-f-liver-results}
##Pulling out the liver M-F results with an alpha of 0.05
liver_con_res <- results(ddsMF_FL_exp, contrast = c("group", "MLiver", "FLiver"), alpha = 0.05)
liver_con_res$trin_geneid <- rownames(liver_con_res)
head(liver_con_res)

summary(liver_con_res)
```

There are many criteria that have been employed previously to denote sex bias. For this study we are classifying sex-biased genes as **genes with a p-value < 0.05 AND a logFC $\ge$ |2|**. With that criteria in the liver there are `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. 

I have then pulled out all of the male-biased, female-biased, and non-biased genes based on the criteria outlined above. I determined non-biased genes as a p-vaule > 0.05.

```{r liver-sex-biased}
#Removing the rows where padj. is NA in results
liver_con_res_noNA <- liver_con_res[!is.na(liver_con_res$padj),]
summary(liver_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating a vector that contains all of the male-biased and female-biased genes in the liver
liver_mal_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange >= 2 & liver_con_res_noNA$padj <= 0.05),]
liver_fem_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange <= -2 & liver_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the liver
liver_non_biased <- liver_con_res_noNA[which(liver_con_res_noNA$padj > 0.05),]

```

I will be generating a table that outlines the genes that are the most male or female biased in each organ type. To do this I need to pull the top 50 differentially expressed genes in males or females, get the Trinity gene IDs and then BLAST the corresponding sequences. Here I am pulling out the Trinity gene IDs for those genes.

```{r top50-liver}
#Creating a subset of the results where the p-value is less than 0.05
liver_con_res_p05 <- liver_con_res_noNA[liver_con_res_noNA$padj <= 0.05, ]


#Pulling the top 50 diff expressed genes in Males
top50_MLiver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(liver_con_res_p05$log2FoldChange, decreasing = TRUE),], n=50)))
#Run once to create the file and then commented out
#write.table(top50_MLiver_trin_genes,
 #           'FL_maleL_top50TRgenes.txt', 
 #            sep = "", 
 #            quote=FALSE, 
 #            row.names = FALSE, 
 #            col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_Fliver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(liver_con_res_p05$log2FoldChange, decreasing = FALSE),], n=50)))
#Run once to create the file and then commented out
#write.table(top50_Fliver_trin_genes,
 #           'FL_femL_top50TRgenes.txt', 
 #           sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)
```


### M-F Gill Comparisson {-}
```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res <- results(ddsMF_FL_exp, contrast = c("group", "MGill", "FGill"), alpha = 0.5)
gill_con_res$trin_geneid <- rownames(gill_con_res)
head(gill_con_res)

summary(gill_con_res)
```

In the gills there were `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

I have then pulled out all of the male-biased, female-biased, and non-biased genes to save them into their own objects.

```{r gill-sex-biased}
#Removing the rows where padj. is NA in results
gill_con_res_noNA <- gill_con_res[!is.na(gill_con_res$padj), ]
summary(gill_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating a vector that contains all of the male-biased and female-biased genes in the gills
gill_mal_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange >= 2 & gill_con_res_noNA$padj <= 0.05),]
gill_fem_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange <= -2 & gill_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gills, p>0.05
gill_non_biased <- gill_con_res_noNA[which(gill_con_res_noNA$padj > 0.05),]
```


Here I am getting the trinity gene IDs that correspond to the top 50 differentially expressed genes for males and females. Because there are only 21 female-biased genes I will only use those.

```{r top50-gill}
#Creating a subset of results where p-value is less than 0.05
gill_con_res_p05 <- gill_con_res_noNA[gill_con_res_noNA$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_MGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(gill_con_res_p05$log2FoldChange, decreasing = TRUE),], n=50)))
#Run once to create the file and then commented out
#write.table(top50_MGill_trin_genes,
 #           'FL_maleG_top50TRgenes.txt', 
 #           sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)

#Pulling the top 21 diff expressed genes in Females - Only 21 female-biased genes in the gills
top21_FGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(gill_con_res_p05$log2FoldChange, decreasing = FALSE),], n=21)))
#Run once to create the file and then commented out
#write.table(top21_FGill_trin_genes,
 #           'FL_femG_top21TRgenes.txt', 
 #           sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)
```

### M-F Gonad Comparisson {-}
```{r m-f-gonads}
##Pulling out the gonad M-F results
gonad_con_res <- results(ddsMF_FL_exp, contrast = c("group", "MGonad", "FGonad"), alpha = 0.05)
gonad_con_res$trin_geneid <- rownames(gonad_con_res)
head(gonad_con_res)

summary(gonad_con_res)
```

Between the ovaries and testis (gonads) there were `r sum(gonad_con_res$padj <= 0.05 & gonad_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gonad_con_res$padj <= 0.05 & gonad_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

From the gonad M-F results I have then pulled out all of the male-biased, female-biased, and non-biased genes and saved them to their own objects.

```{r gonad-sex-biased}
#Removing the rows where padj. is NA in results
gonad_con_res_noNA <- gonad_con_res[!is.na(gonad_con_res$padj), ]
summary(gonad_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating an object that contains all of the male-biased and female-biased genes in the gonads
gonad_mal_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$log2FoldChange >= 2 & gonad_con_res_noNA$padj <= 0.05),]
gonad_fem_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$log2FoldChange <= -2 & gonad_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gonads, p>0.05
gonad_non_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$padj > 0.05),]
```

I then pulled out the trinity gene IDs that corresponded to the top 50 differentially expressed genes in males and females so that I could BLAST the sequences.
```{r top50-gonad}
#Creating a subset of the results where the p-value is less than 0.05
gonad_con_res_p05 <- gonad_con_res_noNA[gonad_con_res_noNA$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_MGonad_trin_genes <- cbind(rownames(head(gonad_con_res_p05[order(gonad_con_res_p05$log2FoldChange, decreasing = TRUE),], n=50)))
#Run once to create the file and then commented out
#write.table(top50_MGonad_trin_genes,
 #           'FL_maleGon_top50TRgenes.txt', 
 #           sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FGonad_trin_genes <- cbind(rownames(head(gonad_con_res_p05[order(gonad_con_res_p05$log2FoldChange, decreasing = FALSE),], n=50)))
#Run once to create the file and then commented out
#write.table(top50_FGonad_trin_genes,
 #           'FL_femGon_top50TRgenes.txt', 
 #           sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)
```

### MA Plots {-}
```{r MA-plot-MF, echo=FALSE, fig.height= 20}
liver_MA <- ggmaplot(liver_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Liver" %->% "Female Liver"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gill_MA <- ggmaplot(gill_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Gill" %->% "Female Gill"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gonad_MA <- ggmaplot(gonad_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Testis" %->% "Ovaries"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

#Merge all the plots into one
plot_grid(liver_MA, gill_MA, gonad_MA,
          ncol = 1,
          nrow = 3,
          labels = c('A', 'B', 'C'),
          label_fontfamily = 'sans',
          label_fontface = 'bold',
          label_size = 18,
          rel_heights = c(4.10, 4.10, 4.10))

```

## Creating an Upset Plot
```{r upset-plot, echo=FALSE}
#Male-biased genes
listInputMB <- list("Testis" = rownames(gonad_mal_biased),
                  "Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased)
                  )
upset(fromList(listInputMB),
      mainbar.y.label = "Number of Shared Male-Biased Genes",
      sets.x.label = "Total Number of Male-Biased Genes",
      point.size = 3)

#Female
listInputFB <- list("Ovaries" = rownames(gonad_fem_biased),
                  "Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased)
                  )
upset(fromList(listInputFB),
      mainbar.y.label = "Number of Shared Female-Biased Genes",
      sets.x.label = "Total Number of Female-Biased Genes",
      point.size = 3)

#Checking there is no overlap between the two sexes
listInputall <- list("Testis" = rownames(gonad_mal_biased),
                  "Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased),
                  "Ovaries" = rownames(gonad_fem_biased),
                  "Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased))
upset(fromList(listInputall),
      mainbar.y.label = "Number of Shared Sex-Biased Genes",
      sets.x.label = "Total Number of Sex-Biased Genes",
      point.size = 3,
      nsets = 6,
      nintersects = NA)
```

## Variation in FC across sex-bias and tissue type
I want to create a plot the highlights the variation we see in fold-change both across the different biases groups (male-biased, female-biased, and non-biased) within one tissue type and across all of the tissue types. To do this I first need to create a "long" dataset in the following format:

| Tissue_type|     Bias      |   Fold-Change   |
|:----------:|:-------------:|:---------------:|
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   fem_bias    |      XXXXX      |
|    Gill    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   fem_bias    |      XXXXX      |
|    Liver   |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   fem_bias    |      XXXXX      |
|   Gonad    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |

That dataset is created here, with the Trinity gene IDs included as well:

```{r logFC-long-dataset}
logFC_long<- data.frame(
  tissue=c(rep("Gill",nrow(gill_fem_biased)),
           rep("Gill", nrow(gill_mal_biased)),
           rep("Gill", nrow(gill_non_biased)),
           rep("Gonad", nrow(gonad_fem_biased)),
           rep("Gonad", nrow(gonad_mal_biased)),
           rep("Gonad", nrow(gonad_non_biased)),
           rep("Liver", nrow(liver_fem_biased)),
           rep("Liver", nrow(liver_mal_biased)),
           rep("Liver", nrow(liver_non_biased))
         ),
  bias=c(rep("FB",nrow(gill_fem_biased)),
         rep("MB",nrow(gill_mal_biased)),
         rep("NB", nrow(gill_non_biased)),
         rep("FB", nrow(gonad_fem_biased)),
         rep("MB", nrow(gonad_mal_biased)),
         rep("NB", nrow(gonad_non_biased)),
         rep("FB", nrow(liver_fem_biased)),
         rep("MB", nrow(liver_mal_biased)),
         rep("NB", nrow(liver_non_biased))
         ),
  logFC=c(gill_fem_biased$log2FoldChange,
          gill_mal_biased$log2FoldChange,
          gill_non_biased$log2FoldChange,
          gonad_fem_biased$log2FoldChange,
          gonad_mal_biased$log2FoldChange,
          gonad_non_biased$log2FoldChange,
          liver_fem_biased$log2FoldChange,
          liver_mal_biased$log2FoldChange,
          liver_non_biased$log2FoldChange
          ),
  geneID=c(rownames(gill_fem_biased),
           rownames(gill_mal_biased),
           rownames(gill_non_biased),
           rownames(gonad_fem_biased),
           rownames(gonad_mal_biased),
           rownames(gonad_non_biased),
           rownames(liver_fem_biased),
           rownames(liver_mal_biased),
           rownames(liver_non_biased)
           )
  
)
```

With the dataset in the proper format now, we can generate the plot
```{r FC-var-plot, echo=FALSE}
#ggplot(logFC_long,
       #aes(x = tissue, y = abs(logFC), fill = bias)) +
  #geom_boxplot() +
  #geom_jitter(color="black", size=0.4, alpha=0.9) +
  #theme_minimal()

my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "NB" = "darkgray")

ggplot(logFC_long,
       aes(x = bias, y = abs(logFC), fill = bias)) +
  geom_boxplot() +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme_minimal() +
  geom_point(data=logFC_long, aes(x=bias, y=abs(logFC)),
             size=0.4, 
             position = position_jitter(width = 0.15)) + 
  labs(x="Bias Level", y="|logFC|") +
  guides(fill = guide_legend(title = "Bias Level", order = 3))
  #scale_fill_discrete(labels=c('Female-Biased', 'Male-Biased', 'Non-Biased'))
  #geom_jitter(color="black", size=0.4, alpha=0.9)
```

There appears to be differences in the logFC across both the organs and the sex-bias groups (male-biased, fem-biased, and non-biased). Let's explore some analyses to see if we can put some statistical power behind those differences.
```{r FC-var-stats, warning=FALSE}
#Looking at some summary statistics for logFC between the different groups
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), mean)
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), sd)

par(mfrow=c(1,1))
interaction.plot(logFC_long$tissue, logFC_long$bias, logFC_long$logFC)

logFC_long$tissue <- as.factor(logFC_long$tissue)
logFC_long$bias <- as.factor(logFC_long$bias)

fc_var_aov <- aov(abs(logFC_long$logFC) ~ logFC_long$bias*logFC_long$tissue)
anova(fc_var_aov)

par(mfrow = c(2,2))
plot(fc_var_aov)

library(car)
leveneTest(fc_var_aov)
par(mfrow=c(1,1))
hist(resid(fc_var_aov))
#shapiro.test(resid(fc_var_aov)) 

plot(TukeyHSD(fc_var_aov))
library(multcompView)
multcompLetters4(fc_var_aov, TukeyHSD(fc_var_aov))
```


## Categorizing sex-specific genes
Now that we have investigated what the sex-bias is like across the different tissue types I want to dive further into genes with **sex-specific expression**. These are genes expressed only in one sex or the other. I am classifying a sex-specific genes within each tissue as ones where the expression of that gene is 0 for all of one sex and there is a median of $\ge 10$ in the other sex.

```{r sex-specific-for-loop}
#Create a vector with the different organ types
organs <- levels(colData(dds_FL)$Organ)

#Create an empty list to store my datasets in
FL_sex_specific_genes <- list()

#Generate the for loop to identify MSpecific and FSpecific genes in each organ based on Medians
for(organ in organs){
  
  #Male-Specific Genes
  ##Pull out all of the rows where fem count = 0
  fem0_organ <- assay(dds_FL)[rowSums(assay(dds_FL)[, dds_FL$Sex == "F" & dds_FL$Organ==organ]) < 1, dds_FL$Organ == organ]
  
  ##Pull out rows where median of male count >=10 for that organ
  mal10_organ <- apply(assay(dds_FL)[, dds_FL$Sex == "M" & dds_FL$Organ == organ],
                       1,
                       function(x) median(x) >=10)
  
  ##Keep only the rows where fem=0 and the Male median>=10
  fem0_mal10_organ <- fem0_organ[rownames(fem0_organ) %in% names(mal10_organ[mal10_organ == TRUE]),]
  
  ##Create a new object with a name based on the organ type
  organ_malsp <- sub("$", "_male_specific", organ)
  FL_sex_specific_genes[[organ_malsp]] <- fem0_mal10_organ
  
  
  
  #Female-Specific Genes
  ##Pull out all of the rows where male count = 0
  mal0_organ <- assay(dds_FL)[rowSums(assay(dds_FL)[, dds_FL$Sex == "M" & dds_FL$Organ==organ]) < 1, dds_FL$Organ == organ]
  
  ##Pull out rows where median of female count >=10 for that organ
  fem10_organ <- apply(assay(dds_FL)[, dds_FL$Sex == "F" & dds_FL$Organ == organ],
                       1,
                       function(x) median(x) >=10)
  
  #Keep only the rows where male=0 and the fem median>=10
  mal0_fem10_organ <- mal0_organ[rownames(mal0_organ) %in% names(fem10_organ[fem10_organ == TRUE]),]
  
  # Create a new object with a name based on the organ type
  organ_femsp <- sub("$", "_female_specific", organ)
  FL_sex_specific_genes[[organ_femsp]] <- mal0_fem10_organ
} 
```

### Investigating the results of the sex-specific subsetting
Let's now take a look at how many sex-specific genes we have in each tissue type:
```{r sex-specific-counts, echo=FALSE, message=FALSE, warning=FALSE}
sex_specific_counts <- data.frame(tissue = rep(c("Gill", "Gonad", "Liver"), 2),
                                  sex = c(rep("M", 3),
                                          rep("F", 3)),
                                  num_genes = c(nrow(FL_sex_specific_genes$Gill_male_specific),
                                                nrow(FL_sex_specific_genes$Gonad_male_specific),
                                                nrow(FL_sex_specific_genes$Liver_male_specific),
                                                nrow(FL_sex_specific_genes$Gill_female_specific),
                                                nrow(FL_sex_specific_genes$Gonad_female_specific),
                                                nrow(FL_sex_specific_genes$Liver_female_specific))
                                  )

sex_specific_counts %>% 
  group_by(tissue, sex) %>% 
  summarise("Number of Sex-specific Genes" = sum(num_genes)) %>%
  kable(caption = "Sex-specific Genes", format = "html",
        col.names = c("Tissue Type", "Sex", "Number of Sex-specific Genes")) %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F, position = "left")
```

To get a better idea what is going on between the sex-specific genes vs the sex-biased genes I first determined how many overlaps there were between our sex-biased and sex-specific genes and then used plotCounts to plot the counts for each gene separately. 

In the **Male Gills** there are `r nrow(FL_sex_specific_genes$Gill_male_specific[rownames(FL_sex_specific_genes$Gill_male_specific) %in% rownames(gill_mal_biased),])` genes that overlap between sex-biased and sex-specific, there are `r nrow(FL_sex_specific_genes$Gonad_male_specific[rownames(FL_sex_specific_genes$Gonad_male_specific) %in% rownames(gonad_mal_biased),])` overlapping in the **gonads** and `r nrow(FL_sex_specific_genes$Liver_male_specific[rownames(FL_sex_specific_genes$Liver_male_specific) %in% rownames(liver_mal_biased),])` overlapping in the **liver**. For females we have `r nrow(FL_sex_specific_genes$Gill_female_specific[rownames(FL_sex_specific_genes$Gill_female_specific) %in% rownames(gill_fem_biased),])` overlapping sex-specific and sex-biased genes in the **gills**, `r nrow(FL_sex_specific_genes$Gonad_female_specific[rownames(FL_sex_specific_genes$Gonad_female_specific) %in% rownames(gonad_fem_biased),])` genes in the **gonads**, and `r nrow(FL_sex_specific_genes$Liver_female_specific[rownames(FL_sex_specific_genes$Liver_female_specific) %in% rownames(liver_fem_biased),])` genes in the **liver**.

```{r gills-sex-specific, eval = FALSE}
## Female sex-specific
fem_sp_geneIDs <- rownames(FL_sex_specific_genes$Gill_female_specific)

##Set the plotting window
par(mfrow=c(1,1))

for(gene in fem_sp_geneIDs){
  
  plotCounts(ddsMF_FL[,ddsMF_FL$Organ == 'Gill'], gene=gene, intgroup="Sex")
  
}

## Female sex-biased
gill_fem_biased_geneIDS <- rownames(gill_fem_biased)

##Set the plotting window
par(mfrow=c(1,1))

for(gene in gill_fem_biased_geneIDS){
  
  plotCounts(ddsMF_FL[,ddsMF_FL$Organ == 'Gill'], gene=gene, intgroup="Sex", col= 'salmon', pch = 19)
  
}
```

## Creating categories and binning the sex-biased genes based on degree of logFC
This binned was only done on the sex-biased genes, will not have a category for the unbiased genes. The cutoffs for the different groups are as follows:  \

    1. Low - biased = LFC 2 - 3
    2. Medium - biased = LFC 3 - 5
    3. High - biased = LFC 5 - 10
    4. Extreme sex bias = LFC > 10

Counts of the sex-specific genes will be added on separately as they were not classified based on fold-change but rather a presence in one sex and an absence in the other sex.

```{r bin-sex-bias}
#Create  subset of our long dataset that does not include the non-biased genes
logFC_long_noNB <- logFC_long[logFC_long$bias != "NB",]

#Make a vector that contains all of the groupings
biased_bins <- c("low", "med", "high", "extreme", "sex-specific")

#Create a new column in the dataset and use ifelse statements to set the category limits
#abs(logFC) was used to account for the fem-biased genes possessing negative values
logFC_long_noNB$bias_cat <- ifelse(abs(logFC_long_noNB$logFC) >= 2 & abs(logFC_long_noNB$logFC) < 3, biased_bins[1],
                                   ifelse(abs(logFC_long_noNB$logFC) >= 3 & abs(logFC_long_noNB$logFC) < 5, biased_bins[2],
                                          ifelse(abs(logFC_long_noNB$logFC) >= 5 & abs(logFC_long_noNB$logFC) < 10,
                                                 biased_bins[3], 
                                                 biased_bins[4])
                                          )
                                   )

#Make a table to count the number of genes in each category for each organ
bias_cat_table <- table(logFC_long_noNB$bias, logFC_long_noNB$bias_cat, logFC_long_noNB$tissue)

#Add the counts of sex-specific genes to the table
bias_cat_gill <- cbind(bias_cat_table[,, "Gill"], "sex-specific" = c(nrow(FL_sex_specific_genes$Gill_female_specific), nrow(FL_sex_specific_genes$Gill_male_specific)))
bias_cat_gonad <- cbind(bias_cat_table[,, "Gonad"], "sex-specific" = c(nrow(FL_sex_specific_genes$Gonad_female_specific), nrow(FL_sex_specific_genes$Gonad_male_specific)))
bias_cat_liver <- cbind(bias_cat_table[,, "Liver"], "sex-specific" = c(nrow(FL_sex_specific_genes$Liver_female_specific), nrow(FL_sex_specific_genes$Liver_male_specific)))

#Plot the counts for each tissue type
par(mfrow=c(1, 3))
barplot(bias_cat_gill[,biased_bins], beside = TRUE, ylim = c(0, max(bias_cat_table) + 1), col = c("#7fc97f", "#beaed4"),
        main = "Gill")
barplot(bias_cat_gonad[,biased_bins], beside = TRUE, ylim = c(0, max(bias_cat_table) + 1), col = c("#7fc97f", "#beaed4"),
        main = "Gonad")
barplot(bias_cat_liver[,biased_bins], beside = TRUE, ylim = c(0, max(bias_cat_table) + 1), col = c("#7fc97f", "#beaed4"),
        main = "Liver")
legend("topright", legend = c("Female-biased", "Male-biased"), fill = c("#7fc97f", "#beaed4"))
```


## Gene Ontology Analysis
We wanted to figure out two things: \  

  1) What are the genes that are sex-biased within each tissue type \  
  2) What are the GO terms for those genes. \  

To accomplish this I first had to pull out the sequences that corresponded to the Trinity gene IDS pulled out above. Once those sequences were fetched they were blasted against the _Syngnathus scovelli_ genome. This was completed with the following bash script on the RCC. 

### Using BLAST in command line
```{bash blastn, eval = FALSE}
#Create a BLAST database - Used Syngnathus scovelli for our BLAST database
makeblastdb -in ../ncbi_dataset/data/GCF_024217435.1/GCF_024217435.1_RoL_Ssco_1.1_genomic.fna -out s_scov_genome -dbtype nucl

```

```{bash blast_pipeline, file = 'bash/blast_pipeline.sh', eval = FALSE}


```

This script was run as `bash bash_scripts/blast_pipeline.sh floridae_high_exp_genes/high_expTG_names other_scripts/subset_fasta_file trinity_supertran_floridae.fasta ../genomes/s_scov/s_scov_genome floridae_high_exp_genes/high_exp_results/`


Once the BLAST results were obtained in the RCC I read them into R for further filtering (keeping only one hit per gene)
```{r readin_blast_results}
#Specify the directory where BLAST results are located
blast_path <- "data/floridae_BLAST/"

#Create a list of the files I want
FL_blast_files <- list.files(blast_path, pattern = "top")

#Create an empty list to store my datasets in
FL_blast_list <- list()

#Create a loop to read in all of the blast results
for(file_name in FL_blast_files){
  # Read the file and store it in a new object
  file_data <- read.delim(file.path(blast_path, file_name), stringsAsFactors = FALSE, quote = "", sep = "\t", header = FALSE)
  
  #Add column names to the dataset
  colnames(file_data) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "scov_gene_info", "scov_gene_start", "scov_gene_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  # Create a new object with a name based on the file name
  blast_name <- sub(".txt$", "", file_name) #Removes the file extension
  FL_blast_list[[blast_name]] <- file_data
}

```

### Filtering Blast results
Because of the way BLAST was run in the RCC, we can have multiple hits for one gene, I want to filter that out so that there is only one hit kept for each gene.
```{r filter-BLAST}
#Use lapply to apply the function to each dataset stored in the list created above
blast_output_filtered <- lapply(FL_blast_list, function(data_frame){
 
  #Pull out the Unique Trinity gene IDs
  uniqueID <- unique(data_frame[1])
  
  #Create an empty dataframe to store intermediate results in
  output <- data.frame(matrix(data = NA, ncol = ncol(data_frame), nrow = 0))
  colnames(output) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "scov_gene_info", "scov_gene_start", "scov_gene_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  #Generate a for loop that pulls out the lowest e-value + highest % identity for each gene
  for(gene in uniqueID$trin_geneid){
    
    #Subset the dataset for each Trinity gene
    this_trin_gene<- subset(data_frame, trin_geneid==gene)
    #Pull out gene with smallest e-value
    uni_gene_subset_lowe <- this_trin_gene[which(this_trin_gene$evalue == min(this_trin_gene$evalue)),]
    #In case mult. genes have same e-value, pull out gene with highest % identity
    uni_gene_subset_lowe_highpid <- uni_gene_subset_lowe[which(uni_gene_subset_lowe$pident == max(uni_gene_subset_lowe$pident)),]
    
    # keep only one of multiple identical rows
    uni_gene_subset_lowe_highpid <- unique(uni_gene_subset_lowe_highpid)

    # check that there is a single gene ID in scovelli, if not, only first row is kept
    if(length(gsub("^.*(GeneID:\\d+)\\].*$","\\1",uni_gene_subset_lowe_highpid$scov_gene_info))>1){
    #   browser()
    # } else{
      uni_gene_subset_lowe_highpid<-uni_gene_subset_lowe_highpid[1,]
    }
    
    #Add the final gene into the empty dataframe from above
    output <- rbind(output, uni_gene_subset_lowe_highpid)
  }
  
  return(output)
})

```

### Merging the BLAST filtered Data frames with the sex-biased information
Not all of the trinity genes had a BLAST hit come back, but I still want to include what those genes are in the dataset. To do this I am merging all of the top 50 genes with the filtered BLAST dataset and if there are genes that did not have any hits (i.e. are not in the filtered BLAST database) then they will receive an "NA".

All genes that had an "NA" come back from blasting against the _S. scovelli_ database will be blasted against the non-redundant BLAST database as well.

```{r merge-blast-results}
#Getting the top 50 diff. expressed genes in Males
top50_MGonad <- head(gonad_con_res_p05[order(gonad_con_res_p05$log2FoldChange, decreasing = TRUE),], 
                     n=50)
top50_MGill <- head(gill_con_res_p05[order(gill_con_res_p05$log2FoldChange, decreasing = TRUE),], 
                     n=50)
top50_MLiver <- head(liver_con_res_p05[order(liver_con_res_p05$log2FoldChange, decreasing = TRUE),], 
                     n=50)

#Pulling the top 50 diff expressed genes in Females
top50_FGonad <- head(gonad_con_res_p05[order(gonad_con_res_p05$log2FoldChange, decreasing = FALSE),],
                     n=50)
top21_FGill <- head(gill_con_res_p05[order(gill_con_res_p05$log2FoldChange, decreasing = FALSE),],
                     n=21)
top50_FLiver <- head(liver_con_res_p05[order(liver_con_res_p05$log2FoldChange, decreasing = FALSE),],
                     n=50)
#Create dataframes for all of the above that contains the trinity gene ID and the logFoldChange
top50_MGonad_df <- as.data.frame(cbind(trin_geneid = top50_MGonad$trin_geneid,
                                       LogFC = top50_MGonad$log2FoldChange))
top50_MGill_df <- as.data.frame(cbind(trin_geneid = top50_MGill$trin_geneid,
                                       LogFC = top50_MGill$log2FoldChange))
top50_MLiver_df <- as.data.frame(cbind(trin_geneid = top50_MLiver$trin_geneid,
                                       LogFC = top50_MLiver$log2FoldChange))
top50_FGonad_df <- as.data.frame(cbind(trin_geneid = top50_FGonad$trin_geneid, 
                                      LogFC = top50_FGonad$log2FoldChange))
top21_FGill_df <- as.data.frame(cbind(trin_geneid = top21_FGill$trin_geneid, 
                                      LogFC = top21_FGill$log2FoldChange))
top50_FLiver_df <- as.data.frame(cbind(trin_geneid = top50_FLiver$trin_geneid, 
                                      LogFC = top50_FLiver$log2FoldChange))

#Merge the above dataframes with the corresponding blast_output_filtered dataframe
blast_output_filtered$FL_femG_top21TRgenes_blast <- merge(top21_FGill_df, blast_output_filtered$FL_femG_top21TRgenes_blast, by = "trin_geneid", all.x = TRUE, all.y = TRUE)
blast_output_filtered$FL_femGon_top50TRgenes_blast <- merge(top50_FGonad_df, blast_output_filtered$FL_femGon_top50TRgenes_blast, by = "trin_geneid", all.x = TRUE, all.y = TRUE)
blast_output_filtered$FL_femL_top50TRgenes_blast <- merge(top50_FLiver_df, blast_output_filtered$FL_femL_top50TRgenes_blast, by = "trin_geneid", all.x = TRUE, all.y = TRUE)
blast_output_filtered$FL_maleG_top50TRgenes_blast <- merge(top50_MGill_df, blast_output_filtered$FL_maleG_top50TRgenes_blast, by = "trin_geneid", all.x = TRUE, all.y = TRUE)
blast_output_filtered$FL_maleGon_top50TRgenes_blast <- merge(top50_MGonad_df, blast_output_filtered$FL_maleGon_top50TRgenes_blast, by = "trin_geneid", all.x = TRUE, all.y = TRUE)
blast_output_filtered$FL_maleL_top50TRgenes_blast <- merge(top50_MLiver_df, blast_output_filtered$FL_maleL_top50TRgenes_blast, by = "trin_geneid", all.x = TRUE, all.y = TRUE)
```

### Blasting against the Zebrafish genome
In order for us to be able to do the GO analysis, we need to BLAST the sequences against one of the PANTHER organisms. The zebrafish, _Danio reiro_ was the only fish in that list.

A BLAST database was generated with the _D. reiro_ proteome as:
`makeblastdb -in ../ncbi_dataset/data/GCF_000002035.6/protein.faa -out d_rerio_prot -dbtyp prot`

The trinity gene IDs that correspond to ALL of the female- or male-biased genes were pulled out and saved to `.txt` files.
```{r biased-trin-ids, eval=FALSE}
write.table(cbind(rownames(gill_fem_biased)),
            'FL_femG_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(gill_mal_biased)),
            'FL_malG_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(gonad_fem_biased)),
            'FL_femGon_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(gonad_mal_biased)),
            'FL_malGon_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(liver_fem_biased)),
            'FL_femL_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(liver_mal_biased)),
            'FL_malL_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```

The `blast_pipeline.sh` script was then used like above to pull out all of the corresponding sequences for each Trinity gene and then to get the BLAST results. The big difference here was that `blastx` was used instead of `blastn` because we are searching nucleotides against a protein database. Once we have the BLAST results we can read and filter them into R as we did before.

```{r readin_blast_results_dreiro, echo=FALSE}
#Specify the directory where BLAST results are located
blast_path <- "data/floridae_BLAST/"

#Create a list of the files I want
FL_blast_files <- list.files(blast_path, pattern = "_TRgenes")

#Create an empty list to store my datasets in
FL_blast_list <- list()

#Create a loop to read in all of the blast results
for(file_name in FL_blast_files){
  # Read the file and store it in a new object
  file_data <- read.delim(file.path(blast_path, file_name), stringsAsFactors = FALSE, quote = "", sep = "\t", header = FALSE)

  #Get rid of the columns we don't need
  file_data <- file_data[,c(1:4,8,10:16)]
  
  #Add column names to the dataset
  colnames(file_data) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "reiro_prot_info", "prot_id", "reiro_prot_start", "reiro_prot_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  # Create a new object with a name based on the file name
  blast_name <- sub(".txt$", "", file_name) #Removes the file extension
  FL_blast_list[[blast_name]] <- file_data
}

```

```{r filter-BLAST-dreiro, echo=FALSE}
#Use lapply to apply the function to each dataset stored in the list created above
blast_output_filtered <- lapply(FL_blast_list, function(data_frame){
 
  #Pull out the Unique Trinity gene IDs
  uniqueID <- unique(data_frame[1])
  
  #Create an empty dataframe to store intermediate results in
  output <- data.frame(matrix(data = NA, ncol = ncol(data_frame), nrow = 0))
  colnames(output) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "reiro_prot_info", "prot_id", "reiro_prot_start", "reiro_prot_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  #Generate a for loop that pulls out the lowest e-value + highest % identity for each gene
  for(gene in uniqueID$trin_geneid){
    
    #Subset the dataset for each Trinity gene
    this_trin_gene<- subset(data_frame, trin_geneid==gene)
    #Pull out gene with smallest e-value
    uni_gene_subset_lowe <- this_trin_gene[which(this_trin_gene$evalue == min(this_trin_gene$evalue)),]
    #In case mult. genes have same e-value, pull out gene with highest % identity
    uni_gene_subset_lowe_highpid <- uni_gene_subset_lowe[which(uni_gene_subset_lowe$pident == max(uni_gene_subset_lowe$pident)),]
    
    # keep only one of multiple identical rows
    uni_gene_subset_lowe_highpid <- unique(uni_gene_subset_lowe_highpid)

    # check that there is a single gene ID in scovelli, if not, only first row is kept
    if(length(gsub("^.*(GeneID:\\d+)\\].*$","\\1",uni_gene_subset_lowe_highpid$reiro_prot_info))>1){
    #   browser()
    # } else{
      uni_gene_subset_lowe_highpid<-uni_gene_subset_lowe_highpid[1,]
    }
    
    #Add the final gene into the empty dataframe from above
    output <- rbind(output, uni_gene_subset_lowe_highpid)
  }
  
  return(output)
})

```

In order to run PANTHER we need IDs that follow one of their supported formats. From NCBI, the gene name works (e.g. rpl27). That information is not automatically supplied with the BLAST output, but we can use the .gff file from NCBI to pull out the gene names that correspond to the proteinID (e.g. NP_956018.1).

```{r gff-gene-names}
#Read in the GFF file for zebrafish that has info about the geneID
reiro_gff <- read.delim("data/d_reiro_genomic.gff",
                        header = FALSE,
                        comment.char = "#")

#Keep only the columns I want and rename them
reiro_gff <- reiro_gff[,c(1:5,9)]
colnames(reiro_gff) <- c("seqname", "source", "feature", "start", "end", "gene_info")

#Subset the dataset to only include the rows we're interested in
genes <- reiro_gff[reiro_gff$feature == "CDS",]

#Pull out the gene name that corresponds to each protein ID
genes$gene_name <- gsub("^(.*;)(gene=)(\\w+\\d*);(.*$)", "\\3", genes$gene_info)
genes$prot_id <- gsub("^(.*;)(protein_id=)(.*)$", "\\3", genes$gene_info)

#Merge the gene names pulled out above with the rest of the BLAST datasets based on the proteinID
blast_output_merged <- lapply(blast_output_filtered, function(dataframe){
  
   output <- merge(dataframe, unique(genes[,7:8]), by = "prot_id")
  
   return(output)
})

```

I am now going to make sure the sex bias information is included with the BLAST results so we can perform statistical overrepresentation and enrichment tests with PANTHER.
```{r merge-blast-logfc}
blast_output_merged$FL_femG_biased_TRgenes_blast <- merge(as.data.frame(gill_fem_biased), blast_output_merged$FL_femG_biased_TRgenes_blast, by = "trin_geneid", all.y = TRUE)

blast_output_merged$FL_femGon_biased_TRgenes_blast <- merge(as.data.frame(gonad_fem_biased), blast_output_merged$FL_femGon_biased_TRgenes_blast, by = "trin_geneid", all.y = TRUE)

blast_output_merged$FL_femL_biased_TRgenes_blast <- merge(as.data.frame(liver_fem_biased), blast_output_merged$FL_femL_biased_TRgenes_blast, by = "trin_geneid", all.y = TRUE)

blast_output_merged$FL_malG_biased_TRgenes_blast <- merge(as.data.frame(gill_mal_biased), blast_output_merged$FL_malG_biased_TRgenes_blast, by = "trin_geneid", all.y = TRUE)

blast_output_merged$FL_malGon_biased_TRgenes_blast <- merge(as.data.frame(gonad_mal_biased), blast_output_merged$FL_malGon_biased_TRgenes_blast, by = "trin_geneid", all.y = TRUE)

blast_output_merged$FL_malL_biased_TRgenes_blast <- merge(as.data.frame(liver_mal_biased), blast_output_merged$FL_malL_biased_TRgenes_blast, by = "trin_geneid", all.y = TRUE)
```


```{r, eval=FALSE}
write.table(c(blast_output_merged$FL_femG_biased_TRgenes_blast$gene_name, blast_output_merged$FL_femGon_biased_TRgenes_blast$gene_name, blast_output_merged$FL_femL_biased_TRgenes_blast$gene_name, blast_output_merged$FL_malG_biased_TRgenes_blast$gene_name, blast_output_merged$FL_malGon_biased_TRgenes_blast$gene_name, blast_output_merged$FL_malL_biased_TRgenes_blast$gene_name),
           'FL_GOnames.txt', 
            sep = " ", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

write.table(rbind(blast_output_merged$FL_femG_biased_TRgenes_blast[,c("gene_name", "log2FoldChange")],
              blast_output_merged$FL_femGon_biased_TRgenes_blast[,c("gene_name", "log2FoldChange")],
              blast_output_merged$FL_femL_biased_TRgenes_blast[,c("gene_name", "log2FoldChange")],
              blast_output_merged$FL_malG_biased_TRgenes_blast[,c("gene_name", "log2FoldChange")],
              blast_output_merged$FL_malGon_biased_TRgenes_blast[,c("gene_name", "log2FoldChange")],
              blast_output_merged$FL_malL_biased_TRgenes_blast[,c("gene_name", "log2FoldChange")]),
            'FL_GOnames_FC.txt',
            sep = "\t",
            quote = FALSE,
            row.names = FALSE,
            col.names = FALSE)
```
