---
title: "Tissue Specificity in _Syngnathus floridae_"
author: "Coley Tosto"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: show
    toc: yes
    toc_float: yes
    number_sections: yes
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
```

``` {r library, include = FALSE}
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
```

``` {r functions}
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.salmon <- readRDS("data/txi.salmon.floride.RDS")

#The samples file generated for tximport
samples <- read.table("FL_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$Sex <- as.factor(samples$Sex)
samples$Organ <- as.factor(samples$Organ)

samples$group <- factor(paste0(samples$Sex, samples$Organ))

#Changing "Gonad" to be more specific to testis or ovaries
samples$Organ <- ifelse(samples$Sex == "F" & samples$Organ =="Gonad",
                        paste0("Ovaries"),
                        ifelse(samples$Sex == "M" & samples$Organ == "Gonad",
                               paste0("Testis"),
                               paste0(samples$Organ))
                        )


#Format colData to be used in the function
colData <- as.data.frame(samples)
rownames(colData) <- samples$ID

```

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_FL <- DESeqDataSetFromTximport(txi.salmon, 
                                   colData = samples,
                                   design = ~ group)

#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_FL)) >= 10
dds_FL <- dds_FL[keep, ]

```
To estimate tissue specificity, the TPM estimates are needed which is the number of transcripts from a particular gene normalized first by gene length, and then by sequencing depth (in millions) in the sample. The output quant.sf files from salmon have the following format:
```
Name | Length | EffectiveLength | TPM | NumReads

```
From these salmon outputs we can then pull out the TPM values for each sample.

```{r getTPMs}
files <- list.files(pattern = ".sf", path = "data/floridae_expression_files", 
                    full.names = TRUE)

tpms <- do.call(cbind, lapply(files, function(file){
  #browser()
  dat <- read.delim(file, row.names = "Name")
  tpm <- dat["TPM"]
  colnames(tpm) <- gsub("data/floridae_expression_files/(.*)_quant.sf","\\1",file)
  
  return(tpm)
}))

#Only keeping the rows that weren't filtered out due to low counts
tpms <- tpms[rownames(tpms) %in% rownames(dds_FL), ]

```

These TPM estimates can then be used to estimate $\tau$, a tissue specificity estimator that can range from 0 to 1. $\tau$ is calculated for each tissue, $i$, as follows:

$$
\tau=\frac{\sum_i{[1-ln(TPM_i)/ln(TPM_{max})]}}{N-1}
$$
where $TPM_{max}$ is the maximum average TPM for a given tissue type, and $TPM_i$ is the average TPM for tissue $i$. If $\tau = 0$, that gene is evenly expressed across tissues; if $\tau=1$, the gene is expressed in an entirely tissue-specific fashion. Because TPM values approach 0 are impacted by sampling stochasticity, any tissues that had an average expression of 0 for a given gene were set to $TPM=2$. 

```{r est_tau_avgsFXN, echo=TRUE}
est_tau_avgs<-function(geneDat,colDat){
  
  tissue_dat<-data.frame(cbind(colDat,
                               geneDat))
  tissue_dat$geneDat[round(tissue_dat$geneDat)==0]<-2
  
  
  tpmMax<-max(tissue_dat$geneDat)
  tissue_avgs<-tapply(tissue_dat$geneDat,tissue_dat$Organ,mean)
  
  tau<-sum(1-(log(tissue_avgs)/log(max(tissue_avgs))))/(length(tissue_avgs)-1)

  return(tau)
}

```


```{r est_tauFXN}
est_tau<-function(geneDat,colDat){
  
  tissue_dat<-data.frame(cbind(colDat,
                               geneDat))
  tissue_dat$geneDat[round(tissue_dat$geneDat)==0]<-2
  
  
  tpmMax<-max(tissue_dat$geneDat)
  tissue_avgs<-tapply(tissue_dat$geneDat,tissue_dat$Organ,mean)
  
  tau<-sum(1-(log(tissue_dat$geneDat)/log(tpmMax)))/(length(tissue_dat$geneDat)-1)

  return(tau)
}

```

```{r calcTau}
tau <- apply(tpms, 1, est_tau_avgs, colDat=colData)


tau_fem<-apply(tpms[,which(colData$Sex=="F")], 1, est_tau,
               colData[which(colData$Sex=="F"),])

tau_mal<-apply(tpms[,which(colData$Sex=="M")], 1, est_tau,
               colData[which(colData$Sex=="M"),])

```

```{r tauHist,fig.cap="Distribution of tissue specificity estimates for only the male samples (left) versus only the female samples versus the male samples and the female samples (right)."}
par(mfrow=c(1,3))
hist(tau_mal,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="")

hist(tau_fem,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="")

hist(tau,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="")
```

Let's now investigate some of the plot counts for the genes with a high tissue specificity index.
```{r ts-plot-counts, eval = FALSE}
tau[which(tau >= 1)]
tpms[rownames(tpms)=="TRINITY_DN316_c5_g1",]
plotCounts(dds_FL, gene="TRINITY_DN316_c5_g1", intgroup="Organ")
tpms[rownames(tpms)=="TRINITY_DN1996_c0_g1",]
plotCounts(dds_FL, gene="TRINITY_DN1996_c0_g1", intgroup="Organ")
tpms[rownames(tpms)=="TRINITY_DN1167_c0_g2",]
plotCounts(dds_FL, gene="TRINITY_DN1167_c0_g2", intgroup="Organ")
tpms[rownames(tpms)=="TRINITY_DN157392_c0_g1",]
plotCounts(dds_FL, gene="TRINITY_DN157392_c0_g1", intgroup="Organ")
tpms[rownames(tpms)=="TRINITY_DN37396_c0_g3",]
plotCounts(dds_FL, gene="TRINITY_DN37396_c0_g3", intgroup="Organ")

tpms[rownames(tpms)=="TRINITY_DN714_c0_g1",]
tpms[rownames(tpms)=="TRINITY_DN29102_c0_g1",]
tpms[rownames(tpms)=="TRINITY_DN316_c5_g1",]
tpms[rownames(tpms)=="TRINITY_DN316_c5_g1",]

plotCounts(ddsMF_FL, gene="TRINITY_DN102386_c0_g3", intgroup="Organ")
```