---
title: "Tissue Specificity in _Stigmatopora nigra_"
author: "Emily Beasley"
date: "`r Sys.Date()`"
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: no
    toc: false
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',
                     fig_path="figs/")
```

``` {r library, message = FALSE, warning = FALSE}
setwd("~/Documents/GitHub/pipefish_sbge/stigmatopora/S_nigra")

#This is a cohesive list of all the libraries used in this document

library(DESeq2)
library(spfTools)

```

``` {r read-data, message = FALSE, warning = FALSE}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.nigra <- readRDS("data/txi.salmon_SN.RDS")

#The samples file generated for tximport
samples <- read.table("data/SN_samples.txt", header = TRUE)

samples$group <- factor(paste0(samples$sex, samples$organ))

#Changing "Gonad" to be more descriptive --> testes or ovaries
samples$organ <- ifelse(samples$sex == "F" & samples$organ == "Gonad",
                        paste0("Ovaries"),
                        ifelse(samples$sex == "M" & samples$organ == "Gonad",
                            paste0("Testes"),
                            paste0(samples$organ))
                  )

#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)

#Format colData to be used in the tau function
colData <- as.data.frame(samples)
rownames(colData) <- samples$ID
```

# Differential Expression Analysis
The main goal of this analysis is to relate measure of pleiotropic constraints (tissue specificity) to sex bias. In order to do that I need to combine my sex-bias data that I generated in a separate .Rmd with the tau values I will be generating in this document.

Here I am reading in the sex-bias datasets that correspond to each of the organ's single-factor analysis. These datasets include all of the genes where the p-value was not equal to "NA".

```{r read-in-sb-datasets}
liver_res <- read.csv("data/liver_res.csv", 
                      row.names = 1, 
                      header = TRUE)
gill_res <- read.csv("data/gill_res.csv", 
                      row.names = 1, 
                      header = TRUE)
gonad_res <- read.csv("data/gonad_res.csv", 
                      row.names = 1, 
                      header = TRUE)
```

# Calculating Tissue Specificity
To estimate tissue specificity, the TPM estimates are needed which is the number of transcripts from a particular gene normalized first by gene length, and then by sequencing depth (in millions) in the sample. The output quant.sf files from salmon have the following format:

```
Name | Length | EffectiveLength | TPM | NumReads

```

## Pulling out TPM values 
From the Salmon outputs I pulled out the TPM values from each sample.

```{r getTPMs}

#Get the list of file names/paths for all of the quant.sf files
files <- list.files(pattern = ".sf", path = "data/nigra_expression_files",
                    full.names = TRUE)

#For each quant.sf file pull out the TPM column
tpms <- do.call(cbind, lapply(files, function(file){

  dat <- read.delim(file, row.names = "Name")
  tpm <- dat["TPM"]
  colnames(tpm) <- gsub("data/nigra_expression_files/(.*)_quant.sf","\\1",file)
  
  return(tpm)
}))

#Remove the sample that was removed in the differential expression analysis 
tpms <- tpms[, !(colnames(tpms) %in% c("S34",
                                        "S44",
                                        "S54"))]
```

## Filtering the TPM dataset
I want to filter the TPM dataset to remove a few things. These include similar things that I filtered out in the DESeq2 datasets prior to performing the differential expression analysis. This includes:
  
  1. Keeping only rows that weren't filtered out in the DEseq2 dataset due to low counts (rowSum $\le$ 10).
  2. Removing the rows that corresponded to genes that were "outliers" in the DEseq2 analysis.
  3. Removing the samples related to individual SNT8F2 (samples S34, S44, S54).

This will be imperfect filtering in this case because I did individual single-factor analysis within each organ type. To try and make life easier, I am going to run the multi-factor analysis on the DESeq2 dataset and then apply the filtering from that onto the TPM dataset. 

It should also be noted that the DESeq2 filtering was done based off of the count data and not the TPMs, but as TPMs are just normalized counts they should be correlated and something that was had low gene counts should also have a low TPM and anything that was considered an outlier in the counts could also be an outlier in terms of TPM, but it may not be perfect.

```{r filterTPMs}
#Running the MF diff. expr. analysis
#Create the DESeq dataset
dds_SN <- DESeqDataSetFromTximport(txi.nigra,
                                   colData = samples,
                                   design = ~ group)

##Remove all S34 organs from the dataset
dds_SN <- dds_SN[, !(dds_SN$ID %in% c("S34",
                                      "S44",
                                      "S54"))]

##Filter the dataset, only keeping rows that have at least 10 reads total
keep <- rowSums(counts(dds_SN)) >= 10 
dds_SN <- dds_SN[keep, ]

#Generate the expression values
dds_SN_exp <- DESeq(dds_SN)

#Filtering the tpm dataset
#Only keeping the rows that weren't filtered out due to low counts
tpms <- tpms[rownames(tpms) %in% rownames(dds_SN), ]

#Pulling out the geneIDs for genes that were categorized as "outliers" by DESeq2
#Calculating the Cooks threshold that would have been used
np <- length(resultsNames(dds_SN_exp))
nsamp <- ncol(dds_SN_exp)
cooks_thresh <- qf(0.99, df1 = np, df2 = nsamp-np)

out_ids <- names(mcols(dds_SN_exp)$maxCooks[mcols(dds_SN_exp)$maxCooks > cooks_thresh])

#Removing the rows in the tpm dataset that were deemed "outliers" by DESeq2
tpms <- tpms[!(rownames(tpms) %in% out_ids), ]

```

## Generating $\tau$
These filtered TPM estimates can then be used to estimate $\tau$, a tissue specificity estimator that can range from 0 to 1. $\tau$ is calculated for each tissue, $i$, as follows:

$$
\tau=\frac{\sum_i{[1-ln(TPM_i)/ln(TPM_{max})]}}{N-1}
$$

where $TPM_{max}$ is the maximum **average** TPM for a given tissue type, and $TPM_i$ is the **average** TPM for tissue $i$. If $\tau = 0$, that gene is evenly expressed across tissues; if $\tau=1$, the gene is expressed in an entirely tissue-specific fashion. Because TPM values approach 0 are impacted by sampling stochasticity, any genes that had an expression approaching 0 were set to $TPM=2$. 

```{r est_tauFXN}
#Function for estimating tau given the TPM matrix and metadata file
est_tau<-function(geneDat,colDat){
  
  #For each row in the TPM matrix cbind it with the metadata file,
  #this attaches organ type information to the TPM values
  tissue_dat<-data.frame(cbind(colDat,
                               geneDat))
  
  #For the TPM values approaching 0, set them to 2
  tissue_dat$geneDat[tissue_dat$geneDat < 1] <- 2
  
  
  #Get the average TPM for each tissue type (TPMi)
  tissue_avgs<-tapply(tissue_dat$geneDat,tissue_dat$organ,mean)
  
  #Get the maximum value from the average TPMS (TPMmax)
  tpmMax <- max(tissue_avgs, na.rm=TRUE)
    
  #IF running tau on JUST males of JUST females, this accounts for the
  #fact that ovary or testis will return an NA in the averaging
  if(length(unique(tissue_dat$organ)) == 3){
    tau <- sum(1-(log(tissue_avgs[unique(tissue_dat$organ)])/log(tpmMax)))/
      (length(unique(tissue_dat$organ))-1)
    
    return(tau)
  }
  
  #IF using the WHOLE dataset, calculate tau
  tau<-sum(1-(log(tissue_avgs)/log(tpmMax)))/(length(unique(tissue_dat$organ))-1)
  
  return(tau)
}

```


I then applied that function across each row in the TPMs matrix with my metadata stored in the object `colData`. The metadata file includes the sample ID, Sex, and Organ type for every column present in the TPM matrix.

I ran the $\tau$ function with the whole dataset, and then for just males and just females.

```{r calcTau}
colData <- colData[!(colData$ID %in% c("S34",
                                       "S44",
                                       "S54")),]

tau <- apply(tpms, 1, est_tau, colDat=colData)


tau_fem<-apply(tpms[,which(colData$sex=="F")], 1, est_tau,
               colData[which(colData$sex=="F"),])

tau_mal<-apply(tpms[,which(colData$sex=="M")], 1, est_tau,
               colData[which(colData$sex=="M"),])

```

```{r tauHist,fig.cap="Distribution of tissue specificity estimates for only the male samples (left) versus only the female samples (middle) versus the male samples and the female samples (right).", echo=FALSE}
par(mfrow=c(1,3))
hist(tau_mal,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="Male-only")

hist(tau_fem,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="Female-only")

hist(tau,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="All samples")
```

We can see that the distribution of $\tau$ doesn't vary drastically when looking at only male/female samples versus when we included all of the samples in the calculation (Fig. \@ref(fig:tauHist)). Because of this, I can be confident in using the $\tau$ with all the samples moving forward.

## Validating the $\tau$ calculations
To make sure $\tau$ is being calculated in a way that makes sense, I am checking some of the TPM values and plotting the corresponding counts for the genes with a high tissue specificity index and a low tissue specificity index.

```{r ts-plot-counts-high, echo=FALSE, fig.height=20, fig.width=15, fig.cap="Plots of the counts from the DESeq2 Dataset for genes with the highest tau", message=FALSE}
#Getting list of genes with the greatest Tau
top_tau <- head(sort(tau, decreasing = TRUE), n=9)

#Looking at the TPMS
tpms[rownames(tpms)== names(top_tau[1]),] #Gill-specific
tpms[rownames(tpms)== names(top_tau[2]),] #Ovary-specific
tpms[rownames(tpms)== names(top_tau[3]),] #Ovary-specific
tpms[rownames(tpms)== names(top_tau[4]),] #Gill-specific
tpms[rownames(tpms)== names(top_tau[5]),] #Ovary-specific
tpms[rownames(tpms)== names(top_tau[6]),] #Ovary-specific
tpms[rownames(tpms)== names(top_tau[7]),] #Ovary-specific
tpms[rownames(tpms)== names(top_tau[8]),] #Liver-specific
tpms[rownames(tpms)== names(top_tau[9]),] #Ovary-specific

#Plotting the gene counts from DESeq2 for the toptau genes
par(mfrow=c(3, 3))
plotCounts(dds_SN, gene= names(top_tau[1]), intgroup="organ")
plotCounts(dds_SN, gene= names(top_tau[2]), intgroup="organ")
plotCounts(dds_SN, gene= names(top_tau[3]), intgroup="organ")
plotCounts(dds_SN, gene= names(top_tau[4]), intgroup="organ")
plotCounts(dds_SN, gene= names(top_tau[5]), intgroup="organ")
plotCounts(dds_SN, gene= names(top_tau[6]), intgroup="organ")
plotCounts(dds_SN, gene= names(top_tau[7]), intgroup="organ")
plotCounts(dds_SN, gene= names(top_tau[8]), intgroup="organ")
plotCounts(dds_SN, gene= names(top_tau[9]), intgroup="organ")
```

```{r ts-plot-counts-low, echo=FALSE, fig.height=20, fig.width=15, fig.cap="Plots of the counts from the DESeq2 Dataset for genes with the lowest tau", message=FALSE}
#Getting list of genes with the lowest Tau
bot_tau <- head(sort(tau, decreasing = FALSE), n=9)

#Looking at the TPMS
tpms[rownames(tpms)== names(bot_tau[1]),]
tpms[rownames(tpms)== names(bot_tau[2]),]
tpms[rownames(tpms)== names(bot_tau[3]),]
tpms[rownames(tpms)== names(bot_tau[4]),]
tpms[rownames(tpms)== names(bot_tau[5]),]
tpms[rownames(tpms)== names(bot_tau[6]),]
tpms[rownames(tpms)== names(bot_tau[7]),]
tpms[rownames(tpms)== names(bot_tau[8]),]
tpms[rownames(tpms)== names(bot_tau[9]),]

#Plotting the gene counts from DESeq2 for the toptau genes
par(mfrow=c(3, 3))
plotCounts(dds_SN, gene= names(bot_tau[1]), intgroup="organ")
plotCounts(dds_SN, gene= names(bot_tau[2]), intgroup="organ")
plotCounts(dds_SN, gene= names(bot_tau[3]), intgroup="organ")
plotCounts(dds_SN, gene= names(bot_tau[4]), intgroup="organ")
plotCounts(dds_SN, gene= names(bot_tau[5]), intgroup="organ")
plotCounts(dds_SN, gene= names(bot_tau[6]), intgroup="organ")
plotCounts(dds_SN, gene= names(bot_tau[7]), intgroup="organ")
plotCounts(dds_SN, gene= names(bot_tau[8]), intgroup="organ")
plotCounts(dds_SN, gene= names(bot_tau[9]), intgroup="organ")

```

We can see that for the genes that have high tau values the plots of the normalized counts show higher expression in just that tissue, whereas for the plots of the lower tau values no clear patterns are showing up in the plots of the counts. This is what we want to see, increasing confidence that the the calculations of tau are accurate.

# Sex Bias and Tissue Specificity
Now that I have calculated $\tau$ I want to relate it back to information about sex bias to see what, if any, relationship exists. I am going to do this separately for the different organs to see if the relationship changes based on organ type.

## $\tau$ and logFC in sex-biased genes for each organ
A gene was determined to be sex-biased if it had a log-fold change $\ge |2|$ AND an adjusted p-value of $< 0.05$. Earlier I read in four datasets that contain information for all of the significantly expressed genes in each organ (i.e. the genes that were not given a p-value of NA from DESeq). 

From these datasets I can pull out the female- and male-biased genes to plot logFC against $\tau$ within the genes that are sex-biased.

```{r tau-v-bias, echo=FALSE, fig.cap="Sex bias (in terms of the square root of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are sex biased in the gonads, liver, skin, and brain."}

#Pull out the sex-biased genes (based on FC of >= |2| and padj of < 0.05) for each organ
liver_biased <- liver_res[which(abs(liver_res$log2FoldChange) >= 2 & liver_res$padj <= 0.05),]
gonad_biased <- gonad_res[which(abs(gonad_res$log2FoldChange) >= 2 & gonad_res$padj <= 0.05),]
gill_biased <- gill_res[which(abs(gill_res$log2FoldChange) >= 2 & gill_res$padj <= 0.05),]

#Merge the sex bias information with the tau values based on row names for each organ
liver_bias_tau <- merge(as.data.frame(liver_biased), as.data.frame(tau), by = "row.names")
gonad_bias_tau <- merge(as.data.frame(gonad_biased), as.data.frame(tau), by = "row.names")
gill_bias_tau <- merge(as.data.frame(gill_biased), as.data.frame(tau), by = "row.names")

#Plot the relationship between FC (bias) and tau for each organ
par(mar=c(5.1, 4.1, 4.1, 8.1), xpd=FALSE)
plot(sqrt(abs(gonad_bias_tau$log2FoldChange)),
     gonad_bias_tau$tau,
     pch=19,
     col = "#EEB42275",
     xlab="Sex bias (|log2 fold change|)",
     ylab=expression(tau["TPM"]))
points(sqrt(abs(liver_bias_tau$log2FoldChange)),
     liver_bias_tau$tau,
     pch=19,
     col="#EE826275")
points(sqrt(abs(gill_bias_tau$log2FoldChange)),
     gill_bias_tau$tau,
     pch=19,
     col = "#20B2AA75")


#Add the best fit lines
abline(lm(gonad_bias_tau$tau ~ sqrt(abs(gonad_bias_tau$log2FoldChange))), col="#EEB422", lwd = 3)
abline(lm(liver_bias_tau$tau ~ sqrt(abs(liver_bias_tau$log2FoldChange))), col = "#EE8262", lwd = 3, lty = 2)
abline(lm(gill_bias_tau$tau ~ sqrt(abs(gill_bias_tau$log2FoldChange))), col = "#20B2AA", lwd = 3, lty = 3)

#Add a legend
outer_legend("right",
             c("Gonad", "Liver", "Gill"),
             col = c("#EEB422", "#EE8262", "#20B2AA"),
             pch = 19,
             lwd = 3,
             bty = 'n',
             cex = 1.5,
             lty = 1:3,
             ncol = 1)


```