---
title: "Differential Expression Analysis in _Stigmatopora nigra_"
output:
  github_document:
    toc: true
  #bookdown::pdf_document2:
    #fig_caption: yes
    #keep_tex: yes
    #number_sections: yes
    #toc: true
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
```

``` {r library, include = TRUE, message = FALSE, warning = FALSE}
setwd("~/Documents/GitHub/pipefish_sbge/stigmatopora/S_nigra")
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
library(spfTools)
library(pdftools)
library(magick)
library(patchwork)
library(ggpubr)
library(cowplot)
library(UpSetR)
library(ggplot2)
library(kableExtra)
library(tidyverse)
library(knitr)
library(patchwork)
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.nigra <- readRDS("data/txi.salmon_SN.RDS")

#The samples file generated for tximport
samples <- read.table("data/SN_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)

```

The package `DESeq2` was used for the differential expression analysis outlined below.

# Single factor analysis - Comparing Males v Females across all organs
To analyse your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions. The model for this single factor analysis was run as counts ~ Sex.

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_SN <- DESeqDataSetFromTximport(txi.nigra, 
                                   colData = samples,
                                   design = ~ sex)

#remove samples with low read mapping
dds_SN <- dds_SN[, !(dds_SN$ID %in% c("S34", "S41"))]
samples <- samples[!(samples$ID %in% c ("S34", "S41")),]

#remove gonad samples 
dds_SN <- dds_SN[dds_SN$organ != "Gonad", ]
samples <- samples[samples$organ != "Gonad", ]
```

The data is then pre-filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

The cutoff here was to remove rows that had counts fewer than 10 across all samples.

```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at least 10 reads total
keep <- rowSums(counts(dds_SN)) >= 10
dds_SN <- dds_SN[keep, ]

```

After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.

```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_SN_exp <- DESeq(dds_SN)

#Compile the results
res <- results(dds_SN_exp)
res
```

Once that has finished we can now start exploring some of the single-factor analysis results
```{r investigate-DESeq-results-SN}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#How many ADJUSTED p-values were less than 0.1?
sum(res$padj < 0.1, na.rm = TRUE)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_SN_exp, alpha = 0.05)
summary(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```


## Visualising the results
### MA-plot - MvF Differntial expression
Generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be colored if the adjusted p-value is less that 0.1.

```{r MA-plot, echo=FALSE, fig.cap="LogFC versus the mean normalised count for all of the genes. Points that are blue have a p-value less than 0.1. The two plots are the same with the only difference coming from an adjusted y-limit. Points that are triangles represent genes with a fold change higher/lower than the y-limit."}

par(mfrow=c(1,2))
plotMA(res, ylim = c(-2,2))
plotMA(res)

```

### Heatmap - overall expression
We can also generate a heat map to look at overall expression levels across our samples. Note, this is not differentially expressed genes. 

```{r heatmap, echo=FALSE, fig.cap= "Heatmap showing the expression level across all organs for the top 20 genes with the highest expression levels."}
#Pull out the top 20 rows with the most expression
df <- as.data.frame(samples[,c("sex", "organ")])
rownames(df) <- samples$ID
df$organ <- as.character(df$organ)

organ_cols<-c("Gill" = "#20B2AA",
              "Liver" = "#EE8262")

sex_cols<-c("F" = "#7fc97f", "M" = "#beaed4")

ann_colors = list(sex=sex_cols,
                  organ=organ_cols)

col_order <- c(rownames(df[df$sex=="F" & df$organ=="Gill",]),
               rownames(df[df$sex=="M" & df$organ=="Gill",]),
               rownames(df[df$sex=="F" & df$organ=="Liver",]),
               rownames(df[df$sex=="M" & df$organ=="Liver",]))


#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(dds_SN_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]


#Transform the data
vsd <- vst(dds_SN_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select, col_order], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df, 
         annotation_colors = ann_colors)
```

From the heatmap we can see that for all of these top 20 expressed genes the highest expression is found in the liver. S34 (female liver) is showing expression patterns that are different from the other livers. There are also a few genes that appear to be highly expressed across all of the tissue types. With this heatmap we can also pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.

```{r heatmap_genes}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))
#Run once to create the file and then commented out
#write.table(heatmap_TG,
 #           'data/no_gon_heatmap_trinitygenes.txt',
  #          sep = "",
   #         quote=FALSE,
    #        row.names = FALSE,
     #       col.names = FALSE)
```

### Sample-dist Heatmap

We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples.

```{r sample-dist, echo=FALSE, fig.cap="Sample-to-sample distances heatmap showing the similarities and differences among all of the samples. The darker the colour, the more similar the samples are. The diagonal line of very dark blue represents the comparissons within the same samples.", warning=FALSE}
vsd <- vsd[, vsd$organ != "Gonad"]

#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$sex, vsd$organ, sep = "-")
colnames(sampleDistsMatrix) <- vsd$ID

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)
```

Here we can see that the highest similarities (aside from same samples comparisons) is between samples from the same organ. Again, S34 is showing unexpected patterns - I may end up removing all organs related to S34. 

### PCA plots
Next we can look at PCA plots for our samples to see how our different samples are grouping. 

From the pairs plot we can see that PC1 appears to be explaining differences between the liver and the other tissues (we can see the one problematic liver sample from above sticking out from all the rest), PC2 appears to be explaining differences between gonadal and somatic tissues, PC3 seems to account for differences between male and female samples, especially in the gonads and in PC4 we can see the one problematic liver sample again sticking out from all the rest. I am going to plot all of PC2 - 4 against PC1 below for a clearer picture.

```{r pca-scaled-organ, echo=FALSE, fig.cap="Principal components analysis reflects that most variation corresponds to differences in expression between organs (green: gill; yellow: gonad; pink: liver), rather than variation due to sex (circle = female; triangle = male). The first axis explains 27% of variation in the dataset, with gills and liver, two types of somatic tissues, sitting on opposite ends of that axis. The second axis explains 22% of the variation in gene expression, with somatic tissues on the opposite side of the axis from the gonads.", warning=FALSE}

samples <- samples[!(samples$ID %in% c ("S34", "S41")),]
samples <- samples[samples$organ != "Gonad", ]
vsd <- vsd[, vsd$organ != "Gonad"]

#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$organ <- samples$organ
pca_plotting_data$sex <- samples$sex
pca_plotting_data$ID <- samples$ID

pca_plotting_data <- pca_plotting_data[pca_plotting_data$organ != "Gonad", ]


#Calculate the percent variation attributed to each axis 
pcaData <- plotPCA(vsd, intgroup = c("organ", "sex"), returnData = TRUE, ntop = nrow(dds_SN))
percentVar <- round(100 * attr(pcaData, "percentVar"))


my_colors <- c("Gill" = "#6E8B3D75", "Liver" = "#EE826275")
my_shapes <- c(16, 17)

ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed') #+ 
  #geom_text(aes(label = ID), vjust = -1, hjust = 0.5)
```


# Multifactor design - Comparing M v F across the diff tissue types
If we investigate the column data of our DESeq dataset we can see that each sample has both a sex and organ type attached to it, we will be using these two factors in our multi-factor analysis. In this multi-factor analysis the model was run as counts ~ group, where group included both the sex and the organ type (i.e. MLiver, FLiver, etc.). The sample "S34" appeared to a bit of an outlier in the above analysis, particularly clear in the sample-dist heatmap and the PCA, so it will be removed from the analysis going forward.

```{r multi-factor-analysis, message=FALSE, warning=FALSE}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.nigra <- readRDS("data/txi.salmon_SN.RDS")

#The samples file generated for tximport
samples <- read.table("data/SN_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)

#Create an additional column the has the two conditions combined(sex and organ type)
samples$group <- factor(paste0(samples$sex, samples$organ))

##Create a copy of the DESeq dataset
ddsMF_SN <- DESeqDataSetFromTximport(txi.nigra,
                                     colData = samples,
                                     design = ~ group)

##Remove S34 and S41 from the dataset
ddsMF_SN <- ddsMF_SN[, !(ddsMF_SN$ID %in% c("S34", "S41"))]

##Filter the dataset, only keeping rows that have at least 10 reads total
keep <- rowSums(counts(ddsMF_SN)) >= 10 #& rowSums(counts(ddsMF_SN)) < 1e6
ddsMF_SN <- ddsMF_SN[keep, ]

#Run the differential expression analysis
ddsMF_SN_exp <- DESeq(ddsMF_SN)
resultsNames(ddsMF_SN_exp)
```

## Invesitgate the results of the differential expression
Thanks to the multi-factor analysis we can now explore differential expression between all of the different combinations:

  - Male Liver v. Female Liver
  - Male Gill v. Female Gill
  - Male Gonad v. Female Gonad
  - All of the within sex tissue comparisons (e.g. Male Liver v. Male Gill, etc.)

### M-F Liver Comparisson
```{r m-f-liver-results}
##Pulling out the liver M-F results with an alpha of 0.05
liver_con_res <- results(ddsMF_SN_exp, contrast = c("group", "MLiver", "FLiver"),
                         alpha = 0.05)
liver_con_res$trin_geneid <- rownames(liver_con_res)
head(liver_con_res)

summary(liver_con_res)
```

There are many criteria that have been employed previously to denote sex bias. For this study we are classifying sex-biased genes as **genes with a p-value < 0.05 AND a logFC $\ge$ |2|**. With that criteria in the liver there are `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. 

I have then pulled out all of the male-biased, female-biased, and non-biased genes based on the criteria outlined above. I determined non-biased genes as a p-vaule > 0.05.

```{r liver-sex-biased}
#Removing the rows where padj. is NA in results
liver_con_res_noNA <- liver_con_res[!is.na(liver_con_res$padj),]
summary(liver_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#write.csv(as.data.frame(liver_con_res_noNA), "data/liver_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the liver
liver_mal_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange >= 2
                                             & liver_con_res_noNA$padj <= 0.05),]
liver_fem_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange <= -2 
                                             & liver_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the liver
liver_non_biased <- liver_con_res_noNA[which(liver_con_res_noNA$padj > 0.05),]

```

I will be generating a table that outlines the genes that are the most male or female biased in each organ type. To do this I need to pull all differentially expressed genes in males or females, get the Trinity gene IDs and then BLAST the corresponding sequences. Here I am pulling out the Trinity gene IDs for those genes.


```{r all-liver}
#Creating a subset of the results where the p-value is less than 0.05
liver_con_res_p05 <- liver_con_res_noNA[liver_con_res_noNA$padj <= 0.05, ]


#Pulling the all diff expressed genes in Males
all_MLiver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(
  liver_con_res_p05$log2FoldChange, decreasing = TRUE),], n=845)))
#Run once to create the file and then commented out
#write.table(all_MLiver_trin_genes,
 #          'data/SN_maleL_allTRgenes.txt', 
  #         sep = "", 
   #        quote=FALSE, 
    #       row.names = FALSE, 
     #      col.names = FALSE)

#Pulling all diff expressed genes in Females
all_Fliver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(
  liver_con_res_p05$log2FoldChange, decreasing = FALSE),], n=665)))

#Run once to create the file and then commented out
#write.table(all_Fliver_trin_genes,
 #           'data/SN_femL_allTRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### M-F Gill Comparisson
```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res <- results(ddsMF_SN_exp, contrast = c("group", "MGill", "FGill"),
                        alpha = 0.05)
gill_con_res$trin_geneid <- rownames(gill_con_res)
head(gill_con_res)

summary(gill_con_res)
```

In the gills there were `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

I have then pulled out all of the male-biased, female-biased, and non-biased genes to save them into their own objects.

```{r gill-sex-biased}
#Removing the rows where padj. is NA in results
gill_con_res_noNA <- gill_con_res[!is.na(gill_con_res$padj), ]
summary(gill_con_res_noNA) #We can now see that there are no outliers or 
                           #low counts since the NAs have been removed
#write.csv(as.data.frame(gill_con_res_noNA), "data/gill_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the gills
gill_mal_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange >= 2 
                                           & gill_con_res_noNA$padj <= 0.05),]
gill_fem_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange <= -2 
                                           & gill_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gills, p>0.05
gill_non_biased <- gill_con_res_noNA[which(gill_con_res_noNA$padj > 0.05),]
```

Here I am getting the trinity gene IDs that correspond to all differentially expressed genes for males and females. Because there are only 22 female-biased and male-biased genes I will only use those.

```{r all-gill}
#Creating a subset of results where p-value is less than 0.05
gill_con_res_p05 <- gill_con_res_noNA[gill_con_res_noNA$padj <= 0.05,]

#Pulling all diff expressed genes in Males
all_MGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(
  gill_con_res_p05$log2FoldChange, decreasing = TRUE),], n=15)))

#Run once to create the file and then commented out
#write.table(all_MGill_trin_genes,
 #           'data/SN_maleG_allTRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)

#Pulling all diff expressed genes in Females 
all_FGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(
  gill_con_res_p05$log2FoldChange, decreasing = FALSE),], n=81)))

#Run once to create the file and then commented out
#write.table(all_FGill_trin_genes,
 #           'data/SN_femG_allTRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```



### MA Plots
```{r MA-plot-MF, echo=FALSE, fig.height= 15, fig.width=8, fig.cap="MA-plots generated for each organ type that compares logFC to the mean expression. Female-biased and male-biased genes are represented by colour (female: green; male: purple) and are determined with a fold change cutoff of 2 and a p-value cutoff of 0.05"}
liver_MA <- ggmaplot(liver_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Liver" %->% "Female Liver"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gill_MA <- ggmaplot(gill_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Gill" %->% "Female Gill"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

#Merge all the plots into one
plot_grid(liver_MA, gill_MA,
          ncol = 1,
          nrow = 2,
          labels = c('A', 'B'),
          label_fontfamily = 'sans',
          label_fontface = 'bold',
          label_size = 18,
          rel_heights = c(4.10, 4.10, 4.10))

```

## Creating an Upset Plot
```{r upset-plot-mal, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in males."}
#Male-biased genes
listInputMB <- list("Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased)
                  )
upset(fromList(listInputMB),
      mainbar.y.label = "Number of Shared Male-Biased Genes",
      sets.x.label = "Total Number of Male-Biased Genes",
      point.size = 3)
```

```{r upset-plot-fem, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in females."}
#Female-biased genes
listInputFB <- list("Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased)
                  )
upset(fromList(listInputFB),
      mainbar.y.label = "Number of Shared Female-Biased Genes",
      sets.x.label = "Total Number of Female-Biased Genes",
      point.size = 3)
```

```{r upset-plot-all, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs across both sexes."}
#Exporting data used to make the Upset plot (run once and them commented out)
write.table(as.data.frame(gill_mal_biased), "data/gill_mal_biased.txt", 
            row.names = TRUE)
write.table(as.data.frame(gill_fem_biased), "data/gill_fem_biased.txt", 
            row.names = TRUE)
write.table(as.data.frame(liver_mal_biased), "data/liver_mal_biased.txt", 
            row.names = TRUE)
write.table(as.data.frame(liver_fem_biased), "data/liver_fem_biased.txt", 
            row.names = TRUE)

#Checking there is no overlap between the two sexes
listInputall <- list("Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased),
                  "Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased))

# Define the desired order of sets
desired_order <- c("Male Gill", "Female Gill", "Male Liver", "Female Liver")

# Create the UpSet object with custom set order
upset(fromList(listInputall),
      sets = desired_order,
      mainbar.y.label = "Number of Shared Sex-Biased Genes",
      sets.x.label = "Total Number of Sex-Biased Genes",
      point.size = 3,
      nsets = 4,
      nintersects = NA,
      order.by = "degree")

#this code creates the same heatmap without the custom set order 
#upset(fromList(listInputall),
 #     mainbar.y.label = "Number of Shared Sex-Biased Genes",
  #    sets.x.label = "Total Number of Sex-Biased Genes",
   #   point.size = 3,
    #  nsets = 4,
     # nintersects = NA)

```

In both males and females the highest number of shared sex-biased genes is found between the gonads and the liver (Fig. \@ref(fig:upset-plot-mal), \@ref(fig:upset-plot-fem)). There are no genes that are shared across the same organ between males and females (e.g., biased in males and females in the liver) which is good (Fig. \@ref(fig:upset-plot-all)). Two sex-biased genes are female-biased in all organs and three genes are male-biased in all three organs (Fig. \@ref(fig:upset-plot-mal), \@ref(fig:upset-plot-fem)).

## Variation in FC across sex-bias and tissue type
I want to create a plot the highlights the variation we see in fold-change both across the different biases groups (male-biased, female-biased, and non-biased) within one tissue type and across all of the tissue types. To do this I first need to create a "long" dataset in the following format:

| Tissue_type|     Bias      |   Fold-Change   |
|:----------:|:-------------:|:---------------:|
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   fem_bias    |      XXXXX      |
|    Gill    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   fem_bias    |      XXXXX      |
|    Liver   |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   fem_bias    |      XXXXX      |
|   Gonad    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |

That dataset is created here, with the Trinity gene IDs included as well:

```{r logFC-long-dataset}
logFC_long<- data.frame(
  tissue=c(rep("gill",nrow(gill_fem_biased)),
           rep("gill", nrow(gill_mal_biased)),
           rep("gill", nrow(gill_non_biased)),
           rep("liver", nrow(liver_fem_biased)),
           rep("liver", nrow(liver_mal_biased)),
           rep("liver", nrow(liver_non_biased))
         ),
  bias=c(rep("FB",nrow(gill_fem_biased)),
         rep("MB",nrow(gill_mal_biased)),
         rep("UB", nrow(gill_non_biased)),
         rep("FB", nrow(liver_fem_biased)),
         rep("MB", nrow(liver_mal_biased)),
         rep("UB", nrow(liver_non_biased))
         ),
  logFC=c(gill_fem_biased$log2FoldChange,
          gill_mal_biased$log2FoldChange,
          gill_non_biased$log2FoldChange,
          liver_fem_biased$log2FoldChange,
          liver_mal_biased$log2FoldChange,
          liver_non_biased$log2FoldChange
          ),
  geneID=c(rownames(gill_fem_biased),
           rownames(gill_mal_biased),
           rownames(gill_non_biased),
           rownames(liver_fem_biased),
           rownames(liver_mal_biased),
           rownames(liver_non_biased)
           )
  
)
```

With the dataset in the proper format now, we can generate the plot

```{r FC-var-plot, echo=FALSE, fig.cap="Absolute value of the logFC for genes that are female-biased, male-biased, and unbiased across each tissue type. Raw fold-change values are added on top of the boxplot as jitters."}
my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "darkgray")

ggplot(logFC_long,
       aes(x = bias, y = abs(logFC), fill = bias)) +
  geom_boxplot() +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_point(data=logFC_long, aes(x=bias, y=abs(logFC)),
             size=0.4, 
             position = position_jitter(width = 0.15)) + 
  labs(x="Bias Level", y="|logFC|") +
  guides(fill = guide_legend(title = "Bias Level", order = 3))

```

# Tissue specificity analysis

``` {r read-data, message = FALSE, warning = FALSE}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.nigra <- readRDS("data/txi.salmon_SN.RDS")

#The samples file generated for tximport
samples <- read.table("data/SN_samples.txt", header = TRUE)

samples$group <- factor(paste0(samples$sex, samples$organ))


#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)

#Format colData to be used in the tau function
colData <- as.data.frame(samples)
rownames(colData) <- samples$ID
```

# Differential Expression Analysis
The main goal of this analysis is to relate measure of pleiotropic constraints (tissue specificity) to sex bias. In order to do that I need to combine my sex-bias data that I generated in a separate .Rmd with the tau values I will be generating in this document.

Here I am reading in the sex-bias datasets that correspond to each of the organ's single-factor analysis. These datasets include all of the genes where the p-value was not equal to "NA".

```{r read-in-sb-datasets}
liver_res <- read.csv("data/liver_res.csv", 
                      row.names = 1, 
                      header = TRUE)
gill_res <- read.csv("data/gill_res.csv", 
                      row.names = 1, 
                      header = TRUE)
```
