---
title: "Stigmatopora nigra analysis"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r knitsetup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../figs/")
```


```{r chunksetup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,out.extra='',fig.pos="H",
                      fig.path = "../figs/",
                      dpi = 300,fig.keep='none',dev='png')
```


## Retrieving Data

Otago University Sequencing centre makes it very easy to download data using Globus ...

Quality scores were also provided by Otago University, so I do not need to run FastQC on all of them.


## Trimming reads

I will run Trimmomatic on my local machine (C001KR). I am running Trimmomatic v. 0.39 and am using the script `run_trimmomatic.sh`, whose contents are below, with the command `nohup run_trimmomatic.sh > trimmomatic.log 2>&1 &`.

```{bash, file='scripts/run_trimmomatic.sh', eval=FALSE}

```

I started this script running on 8 Nov 2022 ~8:48am, PID 14781.

### Checking quality of trimmed reads

I will use `multiqc` (at Coley's recommendation) to quality check the trimmed reads. First, I need to run FastQC on all of my files. This is done using the script `run_fastqc_trimmed.sh` with the command `nohup run_fastqc_trimmed.sh > fasqc_trimmed.log 2>&1 &`.

```{bash, file='scripts/run_trimmomatic.sh', eval=FALSE}

```


Then (once the program is installed using `pip`) all I need to do is go to `OG7926/` and run `multiqc trimmed_qc/ --outdir ../stigmatopora/results/`.

Taking a look at these -- both the paired and unpaired reads -- it looks ok. There are some pretty high percentages of duplicated reads, but this is expected I believe because of the fact that there should be some high expression genes.


## De novo transcriptome assembly using Trinity

### Setup

The first step is to create the sample information files. These should use the paths and filenames for only the trimmed files with pairs (i.e., in `trimmed/` and with the component `_paired` in the filename). I only need information on the 'category', the sample name, and then read one and read 2. 

I'm doing the analysis initially only on the gonads, so I will write those to file separately. 

```{r createSampleFile, eval=FALSE}
samples<-read.csv("OG7629_sample_info.csv")
snigra_samples<-samples[which(samples$Organism=="Stigmatopora nigra"),]

snigra_samples$R1<-unlist(lapply(snigra_samples$OGF_ID,
                                   grep,
                                   x=list.files(path="../OG7629/trimmed",
                                                pattern="_paired_R1",
                                                full.names = TRUE),
                                   value=TRUE))
snigra_samples$R2<-gsub("R1","R2",snigra_samples$R1)

# get the 'treatment' level
snigra_samples$category<-gsub("SNT\\d+(\\w)(\\w).*$",
                              "SN_\\1_\\2",
                              snigra_samples$sample_name)


# reorder it so it is
## SN_TT (species, tissue type) \t SN_TT_ID \t R1 \t R2
write.table(snigra_samples[,c("category","sample_name","R1","R2")],
            "nigra_samples_trinity.txt",
            sep='\t',
            col.names = FALSE,
            row.names = FALSE,
            quote=FALSE)

# output only the gonads

write.table(snigra_samples[snigra_samples$Source %in% c("ovary","testes"),
                           c("category","sample_name","R1","R2")],
            "nigra_gonads_trinity.txt",
            sep='\t',
            col.names = FALSE,
            row.names = FALSE,
            quote=FALSE)
```


### Transferring files

I will be using the RCC `132.181.102.45` for trinity and RSEM, so I need to get my trimmed files there and set up the directories nicely on that machine. There is already an `OG7629/` directory in `/home` on that machine, so I will use that and create a `stigmatopora` directory there. To make life easier, I created a new ssh token to correspond to my github account (spflanagan@github.com) so that I can easily push and pull scripts, same as I would on my local machine. After adding my key to my github account (following the steps from [this webpage](https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/github-clone-with-ssh-keys)), I then cloned my repository using ssh. Then I moved `OG7629/` into `pipefish_sbge/` to match the directory structure on my local machine.

Next, I need to transfer the trimmed reads plus the trinity sample file I want to use. I only need to transfer the trimmed reads corresponding to the gonads. Usually I use `scp` but some googling suggests that `rsync` might be a better option, because it will transfer files specified in a text file, like so:

`rsync -av --progress --files-from=file_list.txt source/ me@example.com:/dest`

That means that I need to save just a list of the files. 

```{r createTransferList, eval=FALSE}
files<-unlist(snigra_samples[snigra_samples$Source %in% c("ovary","testes"),c("R1","R2")])
# remove paths
files<-gsub("../OG7629/trimmed/(.*)","\\1",files)
write.table(files,
            "nigra_gonads_files.txt",
            sep='\t',
            col.names = FALSE,
            row.names = FALSE,
            quote=FALSE)
```

Then, the rsync command should be: `rsync -av --progress --files-from=nigra_gonads_files.txt /mnt/BigData/OG7629/trimmed/ trinity-em:~/pipefish_sbge/OG7629/trimmed/`, from the `pipefish_sbge/stigmatopora/` directory -- note that it needs to use the actual location, not the symbolic link to OG7629.

Then I need to send the samples file to the RCC (from `pipefish_sbge/stigmatopora`: `rsync -av --progress nigra_gonads_trinity.txt trinity-em:~/pipefish_sbge/stigmatopora/` 

I also created a directory on the RCC for the trinity results, called `trinity_nigra_gonads` within the `results/` directory (so, `pipefish_sbge/results/trinity_nigra_gonads`).

It turns out that the sample names can't be relative paths because the docker is trying to run trinity from within the `/usr/local/bin/Trinity` directory. So I need to replace the `../` with the path name, which I'll do using a regex substitution via a perl one-liner: `perl -pe 's/\.\./\/home\/rccuser\/pipefish_sbge/g' nigra_gonads_trinity.txt > nigra_gonads_trinity_rcc.txt`.


### Running Trinity


I am using the script `run_trinity.sh` to run Trinity in the docker. It requires two arguments: one of which is the sample file and the other is the output directory. I'm running it like so (from the `pipefish_sbge/stigmatopora/` folder): `nohup ./scripts/run_trinity.sh nigra_gonads_trinity_rcc.txt results/trinity_nigra_gonads/ > trinity_gonads.log 2>&1 &`.



