---
title: "Stigmatopora nigra analysis"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r knitsetup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../figs/")
```


```{r chunksetup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,out.extra='',fig.pos="H",
                      fig.path = "../figs/",
                      dpi = 300,fig.keep='none',dev='png')
```


## Retrieving Data

Otago University Sequencing centre makes it very easy to download data using Globus ...

Quality scores were also provided by Otago University, so I do not need to run FastQC on all of them.


## Trimming reads

I will run Trimmomatic on my local machine (C001KR). I am running Trimmomatic v. 0.39 and am using the script `run_trimmomatic.sh`, whose contents are below, with the command `nohup run_trimmomatic.sh > trimmomatic.log 2>&1 &`.

```{bash, file='scripts/run_trimmomatic.sh', eval=FALSE}

```

I started this script running on 8 Nov 2022 ~8:48am, PID 14781.

### Checking quality of trimmed reads

I will use `multiqc` (at Coley's recommendation) to quality check the trimmed reads. First, I need to run FastQC on all of my files. This is done using the script `run_fastqc_trimmed.sh` with the command `nohup run_fastqc_trimmed.sh > fasqc_trimmed.log 2>&1 &`.

```{bash, file='scripts/run_trimmomatic.sh', eval=FALSE}

```


Then (once the program is installed using `pip`) all I need to do is go to `OG7926/` and run `multiqc trimmed_qc/ --outdir ../stigmatopora/results/`.

Taking a look at these -- both the paired and unpaired reads -- it looks ok. There are some pretty high percentages of duplicated reads, but this is expected I believe because of the fact that there should be some high expression genes.


## De novo transcriptome assembly using Trinity

### Setup

The first step is to create the sample information files. These should use the paths and filenames for only the trimmed files with pairs (i.e., in `trimmed/` and with the component `_paired` in the filename). I only need information on the 'category', the sample name, and then read one and read 2. 

I'm doing the analysis initially only on the gonads, so I will write those to file separately. 

```{r createSampleFile, eval=FALSE}
samples<-read.csv("OG7629_sample_info.csv")
snigra_samples<-samples[which(samples$Organism=="Stigmatopora nigra"),]

snigra_samples$R1<-unlist(lapply(snigra_samples$OGF_ID,
                                   grep,
                                   x=list.files(path="../OG7629/trimmed",
                                                pattern="_paired_R1",
                                                full.names = TRUE),
                                   value=TRUE))
snigra_samples$R2<-gsub("R1","R2",snigra_samples$R1)

# get the 'treatment' level
snigra_samples$category<-gsub("SNT\\d+(\\w)(\\w).*$",
                              "SN_\\1_\\2",
                              snigra_samples$sample_name)


# reorder it so it is
## SN_TT (species, tissue type) \t SN_TT_ID \t R1 \t R2
write.table(snigra_samples[,c("category","sample_name","R1","R2")],
            "nigra_samples_trinity.txt",
            sep='\t',
            col.names = FALSE,
            row.names = FALSE,
            quote=FALSE)

# output only the gonads

write.table(snigra_samples[snigra_samples$Source %in% c("ovary","testes"),
                           c("category","sample_name","R1","R2")],
            "nigra_gonads_trinity.txt",
            sep='\t',
            col.names = FALSE,
            row.names = FALSE,
            quote=FALSE)
```


### Transferring files

I will be using the RCC `132.181.102.45` for trinity and RSEM, so I need to get my trimmed files there and set up the directories nicely on that machine. There is already an `OG7629/` directory in `/home` on that machine, so I will use that and create a `stigmatopora` directory there. To make life easier, I created a new ssh token to correspond to my github account (spflanagan@github.com) so that I can easily push and pull scripts, same as I would on my local machine. After adding my key to my github account (following the steps from [this webpage](https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/github-clone-with-ssh-keys)), I then cloned my repository using ssh. Then I moved `OG7629/` into `pipefish_sbge/` to match the directory structure on my local machine.

Next, I need to transfer the trimmed reads plus the trinity sample file I want to use. I only need to transfer the trimmed reads corresponding to the gonads. Usually I use `scp` but some googling suggests that `rsync` might be a better option, because it will transfer files specified in a text file, like so:

`rsync -av --progress --files-from=file_list.txt source/ me@example.com:/dest`

That means that I need to save just a list of the files. 

```{r createTransferList, eval=FALSE}
files<-unlist(snigra_samples[snigra_samples$Source %in% c("ovary","testes"),c("R1","R2")])
# remove paths
files<-gsub("../OG7629/trimmed/(.*)","\\1",files)
write.table(files,
            "nigra_gonads_files.txt",
            sep='\t',
            col.names = FALSE,
            row.names = FALSE,
            quote=FALSE)
```

Then, the rsync command should be: `rsync -av --progress --files-from=nigra_gonads_files.txt /mnt/BigData/OG7629/trimmed/ trinity-em:~/pipefish_sbge/OG7629/trimmed/`, from the `pipefish_sbge/stigmatopora/` directory -- note that it needs to use the actual location, not the symbolic link to OG7629.

Then I need to send the samples file to the RCC (from `pipefish_sbge/stigmatopora`: `rsync -av --progress nigra_gonads_trinity.txt trinity-em:~/pipefish_sbge/stigmatopora/` 

I also created a directory on the RCC for the trinity results, called `trinity_nigra_gonads` within the `results/` directory (so, `pipefish_sbge/results/trinity_nigra_gonads`).

It turns out that the sample names can't be relative paths because the docker is trying to run trinity from within the `/usr/local/bin/Trinity` directory. So I need to replace the `../` with the path name, which I'll do using a regex substitution via a perl one-liner: `perl -pe 's/\.\./\/home\/rccuser\/pipefish_sbge/g' nigra_gonads_trinity.txt > nigra_gonads_trinity_rcc.txt`.


### Running Trinity


I am using the script `run_trinity.sh` to run Trinity in the docker. It requires two arguments: one of which is the sample file and the other is the output directory. I'm running it like so (from the `pipefish_sbge/stigmatopora/` folder): `nohup ./scripts/run_trinity.sh nigra_gonads_trinity_rcc.txt results/trinity_nigra_gonads/ > trinity_gonads.log 2>&1 &`.

```{bash, file='scripts/run_trinity.sh', eval=FALSE}

```


It's important to make sure the docker instance is referred to correctly -- in the example scripts, it's at `pwd`:`pwd`, but for me that needed to be adjusted to `/home/rccuser/:/home/rccuser` because I'm not running it from the home directory but rather from the project directory. 

I ran into one issue that was the same as the one Coley ran into, indicating that RAM was an issue: 

```
bash: line 1:   341 Killed                  /usr/local/bin/util/..//Inchworm/bin/fastaToKmerCoverageStats --reads left.fa --kmers jellyfish.K25.min2.kmers.fa --kmer_size 25 --num_threads 6 --DS > left.fa.K25.stats
Thread 3 terminated abnormally: Error, cmd: /usr/local/bin/util/..//Inchworm/bin/fastaToKmerCoverageStats --reads left.fa --kmers jellyfish.K25.min2.kmers.fa --kmer_size 25  --num_threads 6  --DS  > left.fa.K25.stats died with ret 35072 at /usr/local/bin/util/insilico_read_normalization.pl line 795.
Error, thread exited with error Error, cmd: /usr/local/bin/util/..//Inchworm/bin/fastaToKmerCoverageStats --reads left.fa --kmers jellyfish.K25.min2.kmers.fa --kmer_size 25  --num_threads 6  --DS  > left.fa.K25.stats died with ret 35072 at /usr/local/bin/util/insilico_read_normalization.pl line 795.

bash: line 1:   344 Killed                  /usr/local/bin/util/..//Inchworm/bin/fastaToKmerCoverageStats --reads right.fa --kmers jellyfish.K25.min2.kmers.fa --kmer_size 25 --num_threads 6 --DS > right.fa.K25.stats
Thread 4 terminated abnormally: Error, cmd: /usr/local/bin/util/..//Inchworm/bin/fastaToKmerCoverageStats --reads right.fa --kmers jellyfish.K25.min2.kmers.fa --kmer_size 25  --num_threads 6  --DS  > right.fa.K25.stats died with ret 35072 at /usr/local/bin/util/insilico_read_normalization.pl line 795.
Error, thread exited with error Error, cmd: /usr/local/bin/util/..//Inchworm/bin/fastaToKmerCoverageStats --reads right.fa --kmers jellyfish.K25.min2.kmers.fa --kmer_size 25  --num_threads 6  --DS  > right.fa.K25.stats died with ret 35072 at /usr/local/bin/util/insilico_read_normalization.pl line 795.

Error, 2 threads errored out at /usr/local/bin/util/insilico_read_normalization.pl line 999.
Error, cmd: /usr/local/bin/util/insilico_read_normalization.pl --seqType fq --JM 30G  --max_cov 200 --min_cov 1 --CPU 12 --output /home/rccuser/pipefish_sbge/stigmatopora/results/trinity_nigra_gonads/insilico_read_normalization --max_CV 10000  --left /home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-51-49-1_S51_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-52-49-1_S52_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-53-49-1_S53_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-54-49-1_S54_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-55-49-1_S55_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-56-49-1_S56_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-57-49-1_S57_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-58-49-1_S58_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-59-49-1_S59_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-60-49-1_S60_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-61-49-1_S61_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-62-49-1_S62_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-63-49-1_S63_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-64-49-1_S64_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-65-49-1_S65_paired_R1.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-66-49-1_S66_paired_R1.fastq.gz --right /home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-51-49-1_S51_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-52-49-1_S52_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-53-49-1_S53_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-54-49-1_S54_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-55-49-1_S55_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-56-49-1_S56_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-57-49-1_S57_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-58-49-1_S58_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-59-49-1_S59_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-60-49-1_S60_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-61-49-1_S61_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-62-49-1_S62_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-63-49-1_S63_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-64-49-1_S64_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-65-49-1_S65_paired_R2.fastq.gz,/home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-66-49-1_S66_paired_R2.fastq.gz --pairs_together  --PARALLEL_STATS   died with ret 7424 at /usr/local/bin/Trinity line 2879.
  main::process_cmd("/usr/local/bin/util/insilico_read_normalization.pl --seqType "...) called at /usr/local/bin/Trinity line 3432
  main::normalize("/home/rccuser/pipefish_sbge/stigmatopora/results/trinity_nigr"..., 200, ARRAY(0x55f1fd311330), ARRAY(0x55f1fd471ad0)) called at /usr/local/bin/Trinity line 3372
  main::run_normalization(200, ARRAY(0x55f1fd311330), ARRAY(0x55f1fd471ad0)) called at /usr/local/bin/Trinity line 1410

```

So Francois upped the memory to 64GB. After that, Trinity made it through at normalizing the first 6 samples but died on the 7th one with this error:


```
Error, cmd: jellyfish count -t 12 -m 25 -s 8683660219  --canonical  both.fa died with ret 9 at /usr/local/bin/util/insilico_read_normalization.pl line 795.
Error, cmd: /usr/local/bin/util/insilico_read_normalization.pl --seqType fq --JM 62G  --max_cov 200 --min_cov 1 --CPU 12 --output /home/rccuser/pipefish_sbge/stigmatopora/results/trinity_nigra_gonads/norm_for_read_set_7 --max_CV 10000  --left /home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-65-49-1_S65_paired_R1.fastq.gz --right /home/rccuser/pipefish_sbge/OG7629/trimmed/AAAVL52HV-7629-65-49-1_S65_paired_R2.fastq.gz --pairs_together  --PARALLEL_STATS   died with ret 512 at /usr/local/bin/Trinity line 2879.
        main::process_cmd("/usr/local/bin/util/insilico_read_normalization.pl --seqType "...) called at /usr/local/bin/Trinity line 3432
        main::normalize("/home/rccuser/pipefish_sbge/stigmatopora/results/trinity_nigr"..., 200, ARRAY(0x55720b1ae838), ARRAY(0x55720b1ab460)) called at /usr/local/bin/Trinity line 3349
        main::run_normalization(200, ARRAY(0x55720ad82070), ARRAY(0x55720af5add8)) called at /usr/local/bin/Trinity line 1410
```

This might be a RAM issue but according to [this issue on GitHub](https://github.com/trinityrnaseq/trinityrnaseq/issues/662) it might be worth re-running with a clean directory. So I've deleted the previous one and am re-running it now, and have upped the memory to 63GB RAM and 14 CPUs (from 62 and 12). 

It was indeed a RAM issue, and Trinity has now successfully run with 128GB of RAM on machine 132.181.102.45. I created a single assembly for the *S. nigra* gonads (note: if the testes and ovary transcriptomes are sufficiently diverged, I might decide to create separate assemblies for males and females and do some mapping across them to enable comparisons).

#### Assessing the assembly

I used the built-in assembly check tool from Trinity, using this command:

```{bash}
sudo docker run -v`pwd`:`pwd` trinityrnaseq/trinityrnaseq /usr/local/bin/util/TrinityStats.pl \
        /home/rccuser/pipefish_sbge/stigmatopora/results/trinity_nigra_gonads.Trinity.fasta \
        >> /home/rccuser/pipefish_sbge/stigmatopora/results/trinity_nigra_gonads.Trinity.stats
```

It took about 30-60s to run but successfully reported the following statistics:

```{bash, file='results/trinity_nigra_gonads.Trinity.stats', eval=FALSE}

```

It's produced an ~307Mb transcriptome, which isn't too crazy. 

### Counting reads using RSEM

The next step is to map the reads back to the transcriptome to establish expression values. We'll do this with RSEM -- and afterwards, these expression levels will be used to help clean and filter the transcriptome.


