---
title: "Differential Expression Analysis in _Stigmatopora macropterygia_"
author: "This code was adapted from Coley Tosto's code by Emily Beasley and Sarah Flanagan."
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: yes
    toc: true
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
always_allow_html: true
---


```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
```
```{r setdir}
knitr::opts_knit$set(root.dir='../',
                     fig_path="../figs/")
```


``` {r library, include = TRUE, message = FALSE, warning = FALSE}
#This is a comprehensive list of all the libraries used in this document
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(ggpubr)
library(cowplot)
library(UpSetR)
library(ggplot2)
library(knitr)
library(kableExtra)
library(tidyverse)
library(magick)
library(multcompView)
library(spfTools)
library(patchwork)
library(ridgeline) #remotes::install_github("R-CoderDotCom/ridgeline@main")

library(car)
library(multcompView)
library(WGCNA)
library(yarrr)
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.macropterygia <- readRDS("data/txi.salmon_SM.RDS")

#The samples file generated for tximport
samples <- read.table("data/SM_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)

```

The package `DESeq2` was used for the differential expression analysis outlined below.

To analyze your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions. If we investigate the column data of our DESeq dataset we can see that each sample has both a sex and organ type attached to it, we will be using these two factors in our multi-factor analysis. 


# Multifactor design - Comparing M v F across the diff tissue types

In this multi-factor analysis the model was run as counts ~ group, where group included both the sex and the organ type (i.e. MLiver, FLiver, etc.). The sample "S10" appeared to a bit of an outlier in the above analysis, particularly clear in the sample-dist heatmap and the PCA, so it will be removed from the analysis going forward.

```{r multi-factor-analysis, message=FALSE, warning=FALSE}
#Create an additional column the has the two conditions combined(sex and organ type)
samples$group <- factor(paste0(samples$sex, samples$organ))

##Create a copy of the DESeq dataset
ddsMF_SM <- DESeqDataSetFromTximport(txi.macropterygia,
                                     colData = samples,
                                     design = ~ group)
```

The data is then filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

The cutoff here was to remove rows that had counts fewer than 10 across all samples.

Sample S23 (Mapping rate = 33.7558%) and S30 (Mapping rate = 11.6745%) both had lower than expected gene counts, so they were filtered out entirely. S10 was clustering in unexpected ways, so they were also removed from further analyses. 

```{r multi-factor-analysis, message=FALSE, warning=FALSE}
##Remove S10, S23, and S30 from the dataset
ddsMF_SM <- ddsMF_SM[, !(ddsMF_SM$ID %in% c("S10", "S23", "S30"))]
samples<-samples[!(samples$ID %in%  c("S10", "S23", "S30")),]

##Filter the dataset, only keeping rows that have at least 10 reads total
keep <- rowSums(counts(ddsMF_SM)) >= 10 
ddsMF_SM <- ddsMF_SM[keep, ]

#Run the differential expression analysis
ddsMF_SM_exp <- DESeq(ddsMF_SM)
resultsNames(ddsMF_SM_exp)
```

## Visualising multifactor results

```{r}
resMF <- results(ddsMF_SM_exp)
```

```{r MAplotMultiFactor, echo=FALSE, fig.cap="LogFC versus the mean normalised count for all of the genes in the multifactor analysis. Points that are blue have a p-value less than 0.1. The two plots are the same with the only difference coming from an adjusted y-limit. Points that are triangles represent genes with a fold change higher/lower than the y-limit."}

par(mfrow=c(1,2))
plotMA(resMF, ylim = c(-2,2))
plotMA(resMF)

```

```{r getHeatmapData, echo=FALSE}
#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(ddsMF_SM_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]
df <- as.data.frame(colData(ddsMF_SM)[,c("sex", "organ")])
```


```{r transformDDS, echo=FALSE, eval=FALSE}
#Transform the data
vsd <- vst(ddsMF_SM_exp, blind=FALSE)
saveRDS(vsd,"data/vsd.RDS")
```
```{r}
vsd<-readRDS("data/vsd.RDS")
```


Let's save the expression data to a file for future investigations.

```{r eval=FALSE}
# following https://fuzzyatelin.github.io/bioanth-stats/module-F21-Group1/module-F21-Group1.html
expression_data<-t(assay(vsd))
write.csv(expression_data,"data/expression_data_for_WGCNA.csv",row.names = TRUE)
```


```{r heatmapMF, echo=FALSE, fig.cap= "Heatmap showing the expression level across all organs for the top 20 genes with the highest expression levels in the multifactor analysis."}
#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df)
```

From the heatmap (Fig. \@ref(fig:heatmap)) we can see that for all of these top 20 expressed genes the highest expression in found in the liver. There are also a few genes that appear to be highly expressed across all of the tissue types. With this heatmap we can also pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.

```{r heatmap_genes}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))
#Run once to create the file and then commented out
#write.table(heatmap_TG,
 #           'data/heatmap_trinitygenes.txt',
  #          sep = "",
   #         quote=FALSE,
    #        row.names = FALSE,
     #       col.names = FALSE)
```

### Sample-dist Heatmap

We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples.

```{r sample-dist, echo=FALSE, fig.cap="Sample-to-sample distances heatmap showing the similarities and differences between all of the samples. The darker the colour, the more similar the samples are. The diagonal line of very dark blue represents the comparisons between the same samples.", warning=FALSE}
#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$sex, vsd$organ, sep = "-")
colnames(sampleDistsMatrix) <- vsd$ID

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)
```

We can see that the highest similarities (aside from same samples comparisons) is between samples from the same organ. After that, the next highest similarities can be seen between male and female livers. 

### PCA plots

Next we can look at PCA plots for our samples to see how our different samples are grouping. 

```{r pca-plot-1v2, echo=FALSE, message=FALSE, warning=FALSE}


#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$organ <- samples$organ
pca_plotting_data$sex <- samples$sex
pca_plotting_data$ID <- samples$ID

#Calculate the percent variation attributed to each axis 
sumpca<-summary(pca)
percentVar <- round(sumpca$importance[2,]*100, digits = 2)


my_colors <- c("Gill" = "#6E8B3D75", "Gonad" = "#EEB42275", "Liver" = "#EE826275")
my_shapes <- c(16, 17)

ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed') + 
  geom_text(aes(label = ID), vjust = -1, hjust = 0.5)
```

From the pairs plot we can see that PC1 appears to be explaining differences between the liver and the other tissues, PC2 appears to be explaining differences between gonadal and somatic tissues, PC3 seems to account for differences between male and female samples, especially in the gonads. I am going to plot all of PC2 - 4 against PC1 below for a clearer picture.


```{r pca1v2-v2, echo=FALSE, warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

```{r pca1v3-v2, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC3.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC3, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC3: ",percentVar[3],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

```{r pca1v4-v2, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC4. This fourth axis explains NA% of the variation.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC4, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC4: ",percentVar[4],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

#### Saving the PCA plots to use in figure creation
I will be including the PCAs describing the relationship between PC1 - 3 in the main text, as they account for the majority of the differences between our samples. The plot with PC1 and PC4 I still want to export, and will likely include as a supplemental to highlight why samples related to S34 will be removed from the subsequent analyses.

```{r setParams-PCAs}
#Setting the shapes I want to show up for the different organs
organ_shapes <- c(Gill = 16, Liver = 17, Gonad = 15)

#Setting the colors I want for each sex
sex_cols <- c("F" = "#7fc97f", "M" = "#beaed4")

```

```{r figPCA1v2, eval = FALSE}
#Create the blank pdf to store the plot in
pdf("figs/Fig_PCA1v2.pdf",height = 6,width=6)

#Set the plotting parameters
par(mar=c(4,5,4,1), oma=c(2,2,2,2))

#Create the plot for PC1 v PC2
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC2,
     col = paste0(sex_cols[pca_plotting_data$sex],"75"),
     pch = organ_shapes[as.character(pca_plotting_data$organ)],
     cex = 3,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC2: ",percentVar[2], "% variance"),
     bty = 'l',
     xpd = TRUE)

#Add a legend describing the sex
outer_legend("top",
             c("Female","Male"),
             pch = 18,
             bty = 'n',
             col=paste0(sex_cols,"75"),
             cex = 1.75,
             ncol = 2,
             pt.cex = 3)

dev.off()
```

```{r figPCA1v3, eval = FALSE}
#Create the blank pdf to store the plot in
pdf("figs/Fig_PCA1v3.pdf",height = 6,width=6)

#Set the plotting parameters
par(mar=c(4,5,4,1), oma=c(2,2,2,2))

#Create the plot for PC1 v PC3
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC3,
     col = paste0(sex_cols[pca_plotting_data$sex],"75"),
     pch = organ_shapes[as.character(pca_plotting_data$organ)],
     cex = 3,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC3: ",percentVar[3], "% variance"),
     bty = 'l',
     xpd = TRUE)

#Add  legend describing the organs
outer_legend("top",
             names(organ_shapes),
             pch=organ_shapes,
             bty = 'n',
             col = 'darkgrey',
             cex = 1.75,
             ncol = 3,
             pt.cex = 3)

dev.off()
```

```{r figPCA1v4, eval = FALSE}
#Create the blank pdf to store the plot in
pdf("figs/Fig_PCA1v4.pdf",height = 6,width=8)

#Set the plotting parameters
par(mar=c(4,5,1,5),oma=c(2,2,2,2))

#Create the plot for PC1 v PC4
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC4,
     col = paste0(sex_cols[pca_plotting_data$sex],"75"),
     pch = organ_shapes[as.character(pca_plotting_data$organ)],
     cex = 3,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC4: ",percentVar[4], "% variance"),
     bty = 'l',
     xpd = TRUE)


#Add  legend describing the organs and sex
outer_legend("right",
             c("Gill","Gonad","Liver"),
             pch = c(16,15,17),
             bty = 'n',
             col = 'darkgrey',
             cex = 1.5,
             ncol = 1,
             pt.cex = 2.25)

outer_legend("topright",
             c("Female","Male"),
             pch = 18,
             bty = 'n',
             col=paste0(sex_cols,"75"),
             cex = 1.5,
             ncol = 1,
             pt.cex = 2.25)
dev.off()
```

#### Creating heatmaps based on the PCA axes
I now want to generate heatmaps based on the genes that have the highest loadings for each of the PCA axes. I will be using a cut off of 0.02 to determine which genes will be included in the heatmaps. Once again I am doing this for PCA1-4, but will likely only include PCA1-3 in a figure for the maintext and PC4 will go in the supplemental.

```{r setParams-heatmaps}

df <- as.data.frame(samples[,c("sex", "organ")])
rownames(df) <- samples$ID
df$organ <- as.character(df$organ)

organ_cols<-c("Gill" = "#20B2AA",
              "Gonad" = "#EEB422", 
              "Liver" = "#EE8262")

ann_colors = list(sex=sex_cols,
                  organ=organ_cols)

col_order <- c(rownames(df[df$sex=="F" & df$organ=="Gill",]),
               rownames(df[df$sex=="M" & df$organ=="Gill",]),
               rownames(df[df$sex=="F" & df$organ=="Gonad",]),
               rownames(df[df$sex=="M" & df$organ=="Gonad",]),
               rownames(df[df$sex=="F" & df$organ=="Liver",]),
               rownames(df[df$sex=="M" & df$organ=="Liver",]))

pca_rotation <- pca$rotation[, 1:4]

```

```{r savePheatmapFxn}

#Function to use to export the heatmaps to pdfs
# from https://stackoverflow.com/questions/43051525/how-to-draw-pheatmap-plot-to-screen-and-also-save-to-file
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

```

```{r figheatmap-PC1, eval = FALSE}

pc1 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,1]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors = ann_colors,
                cellwidth = 9,
                fontsize = 16,
                annotation_legend = FALSE,
                main = "Top loading genes on PC1")


save_pheatmap_pdf(pc1, "figs/Fig_pc1_heatmap.pdf",
                 width=6,
                height=6)

```

```{r figheatmap-PC2, eval=FALSE}

pc2 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,2]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                annotation_legend = FALSE,
                main = "Top loading genes on PC2")

save_pheatmap_pdf(pc2, "figs/Fig_pc2_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC3, eval=FALSE}

pc3 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,3]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                border_color = NA,
                main = "Top loading genes on PC3")

save_pheatmap_pdf(pc3, "figs/Fig_pc3_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC4, eval=FALSE}

pc4 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,4]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                border_color = NA,
                main = "Top loading genes on PC4")

save_pheatmap_pdf(pc4, "figs/Fig_pc4_heatmap.pdf",
                  width=6,
                  height=6)

```



## Invesitgate the results of the differential expression
Thanks to the multi-factor analysis we can now explore differential expression between all of the different combinations:

  - Male Liver v. Female Liver
  - Male Gill v. Female Gill
  - Male Gonad v. Female Gonad
  - All of the within sex tissue comparisons (e.g. Male Liver v. Male Gill, etc.)

### M-F Liver Comparisson
```{r m-f-liver-results}
##Pulling out the liver M-F results with an alpha of 0.05
liver_con_res <- results(ddsMF_SM_exp, contrast = c("group", "MLiver", "FLiver"),
                         alpha = 0.05)
liver_con_res$trin_geneid <- rownames(liver_con_res)
head(liver_con_res)

summary(liver_con_res)
```

There are many criteria that have been employed previously to denote sex bias. For this study we are classifying sex-biased genes as **genes with a p-value < 0.05 AND a logFC $\ge$ |2|**. With that criteria in the liver there are `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. 

I have then pulled out all of the male-biased, female-biased, and non-biased genes based on the criteria outlined above. I determined non-biased genes as a p-vaule > 0.05.

```{r liver-sex-biased}
#Removing the rows where padj. is NA in results
liver_con_res_noNA <- liver_con_res[!is.na(liver_con_res$padj),]
summary(liver_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed
#Creating a vector that contains all of the male-biased and female-biased genes in the liver
liver_mal_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange >= 2
                                             & liver_con_res_noNA$padj <= 0.05),]
liver_fem_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange <= -2 
                                             & liver_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the liver
liver_non_biased <- liver_con_res_noNA[which(liver_con_res_noNA$padj > 0.05),]

```


```{r save-liver-sex-biased, eval=FALSE}
write.csv(as.data.frame(liver_con_res_noNA), "data/liver_res.csv", row.names = TRUE)
```



I will be generating a table that outlines the genes that are the most male or female biased in each organ type. To do this I need to pull all differentially expressed genes in males or females, get the Trinity gene IDs and then BLAST the corresponding sequences. Here I am pulling out the Trinity gene IDs for those genes.

```{r all-liver}
#Creating a subset of the results where the p-value is less than 0.05
liver_con_res_p05 <- liver_con_res_noNA[liver_con_res_noNA$padj <= 0.05, ]


#Pulling the all diff expressed genes in Males
all_MLiver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(
  liver_con_res_p05$log2FoldChange, decreasing = TRUE),], n=4979)))
#Run once to create the file and then commented out
#write.table(all_MLiver_trin_genes,
 #          'data/SM_maleL_allTRgenes.txt', 
  #         sep = "", 
   #        quote=FALSE, 
    #         row.names = FALSE, 
     #        col.names = FALSE)

#Pulling all diff expressed genes in Females
all_Fliver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(
  liver_con_res_p05$log2FoldChange, decreasing = FALSE),], n=2571)))

#Run once to create the file and then commented out
#write.table(all_Fliver_trin_genes,
 #           'data/SM_femL_allTRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### M-F Gill Comparisson
```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res <- results(ddsMF_SM_exp, contrast = c("group", "MGill", "FGill"),
                        alpha = 0.05)
gill_con_res$trin_geneid <- rownames(gill_con_res)
head(gill_con_res)

summary(gill_con_res)
```

In the gills there were `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

I have then pulled out all of the male-biased, female-biased, and non-biased genes to save them into their own objects.

```{r gill-sex-biased}
#Removing the rows where padj. is NA in results
gill_con_res_noNA <- gill_con_res[!is.na(gill_con_res$padj), ]
summary(gill_con_res_noNA) #We can now see that there are no outliers or 
                           #low counts since the NAs have been removed
#Creating a vector that contains all of the male-biased and female-biased genes in the gills
gill_mal_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange >= 2 
                                           & gill_con_res_noNA$padj <= 0.05),]
gill_fem_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange <= -2 
                                           & gill_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gills, p>0.05
gill_non_biased <- gill_con_res_noNA[which(gill_con_res_noNA$padj > 0.05),]
```
```{r save-gill-sex-biased, eval=FALSE}
write.csv(as.data.frame(gill_con_res_noNA), "data/gill_res.csv", row.names = TRUE)
```


Here I am getting the trinity gene IDs that correspond to all differentially expressed genes for males and females. 

```{r all-gill}
#Creating a subset of results where p-value is less than 0.05
gill_con_res_p05 <- gill_con_res_noNA[gill_con_res_noNA$padj <= 0.05,]

#Pulling all diff expressed genes in Males
all_MGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(
  gill_con_res_p05$log2FoldChange, decreasing = TRUE),], n=137)))

#Run once to create the file and then commented out
#write.table(all_MGill_trin_genes,
 #           'data/SM_maleG_allTRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)

#Pulling all diff expressed genes in Females 
all_FGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(
  gill_con_res_p05$log2FoldChange, decreasing = FALSE),], n=16)))

#Run once to create the file and then commented out
#write.table(all_FGill_trin_genes,
 #           'data/SM_femG_allTRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### M-F Gonad Comparisson
```{r m-f-gonads}
##Pulling out the gonad M-F results
gonad_con_res <- results(ddsMF_SM_exp, 
                         contrast = c("group", "MGonad", "FGonad"),
                         alpha = 0.05)
gonad_con_res$trin_geneid <- rownames(gonad_con_res)
head(gonad_con_res)

summary(gonad_con_res)
```

Between the ovaries and testes (gonads) there were `r sum(gonad_con_res$padj <= 0.05 & gonad_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gonad_con_res$padj <= 0.05 & gonad_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

From the gonad M-F results I have then pulled out all of the male-biased, female-biased, and non-biased genes and saved them to their own objects.

```{r gonad-sex-biased}
#Removing the rows where padj. is NA in results
gonad_con_res_noNA <- gonad_con_res[!is.na(gonad_con_res$padj), ]
summary(gonad_con_res_noNA) #We can now see that there are no outliers or 
                            #low counts since the NAs have been removed

#Creating an object that contains all of the male-biased and female-biased genes in the gonads
gonad_mal_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$log2FoldChange >= 2 
                                             & gonad_con_res_noNA$padj <= 0.05),]
gonad_fem_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$log2FoldChange <= -2 
                                             & gonad_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gonads, p>0.05
gonad_non_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$padj > 0.05),]
```
```{r save-gonad-sex-biased}
write.csv(as.data.frame(gonad_con_res_noNA), "data/gonad_res.csv", row.names = TRUE)
```

I then pulled out the trinity gene IDs that corresponded to all differentially expressed genes in males and females so that I could BLAST the sequences.


```{r all-gonad}
#Creating a subset of the results where the p-value is less than 0.05
gonad_con_res_p05 <- gonad_con_res_noNA[gonad_con_res_noNA$padj <= 0.05,]

#Pulling all diff expressed genes in Males
all_MGonad_trin_genes <- cbind(rownames(head(gonad_con_res_p05[order(
  gonad_con_res_p05$log2FoldChange, decreasing = TRUE),], n=4781)))

#Run once to create the file and then commented out
#write.table(all_MGonad_trin_genes,
 #        'data/SM_maleGon_allTRgenes.txt', 
  #        sep = "", 
   #       quote=FALSE, 
    #      row.names = FALSE, 
     #     col.names = FALSE)

#Pulling all diff expressed genes in Females
all_FGonad_trin_genes <- cbind(rownames(head(gonad_con_res_p05[order(
  gonad_con_res_p05$log2FoldChange, decreasing = FALSE),], n=2466)))

#Run once to create the file and then commented out
#write.table(all_FGonad_trin_genes,
 #           'data/SM_femGon_allTRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### MA Plots
```{r MA-plot-MF, echo=FALSE, fig.height= 15, fig.width=8, fig.cap="MA-plots generated for each organ type that compares logFC to the mean expression. Female-biased and male-biased genes are represented by colour (female: green; male: purple) and are determined with a fold change cutoff of 2 and a p-value cutoff of 0.05"}
liver_MA <- ggmaplot(liver_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Liver" %->% "Female Liver"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gill_MA <- ggmaplot(gill_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Gill" %->% "Female Gill"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gonad_MA <- ggmaplot(gonad_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Testis" %->% "Ovaries"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

#Merge all the plots into one
plot_grid(liver_MA, gill_MA, gonad_MA,
          ncol = 1,
          nrow = 3,
          labels = c('A', 'B', 'C'),
          label_fontfamily = 'sans',
          label_fontface = 'bold',
          label_size = 18,
          rel_heights = c(4.10, 4.10, 4.10))

```

From the MA plots we can see the highest number of biased genes are found in the gonads (Fig. \@ref(fig:MA-plot-MF)) and then the liver and lastly the gills. In both the liver and the gonads there are more female-biased genes than male-biased genes.

## Variation in FC across sex-bias and tissue type
I want to create a plot the highlights the variation we see in fold-change both across the different biases groups (male-biased, female-biased, and non-biased) within one tissue type and across all of the tissue types. To do this I first need to create a "long" dataset in the following format:

| Tissue_type|     Bias      |   Fold-Change   |
|:----------:|:-------------:|:---------------:|
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   fem_bias    |      XXXXX      |
|    Gill    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   fem_bias    |      XXXXX      |
|    Liver   |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   fem_bias    |      XXXXX      |
|   Gonad    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |

That dataset is created here, with the Trinity gene IDs included as well:

```{r logFC-long-dataset, eval=FALSE}
logFC_long<- data.frame(
  tissue=c(rep("Gill",nrow(gill_fem_biased)),
           rep("Gill", nrow(gill_mal_biased)),
           rep("Gill", nrow(gill_non_biased)),
           rep("Gonad", nrow(gonad_fem_biased)),
           rep("Gonad", nrow(gonad_mal_biased)),
           rep("Gonad", nrow(gonad_non_biased)),
           rep("Liver", nrow(liver_fem_biased)),
           rep("Liver", nrow(liver_mal_biased)),
           rep("Liver", nrow(liver_non_biased))
         ),
  bias=c(rep("FB",nrow(gill_fem_biased)),
         rep("MB",nrow(gill_mal_biased)),
         rep("UB", nrow(gill_non_biased)),
         rep("FB", nrow(gonad_fem_biased)),
         rep("MB", nrow(gonad_mal_biased)),
         rep("UB", nrow(gonad_non_biased)),
         rep("FB", nrow(liver_fem_biased)),
         rep("MB", nrow(liver_mal_biased)),
         rep("UB", nrow(liver_non_biased))
         ),
  logFC=c(gill_fem_biased$log2FoldChange,
          gill_mal_biased$log2FoldChange,
          gill_non_biased$log2FoldChange,
          gonad_fem_biased$log2FoldChange,
          gonad_mal_biased$log2FoldChange,
          gonad_non_biased$log2FoldChange,
          liver_fem_biased$log2FoldChange,
          liver_mal_biased$log2FoldChange,
          liver_non_biased$log2FoldChange
          ),
  geneID=c(rownames(gill_fem_biased),
           rownames(gill_mal_biased),
           rownames(gill_non_biased),
           rownames(gonad_fem_biased),
           rownames(gonad_mal_biased),
           rownames(gonad_non_biased),
           rownames(liver_fem_biased),
           rownames(liver_mal_biased),
           rownames(liver_non_biased)
           )
  
)

#Exporting this data, run once and then just read in the file
write.csv(logFC_long, "data/logFC_long_sexbias.csv", row.names = FALSE)
```
```{r}
logFC_long <- read.csv("data/logFC_long_sexbias.csv")
```


With the dataset in the proper format now, we can generate the plot


```{r FC-var-plot, echo=FALSE, fig.cap="Absolute value of the logFC for genes that are female-biased, male-biased, and unbiased across each tissue type. Raw fold-change values are added on top of the boxplot as jitters."}
my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "darkgray")

ggplot(logFC_long,
       aes(x = bias, y = abs(logFC), fill = bias)) +
  geom_boxplot() +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_point(data=logFC_long, aes(x=bias, y=abs(logFC)),
             size=0.4, 
             position = position_jitter(width = 0.15)) + 
  labs(x="Bias Level", y="|logFC|") +
  guides(fill = guide_legend(title = "Bias Level", order = 3))

```


Range -30.33088  46.35090

### Statistical analysis


There appears to be differences in the logFC across both the organs and the sex-bias groups (male-biased, fem-biased, and non-biased) (Fig. \@ref(fig:FC-var-plot)). Let's explore some analyses to see if we can put some statistical power behind those differences.

```{r FC-var-summary-stats, warning=FALSE, eval=FALSE}
#Looking at some summary statistics for logFC between the different groups
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), mean)
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), sd)
```


We'll run an ANOVA on rank-transformed data

```{r FC-var-anova, warning=FALSE, eval=FALSE}

logFC_long$tissue <- as.factor(logFC_long$tissue)
logFC_long$bias <- as.factor(logFC_long$bias)

fc_var_aov <- aov(rank(abs(logFC_long$logFC)) ~ logFC_long$bias*logFC_long$tissue)

par(mfrow = c(2,2))
plot(fc_var_aov) 
leveneTest(fc_var_aov) # not equal variances
par(mfrow=c(1,1))
hist(resid(fc_var_aov))
```


```{r FC-var-anova, warning=FALSE, eval=FALSE}
anova(fc_var_aov)
plot(TukeyHSD(fc_var_aov))

multcompLetters4(fc_var_aov, TukeyHSD(fc_var_aov))

```

```{r}
par(mfrow=c(1,1))
interaction.plot(logFC_long$tissue, logFC_long$bias, logFC_long$logFC)

```



"sex-biased genes displayed a broad range of log2 fold-change values, generally exceeding those observed in the liver or gills "

### Visualising male vs female expression

```{r mvf-expression}
#Looking at the relationship between male expression and female expression
organs <- c(unique(logFC_long$tissue))

dds_SM_exp$organ <- ifelse(dds_SM_exp$organ %in% c("Ovaries", "Testis"),
                           paste0("Gonad"), 
                           paste0(dds_SM_exp$organ))

bias_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "#A9A9A975") 

#pdf("figs/Fig_MFexpression.pdf", width = 10, height=4)

par(mfrow=c(1,3), oma=c(4,4,2,8), mar=c(1,1,1,0))

for (organ in organs) {
  
  #browser()
  tmp <- as.data.frame(counts(dds_SM_exp, normalized = TRUE)[, dds_SM_exp$organ == organ])
  tmpF <- as.data.frame(counts(dds_SM_exp, normalized = TRUE)[, dds_SM_exp$organ == organ & dds_SM_exp$sex == "F"])
  tmpM <- as.data.frame(counts(dds_SM_exp, normalized = TRUE)[, dds_SM_exp$organ == organ & dds_SM_exp$sex == "M"])
  
  tmp$Fmedian <- apply(tmpF, 1, median)
  tmp$Mmedian <- apply(tmpM, 1, median)
  
  tmp <- tmp[rownames(tmp) %in% logFC_long[logFC_long$tissue == organ, ]$geneID, ]
  
  tmp <- merge(tmp, logFC_long[logFC_long$tissue == organ, ],
               by.x = 0, 
               by.y = "geneID", 
               all = TRUE)
  ymax <- max(tmp$Mmedian, na.rm = TRUE) + 1
  xmax <- max(tmp$Fmedian, na.rm = TRUE) + 1
  
  plot(tmp$Fmedian, tmp$Mmedian,
       #ylab = expression('log'[2] * '(Male Expression)'),
       #xlab = expression('log'[2] * '(Female Expression)'),
       #axes=FALSE,
       cex.main=2,
       #main = organ,
       col = bias_colors[tmp$bias],
       pch = 19,
       cex = ifelse(organ != "Gonad" & tmp$bias %in% c("MB", "FB"),
                    0.75,
                    0.25),
       ylim = c(0, 100),
       xlim = c(0, 100))
  
  
  
  #Make the axis lines longer and thicker
  axis(1, labels=seq(-2, xmax, 2), 
       line=NA, at=seq(-2, xmax, 2), 
       lwd=2,xlim=c(0, xmax),cex.axis=1.5,las=1)
  axis(2, labels=seq(-2, ymax, 2),
       line=NA, at=seq(-2, ymax, 2), 
       lwd=2, ylim=c(0,ymax),cex.axis=1.5,las=1)
  
  mtext(organ, cex=1.5,outer=FALSE, line=-1)
  
}

# add x axis label
mtext(expression('log'[2] * '(Female Expression)'), side = 1, cex=1.25, outer=TRUE, line=2)

# add y axis label
mtext(expression('log'[2] * '(Male Expression)'), side = 2, cex=1.25, outer=TRUE, line=2)

# add legend
spfTools::outer_legend("right",
                       legend=c("Female\nbiased",
                                "Male\nbiased",
                                "Unbiased"),
                       pt.bg=bias_colors,
                       pch=22,
                       bty='n',
                       ncol=1,
                       cex=2,
                       y.intersp = 1.5,
                       pt.cex=2)

#dev.off()


```

```{r mean-mvf-expression, fig.height=7, fig.width=15,dpi=300}
#Looking at the relationship between male expression and female expression
organs <- c(unique(logFC_long$tissue))

ddsMF_SM_exp$organ <- ifelse(ddsMF_SM_exp$organ %in% c("Ovaries", "Testes"),
                           paste0("Gonad"), 
                           paste0(ddsMF_SM_exp$organ))

bias_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "#A9A9A975") 

#dev.new(width=15, height=7)

par(mfrow=c(1,3), oma=c(4,4,2,8), mar=c(10,5,5,0))

for (organ in organs) {
  
  #browser()
  tmp <- as.data.frame(counts(ddsMF_SM_exp, normalized = TRUE)[, ddsMF_SM_exp$organ == organ])
  tmpF <- as.data.frame(counts(ddsMF_SM_exp, normalized = TRUE)[, ddsMF_SM_exp$organ == organ & ddsMF_SM_exp$sex == "F"])
  tmpM <- as.data.frame(counts(ddsMF_SM_exp, normalized = TRUE)[, ddsMF_SM_exp$organ == organ & ddsMF_SM_exp$sex == "M"])
  
  tmp$Fmean <- apply(tmpF, 1, mean)
  tmp$Mmean <- apply(tmpM, 1, mean)
  
  tmp <- tmp[rownames(tmp) %in% logFC_long[logFC_long$tissue == organ, ]$geneID, ]
  
  tmp <- merge(tmp, logFC_long[logFC_long$tissue == organ, ],
               by.x = 0, 
               by.y = "geneID", 
               all = TRUE)
  ymax <- max(tmp$Mmean, na.rm = TRUE) + 1
  xmax <- max(tmp$Fmean, na.rm = TRUE) + 1
  
  plot(tmp$Fmean, tmp$Mmean,
       xlab="",
       ylab="",
       #ylab = expression('log'[2] * '(Male Expression)'),
       #xlab = expression('log'[2] * '(Female Expression)'),
       #axes=FALSE,
       cex.main=2,
       #main = organ,
       col = bias_colors[tmp$bias],
       pch = 19,
       cex = ifelse(organ != "Gonad" & tmp$bias %in% c("MB", "FB"),
                    0.75,
                    0.25),
       ylim = c(0, 100),
       xlim = c(0, 100))
  
  
  
  #Make the axis lines longer and thicker
  axis(1, labels=seq(-2, xmax, 2), 
       line=NA, at=seq(-2, xmax, 2), 
       lwd=2,xlim=c(0, xmax),cex.axis=1.5,las=1)
  axis(2, labels=seq(-2, ymax, 2),
       line=NA, at=seq(-2, ymax, 2), 
       lwd=2, ylim=c(0,ymax),cex.axis=1.5,las=1)
  
  mtext(organ, cex=1.5,outer=FALSE, line=-1)
  
}

# add x axis label
mtext(expression('log'[2] * '(Female Expression)'), side = 1, cex=1.25, outer=TRUE, line=2)

# add y axis label
mtext(expression('log'[2] * '(Male Expression)'), side = 2, cex=1.25, outer=TRUE, line=2)

# add legend
spfTools::outer_legend("right",
                       legend=c("Female\nbiased",
                                "Male\nbiased",
                                "Unbiased"),
                       pt.bg=bias_colors,
                       pch=22,
                       bty='n',
                       ncol=1,
                       cex=2,
                       y.intersp = 1.5,
                       pt.cex=2)


```



## Categorising sex-specific genes
Now that we have investigated what the sex-bias is like across the different tissue types I want to dive further into genes with **sex-specific expression**. These are genes expressed only in one sex or the other. I am classifying a sex-specific genes within each tissue as ones where the expression of that gene is less than 10 for all of one sex and there is a median of $\ge 20$ in the other sex. 

Prior to running the for loop, any "outliers" that were determined by DESeq were removed from the pool of genes. Additionally, normalized counts were used for the classification of sex-specific genes as DESeq2 uses the moralized counts for all of the logFC/sexbias calculations.

```{r sex-specific-for-loop}
#Pulling out the geneIDs for genes that were categorized as "outliers" by DESeq2
#Calculating the Cooks threshold that would have been used
np <- length(resultsNames(ddsMF_SM_exp))
nsamp <- ncol(ddsMF_SM_exp)
cooks_thresh <- qf(0.99, df1 = np, df2 = nsamp-np)

out_ids <- names(mcols(ddsMF_SM_exp)$maxCooks[mcols(ddsMF_SM_exp)$maxCooks >
                                                cooks_thresh])

#Checking to make sure we have correct number of outliers
length(out_ids)

#Filtering out the dds dataset to remove the outliers determined by DESeq
ddsMF_SM_exp_filtered <- ddsMF_SM_exp[!(rownames(ddsMF_SM_exp) %in% out_ids), ]

#Create a vector with the different organ types
organs <- levels(colData(ddsMF_SM_exp_filtered)$organ)

#Create an empty list to store my datasets in
SM_sex_specific_genes <- list()

#Generate the for loop to identify MSpecific and FSpecific genes in each 
#organ based on Medians
for(organ in organs){
  
  #Male-Specific Genes
  ##Pull out all of the rows where fem count <=10 in every female sample
  fem0_organ_names <- which(rowSums(t(apply(counts(ddsMF_SM_exp_filtered, 
                                                   normalized = TRUE)[, ddsMF_SM_exp_filtered$sex == "F" 
                                                                      & 
                                                                        ddsMF_SM_exp_filtered$organ == organ], 
                                            1, 
                                            function(x) x <= 10)
                                      )
                                ) == ncol(counts(ddsMF_SM_exp_filtered, 
                                                 normalized = TRUE)[, ddsMF_SM_exp_filtered$sex == "F" &
                                                                      ddsMF_SM_exp_filtered$organ == organ])
                      )
  
  fem0_organ <- counts(ddsMF_SM_exp_filtered, 
                       normalized = TRUE)[rownames(counts(ddsMF_SM_exp_filtered, 
                                                          normalized = TRUE)) %in% 
                                            names(fem0_organ_names), 
                                          ddsMF_SM_exp_filtered$organ == organ]
                                                         
  
  ##Pull out rows where median of male count >=20 for that organ
  mal10_organ <- apply(counts(ddsMF_SM_exp_filtered, 
                              normalized = TRUE)[, ddsMF_SM_exp_filtered$sex == "M" 
                                                 & ddsMF_SM_exp_filtered$organ == organ], 
                       1,
                       function(x) median(x) >=20)
  
  ##Keep only the rows where all fem samples <=10 and the Male median>=20
  fem0_mal10_organ <- fem0_organ[rownames(fem0_organ) %in% 
                                   names(mal10_organ[mal10_organ == TRUE]),
                                 ]
  
  ##Create a new object with a name based on the organ type
  organ_malsp <- sub("$", "_male_specific", organ)
  SM_sex_specific_genes[[organ_malsp]] <- fem0_mal10_organ
  
  
  
  #Female-Specific Genes
  ##Pull out all of the rows where male count <=10 in every male sample
  mal0_organ_names <- which(rowSums(t(apply(counts(ddsMF_SM_exp_filtered, 
                                                   normalized = TRUE)[, ddsMF_SM_exp_filtered$sex == "M" 
                                                                      & 
                                                                        ddsMF_SM_exp_filtered$organ == organ],
                                            1, 
                                            function(x) x <= 10)
                                      )
                                    ) == ncol(counts(ddsMF_SM_exp_filtered, 
                                                     normalized = TRUE)[, ddsMF_SM_exp_filtered$sex == "M" & 
                                                                          ddsMF_SM_exp_filtered$organ == organ])
                            )
  
  mal0_organ <- counts(ddsMF_SM_exp_filtered, 
                       normalized = TRUE)[rownames(counts(ddsMF_SM_exp_filtered, 
                                                          normalized = TRUE)) %in% 
                                            names(mal0_organ_names), 
                                          ddsMF_SM_exp_filtered$organ == organ]
  
  ##Pull out rows where median of female count >=20 for that organ
  fem10_organ <- apply(counts(ddsMF_SM_exp_filtered, 
                              normalized = TRUE)[, ddsMF_SM_exp_filtered$sex == "F" &
                                                   ddsMF_SM_exp_filtered$organ == organ],
                       1,
                       function(x) median(x) >=20)
  
  #Keep only the rows where male=0 and the fem median>=10
  mal0_fem10_organ <- mal0_organ[rownames(mal0_organ) %in% 
                                   names(fem10_organ[fem10_organ == TRUE]),
                                 ]
  
  # Create a new object with a name based on the organ type
  organ_femsp <- sub("$", "_female_specific", organ)
  SM_sex_specific_genes[[organ_femsp]] <- mal0_fem10_organ
  
} 
```

### Investigating the results of the sex-specific subsetting
Let's now take a look at how many sex-specific genes we have in each tissue type:
```{r sex-specific-counts, echo=FALSE, message=FALSE, warning=FALSE}
sex_specific_counts <- data.frame(tissue = rep(c("Gill", "Gonad", "Liver"), 2),
                                  sex = c(rep("M", 3),
                                          rep("F", 3)),
                                  num_genes = c(nrow(SM_sex_specific_genes$Gill_male_specific),
                                                nrow(SM_sex_specific_genes$Gonad_male_specific),
                                                nrow(SM_sex_specific_genes$Liver_male_specific),
                                                nrow(SM_sex_specific_genes$Gill_female_specific),
                                                nrow(SM_sex_specific_genes$Gonad_female_specific),
                                                nrow(SM_sex_specific_genes$Liver_female_specific))
                                  )

sex_specific_counts %>% 
  group_by(tissue, sex) %>% 
  summarise("Number of Sex-specific Genes" = sum(num_genes)) %>%
  kable(caption = "Sex-specific Genes", format = "html",
        col.names = c("Tissue Type", "Sex", "Number of Sex-specific Genes")) %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F, position = "left")
```

To get a better idea what is going on between the sex-specific genes vs the sex-biased genes I first determined how many overlaps there were between our sex-biased and sex-specific genes and then used plotCounts to plot the counts for each gene separately. 

In the **Male Gills** there are `r nrow(SM_sex_specific_genes$Gill_male_specific[rownames(SM_sex_specific_genes$Gill_male_specific) %in% rownames(gill_mal_biased),])` genes that overlap between sex-biased and sex-specific, there are `r nrow(SM_sex_specific_genes$Gonad_male_specific[rownames(SM_sex_specific_genes$Gonad_male_specific) %in% rownames(gonad_mal_biased),])` overlapping in the **gonads** and `r nrow(SM_sex_specific_genes$Liver_male_specific[rownames(SM_sex_specific_genes$Liver_male_specific) %in% rownames(liver_mal_biased),])` overlapping in the **liver**. For females we have `r nrow(SM_sex_specific_genes$Gill_female_specific[rownames(SM_sex_specific_genes$Gill_female_specific) %in% rownames(gill_fem_biased),])` overlapping sex-specific and sex-biased genes in the **gills**, `r nrow(SM_sex_specific_genes$Gonad_female_specific[rownames(SM_sex_specific_genes$Gonad_female_specific) %in% rownames(gonad_fem_biased),])` genes in the **gonads**, and `r nrow(SM_sex_specific_genes$Liver_female_specific[rownames(SM_sex_specific_genes$Liver_female_specific) %in% rownames(liver_fem_biased),])` genes in the **liver**.

```{r gills-sex-specific, eval = FALSE}
## Female sex-specific
fem_sp_geneIDs <- rownames(SM_sex_specific_genes$Gill_female_specific)

##Set the plotting window
par(mfrow=c(1,1))

for(gene in fem_sp_geneIDs){
  
  plotCounts(ddsMF_SM[,ddsMF_SM$organ == 'Gill'], gene=gene, intgroup="sex")
  
}

## Female sex-biased
gill_fem_biased_geneIDS <- rownames(gill_fem_biased)

##Set the plotting window
par(mfrow=c(1,1))

for(gene in gill_fem_biased_geneIDS){
  
  plotCounts(ddsMF_SM[,ddsMF_SM$organ == 'Gill'], gene=gene, intgroup="sex", col= 'salmon', pch = 19)
  
}
```

### Pulling out the gene IDs for all of the sex-specific genes
```{r, eval=FALSE}
#Run once to create the file and then commented out
write.table(cbind(rownames(SM_sex_specific_genes$Gill_male_specific)),
          'data/SM_malG_specific_TRgenes.txt',
         sep = "",
        quote=FALSE,
       row.names = FALSE,
      col.names = FALSE)
write.table(cbind(rownames(SM_sex_specific_genes$Gill_female_specific)),
          'data/SM_femG_specific_TRgenes.txt',
         sep = "",
        quote=FALSE,
       row.names = FALSE,
      col.names = FALSE)
write.table(cbind(rownames(SM_sex_specific_genes$Gonad_male_specific)),
          'data/SM_malGon_specific_TRgenes.txt',
         sep = "",
        quote=FALSE,
       row.names = FALSE,
      col.names = FALSE)
write.table(cbind(rownames(SM_sex_specific_genes$Gonad_female_specific)),
          'data/SM_femGon_specific_TRgenes.txt',
         sep = "",
        quote=FALSE,
       row.names = FALSE,
      col.names = FALSE)
write.table(cbind(rownames(SM_sex_specific_genes$Liver_male_specific)),
          'data/SM_malL_specific_TRgenes.txt',
         sep = "",
        quote=FALSE,
       row.names = FALSE,
      col.names = FALSE)
write.table(cbind(rownames(SM_sex_specific_genes$Liver_female_specific)),
          'data/SM_femL_specific_TRgenes.txt',
         sep = "",
        quote=FALSE,
       row.names = FALSE,
      col.names = FALSE)
```


## Creating categories and binning the sex-biased genes based on degree of logFC
This binned was only done on the sex-biased genes, will not have a category for the unbiased genes. The cutoffs for the different groups are as follows:

    1. Low - biased = LFC 2 - 3
    2. Medium - biased = LFC 3 - 4
    3. High - biased = LFC 4 - 8
    4. Extreme sex bias = LFC > 8

Counts of the sex-specific genes will be added on separately as they were not classified based on fold-change but rather a presence in one sex and an absence in the other sex.

```{r bin-sex-bias}
#Make a vector that contains all of the groupings
biased_bins <- c("Unbiased", "Low", "Med", "High", "Extreme", "Sex-specific")
```


```{r bin-sex-bias-cat, eval=FALSE}
#Create a new column in the dataset and use ifelse statements to set the category limits
#abs(logFC) was used to account for the fem-biased genes possessing negative values
logFC_long$bias_cat <- ifelse(logFC_long$bias == "UB",
                              biased_bins[1],
                              ifelse(abs(logFC_long$logFC) >= 2 & abs(logFC_long$logFC) < 3,
                                     biased_bins[2],
                                     ifelse(abs(logFC_long$logFC) >= 3 & abs(logFC_long$logFC) < 5,
                                            biased_bins[3],
                                            ifelse(abs(logFC_long$logFC) >= 5 & abs(logFC_long$logFC) < 10,
                                                   biased_bins[4],
                                                   biased_bins[5])
                                            )
                                     )
                              )

#Making sure the genes we categorized as sex-specific are labeled as sex-specific for their 
#bias cat. in the dataset
organs <- c("Gill", "Gill", "Gonad", "Gonad", "Liver", "Liver")
bias <- c("MB", "FB", "MB", "FB", "MB", "FB")

for(i in 1:length(SM_sex_specific_genes)){
  tmp <- SM_sex_specific_genes[[i]]
  tmp <- as.data.frame(tmp)
  tmp$geneID <- rownames(tmp)
  
  for(j in 1:nrow(tmp)){
    
    one_gene <- tmp[j, ]
    
    if(one_gene[["geneID"]] %in%
       logFC_long[logFC_long$tissue == organs[[i]] &
                  logFC_long$bias == bias[[i]],"geneID"]){
       
      logFC_long[logFC_long$geneID == one_gene[["geneID"]] &
                   logFC_long$tissue == organs[[i]] &
                   logFC_long$bias == bias[[i]],
                 "bias_cat"] <- "Sex-specific"
    }else{
      
      one_gene_dat <- data.frame(matrix(ncol= ncol(logFC_long),
                                        nrow=1))
      colnames(one_gene_dat) <- colnames(logFC_long)
      
      one_gene_dat$tissue <- organs[[i]]
      one_gene_dat$geneID <- one_gene[["geneID"]]
      one_gene_dat$bias <- bias[[i]]
      one_gene_dat$bias_cat <- "Sex-specific"
      rownames(one_gene_dat) <- NULL
      
      logFC_long <- rbind(one_gene_dat, logFC_long)
      
      rownames(logFC_long) <- NULL
    }
  }
 }

#Create  subset of our long dataset that does not include the non-biased genes
logFC_long_noNB <- logFC_long[logFC_long$bias_cat != "Unbiased",]

#Make a table to count the number of genes in each category for each organ
bias_cat_table <- table(logFC_long_noNB$bias, logFC_long_noNB$bias_cat, logFC_long_noNB$tissue)

#Add the counts of sex-specific genes to the table
bias_cat_gill <- bias_cat_table[,, "Gill"]
write.table(bias_cat_gill, "data/bias_cat_gills.txt", row.names = TRUE)

bias_cat_gonad <- bias_cat_table[,, "Gonad"]
write.table(bias_cat_gonad, "data/bias_cat_gonad.txt", row.names = TRUE)

bias_cat_liver <- bias_cat_table[,, "Liver"]
write.table(bias_cat_liver, "data/bias_cat_liver.txt", row.names = TRUE)
```
```{r}
bias_cat_gill<-read.table("data/bias_cat_gills.txt",
                          row.names = 1)
bias_cat_gonad<-read.table("data/bias_cat_gonad.txt",
                          row.names = 1)
bias_cat_liver<-read.table("data/bias_cat_liver.txt",
                          row.names = 1)
```


```{r bin-sex-bias}
bins_to_use<-biased_bins[biased_bins!="Unbiased"]
bins_to_use[bins_to_use == "Sex-specific"]<-"Sex.specific"
#Plot the counts for each tissue type
png("figs/barplots_S1.png",height=5,width=15,units = "in", res=300)
par(mfrow=c(1, 3))
barplot(as.matrix(bias_cat_gill[,bins_to_use]), beside = TRUE, ylim = c(0, max(bias_cat_gill) + 1), col = c("#7fc97f", "#beaed4"),
        main = "Gill", ylab = "Number of Genes", xlab = "Bias Category")
barplot(as.matrix(bias_cat_gonad[,bins_to_use]), beside = TRUE, ylim = c(0, max(bias_cat_gonad) + 1), col = c("#7fc97f", "#beaed4"),
        main = "Gonad")
barplot(as.matrix(bias_cat_liver[,bins_to_use]), beside = TRUE, ylim = c(0, max(bias_cat_liver) + 1), col = c("#7fc97f", "#beaed4"),
        main = "Liver")
legend("topright", legend = c("Female-biased", "Male-biased"), fill = c("#7fc97f", "#beaed4"))
dev.off()
```




# Patterns across genes

## Creating an Upset Plot
```{r upset-plot-mal, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in males."}
#Male-biased genes
listInputMB <- list("Testis" = rownames(gonad_mal_biased),
                  "Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased)
                  )
upset(fromList(listInputMB),
      mainbar.y.label = "Number of Shared Male-Biased Genes",
      sets.x.label = "Total Number of Male-Biased Genes",
      point.size = 3)
```

```{r upset-plot-fem, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in females."}
#Female-biased genes
listInputFB <- list("Ovaries" = rownames(gonad_fem_biased),
                  "Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased)
                  )
upset(fromList(listInputFB),
      mainbar.y.label = "Number of Shared Female-Biased Genes",
      sets.x.label = "Total Number of Female-Biased Genes",
      point.size = 3)
```

```{r exportUpsetData, eval=FALSE}
#Exporting data used to make the Upset plot (run once and them commented out)
write.table(as.data.frame(gonad_mal_biased), "data/gonad_mal_biased.txt", 
           row.names = TRUE)
write.table(as.data.frame(gonad_fem_biased), "data/gonad_fem_biased.txt", 
           row.names = TRUE)
write.table(as.data.frame(gill_mal_biased), "data/gill_mal_biased.txt", 
           row.names = TRUE)
write.table(as.data.frame(gill_fem_biased), "data/gill_fem_biased.txt", 
           row.names = TRUE)
write.table(as.data.frame(liver_mal_biased), "data/liver_mal_biased.txt", 
           row.names = TRUE)
write.table(as.data.frame(liver_fem_biased), "data/liver_fem_biased.txt", 
           row.names = TRUE)
```

```{r upset-plot-all, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs across both sexes."}


#Checking there is no overlap between the two sexes
listInputall <- list("Testes" = rownames(gonad_mal_biased),
                  "Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased),
                  "Ovaries" = rownames(gonad_fem_biased),
                  "Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased))

# Define the desired order of sets
desired_order <- c("Male Gill", "Female Gill", "Male Liver", "Female Liver", "Testes", "Ovaries")

# Create the UpSet object with custom set order
pdf("figs/upset.pdf",height=4,width=10)
upset(fromList(listInputall),
      sets = desired_order,
      mainbar.y.label = "Number of Shared Sex-Biased Genes",
      sets.x.label = "Total Number of Sex-Biased Genes",
      point.size = 3,
      nsets = 6,
      nintersects = NA,
      keep.order = TRUE,
      order.by = "degree",
      sets.bar.color  = rev(c(organ_cols["Gonad"],
                          organ_cols["Gonad"],
                          organ_cols["Liver"],
                          organ_cols["Liver"],
                          organ_cols["Gill"],
                          organ_cols["Gill"])),
      shade.color = sex_cols["M"],
      main.bar.color=c(rep("gray23",11),
                       c(organ_cols["Gonad"],
                          organ_cols["Gonad"],
                          organ_cols["Liver"],
                          organ_cols["Liver"],
                          organ_cols["Gill"],
                          organ_cols["Gill"]))
)
dev.off()
```

In both males and females the highest number of shared sex-biased genes is found between the gonads and the liver (Fig. \@ref(fig:upset-plot-mal), \@ref(fig:upset-plot-fem)). There are no genes that are shared across the same organ between males and females (e.g., biased in males and females in the liver) which is good (Fig. \@ref(fig:upset-plot-all)). Two sex-biased genes are female-biased in all organs and three genes are male-biased in all three organs (Fig. \@ref(fig:upset-plot-mal), \@ref(fig:upset-plot-fem)).

# Gene Ontology Analysis

We wanted to figure out two things:   

  1) What are the genes that are sex-biased within each tissue type   
  2) What are the GO terms for those genes.   

To accomplish this I first had to pull out the sequences that corresponded to the Trinity gene IDS for sex-biased genes (above). Once those sequences were fetched they were blasted against the _Stigmatopora nigra_ genome, implemented on the RCH.

```{bash, eval=FALSE}
#!/bin/bash

#
#SBATCH --job-name=diamondblastx
#
#SBATCH --ntasks=1
#SBATCH --time=25:00:00
#SBATCH --mem-per-cpu=3GB
#SBATCH --cpus-per-task=8
#SBATCH --error=job_%x_%j.err
#SBATCH --output=job_%x_%j.out


module load prefix/latest
module load DIAMOND/2.1.11-GCC-13.3.0

cd /home/spf50/S_macropterygia/

# process highly tissue specific and sex biased genes
./subset_fasta_file -c -f SMA_supertranscriptome.fasta -l sex_biased_trinity_genes.txt
mv SMA_supertranscriptome.subset..fasta sex_biased_genes.fasta


diamond blastx --threads 8 --db s_nigra --out sex_biased_genes.dmdblastx --query sex_biased_genes.fasta --outfmt 6
```





### Filtering Blast results
Because of the way BLAST was run in the RCC, we can have multiple hits for one gene, I want to filter that out so that there is only one hit kept for each gene.


```{r}
filter_blast<-function(data_frame){
 
  #Pull out the Unique Trinity gene IDs
  uniqueID <- unique(data_frame[,1])
  
  #Create an empty dataframe to store intermediate results in
  output <- data.frame(matrix(data = NA, ncol = ncol(data_frame), nrow = 0))
  colnames(output) <-  colnames(data_frame)
  
  #Generate a for loop that pulls out the lowest e-value + highest % identity for each gene
  for(gene in uniqueID){
    
    #Subset the dataset for each Trinity gene
    this_trin_gene<- data_frame[which(data_frame[,1] == gene),]
    #Pull out gene with smallest e-value
    uni_gene_subset_lowe <- this_trin_gene[which(this_trin_gene$evalue == min(this_trin_gene$evalue)),]
    #In case mult. genes have same e-value, pull out gene with highest % identity
    uni_gene_subset_lowe_highpid <- uni_gene_subset_lowe[which(uni_gene_subset_lowe$pident == max(uni_gene_subset_lowe$pident)),]
    
    # keep only one of multiple identical rows
    uni_gene_subset_lowe_highpid <- unique(uni_gene_subset_lowe_highpid)
    
    #Add the final gene into the empty dataframe from above
    output <- rbind(output, uni_gene_subset_lowe_highpid)
  }
  
  return(output)
}
parse_gff<-function(filename){
  gff <- read.delim(filename,
                  comment.char = "#", 
                  header = FALSE)
  colnames(gff) <- c("chromosome","method","type","start","stop","x","strand","y","description")
  return(gff)
}
```



```{r, eval=FALSE}
gff<-parse_gff("stigmatopora_nigra/genomic.gff")
annotations<-read.delim("stigmatopora_nigra/GCF_051989575.1-RS_2025_11_gene_ontology.gaf",
                        comment.char="!", 
                        header = FALSE)
colnames(annotations)<-c(
  "DB","GeneID","Symbol","Qualifier","GO_ID","Reference","Evidence_Code","With_From","Aspect","Gene_Name","Gene_Synonym","Type","Taxon","Date","Assigned_By","Annot_Ext","Gene_Product_Form_ID"
)
library(GO.db)
# get keys for selection
go <- keys(GO.db, keytype="GOID")
goids <- select(GO.db,
                columns=c("GOID","TERM","ONTOLOGY","DEFINITION"),
                keys=go, 
                keytype="GOID")
annotations<-merge(annotations,
                   goids,
                   by.x="GO_ID",
                   by.y="GOID",
                   all.x=TRUE)
write.csv(annotations,
          "stigmatopora_nigra/go_annotations.csv",
          row.names=FALSE)
```

```{r filterBlast, eval=FALSE}
blast_filtered<-filter_blast(blast_output_all)

write.csv(all_blast_filtered,"sex_bias_blast_filtered.csv")
```


```{r blastAnnotate, eval=FALSE}
# run on RCH
all_blast_filtered <- read.csv("data/sex_bias_blast_filtered.csv", row.names = 1)
annotations<-read.csv("stigmatopora_nigra/go_annotations.csv")

all_blast_info<-as.data.frame(do.call(rbind, lapply(all_blast_filtered$sseqid, function(sid,gff){
  this_gff<-gff[grep(sid, gff$description),]
  product<-paste(unique(gsub("^.*product=(.*)\\;protein_id.*$","\\1",this_gff$description)),collapse=";")
  gene_id<-paste(unique(gsub("^.*GeneID\\:(\\d+)\\,.*$","\\1",this_gff$description)),collapse=";")
  if(length(product)<1){ browser() }
  return(c(product, gene_id))
},gff=gff)))
all_blast_info$V2<-round(as.numeric(all_blast_info$V2),2)

all_blast_filtered$GeneID<-all_blast_info$V2
all_blast_filtered$product<-all_blast_info$V1

# get the GO annotations
go_anns<-apply(all_blast_filtered, 1, function(this_gene, annotations){
  
  this_ann<-annotations[annotations$GeneID==as.numeric(this_gene["GeneID"]),]
  
  return(paste(paste(this_ann$Qualifier, this_ann$TERM),collapse=";"))
  
}, annotations=annotations)
all_blast_filtered$GO<-go_anns
write.csv(all_blast_filtered,
          "data/sex_bias_blast_filtered_annotated.csv",
          row.names=FALSE)
```




### Merging the BLAST filtered Data frames with the sex-biased information

Not all of the trinity genes had a BLAST hit come back, but I still want to include what those genes are in the dataset. To do this I am merging all of the genes with the filtered BLAST dataset and if there are genes that did not have any hits (i.e. are not in the filtered BLAST database) then they will receive an "NA".

```{r merge-SB-blast}
all_blast_filtered<-read.csv("data/sex_bias_blast_filtered_annotated.csv")
# get the sex bias and tau info
logFC_long_all <- read.csv("data/logFC_long_all.csv")


blast_merged <- merge(logFC_long_all, all_blast_filtered, by.x = "geneID", by.y="qseqid", all = TRUE)


#write once then comment out
write.csv(as.data.frame(blast_merged), 
          "data/SuppTable_SM_blast_annotated.csv", 
          row.names = FALSE)
```


