---
title: "WGCNA and cluster evolution in _S. macropterygia_"
author: Sarah P. Flanagan
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: yes
    toc: true
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
```
```{r setdir}
knitr::opts_knit$set(root.dir='../',
                     fig_path="../figs/")
```


``` {r library, include = TRUE, message = FALSE, warning = FALSE}
#This is a comprehensive list of all the libraries used in this document
library(knitr)
library(kableExtra)
library(spfTools)
library(WGCNA)
library(vioplot)
library(multcompView)
library(RColorBrewer)
sex_cols <- c("F" = "#7fc97f", "M" = "#beaed4")
```

``` {r traitData, eval=FALSE}
#The samples file generated for tximport
samples <- read.table("data/SM_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)
trait_data<-data.frame(
  female=ifelse(samples$sex=="F",1,0),
  gill=ifelse(samples$organ=="Gill",1,0),
  gonad=ifelse(samples$organ=="Gonad",1,0),
  liver=ifelse(samples$organ=="Liver",1,0)
)
rownames(trait_data)<-samples$ID
write.csv(trait_data,"data/trait_data_for_WGCNA.csv",row.names = TRUE)
```
```{r readData}
trait_data<-read.csv("data/trait_data_for_WGCNA.csv",
                     row.names = 1)
expression_data<-read.csv("data/expression_data_for_WGCNA.csv", row.names = 1)
logFC_long <- read.csv("data/logFC_long_sexbias.csv")
```


## WGCNA

I'm going to subset the dataset to only include the genes that showed some form of bias -- otherwise we have >90,000 genes and that's difficult to work with.

```{r}
expression_data_subset<-expression_data[,logFC_long$geneID[logFC_long$bias!="UB"]]
# make sure we have trait data only for the individuals in the analysis
trait_data<-trait_data[rownames(trait_data) %in% rownames(expression_data),]
```


The first step is to check that the dataset does not contain any outliers or problematic samples (it shouldn't, since we've already cleaned it up for differential expression analysis). Note that WGCNA needs each column to be the genes -- so it's the transposed version of the DESeq dataset.

```{r}
# following the approaches from:
## https://bioinformaticsworkbook.org/tutorials/wgcna.html#gsc.tab=0
## https://cloud.wikis.utexas.edu/wiki/spaces/bioiteam/pages/47714689/Clustering+using+WGCNA
gsg <-goodSamplesGenes(expression_data_subset)
summary(gsg)
gsg$allOK
```

Since everything is ok, we can move on to picking a threshold. 


```{r}
spt <- pickSoftThreshold(expression_data_subset)
```

With this threshold, we want to maximize the $R^2$ value and minimizing mean connectivity.

```{r}
par(mar=c(4,4.5,1,1))
plot(spt$fitIndices[,1],spt$fitIndices[,2],
xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
main = paste("Scale independence"))
text(spt$fitIndices[,1],spt$fitIndices[,2],col="red")
abline(h=0.80,col="red")
```


```{r}
par(mar=c(4,4.5,1,1))
plot(spt$fitIndices[,1], spt$fitIndices[,5],
xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
main = paste("Mean connectivity"))
text(spt$fitIndices[,1], spt$fitIndices[,5], labels= spt$fitIndices[,1],col="red")
```

I think we want to go with 6. 

```{r}
soft_power<-6
# this creates a big matrix
adjacency <- adjacency(expression_data_subset, power = soft_power)
# get the similarity
TOM <- TOMsimilarity(adjacency)
# turn it into dissimilarity
TOM.dissimilarity <- 1-TOM


#creating the dendrogram 
geneTree <- hclust(as.dist(TOM.dissimilarity), 
                   method = "average") 

# identify modules
Modules <- cutreeDynamic(dendro = geneTree, 
                         distM = TOM.dissimilarity, 
                         deepSplit = 2,
                         pamRespectsDendro = FALSE, 
                         minClusterSize = 30)
# assign each module number a color
ModuleColors <- labels2colors(Modules) 
table(ModuleColors)
```


```{r}
# plotting the dendrogram
sizeGrWindow(12,9)

plotDendroAndColors(geneTree, 
                    ModuleColors,"Module",
                    dendroLabels = FALSE, 
                    hang = 0.03,
                    addGuide = TRUE, 
                    guideHang = 0.05,
                    main = "Gene dendrogram and module colors")
```

Now we need to correlate expression modules with traits. 


```{r}
MElist <- moduleEigengenes(expression_data_subset,
                           colors = ModuleColors) 
MEs0 <- MElist$eigengenes 

MEs <- orderMEs(MEs0)
moduleTraitCor <- WGCNA::cor(MEs, trait_data, use= "p")
moduleTraitPvalue <- corPvalueStudent(moduleTraitCor,
                                      nrow(trait_data))
```

```{r}
#Print correlation heatmap between modules and traits
textMatrix<- paste(signif(moduleTraitCor, 2), "\n(", 
						signif(moduleTraitPvalue, 1), ")", sep= "")
dim(textMatrix)<- dim(moduleTraitCor)

# get module counts in correct order
moduleCounts<-table(ModuleColors)
moduleCounts<-moduleCounts[gsub("ME","",dimnames(moduleTraitCor)[[1]])]
```


```{r}
par(mar= c(6, 8.5, 3, 3))
#display the correlation values with a heatmap plot

labeledHeatmap(Matrix= moduleTraitCor, 
			xLabels= names(trait_data), 
			yLabels= names(MEs), 
			ySymbols= rep("",nrow(moduleTraitCor)), 
			colorLabels= TRUE, 
			colors= colorRampPalette(c("#af8dc3","#f7f7f7","#7fbf7b"))(50), 
			textMatrix= textMatrix, 
			setStdMargins= FALSE, 
			cex.text= 0.75, 
			zlim= c(-1,1), 
			main="")

```

```{r moduleSexHeatmap, fig.width=350, fig.height=550}
pdf("figs/moduleSexHeatmap.pdf",height = 5.5, width=3.5)
par(mar= c(3, 5, 1, 5))
#display the correlation values with a heatmap plot

labeledHeatmap(Matrix= as.matrix(moduleTraitCor[,1]), 
			xLabels= "", 
			yLabels= names(MEs), 
			ySymbols= paste0("n=",moduleCounts), 
			colorLabels= TRUE, 
			colors= colorRampPalette(c("#af8dc3","#f7f7f7","#7fbf7b"))(50), 
			textMatrix= textMatrix[,1], 
			setStdMargins= FALSE, 
			cex.text= 1, 
			zlim= c(-1,1),
			main="")
text(x=c(1.4,1.4),
     y=c(0.965, 0.05),
     labels = c("female","male"),
     xpd=TRUE)
dev.off()
```

```{r moduleOrganHeatmap, fig.width=500, fig.height=550}
pdf("figs/moduleOrganHeatmap.pdf", height=5.5, width=5.5)
par(mar= c(3,5,1,3))
#display the correlation values with a heatmap plot

labeledHeatmap(Matrix= moduleTraitCor[,-1], 
			xLabels= names(trait_data[,-1]), 
			yLabels= names(MEs), 
			ySymbols= paste0("n=",moduleCounts), 
			colorLabels= TRUE, 
			colors= colorRampPalette(c("#edf8b1","#7fcdbb","#2c7fb8"))(50), 
			textMatrix= textMatrix[,-1], 
			setStdMargins= FALSE, 
			cex.text= 1, 
			zlim= c(-1,1), 
			main="")
dev.off()
```

## Compare to other values, logFC and tau

```{r}
logFC_long$module<-NA


for(col in unique(ModuleColors)){
  genes<-colnames(expression_data_subset)[ModuleColors==col]

  logFC_long$module[logFC_long$geneID%in%genes]<-col
  
}

logFC_long$module<-as.factor(logFC_long$module)
```


```{r}
mod_lm<-lm(rank(logFC_long$logFC)~logFC_long$tissue *logFC_long$module)
anova(mod_lm)
mod_labs<-multcompLetters4(aov(mod_lm), 
                           TukeyHSD(aov(mod_lm)))
mod_labs
```
```{r}
par(mfrow=c(3,1))
plot(TukeyHSD(aov(mod_lm)))
```
```{r}
tukey_tab<-TukeyHSD(aov(mod_lm))
tukey_tab[[3]][grep("black",rownames(tukey_tab[[3]])),]
```


```{r}
module_medians<-tapply(logFC_long$logFC,
       paste(logFC_long$tissue, logFC_long$module),
       median)
module_medians[unlist(lapply(c("black","blue","red","turquoise"),
                      grep,
                      x=names(module_medians)))]

lapply(names(table(ModuleColors))[!names(table(ModuleColors)) %in% c("black","blue","red","turquoise")],
       function(col){
         max(module_medians[grep(col, names(module_medians))])
       })
```


```{r, fig.width=10, fig.height=4.5}
pdf("figs/WGCNA_logFC.pdf",height = 4,width = 10)
plot_cols<-unlist(lapply(levels(as.factor(ModuleColors)),rep,times=3))

par(mar=c(4,4,1,1))
plot(x=0:34,
     y=seq(-30,50,length.out=35),
      xaxt='n',
        xlab="",
        ylab="logFC",
     type='n',
     las=1,
     bty='n')
abline(h=0,lty=1,lwd=1)
abline(h=2,lty=2,lwd=2,col="grey")
abline(h=-2,lty=2,lwd=2,col="grey")
vioplot(logFC_long$logFC~logFC_long$tissue + logFC_long$module,
        col=plot_cols,
        add=TRUE,
        bty='n',
        xaxt='n',
        axes=FALSE,
        border="black",
        rectCol=plot_cols,
        lineCol=plot_cols,
        colMed="white",
        plotCentre=TRUE)
points(x=13:15,
       y=logFC_long$logFC[which(logFC_long$module=="grey")],
       col="black",
       bg="grey",
       pch=23,
       cex=1.25)
axis(1,at=1:33,labels=rep(c("Gill","Gonad","Liver"),11),las=2)
axis(2,
     at=c(-30,30),
     c("female-biased","male-biased"),
     tick = 0,pos = -2.5)  
     
    

# get the correlation coefficients
cor_vals<-signif(moduleTraitCor, 2)[,1]
names(cor_vals)<-gsub("ME","",names(cor_vals))
cor_vals<-cor_vals[levels(logFC_long$module)]

sig_sex_cor<-which(moduleTraitPvalue[,1]<=0.05)
sig_x<-which(plot_cols %in% gsub("ME","",names(sig_sex_cor)))
sig_x<-sig_x[seq(2,length(sig_x),by=3)]
non_sex_cor<-which(moduleTraitPvalue[,1]>0.05)
non_x<-which(plot_cols %in% gsub("ME","",names(non_sex_cor)))
non_x<-non_x[seq(2,length(non_x),by=3)]
text(x=sig_x,
     y=rep(-30,length(sig_x)),
     labels=cor_vals[names(cor_vals) %in% gsub("ME","",names(sig_sex_cor))],
     font = 2)

dev.off()
```


### Compare to tau


```{r read-logFC-long}
logFC_tau_all<-read.csv("data/logFC_tau_all.csv")
logFC_tau_all$tissue<-factor(logFC_tau_all$tissue, levels=c("Gill","Liver","Gonad"))
logFC_tau_all$module<-NA


for(col in unique(ModuleColors)){
  genes<-colnames(expression_data_subset)[ModuleColors==col]

  logFC_tau_all$module[logFC_tau_all$geneID%in%genes]<-col
  
}

logFC_tau_all$module<-as.factor(logFC_tau_all$module)

logFC_tau_all$grp<-paste0(logFC_tau_all$module,"_",logFC_tau_all$tissue)
logFC_tau_all$grp<-as.factor(logFC_tau_all$grp)
```

```{r}
tau_col_dat<-logFC_tau_all[which(logFC_tau_all$module != "grey"),]
tau_col_dat$module <- factor(tau_col_dat$module)
write.csv(tau_col_dat, "data/tau_col_dat.csv", row.names = FALSE, quote=FALSE)
```
```{r}
tau_col_dat<-read.csv("data/tau_col_dat.csv")
```


```{r}

library(betareg)
library(emmeans)
library(multcomp)
tau_mod<-betareg(tau~tissue *module, dat=tau_col_dat)
```
```{r}
par(mfrow=c(3,2))
plot(tau_mod)
```
```{r}
joint_tests(tau_mod)

emm <- emmeans(tau_mod, ~ tissue * module, type="response")
pairwise_emm<-pairs(emm, adjust = "tukey")
tau_mod_lab<-multcomp::cld(emm, Letters = letters)
write.csv(tau_mod_lab, "data/tau_by_module_tissue_pairwise_letters.csv")
```

Call:
betareg(formula = tau ~ tissue * module, data = tau_col_dat)

Randomized quantile residuals:
    Min      1Q  Median      3Q     Max 
-3.6939 -0.6212  0.0656  0.7097  3.3496 

Coefficients (mu model with logit link):
                              Estimate Std. Error z value Pr(>|z|)
(Intercept)                  9.990e-02  4.102e-02   2.435 0.014872
tissueGonad                 -1.956e-01  5.312e-02  -3.682 0.000231
tissueLiver                  2.758e-08  5.801e-02   0.000 1.000000
moduleblue                  -4.816e-03  4.463e-02  -0.108 0.914060
modulebrown                 -5.321e-01  4.871e-02 -10.923  < 2e-16
modulegreen                 -7.044e-01  5.327e-02 -13.224  < 2e-16
modulemagenta               -1.061e+00  7.597e-02 -13.971  < 2e-16
modulepink                  -2.611e-01  6.483e-02  -4.027 5.64e-05
modulepurple                -1.063e+00  7.609e-02 -13.966  < 2e-16
modulered                   -1.049e+00  6.047e-02 -17.354  < 2e-16
moduleturquoise             -7.815e-01  4.347e-02 -17.979  < 2e-16
moduleyellow                -3.039e-01  5.071e-02  -5.992 2.08e-09
tissueGonad:moduleblue      -4.289e-02  5.737e-02  -0.748 0.454671
tissueLiver:moduleblue      -5.577e-08  6.311e-02   0.000 0.999999
tissueGonad:modulebrown      6.778e-02  6.403e-02   1.059 0.289812
tissueLiver:modulebrown      1.542e-08  6.872e-02   0.000 1.000000
tissueGonad:modulegreen      1.647e-01  6.983e-02   2.358 0.018370
tissueLiver:modulegreen     -1.348e-07  7.520e-02   0.000 0.999999
tissueGonad:modulemagenta    1.145e-01  9.843e-02   1.163 0.244739
tissueLiver:modulemagenta   -4.808e-08  1.073e-01   0.000 1.000000
tissueGonad:modulepink       7.621e-02  8.691e-02   0.877 0.380576
tissueLiver:modulepink      -2.431e-09  9.168e-02   0.000 1.000000
tissueGonad:modulepurple     1.530e-01  1.003e-01   1.525 0.127218
tissueLiver:modulepurple    -5.199e-08  1.073e-01   0.000 1.000000
tissueGonad:modulered        9.594e-02  7.680e-02   1.249 0.211613
tissueLiver:modulered       -9.952e-08  8.527e-02   0.000 0.999999
tissueGonad:moduleturquoise  9.744e-02  5.619e-02   1.734 0.082914
tissueLiver:moduleturquoise -7.446e-08  6.129e-02   0.000 0.999999
tissueGonad:moduleyellow     1.161e-01  6.663e-02   1.743 0.081356
tissueLiver:moduleyellow    -7.864e-08  7.170e-02   0.000 0.999999
                               
(Intercept)                 *  
tissueGonad                 ***
tissueLiver                    
moduleblue                     
modulebrown                 ***
modulegreen                 ***
modulemagenta               ***
modulepink                  ***
modulepurple                ***
modulered                   ***
moduleturquoise             ***
moduleyellow                ***
tissueGonad:moduleblue         
tissueLiver:moduleblue         
tissueGonad:modulebrown        
tissueLiver:modulebrown        
tissueGonad:modulegreen     *  
tissueLiver:modulegreen        
tissueGonad:modulemagenta      
tissueLiver:modulemagenta      
tissueGonad:modulepink         
tissueLiver:modulepink         
tissueGonad:modulepurple       
tissueLiver:modulepurple       
tissueGonad:modulered          
tissueLiver:modulered          
tissueGonad:moduleturquoise .  
tissueLiver:moduleturquoise    
tissueGonad:moduleyellow    .  
tissueLiver:moduleyellow       

Phi coefficients (phi model with identity link):
      Estimate Std. Error z value Pr(>|z|)    
(phi)  10.6634     0.2186   48.78   <2e-16 ***

Exceedence parameter (extended-support xbetax model):
        Estimate Std. Error z value Pr(>|z|)    
Log(nu)   -3.152      0.078  -40.41   <2e-16 ***

Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 

Exceedence parameter nu: 0.0428
Type of estimator: ML (maximum likelihood)
Log-likelihood:  8305 on 32 Df
Number of iterations in BFGS optimization: 47 


```{r}
png("figs/tau_modules_vioplot.png",height = 4.5,width=10,units="in",res=300)

plot_cols<-unlist(lapply(levels(as.factor(ModuleColors[ModuleColors!="grey"])),rep,times=3))
par(mar=c(4,4,1,1))
plot(x=0:length(plot_cols),
     y=seq(0,1,length.out=length(plot_cols)+1),
      xaxt='n',
        xlab="",
        ylab=expression(tau),
     type='n',
     las=1)

vioplot(tau_col_dat$tau~tau_col_dat$tissue+tau_col_dat$module,
        col=plot_cols,
        add=TRUE)

axis(1,
     at=1:length(plot_cols),
     labels=rep(levels(tau_col_dat$tissue),
                length(levels(tau_col_dat$module))),
     las=2,tick = FALSE)

     
# get the correlation coefficients with sex
cor_vals<-signif(moduleTraitCor, 2)[,1]
names(cor_vals)<-gsub("ME","",names(cor_vals))
cor_vals<-cor_vals[levels(tau_col_dat$module)]

text(x=seq(2,length(plot_cols)-1,length.out=length(levels(tau_col_dat$module))),
     y=rep(1,length(levels(tau_col_dat$module))),
     paste0("r=",cor_vals))

# add the tissue specific correlation coefficients?
moduleTraitCorPlots<-signif(moduleTraitCor, 2)
moduleTraitCorPlots<-moduleTraitCorPlots[,-1]
dimnames(moduleTraitCorPlots)[[1]]<-gsub("ME","",dimnames(moduleTraitCorPlots)[[1]])
moduleTraitCorPlots<-moduleTraitCorPlots[levels(tau_col_dat$module),]
moduleTraitCorPlots<-t(moduleTraitCorPlots)
cor_vec<-unlist(as.list(moduleTraitCorPlots))
cor_vec<-cor_vec

text(x=1:length(plot_cols),
     y=rep(-0.06,length(plot_cols)),
     cor_vec, 
     xpd=TRUE,
     cex=0.5)


# get labels
vio_labels<-tau_mod_lab[,c("tissue","module",".group")]
vio_labels$.group<-gsub("\\s+","",vio_labels$.group)
vio_labels<-vio_labels[order(vio_labels$module, vio_labels$tissue),]
text(x=1:length(plot_cols),
     y=rep(0.95, length(plot_cols)),
     vio_labels$.group)
dev.off()
```

the two most highly correlated modules with male gene expression have significantly higher tissue specificity than the other modules; but the same was not true for the other significantly correlated with males module, nor for the female-correlated module. Strong correlations with up- or down-regulation in tissues was not associated with tissue specificity.

### identifying hubs

First need to get gene significance and module membership

```{r}
# based on https://ngs101.com/how-to-build-gene-co-expression-networks-from-rna-seq-data-using-wgcna-complete-step-by-step-guide-for-absolute-beginners/
# Calculate gene significance (correlation with trait) for sex
gene_trait_cor <- cor(expression_data_subset, 
                      trait_data$female, 
                      use = "pairwise.complete.obs")
gene_trait_pvalue <- corPvalueStudent(as.numeric(gene_trait_cor), nrow(trait_data))
 
# Add gene names to the p-value vector
names(gene_trait_cor) <- colnames(expression_data_subset)
names(gene_trait_pvalue) <- colnames(expression_data_subset)

# create objects to put data into
all_hub_gene_info<-data.frame() 
all_hub_genes<-data.frame()

for(sex_module in c("black","blue","red","turquoise")){

  # Get genes in this module
  module_genes <- colnames(expression_data_subset)[ModuleColors == gsub("ME","",sex_module)]
  cat("Number of genes in", sex_module, "module:", length(module_genes), "\n")
  
   
  # Calculate module membership for these genes
  # MM measures how correlated each gene is with the module eigengene
  gene_module_membership <- cor(expression_data_subset[, module_genes],
                                 MEs[, paste0("ME",sex_module)],
                                 use = "pairwise.complete.obs")
   
  # Create a data frame with gene information
  module_gene_info <- data.frame(
    Gene = module_genes,
    ModuleMembership = as.numeric(gene_module_membership),
    GeneSignificance = gene_trait_cor[module_genes],
    MM_pvalue = corPvalueStudent(as.numeric(gene_module_membership), nrow(trait_data)),
    GS_pvalue = gene_trait_pvalue[module_genes],
    stringsAsFactors = FALSE
  )
   
  # Sort by module membership
  module_gene_info <- module_gene_info[order(-abs(module_gene_info$ModuleMembership)), ]
   
  
  # Calculate connectivity (intramodular connectivity)
  # First, calculate adjacency matrix for genes in this module
  adjacency_matrix <- adjacency(
    expression_data_subset[, module_genes],
    power = soft_power,
    type = "signed"
  )
   
  # Calculate connectivity: sum of connection weights for each gene
  # (subtract 1 to exclude self-connection)
  connectivity <- rowSums(adjacency_matrix) - 1
   
  # Combine all metrics for hub gene identification
  hub_gene_info <- data.frame(
    Gene = module_genes,
    Connectivity = connectivity,
    ModuleMembership = as.numeric(gene_module_membership),
    GeneSignificance =  gene_trait_cor[module_genes],
    stringsAsFactors = FALSE
  )
   
  # Sort by connectivity to identify hubs
  hub_gene_info <- hub_gene_info[order(-hub_gene_info$Connectivity), ]
  hub_gene_info$module <- sex_module 
  
  hub_genes<-hub_gene_info[which(hub_gene_info$Connectivity > quantile(hub_gene_info$Connectivity,0.9) &
                                   hub_gene_info$ModuleMembership > 0.7 &
                                   abs(hub_gene_info$GeneSignificance) > moduleTraitCor[paste0("ME",sex_module),1]),]
  
  # save to bigger data frames
  all_hub_gene_info <- rbind(all_hub_gene_info, hub_gene_info)
  all_hub_genes<-rbind(all_hub_genes, hub_genes)
}
write.table(all_hub_genes$Gene,"data/hub_gene_IDs.txt",
          row.names=FALSE,
          col.names=FALSE,
          quote=FALSE)
```
```{r}
table(all_hub_genes$module)
```

```{r}
all_hub_gene_info<-merge(all_hub_gene_info, 
                         logFC_long_all,
                         by.x="Gene",
                         by.y="geneID",
                         all.x=TRUE
                         )
write.csv(all_hub_gene_info,"all_hub_gene_info.csv",row.names = FALSE)
```


```{r}
my_path<-"\\file\Research\FlanaganLab\intermediate_data\SMA_DE_pipeline_analysis_2024\07_TRgenes\cory_BLAST"
blast_results_files<-list.files(pattern="blast.txt", path=my_path)
```



Or CoDiNA:

```{r}
library(wTO)
library(CoDiNA)

expression_matrix<-as.data.frame(assay(vsd))

wto_sma<-CorrelationOverlap(Data = expression_matrix,
                               Overlap = rownames(expression_matrix), 
                               method = 's') %>% 
  wTO(., sign = 'abs')

wto_complete<- wTO.Complete(k = 2, 
                            n = 100,
                            Data = expression_matrix, 
                            Overlap = rownames(expression_matrix), 
                            method = 's', 
                            method_resampling = 'Bootstrap',
                            pvalmethod = 'BH', 
                            savecor = TRUE, 
                            expected.diff = 0.2, 
                            plot = TRUE)
```

## Evolutionary analysis

First, I ran transdecoder on the supertranscriptome.

Then I ran orthofinder using the following files:

corythoichthys_intestinalis/data/GCF_030265065.1/protein.faa -> orthofinder/corythoichthys_intestinalis.protein.faa
stigmatopora_argus/data/GCF_051989625.1/protein.faa -> orthofinder/stigmatopora_argus.protein.faa
stigmatopora_nigra/data/GCF_051989575.1/protein.faa -> orthofinder/stigmatopora_nigra.protein.faa
SMA_supertranscriptome.fasta.transdecoder.pep

(C. haematopterus has a published genome sequence but no protein fasta, so I will discard it fo rnow)
```{bash, eval=FALSE}
# on the RCH
#!/bin/bash -l
#
#SBATCH --job-name=orthofinder
#
#SBATCH --ntasks=1
#SBATCH --time=5:00:00
#SBATCH --mem-per-cpu=3GB
#SBATCH --cpus-per-task=8
#SBATCH --error=job_%x_%j.err
#SBATCH --output=job_%x_%j.out

module load prefix/latest
module load OrthoFinder/2.5.5-foss-2023a
cd /home/spf50/S_macropterygia/
orthofinder -t 8 -f orthofinder/
```


