---
title: "Differential Expression Analysis in Australian _Stigmatopora nigra_"
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: yes
    toc: true
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
```

``` {r library, include = TRUE, message = FALSE, warning = FALSE}
setwd("~/Documents/GitHub/pipefish_sbge/stigmatopora/S_nigra_aus")
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
library(ggpubr)
library(cowplot)
library(UpSetR)
library(ggplot2)
library(kableExtra)
library(tidyverse)
library(knitr)
library(magick)
library(multcompView)
library(spfTools)
library(patchwork)
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.aus <- readRDS("data/txi.salmon_aus.RDS")

#The samples file generated for tximport
samples <- read.table("data/aus_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)

```

The package `DESeq2` was used for the differential expression analysis outlined below.

# Single factor analysis - Comparing Males v Females across all organs
To analyze your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions. The model for this single factor analysis was run as counts ~ Sex.

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_aus <- DESeqDataSetFromTximport(txi.aus, 
                                   colData = samples,
                                   design = ~ sex)

dds_aus <- dds_aus[, !(dds_aus$ID %in% c("S3", "S10","S13","S16"))]

samples <- samples[!(dds_aus$ID %in% c("S3", "S10","S13","S16")), ]
```

The data is then pre-filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

The cutoff here was to remove rows that had counts fewer than 10 across all samples.

```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at least 10 reads total
keep <- rowSums(counts(dds_aus)) >= 10
dds_aus <- dds_aus[keep, ]

```

After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.

```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_aus_exp <- DESeq(dds_aus)

#Compile the results
res <- results(dds_aus_exp)
res
```

Once that has finished we can now start exploring some of the single-factor analysis results
```{r investigate-DESeq-results-SN}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#How many ADJUSTED p-values were less than 0.1?
sum(res$padj < 0.1, na.rm = TRUE)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_aus_exp, alpha = 0.05)
summary(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```

## Visualising the results
### MA-plot - MvF Differntial expression
Generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be colored if the adjusted p-value is less that 0.1.

```{r MA-plot, echo=FALSE, fig.cap="LogFC versus the mean normalised count for all of the genes. Points that are blue have a p-value less than 0.1. The two plots are the same with the only difference coming from an adjusted y-limit. Points that are triangles represent genes with a fold change higher/lower than the y-limit."}

par(mfrow=c(1,2))
plotMA(res, ylim = c(-2,2))
plotMA(res)

```


### Heatmap - overall expression
We can also generate a heat map to look at overall expression levels across our samples. Note, this is not differentially expressed genes.

```{r heatmap, echo=FALSE, fig.cap= "Heatmap showing the expression level across all organs for the top 20 genes with the highest expression levels."}
df <- as.data.frame(samples[,c("sex", "organ")])
rownames(df) <- samples$ID
df$organ <- as.character(df$organ)

organ_cols<-c("gill" = "#20B2AA",
              "gonad" = "#EEB422", 
              "liver" = "#EE8262")

sex_cols<-c("F" = "#7fc97f", "M" = "#beaed4")

ann_colors = list(sex=sex_cols,
                  organ=organ_cols)

col_order <- c(rownames(df[df$sex=="F" & df$organ=="gill",]),
               rownames(df[df$sex=="M" & df$organ=="gill",]),
               rownames(df[df$sex=="F" & df$organ=="liver",]),
               rownames(df[df$sex=="M" & df$organ=="liver",]))


#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(dds_aus_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]


#Transform the data
vsd <- vst(dds_aus_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select, col_order], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df, 
         annotation_colors = ann_colors)
```

With this heatmap we can also pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.

```{r heatmap_genes}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))
#Run once to create the file and then commented out
#write.table(heatmap_TG,
 #           'data/heatmap_trinitygenes.txt',
  #          sep = "",
   #         quote=FALSE,
    #        row.names = FALSE,
     #       col.names = FALSE)
```

### Sample-dist Heatmap

We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples. The darker the colour, the more similar the samples are. The diagonal line of very dark blue represents the comparison within the same samples.

```{r sample-dist, echo=FALSE, fig.cap="Sample-to-sample distances heatmap showing the similarities and differences between all of the samples. The darker the colour, the more similar the samples are. The diagonal line of very dark blue represents the comparisons between the same samples.", warning=FALSE}
#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$sex, vsd$organ, sep = "-")
colnames(sampleDistsMatrix) <- vsd$ID

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)
```

We can see that the highest similarities (aside from within samples comparisons) is between samples from the same organ. 

### PCA plots

```{r pca-pairs-plot, echo=FALSE, fig.cap="Principal components analysis pairsplot showing PCA axis 1 - 4."}
#Generate the PCA data to create the pairs plot, using the PCAtools package
p <- pca(assay(vsd), 
         metadata = colData(dds_aus))

#Create the pairs plot
pairsplot(p,
          components = getComponents(p, c(1:4)),
          colby = "organ",
          triangle = FALSE,
          hline = 0, vline = 0,
          colkey = c("gill" = "#6E8B3D75", 
                     "liver" = "#EE826275"),
          shape = "sex",
          shapekey = c(16, 17),
          pointSize = 3,
          margingaps = unit(c(0.1, 0.1, 0.1, 0.1), 'cm'),
          legendPosition = "left", legendIconSize = 3, legendLabSize = 10)

```

Next we can look at PCA plots for our samples to see how our different samples are grouping.
```{r pca-plot, echo=FALSE, message=FALSE, warning=FALSE}

samples <- samples[!(samples$ID %in% c("S3", "S10","S13","S16")),]


#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$organ <- samples$organ
pca_plotting_data$sex <- samples$sex
pca_plotting_data$ID <- samples$ID

#Calculate the percent variation attributed to each axis 
percentVar <- round(p$variance, digits = 2)


my_colors <- c("gill" = "#6E8B3D75", "liver" = "#EE826275")
my_shapes <- c(16, 17)

ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed') #+ 
  #geom_text(aes(label = ID), vjust = -1, hjust = 0.5)
```

S7 and S2 are clustering quite far away from any of the other samples. They both had lower mapping rates (69% and 78% respectively). We may remove them from downstream analyses. 

```{r pca1v3, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC3. This third axis explains 9.08% of the variation, largely attributed to sex differences between the gonads samples.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC3, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC3: ",percentVar[3],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed') #+ 
  #geom_text(aes(label = ID), vjust = -1, hjust = 0.5)
```

```{r pca1v4, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC4. This fourth axis explains 6.63% of the variation, largely attributed to three of the gill samples.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC4, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC4: ",percentVar[4],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed') #+ 
  #geom_text(aes(label = ID), vjust = -1, hjust = 0.5)
```

#### Saving the PCA plots to use in figure creation
I will be including the PCAs describing the relationship between PC1 - 3 in the main text, as they account for the majority of the differences between our samples. The plot with PC1 and PC4 I still want to export, and will likely include as a supplemental to highlight why samples related to S34 will be removed from the subsequent analyses.

```{r setParams-PCAs}
#Setting the shapes I want to show up for the different organs
organ_shapes <- c(gill = 16, liver = 17)

#Setting the colors I want for each sex
sex_cols <- c("F" = "#7fc97f", "M" = "#beaed4")

```

```{r figPCA1v2, eval = FALSE}
#Create the blank pdf to store the plot in
pdf("figs/Fig_PCA1v2.pdf",height = 6,width=6)

#Set the plotting parameters
par(mar=c(4,5,4,1), oma=c(2,2,2,2))

#Create the plot for PC1 v PC2
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC2,
     col = paste0(sex_cols[pca_plotting_data$sex],"75"),
     pch = organ_shapes[pca_plotting_data$organ],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC2: ",percentVar[2], "% variance"),
     bty = 'l',
     xpd = TRUE)

#Add a legend describing the sex
outer_legend("top",
             c("Female","Male"),
             pch = 18,
             bty = 'n',
             col=paste0(sex_cols,"75"),
             cex = 1.75,
             ncol = 2,
             pt.cex = 3)

dev.off()
```

```{r figPCA1v3, eval = FALSE}
#Create the blank pdf to store the plot in
pdf("figs/Fig_PCA1v3.pdf",height = 6,width=6)

#Set the plotting parameters
par(mar=c(4,5,4,1), oma=c(2,2,2,2))

#Create the plot for PC1 v PC3
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC3,
     col = paste0(sex_cols[pca_plotting_data$sex],"75"),
     pch = organ_shapes[pca_plotting_data$organ],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC3: ",percentVar[3], "% variance"),
     bty = 'l',
     xpd = TRUE)

#Add  legend describing the organs
outer_legend("top",
             c("Gill","Liver"),
             pch = c(16,15,17),
             bty = 'n',
             col = 'darkgrey',
             cex = 1.75,
             ncol = 3,
             pt.cex = 3)

dev.off()
```

```{r figPCA1v4, eval = FALSE}
#Create the blank pdf to store the plot in
pdf("figs/Fig_PCA1v4.pdf",height = 6,width=8)

#Set the plotting parameters
par(mar=c(4,5,1,5),oma=c(2,2,2,2))

#Create the plot for PC1 v PC4
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC4,
     col = paste0(sex_cols[pca_plotting_data$sex],"75"),
     pch = organ_shapes[pca_plotting_data$organ],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC4: ",percentVar[4], "% variance"),
     bty = 'l',
     xpd = TRUE)

text(pca_plotting_data$PC1[pca_plotting_data$ID == "S3", "S10","S13","S16"] + 50,
     pca_plotting_data$PC4[pca_plotting_data$ID == "S3", "S10","S13","S16"] - 25,
     labels = pca_plotting_data$ID[pca_plotting_data$ID == "S3", "S10","S13","S16"])

#Add  legend describing the organs and sex
outer_legend("right",
             c("Gill","Liver"),
             pch = c(16,15,17),
             bty = 'n',
             col = 'darkgrey',
             cex = 1.5,
             ncol = 1,
             pt.cex = 2.25)

outer_legend("topright",
             c("Female","Male"),
             pch = 18,
             bty = 'n',
             col=paste0(sex_cols,"75"),
             cex = 1.5,
             ncol = 1,
             pt.cex = 2.25)
dev.off()
```


#### Creating heatmaps based on the PCA axes
I now want to generate heatmaps based on the genes that have the highest loadings for each of the PCA axes. I will be using a cut off of 0.02 to determine which genes will be included in the heatmaps. Once again I am doing this for PCA1-4, but will likely only inlcude PCA1-3 in a figure for the maintext and PC4 will go in the supplemental.

```{r setParams-heatmaps}

df <- as.data.frame(samples[,c("sex", "organ")])
rownames(df) <- samples$ID
df$organ <- as.character(df$organ)

organ_cols<-c("gill" = "#20B2AA",
              "liver" = "#EE8262")

ann_colors = list(sex=sex_cols,
                  organ=organ_cols)

col_order <- c(rownames(df[df$sex=="F" & df$organ=="gill",]),
               rownames(df[df$sex=="M" & df$organ=="gill",]),
               rownames(df[df$sex=="F" & df$organ=="liver",]),
               rownames(df[df$sex=="M" & df$organ=="liver",]))

pca_rotation <- pca$rotation[, 1:4]

```

```{r savePheatmapFxn}

#Function to use to export the heatmaps to pdfs
# from https://stackoverflow.com/questions/43051525/how-to-draw-pheatmap-plot-to-screen-and-also-save-to-file
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

```

```{r figheatmap-PC1, eval = FALSE}

pc1 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,1]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors = ann_colors,
                cellwidth = 9,
                fontsize = 16,
                annotation_legend = FALSE,
                main = "Top loading genes on PC1")


save_pheatmap_pdf(pc1, "figs/Fig_pc1_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC2, eval=FALSE}

pc2 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,2]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                annotation_legend = FALSE,
                main = "Top loading genes on PC2")

save_pheatmap_pdf(pc2, "figs/Fig_pc2_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC3, eval=FALSE}

pc3 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,3]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                border_color = NA,
                main = "Top loading genes on PC3")

save_pheatmap_pdf(pc3, "figs/Fig_pc3_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC4, eval=FALSE}

pc4 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,4]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                border_color = NA,
                main = "Top loading genes on PC4")

save_pheatmap_pdf(pc4, "figs/Fig_pc4_heatmap.pdf",
                  width=6,
                  height=6)

```

#### Making the combined figure

```{r assembleFigPCA, message=FALSE}
figPCAa <- image_ggplot(image_read_pdf('figs/Fig_PCA1v2.pdf'),
                        interpolate = TRUE)
figPCAb <- image_ggplot(image_read_pdf('figs/Fig_PCA1v3.pdf'),
                        interpolate = TRUE)
figPCAc <- image_ggplot(image_read_pdf('figs/Fig_pc1_heatmap.pdf'),
                        interpolate = TRUE)
figPCAd <- image_ggplot(image_read_pdf('figs/Fig_pc2_heatmap.pdf'),
                        interpolate = TRUE)
figPCAe <- image_ggplot(image_read_pdf('figs/Fig_pc3_heatmap.pdf'),
                        interpolate = TRUE)

# make two patchworks
pcas <- figPCAa + figPCAb
hms <- figPCAc + figPCAd + figPCAe

# put the patchworks together
figPCA <-  wrap_plots(pcas,
                      hms,
                      ncol=1) + 
  plot_annotation(tag_levels = 'A') 

ggsave("figs/FigPCA.pdf", figPCA, height=4, width=5)
ggsave("figs/FigPCA.png", figPCA, height=4, width=5) # also save as a png
```

```{r assembleFigPCA-supp, message=FALSE}
figPCA4 <- image_ggplot(image_read_pdf('figs/Fig_PCA1v4.pdf'),
                        interpolate = TRUE)
figPCA4_hm <- image_ggplot(image_read_pdf('figs/Fig_pc4_heatmap.pdf'),
                           interpolate = TRUE)

# put the patchworks together
figPCA_supp <-  wrap_plots(figPCA4, figPCA4_hm, ncol=2) + 
  plot_annotation(tag_levels = 'A') 

ggsave("figs/FigPCA_supp.pdf", figPCA_supp, height=4, width=8)
ggsave("figs/FigPCA_supp.png", figPCA_supp, height=4, width=8) # also save as a png
```

```{r showAssembledFigPCA, eval=TRUE}
figPCA
```

# Multifactor design - Comparing M v F across the different tissue types 
If we investigate the column of our DESeq dataset, we can see that each sample has both a sex and an organ type attached to it - we will be using these two factors in our multifactor analysis. 

```{r multi-factor-analysis, message=FALSE, warning=FALSE}
#import samples.txt 
samples <- read.table("data/aus_samples.txt", header = TRUE)

#Create an additional column the has the two conditions combined(sex and organ type)
samples$group <- factor(paste0(samples$sex, samples$organ))

##Create a copy of the DESeq dataset
ddsMF_aus <- DESeqDataSetFromTximport(txi.aus,
                                     colData = samples,
                                     design = ~ group)

##Filter the dataset, only keeping rows that have at least 10 reads total, but less than 1,000,000
keep <- rowSums(counts(ddsMF_aus)) >= 10 & rowSums(counts(ddsMF_aus)) < 1e6
ddsMF_aus <- ddsMF_aus[keep, ]

#Run the differential expression analysis
ddsMF_aus_exp <- DESeq(ddsMF_aus)
resultsNames(ddsMF_aus_exp)

```

## Invesitgate the results of the differential expression analysis
Thanks to the multi-factor analysis we can now explore differential expression between all of the different combinations:   

  - Male Liver v. Female Liver 
  - Male Gill v. Female Gill  
  - All of the within sex tissue comparisons (e.g. Male Liver v. Male Gill, etc.)

### M-F liver comparison
```{r m-f-liver-results}
##Pulling out the liver M-F results with an alpha of 0.05
liver_con_res <- results(ddsMF_aus_exp, contrast = c("group", "Mliver", "Fliver"),
                         alpha = 0.05)
liver_con_res$trin_geneid <- rownames(liver_con_res)
head(liver_con_res)

summary(liver_con_res)
```

There are many criteria that have been employed previously to denote sex bias. For this study we are classifying sex-biased genes as **genes with a p-value < 0.05 AND a logFC $\ge$ |2|**. With that criteria in the liver there are `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. 

I have then pulled out all of the male-biased, female-biased, and non-biased genes based on the criteria outlined above. I determined non-biased genes as a p-vaule > 0.05.

```{r liver-sex-biased}
#Removing the rows where padj. is NA in results
liver_con_res_noNA <- liver_con_res[!is.na(liver_con_res$padj),]
summary(liver_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#write.csv(as.data.frame(liver_con_res_noNA), "data/liver_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the liver
liver_mal_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange >= 2
                                             & liver_con_res_noNA$padj <= 0.05),]
liver_fem_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange <= -2 
                                             & liver_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the liver
liver_non_biased <- liver_con_res_noNA[which(liver_con_res_noNA$padj > 0.05),]

```

I will be generating a table that outlines the genes that are the most male or female biased in each organ type. To do this I need to pull all differentially expressed genes in males or females, get the Trinity gene IDs and then BLAST the corresponding sequences. Here I am pulling out the Trinity gene IDs for those genes.


```{r all-liver}
#Creating a subset of the results where the p-value is less than 0.05
liver_con_res_p05 <- liver_con_res_noNA[liver_con_res_noNA$padj <= 0.05, ]


#Pulling the all diff expressed genes in Males
MLiver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(
  liver_con_res_p05$log2FoldChange, decreasing = TRUE),], n=547)))
#Run once to create the file and then commented out
#write.table(MLiver_trin_genes,
 #           'data/aus_maleL_TRgenes.txt', 
  #           sep = "", 
   #          quote=FALSE, 
    #         row.names = FALSE, 
     #        col.names = FALSE)

#Pulling all diff expressed genes in Females
Fliver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(
  liver_con_res_p05$log2FoldChange, decreasing = FALSE),], n=1336)))

#Run once to create the file and then commented out
#write.table(Fliver_trin_genes,
 #           'data/aus_femL_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### M-F Gill Comparisson
```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res <- results(ddsMF_aus_exp, contrast = c("group", "Mgill", "Fgill"),
                        alpha = 0.05)
gill_con_res$trin_geneid <- rownames(gill_con_res)
head(gill_con_res)

summary(gill_con_res)
```

In the gills there were `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

I have then pulled out all of the male-biased, female-biased, and non-biased genes to save them into their own objects.

```{r gill-sex-biased}
#Removing the rows where padj. is NA in results
gill_con_res_noNA <- gill_con_res[!is.na(gill_con_res$padj), ]
summary(gill_con_res_noNA) #We can now see that there are no outliers or 
                           #low counts since the NAs have been removed
write.csv(as.data.frame(gill_con_res_noNA), "data/gill_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the gills
gill_mal_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange >= 2 
                                           & gill_con_res_noNA$padj <= 0.05),]
gill_fem_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange <= -2 
                                           & gill_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gills, p>0.05
gill_non_biased <- gill_con_res_noNA[which(gill_con_res_noNA$padj > 0.05),]
```

Here I am getting the trinity gene IDs that correspond to all differentially expressed genes for males and females.

```{r all-gill}
#Creating a subset of results where p-value is less than 0.05
gill_con_res_p05 <- gill_con_res_noNA[gill_con_res_noNA$padj <= 0.05,]

#Pulling all diff expressed genes in Males
MGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(
  gill_con_res_p05$log2FoldChange, decreasing = TRUE),], n=25)))

#Run once to create the file and then commented out
#write.table(MGill_trin_genes,
 #           'data/aus_maleG_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)

#Pulling all diff expressed genes in Females 
FGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(
  gill_con_res_p05$log2FoldChange, decreasing = FALSE),], n=12)))

#Run once to create the file and then commented out
#write.table(FGill_trin_genes,
 #           'data/aus_femG_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### MA Plots
```{r MA-plot-MF, echo=FALSE, fig.height= 15, fig.width=8, fig.cap="MA-plots generated for each organ type that compares logFC to the mean expression. Female-biased and male-biased genes are represented by colour (female: green; male: purple) and are determined with a fold change cutoff of 2 and a p-value cutoff of 0.05"}
liver_MA <- ggmaplot(liver_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Liver" %->% "Female Liver"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gill_MA <- ggmaplot(gill_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Gill" %->% "Female Gill"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

#Merge all the plots into one
plot_grid(liver_MA, gill_MA,
          ncol = 1,
          nrow = 2,
          labels = c('A', 'B'),
          label_fontfamily = 'sans',
          label_fontface = 'bold',
          label_size = 18,
          rel_heights = c(4.10, 4.10))

```

From the MA plots we can see the highest number of biased genes are found in  the liver and then the gills. 

## Creating an Upset Plot
```{r upset-plot-mal, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in males."}
#Male-biased genes
listInputMB <- list("Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased)
                  )
upset(fromList(listInputMB),
      mainbar.y.label = "Number of Shared Male-Biased Genes",
      sets.x.label = "Total Number of Male-Biased Genes",
      point.size = 3)
```

```{r upset-plot-fem, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in females."}
#Female-biased genes
listInputFB <- list("Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased)
                  )
upset(fromList(listInputFB),
      mainbar.y.label = "Number of Shared Female-Biased Genes",
      sets.x.label = "Total Number of Female-Biased Genes",
      point.size = 3)
```

```{r upset-plot-all, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs across both sexes."}
#Exporting data used to make the Upset plot (run once and them commented out)
write.table(as.data.frame(gill_mal_biased), "data/gill_mal_biased.txt", 
            row.names = TRUE)
write.table(as.data.frame(gill_fem_biased), "data/gill_fem_biased.txt", 
            row.names = TRUE)
write.table(as.data.frame(liver_mal_biased), "data/liver_mal_biased.txt", 
            row.names = TRUE)
write.table(as.data.frame(liver_fem_biased), "data/liver_fem_biased.txt", 
            row.names = TRUE)

#Checking there is no overlap between the two sexes
listInputall <- list("Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased),
                  "Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased))

# Define the desired order of sets
desired_order <- c("Male Gill", "Female Gill", "Male Liver", "Female Liver")

# Create the UpSet object with custom set order
upset(fromList(listInputall),
      sets = desired_order,
      mainbar.y.label = "Number of Shared Sex-Biased Genes",
      sets.x.label = "Total Number of Sex-Biased Genes",
      point.size = 3,
      nsets = 4,
      nintersects = NA,
      order.by = "degree")

#this code creates the same heatmap without the custom set order 
#upset(fromList(listInputall),
 #     mainbar.y.label = "Number of Shared Sex-Biased Genes",
  #    sets.x.label = "Total Number of Sex-Biased Genes",
   #   point.size = 3,
    #  nsets = 4,
     # nintersects = NA)

```

## Variation in FC across sex-bias and tissue type
I want to create a plot the highlights the variation we see in fold-change both across the different biases groups (male-biased, female-biased, and non-biased) within one tissue type and across all of the tissue types. To do this I first need to create a "long" dataset in the following format:

| Tissue_type|     Bias      |   Fold-Change   |
|:----------:|:-------------:|:---------------:|
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   fem_bias    |      XXXXX      |
|    Gill    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   fem_bias    |      XXXXX      |
|    Liver   |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |

That dataset is created here, with the Trinity gene IDs included as well:

```{r logFC-long-dataset}
logFC_long<- data.frame(
  tissue=c(rep("gill",nrow(gill_fem_biased)),
           rep("gill", nrow(gill_mal_biased)),
           rep("gill", nrow(gill_non_biased)),
           rep("liver", nrow(liver_fem_biased)),
           rep("liver", nrow(liver_mal_biased)),
           rep("liver", nrow(liver_non_biased))
         ),
  bias=c(rep("FB",nrow(gill_fem_biased)),
         rep("MB",nrow(gill_mal_biased)),
         rep("UB", nrow(gill_non_biased)),
         rep("FB", nrow(liver_fem_biased)),
         rep("MB", nrow(liver_mal_biased)),
         rep("UB", nrow(liver_non_biased))
         ),
  logFC=c(gill_fem_biased$log2FoldChange,
          gill_mal_biased$log2FoldChange,
          gill_non_biased$log2FoldChange,
          liver_fem_biased$log2FoldChange,
          liver_mal_biased$log2FoldChange,
          liver_non_biased$log2FoldChange
          ),
  geneID=c(rownames(gill_fem_biased),
           rownames(gill_mal_biased),
           rownames(gill_non_biased),
           rownames(liver_fem_biased),
           rownames(liver_mal_biased),
           rownames(liver_non_biased)
           )
  
)

#Exporting this data, run once and then commented out
#write.csv(logFC_long, "data/logFC_long_sexbias.csv", row.names = FALSE)
```

With the dataset in the proper format now, we can generate the plot

```{r FC-var-plot, echo=FALSE, fig.cap="Absolute value of the logFC for genes that are female-biased, male-biased, and unbiased across each tissue type. Raw fold-change values are added on top of the boxplot as jitters."}
my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "darkgray")

ggplot(logFC_long,
       aes(x = bias, y = abs(logFC), fill = bias)) +
  geom_boxplot() +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_point(data=logFC_long, aes(x=bias, y=abs(logFC)),
             size=0.4, 
             position = position_jitter(width = 0.15)) + 
  labs(x="Bias Level", y="|logFC|") +
  guides(fill = guide_legend(title = "Bias Level", order = 3))

```

```{r FC-vio-plot-G-L}

  
#Make the plot
my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "darkgray")

ggplot(logFC_long,
       aes(x = bias, y = abs(logFC), fill = bias)) +
  geom_violin() +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_point(data=logFC_long, aes(x=bias, y=abs(logFC)),
             size=0.4, 
             position = position_jitter(width = 0.15)) + 
  labs(x="Bias Level", y="|logFC|") +
  guides(fill = guide_legend(title = "Bias Level", order = 3))

```

There appears to be differences in the logFC across both the organs and the sex-bias groups (male-biased, fem-biased, and non-biased) (Fig. \@ref(fig:FC-var-plot)). Let's explore some analyses to see if we can put some statistical power behind those differences.

```{r FC-var-stats, warning=FALSE, eval=FALSE}
#Looking at some summary statistics for logFC between the different groups
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), mean)
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), sd)

par(mfrow=c(1,1))
interaction.plot(logFC_long$tissue, logFC_long$bias, logFC_long$logFC)

logFC_long$tissue <- as.factor(logFC_long$tissue)
logFC_long$bias <- as.factor(logFC_long$bias)

fc_var_aov <- aov(abs(logFC_long$logFC) ~ logFC_long$bias*logFC_long$tissue)
anova(fc_var_aov)

par(mfrow = c(2,2))
plot(fc_var_aov)

library(car)
leveneTest(fc_var_aov)
par(mfrow=c(1,1))
hist(resid(fc_var_aov))
#shapiro.test(resid(fc_var_aov)) 

plot(TukeyHSD(fc_var_aov))
library(multcompView)
multcompLetters4(fc_var_aov, TukeyHSD(fc_var_aov))

##Fitting a negative binomial model
library(MASS)
model <- glm.nb(abs(logFC_long$logFC) ~ logFC_long$tissue * logFC_long$bias)

linear_model <- lm(abs(logFC_long$logFC)~ logFC_long$tissue * logFC_long$bias)
```

## Categorising sex-specific genes
Now that we have investigated what the sex-bias is like across the different tissue types I want to dive further into genes with **sex-specific expression**. These are genes expressed only in one sex or the other. I am classifying a sex-specific genes within each tissue as ones where the expression of that gene is less than 10 for all of one sex and there is a median of $\ge 20$ in the other sex. 

Prior to running the for loop, any "outliers" that were determined by DESeq were removed from the pool of genes. Additionally, normalised counts were used for the classification of sex-specific genes as DESeq2 uses the normalised counts for all of the logFC/sexbias calculations.

```{r sex-specific-for-loop}
#Pulling out the geneIDs for genes that were categorised as "outliers" by DESeq2
#Calculating the Cooks threshold that would have been used
ddsMF_aus_exp <- ddsMF_aus_exp[, !(ddsMF_aus_exp$ID %in% c("S3", "S10","S13","S16"))]

np <- length(resultsNames(ddsMF_aus_exp))
nsamp <- ncol(ddsMF_aus_exp)
cooks_thresh <- qf(0.99, df1 = np, df2 = nsamp-np)

out_ids <- names(mcols(ddsMF_aus_exp)$maxCooks[mcols(ddsMF_aus_exp)$maxCooks >
                                                cooks_thresh])

#Checking to make sure we have correct number of outliers
length(out_ids)

#Filtering out the dds dataset to remove the outliers determined by DESeq
ddsMF_aus_exp_filtered <- ddsMF_aus_exp[!(rownames(ddsMF_aus_exp) %in% out_ids), ]

#Create a vector with the different organ types
organs <- unique(colData(ddsMF_aus_exp_filtered)$organ)

#Create an empty list to store my datasets in
aus_sex_specific_genes <- list()

#Generate the for loop to identify MSpecific and FSpecific genes in each 
#organ based on Medians
for(organ in organs){
  #Male-Specific Genes
  ##Pull out all of the rows where fem count <=10 in every female sample
  fem0_organ_names <- which(rowSums(t(apply(counts(ddsMF_aus_exp_filtered, 
                                                   normalized = TRUE)[, ddsMF_aus_exp_filtered$sex == "F" 
                                                                      & 
                                                                        ddsMF_aus_exp_filtered$organ == organ], 
                                            1, 
                                            function(x) x <= 10)
                                      )
                                ) == ncol(counts(ddsMF_aus_exp_filtered, 
                                                 normalized = TRUE)[, ddsMF_aus_exp_filtered$sex == "F" &
                                                                      ddsMF_aus_exp_filtered$organ == organ])
                      )
  
  fem0_organ <- counts(ddsMF_aus_exp_filtered, 
                       normalized = TRUE)[rownames(counts(ddsMF_aus_exp_filtered, 
                                                          normalized = TRUE)) %in% 
                                            names(fem0_organ_names), 
                                          ddsMF_aus_exp_filtered$organ == organ]
                                                         
  
  ##Pull out rows where median of male count >=20 for that organ
  mal10_organ <- apply(counts(ddsMF_aus_exp_filtered, 
                              normalized = TRUE)[, ddsMF_aus_exp_filtered$sex == "M" 
                                                 & ddsMF_aus_exp_filtered$organ == organ], 
                       1,
                       function(x) median(x) >=20)
  
  ##Keep only the rows where all fem samples <=10 and the Male median>=20
  fem0_mal10_organ <- fem0_organ[rownames(fem0_organ) %in% 
                                   names(mal10_organ[mal10_organ == TRUE]),
                                 ]
  
  ##Create a new object with a name based on the organ type
  organ_malsp <- sub("$", "_male_specific", organ)
  aus_sex_specific_genes[[organ_malsp]] <- fem0_mal10_organ
  
  
  
  #Female-Specific Genes
  ##Pull out all of the rows where male count <=10 in every male sample
  mal0_organ_names <- which(rowSums(t(apply(counts(ddsMF_aus_exp_filtered, 
                                                   normalized = TRUE)[, ddsMF_aus_exp_filtered$sex == "M" 
                                                                      & 
                                                                        ddsMF_aus_exp_filtered$organ == organ],
                                            1, 
                                            function(x) x <= 10)
                                      )
                                    ) == ncol(counts(ddsMF_aus_exp_filtered, 
                                                     normalized = TRUE)[, ddsMF_aus_exp_filtered$sex == "M" & 
                                                                          ddsMF_aus_exp_filtered$organ == organ])
                            )
  
  mal0_organ <- counts(ddsMF_aus_exp_filtered, 
                       normalized = TRUE)[rownames(counts(ddsMF_aus_exp_filtered, 
                                                          normalized = TRUE)) %in% 
                                            names(mal0_organ_names), 
                                          ddsMF_aus_exp_filtered$organ == organ]
  
  ##Pull out rows where median of female count >=10 for that organ
  fem10_organ <- apply(counts(ddsMF_aus_exp_filtered, 
                              normalized = TRUE)[, ddsMF_aus_exp_filtered$sex == "F" &
                                                   ddsMF_aus_exp_filtered$organ == organ],
                       1,
                       function(x) median(x) >=20)
  
  #Keep only the rows where male=0 and the fem median>=10
  mal0_fem10_organ <- mal0_organ[rownames(mal0_organ) %in% 
                                   names(fem10_organ[fem10_organ == TRUE]),
                                 ]
  
  # Create a new object with a name based on the organ type
  organ_femsp <- sub("$", "_female_specific", organ)
  aus_sex_specific_genes[[organ_femsp]] <- mal0_fem10_organ
  
} 
```


### Investigating the results of the sex-specific subsetting
Let's now take a look at how many sex-specific genes we have in each tissue type:
```{r sex-specific-counts, message=FALSE, warning=FALSE, eval = FALSE}
sex_specific_counts <- data.frame(tissue = rep(c("Gill", "Liver"), 2),
                                  sex = c(rep("M", 2),
                                          rep("F", 2)),
                                  num_genes = c(nrow(aus_sex_specific_genes$gill_male_specific),
                                                nrow(aus_sex_specific_genes$liver_male_specific),
                                                nrow(aus_sex_specific_genes$gill_female_specific),
                                                nrow(aus_sex_specific_genes$liver_female_specific))
                                  )

sex_specific_counts %>% 
  group_by(tissue, sex) %>% 
  summarise("Number of Sex-specific Genes" = sum(num_genes)) %>%
  kable(caption = "Sex-specific Genes", format = "html",
        col.names = c("Tissue Type", "Sex", "Number of Sex-specific Genes")) %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F, position = "left")
```

To get a better idea what is going on between the sex-specific genes vs the sex-biased genes I first determined how many overlaps there were between our sex-biased and sex-specific genes and then used plotCounts to plot the counts for each gene separately. 

In the **Male Gills** there are `r nrow(aus_sex_specific_genes$gill_male_specific[rownames(aus_sex_specific_genes$gill_male_specific) %in% rownames(gill_mal_biased),])` genes that overlap between sex-biased and sex-specific, and `r nrow(aus_sex_specific_genes$liver_male_specific[rownames(aus_sex_specific_genes$liver_male_specific) %in% rownames(liver_mal_biased),])` overlapping in the **liver**. For females we have `r nrow(aus_sex_specific_genes$gill_female_specific[rownames(aus_sex_specific_genes$gill_female_specific) %in% rownames(gill_fem_biased),])` overlapping sex-specific and sex-biased genes in the **gills**, and `r nrow(aus_sex_specific_genes$liver_female_specific[rownames(aus_sex_specific_genes$liver_female_specific) %in% rownames(liver_fem_biased),])` genes in the **liver**.

### Pulling out the gene IDs for all of the sex-specific genes
```{r}
#Run once to create the file and then commented out
#write.table(cbind(rownames(aus_sex_specific_genes$gill_male_specific)),
 #           'data/aus_malG_specific_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
#write.table(cbind(rownames(aus_sex_specific_genes$gill_female_specific)),
 #           'data/aus_femG_specific_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)

#write.table(cbind(rownames(aus_sex_specific_genes$liver_male_specific)),
 #           'data/aus_malL_specific_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
#write.table(cbind(rownames(aus_sex_specific_genes$liver_female_specific)),
 #           'data/aus_femL_specific_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

## Creating categories and binning the sex-biased genes based on degree of logFC
This binned was only done on the sex-biased genes, will not have a category for the unbiased genes. The cutoffs for the different groups are as follows:

    1. Low - biased = LFC 2 - 3
    2. Medium - biased = LFC 3 - 5
    3. High - biased = LFC 5 - 10
    4. Extreme sex bias = LFC > 10

Counts of the sex-specific genes will be added on separately as they were not classified based on fold-change but rather a presence in one sex and an absence in the other sex.

```{r bin-sex-bias, fig.width=15}

#Make a vector that contains all of the groupings
biased_bins <- c("Unbiased", "Low", "Med", "High", "Extreme", "Sex-specific")

#Create a new column in the dataset and use ifelse statements to set the category limits
#abs(logFC) was used to account for the fem-biased genes possessing negative values
logFC_long$bias_cat <- ifelse(logFC_long$bias == "UB",
                              biased_bins[1],
                              ifelse(abs(logFC_long$logFC) >= 2 & abs(logFC_long$logFC) < 3,
                                     biased_bins[2],
                                     ifelse(abs(logFC_long$logFC) >= 3 & abs(logFC_long$logFC) < 5,
                                            biased_bins[3],
                                            ifelse(abs(logFC_long$logFC) >= 5 & abs(logFC_long$logFC) < 10, 
                                                   biased_bins[4],
                                                   biased_bins[5])
                                            )
                                     )
                              ) 

#Making sure the genes we categorized as sex-specific are labeled as sex-specific for their 
#bias cat. in the dataset
organs <- c("gill", "gill", "liver", "liver")
bias <- c("MB", "FB", "MB", "FB", "MB", "FB")

for(i in 1:length(aus_sex_specific_genes)){
  tmp <- aus_sex_specific_genes[[i]]
  tmp <- as.data.frame(tmp)
  tmp$geneID <- rownames(tmp)
  
  for(j in 1:nrow(tmp)){
    
    one_gene <- tmp[j, ]
    
    if(one_gene[["geneID"]] %in%
       logFC_long[logFC_long$tissue == organs[[i]] &
                  logFC_long$bias == bias[[i]],"geneID"]){
       
      logFC_long[logFC_long$geneID == one_gene[["geneID"]] &
                   logFC_long$tissue == organs[[i]] &
                   logFC_long$bias == bias[[i]],
                 "bias_cat"] <- "Sex-specific"
    }else{
      
      one_gene_dat <- data.frame(matrix(ncol= ncol(logFC_long),
                                        nrow=1))
      colnames(one_gene_dat) <- colnames(logFC_long)
      
      one_gene_dat$tissue <- organs[[i]]
      one_gene_dat$geneID <- one_gene[["geneID"]]
      one_gene_dat$bias <- bias[[i]]
      one_gene_dat$bias_cat <- "Sex-specific"
      rownames(one_gene_dat) <- NULL
      
      logFC_long <- rbind(one_gene_dat, logFC_long)
      
      rownames(logFC_long) <- NULL
    }
  }
 }

#Create  subset of our long dataset that does not include the non-biased genes
logFC_long_noNB <- logFC_long[logFC_long$bias_cat != "Unbiased",]

#Make a table to count the number of genes in each category for each organ
bias_cat_table <- table(logFC_long_noNB$bias, logFC_long_noNB$bias_cat, logFC_long_noNB$tissue)

#Add the counts of sex-specific genes to the table
bias_cat_gill <- bias_cat_table[,, "gill"]
#write.table(bias_cat_gill, "data/bias_cat_gills.txt", row.names = TRUE)

bias_cat_liver <- bias_cat_table[,, "liver"]
#write.table(bias_cat_liver, "data/bias_cat_liver.txt", row.names = TRUE)

#Plot the counts for each tissue type
par(mfrow=c(1, 2))
barplot(bias_cat_gill[,biased_bins[biased_bins!="Unbiased"]], beside = TRUE, ylim = c(0, max(bias_cat_liver) + 1), col = c("#7fc97f", "#beaed4"),
        main = "gill", ylab = "Number of Genes", xlab = "Bias Category")
barplot(bias_cat_liver[,biased_bins[biased_bins!="Unbiased"]], beside = TRUE, ylim = c(0, max(bias_cat_liver) + 1), col = c("#7fc97f", "#beaed4"),
        main = "liver")
legend("topright", legend = c("Female-biased", "Male-biased"), fill = c("#7fc97f", "#beaed4"))


```

## Gene Ontology Analysis
We wanted to figure out two things:   

  1) What are the genes that are sex-biased within each tissue type   
  2) What are the GO terms for those genes.   

To accomplish this I first had to pull out the sequences that corresponded to the Trinity gene IDS pulled out above. Once those sequences were fetched they were blasted against the _Corythoichthys intestinalis_ genome. This was completed with the following bash script on the RCC. 

### Using BLAST in command line
```{bash, eval = FALSE}
#Create a BLAST database - Used Corythoichthys intestinalis for our BLAST database
makeblastdb -in /home/rccuser/shared/emily_files/australia_nigra_raw_reads/c_intestinalis_data/data/GCF_030265065.1/GCF_030265065.1_ASM3026506v1_genomic.fna -out c_intestinalis_genome -dbtype nucl
```

```{bash, eval = FALSE}
#!/bin/bash

#Create the arguements
input_dir_TR=$1 #Location of the .txt files that contain the trinity gene IDs
subset_fasta=$2 #Path to the subset_fasta_file script
assembly_file=$3 #Name/location of the de novo assembly
blast_database=$4
output_dir=$5 #Desired output directory for .fasta and BLAST files

## Loop through all the Trinity gene ID .txt files
for file in $input_dir_TR/*TRgenes.txt
        do

        #Extract the sample name from the file name
        sample=$(basename $file .txt)

        #Get the corresponding sequences for the Trinity Gene IDs
        echo "Running subset_fasta_file for ${sample}..."
        $subset_fasta -c -f $assembly_file -l $input_dir_TR/${sample}.txt

        #Rename the automated output from the subset_fasta_file script
        fasta_out=$(basename $assembly_file fasta)
        mv $fasta_out.subset.fasta ${sample}.fasta

        #Blast the sequences
        echo "Running BLAST for ${sample}..."
        blastn -db $blast_database -query ${sample}.fasta -out ${sample}_corythoicthys_blast.txt \
                -evalue 0.001 \
                -num_threads 12 \
                -outfmt "6 qseqid qstart qend stitle salltitles sseqid sallseqid sacc sblastnames sstart send evalue bitscore length pident gaps"       

        #Move the outputs into desired output directory
        mv ${sample}* $output_dir

done

```

This script was run as `bash ../scripts/blast_pipeline.sh TRgenes/ ../scripts/subset_fasta_file supertranscriptome.fasta c_intestinalis_genome/ cory_blast/`

There were no hits on the _C. intestinalis_ genome, so I'll try the _D. rerio_ proteom now. 

### Blasting against the Zebrafish genome
In order for us to be able to do the GO analysis, we need to BLAST the sequences against one of the PANTHER organisms. The zebrafish, _Danio reiro_ was the only fish in that list.

A BLAST database was generated with the _D. reiro_ proteome as:
`makeblastdb -in ../ncbi_dataset/data/GCF_000002035.6/protein.faa -out d_rerio_prot -dbtyp prot`

The `blastx_pipeline.sh` script was then used like above to pull out all of the corresponding sequences for each Trinity gene and then to get the BLAST results. The big difference here was that `blastx` was used instead of `blastn` because we are searching nucleotides against a protein database. Once we have the BLAST results we can read and filter them into R as we did before: `bash ../scripts/blastx_pipeline.sh TRgenes/ ../scripts/subset_fasta_file super_trinity_genes.fasta ../../genomes/d_rerio/d_rerio_prot/ TRgenes`


