---
title: "Differential Expression Analysis in _Stigmatopora argus_"
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: yes
    toc: true
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
```

``` {r library, include = TRUE, message = FALSE, warning = FALSE}
setwd("~/Documents/GitHub/pipefish_sbge/stigmatopora/S_argus")
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
library(ggpubr)
library(cowplot)
library(UpSetR)
library(ggplot2)
library(kableExtra)
library(tidyverse)
library(knitr)
library(magick)
library(multcompView)
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.argus <- readRDS("data/txi.salmon_SA.RDS")

#The samples file generated for tximport
samples <- read.table("data/SA_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)

```

The package `DESeq2` was used for the differential expression analysis outlined below.

# Single factor analysis - Comparing Males v Females across all organs
To analyze your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions. The model for this single factor analysis was run as counts ~ Sex.

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_SA <- DESeqDataSetFromTximport(txi.argus, 
                                   colData = samples,
                                   design = ~ sex)

```

The data is then pre-filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

The cutoff here was to remove rows that had counts fewer than 10 across all samples.

```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_SA)) >= 10
dds_SA <- dds_SA[keep, ]

```

After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.

```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_SA_exp <- DESeq(dds_SA)

#Compile the results
res <- results(dds_SA_exp)
res
```

Once that has finished we can now start exploring some of the single-factor analysis results
```{r investigate-DESeq-results-SA}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#How many ADJUSTED p-values were less than 0.1?
sum(res$padj < 0.1, na.rm = TRUE)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_SA_exp, alpha = 0.05)
summary(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```

## Visualizing the results
### MA-plot - MvF Differntial expression
Generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be colored if the adjusted p-value is less that 0.1.

```{r MA-plot, echo=FALSE, fig.cap="LogFC versus the mean normalized count for all of the genes. Points that are blue have a p-value less than 0.1. The two plots are the same with the only difference coming from an adjusted y-limit. Points that are triangles represent genes with a fold change higher/lower than the y-limit."}

par(mfrow=c(1,2))
plotMA(res, ylim = c(-2,2))
plotMA(res)

```

### Heatmap - overall expression
We can also generate a heat map to look at overall expression levels across our samples. Note, this is not differentially expressed genes.

```{r heatmap, echo=FALSE, fig.cap= "Heatmap showing the expression level across all organs for the top 20 genes with the highest expression levels."}
df <- as.data.frame(samples[,c("sex", "organ")])
rownames(df) <- samples$ID
df$organ <- as.character(df$organ)

organ_cols<-c("GILL" = "#20B2AA",
              "GONAD" = "#EEB422", 
              "LIVER" = "#EE8262")

sex_cols<-c("F" = "#7fc97f", "M" = "#beaed4")

ann_colors = list(sex=sex_cols,
                  organ=organ_cols)

col_order <- c(rownames(df[df$sex=="F" & df$organ=="GILL",]),
               rownames(df[df$sex=="M" & df$organ=="GILL",]),
               rownames(df[df$sex=="F" & df$organ=="GONAD",]),
               rownames(df[df$sex=="M" & df$organ=="GONAD",]),
               rownames(df[df$sex=="F" & df$organ=="LIVER",]),
               rownames(df[df$sex=="M" & df$organ=="LIVER",]))


#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(dds_SA_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]


#Transform the data
vsd <- vst(dds_SA_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select, col_order], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df, 
         annotation_colors = ann_colors)
```

From the heatmap (Fig. \@ref(fig:heatmap)) we can see that for all of these top 20 expressed genes the highest expression in found in the liver. There are also a few genes that appear to be highly expressed across all of the tissue types. With this heatmap we can also pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.

```{r heatmap_genes}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))
#Run once to create the file and then commented out
#write.table(heatmap_TG,
 #           'heatmap_trinitygenes.txt',
  #          sep = "",
   #         quote=FALSE,
    #        row.names = FALSE,
     #       col.names = FALSE)
```

### Sample-dist Heatmap

We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples.

```{r sample-dist, echo=FALSE, fig.cap="Sample-to-sample distances heatmap showing the similarities and differences between all of the samples. The darker the color, the more similar the samples are. The diagonal line of very dark blue represents the comparissons between the same samples.", warning=FALSE}
#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$sex, vsd$organ, sep = "-")
colnames(sampleDistsMatrix) <- vsd$ID

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)
```

### PCA plots
Next we can look at PCA plots for our samples to see how our different samples are grouping. The size of the points for these plots will be scaled by the mating success of the individual to see if that may be at all correlated with the sample grouping.

```{r pca-scaled-organ, echo=FALSE, fig.cap="Principal components analysis reflects that most variation corresponds to differences in expression between organs (green: gill; yellow: gonad; pink: liver), rather than variation due to sex (circle = female; triangle = male) or mating status (the size of points reflects an individual’s mating success). The first axis explains 32% of variation in the dataset, with gills and liver, two types of somatic tissues, sitting on opposite ends of that axis. The second axis explains 23% of the variation in gene expression, with somatic tissues on the opposite side of the axis from the gonads.", warning=FALSE}

#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$organ <- samples$organ
pca_plotting_data$sex <- samples$sex

#Calculate the percent variation attributed to each axis 
pcaData <- plotPCA(vsd, intgroup = c("organ", "sex"), returnData = TRUE, ntop = nrow(dds_SA))
percentVar <- round(100 * attr(pcaData, "percentVar"))


my_colors <- c("GILL" = "#6E8B3D75", "GONAD" = "#EEB42275", "LIVER" = "#EE826275")
my_shapes <- c(16, 17)

par(mfrow=c(1,2))
ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"))
```

```{r pca-scaled-sex, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with colour now representing sex and shape representing organ type (left) as well as the comparison of PC1 and PC3 which shows variation in expression patterns between males and females, particularly in the gonads, explains approximately 10% of the variance in gene expression data, as reflected by the third principal components axis."}

par(mfrow=c(1,2))
my_colors2 <- c("F" = "#7fc97f75", "M" = "#beaed475" )
my_shapes2 <- c(16, 17, 15)
pc1v2 <- ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = sex, shape = organ)) +
  scale_color_manual(values = my_colors2) +
  scale_shape_manual(values = my_shapes2) +
  geom_point(aes(size = organ)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"))

pc1v2

pc1v3 <- ggplot(pca_plotting_data, aes(x=PC1, y=PC3, color = sex, shape = organ)) +
  scale_color_manual(values = my_colors2) +
  scale_shape_manual(values = my_shapes2) +
  geom_point(aes(size = organ)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC3: 10% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"))

pc1v3

```


```{r pca-scaled-organ, echo=FALSE, fig.cap="Principal components analysis reflects that most variation corresponds to differences in expression between organs (green: gill; yellow: gonad; pink: liver), rather than variation due to sex (circle = female; triangle = male) or mating status (the size of points reflects an individual’s mating success). The first axis explains 31% of variation in the dataset, with gills and liver, two types of somatic tissues, sitting on opposite ends of that axis. The second axis explains 20% of the variation in gene expression, with somatic tissues on the opposite side of the axis from the gonads.", warning=FALSE}

#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$organ <- samples$organ
pca_plotting_data$sex <- samples$sex

#Calculate the percent variation attributed to each axis 
pcaData <- plotPCA(vsd, intgroup = c("organ", "sex"), returnData = TRUE, ntop = nrow(dds_SA))
percentVar <- round(100 * attr(pcaData, "percentVar"))


my_colors <- c("GILL" = "#6E8B3D75", "GONAD" = "#EEB42275", "LIVER" = "#EE826275")
my_shapes <- c(16, 17)

par(mfrow=c(1,2))
ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(aes(size = organ)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"))
```

```{r pca-scaled-sex, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with colour now representing sex and shape representing organ type (left) as well as the comparison of PC1 and PC3 which shows variation in expression patterns between males and females, particularly in the gonads, explains approximately 10% of the variance in gene expression data, as reflected by the third principal components axis."}

par(mfrow=c(1,2))
my_colors2 <- c("F" = "#7fc97f75", "M" = "#beaed475" )
my_shapes2 <- c(16, 17, 15)
pc1v2 <- ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = sex, shape = organ)) +
  scale_color_manual(values = my_colors2) +
  scale_shape_manual(values = my_shapes2) +
  geom_point(aes(size = organ)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"))

pc1v2

pc1v3 <- ggplot(pca_plotting_data, aes(x=PC1, y=PC3, color = sex, shape = organ)) +
  scale_color_manual(values = my_colors2) +
  scale_shape_manual(values = my_shapes2) +
  geom_point(aes(size = organ)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC3: 10% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"))

pc1v3

```


# Multifactor design - Comparing M v F across the diff tissue types
If we investigate the column data of our DESeq dataset we can see that each sample has both a sex and organ type attached to it, we will be using these two factors in our multi-factor analysis. In this multi-factor analysis the model was run as counts ~ group, where group included both the sex and the organ type (i.e. MLiver, FLiver, etc.). The sample "S34" appeared to a bit of an outlier in the above analysis, particularly clear in the sample-dist heatmap and the PCA, so it will be removed from the analysis going forward.

```{r multi-factor-analysis, message=FALSE, warning=FALSE}
#Create an additional column the has the two conditions combined(sex and organ type)
samples$group <- factor(paste0(samples$sex, samples$organ))

##Create a copy of the DESeq dataset
ddsMF_SA <- DESeqDataSetFromTximport(txi.argus,
                                     colData = samples,
                                     design = ~ group)


##Filter the dataset, only keeping rows that have at least 10 reads total
keep <- rowSums(counts(ddsMF_SA)) >= 10 #& rowSums(counts(ddsMF_SA)) < 1e6
ddsMF_SA <- ddsMF_SA[keep, ]

#Run the differential expression analysis
ddsMF_SA_exp <- DESeq(ddsMF_SA)
resultsNames(ddsMF_SA_exp)
```

## Invesitgate the results of the differential expression
Thanks to the multi-factor analysis we can now explore differential expression between all of the different combinations:

  - Male Liver v. Female Liver
  - Male Gill v. Female Gill
  - Male Gonad v. Female Gonad
  - All of the within sex tissue comparisons (e.g. Male Liver v. Male Gill, etc.)

### M-F Liver Comparisson
```{r m-f-liver-results}
##Pulling out the liver M-F results with an alpha of 0.05
liver_con_res <- results(ddsMF_SA_exp, contrast = c("group", "MLIVER", "FLIVER"),
                         alpha = 0.05)
liver_con_res$trin_geneid <- rownames(liver_con_res)
head(liver_con_res)

summary(liver_con_res)
```

There are many criteria that have been employed previously to denote sex bias. For this study we are classifying sex-biased genes as **genes with a p-value < 0.05 AND a logFC $\ge$ |2|**. With that criteria in the liver there are `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. 

I have then pulled out all of the male-biased, female-biased, and non-biased genes based on the criteria outlined above. I determined non-biased genes as a p-vaule > 0.05.

```{r liver-sex-biased}
#Removing the rows where padj. is NA in results
liver_con_res_noNA <- liver_con_res[!is.na(liver_con_res$padj),]
summary(liver_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating a vector that contains all of the male-biased and female-biased genes in the liver
liver_mal_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange >= 2
                                             & liver_con_res_noNA$padj <= 0.05),]
liver_fem_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange <= -2 
                                             & liver_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the liver
liver_non_biased <- liver_con_res_noNA[which(liver_con_res_noNA$padj > 0.05),]

```

I will be generating a table that outlines the genes that are the most male or female biased in each organ type. To do this I need to pull the top 50 differentially expressed genes in males or females, get the Trinity gene IDs and then BLAST the corresponding sequences. Here I am pulling out the Trinity gene IDs for those genes.

```{r top50-liver}
#Creating a subset of the results where the p-value is less than 0.05
liver_con_res_p05 <- liver_con_res_noNA[liver_con_res_noNA$padj <= 0.05, ]


#Pulling the top 50 diff expressed genes in Males
top50_MLiver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(
  liver_con_res_p05$log2FoldChange, decreasing = TRUE),], n=50)))
#Run once to create the file and then commented out
#write.table(top50_MLiver_trin_genes,
 #          'SA_maleL_top50TRgenes.txt', 
  #         sep = "", 
   #        quote=FALSE, 
    #         row.names = FALSE, 
     #        col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_Fliver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(
  liver_con_res_p05$log2FoldChange, decreasing = FALSE),], n=50)))

#Run once to create the file and then commented out
#write.table(top50_Fliver_trin_genes,
 #           'SA_femL_top50TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### M-F Gill Comparisson
```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res <- results(ddsMF_SA_exp, contrast = c("group", "MGILL", "FGILL"),
                        alpha = 0.05)
gill_con_res$trin_geneid <- rownames(gill_con_res)
head(gill_con_res)

summary(gill_con_res)
```

In the gills there were `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

I have then pulled out all of the male-biased, female-biased, and non-biased genes to save them into their own objects.

```{r gill-sex-biased}
#Removing the rows where padj. is NA in results
gill_con_res_noNA <- gill_con_res[!is.na(gill_con_res$padj), ]
summary(gill_con_res_noNA) #We can now see that there are no outliers or 
                           #low counts since the NAs have been removed

#Creating a vector that contains all of the male-biased and female-biased genes in the gills
gill_mal_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange >= 2 
                                           & gill_con_res_noNA$padj <= 0.05),]
gill_fem_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange <= -2 
                                           & gill_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gills, p>0.05
gill_non_biased <- gill_con_res_noNA[which(gill_con_res_noNA$padj > 0.05),]
```

Here I am getting the trinity gene IDs that correspond to the top 50 differentially expressed genes for males and females. Because there are only 27 female-biased and male-biased genes I will only use those.

```{r top50-gill}
#Creating a subset of results where p-value is less than 0.05
gill_con_res_p05 <- gill_con_res_noNA[gill_con_res_noNA$padj <= 0.05,]

#Pulling the top 27 diff expressed genes in Males
top27_MGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(
  gill_con_res_p05$log2FoldChange, decreasing = TRUE),], n=27)))

#Run once to create the file and then commented out
#write.table(top27_MGill_trin_genes,
 #           'SA_maleG_top27TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)

#Pulling the top 27 diff expressed genes in Females 
top27_FGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(
  gill_con_res_p05$log2FoldChange, decreasing = FALSE),], n=27)))

#Run once to create the file and then commented out
#write.table(top27_FGill_trin_genes,
 #           'SA_femG_top27TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### M-F Gonad Comparisson
```{r m-f-gonads}
##Pulling out the gonad M-F results
gonad_con_res <- results(ddsMF_SA_exp, 
                         contrast = c("group", "MGONAD", "FGONAD"),
                         alpha = 0.05)
gonad_con_res$trin_geneid <- rownames(gonad_con_res)
head(gonad_con_res)

summary(gonad_con_res)
```

Between the ovaries and testes (gonads) there were `r sum(gonad_con_res$padj <= 0.05 & gonad_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gonad_con_res$padj <= 0.05 & gonad_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

From the gonad M-F results I have then pulled out all of the male-biased, female-biased, and non-biased genes and saved them to their own objects.

```{r gonad-sex-biased}
#Removing the rows where padj. is NA in results
gonad_con_res_noNA <- gonad_con_res[!is.na(gonad_con_res$padj), ]
summary(gonad_con_res_noNA) #We can now see that there are no outliers or 
                            #low counts since the NAs have been removed

#Creating an object that contains all of the male-biased and female-biased genes in the gonads
gonad_mal_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$log2FoldChange >= 2 
                                             & gonad_con_res_noNA$padj <= 0.05),]
gonad_fem_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$log2FoldChange <= -2 
                                             & gonad_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gonads, p>0.05
gonad_non_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$padj > 0.05),]
```

I then pulled out the trinity gene IDs that corresponded to the top 50 differentially expressed genes in males and females so that I could BLAST the sequences.
```{r top50-gonad}
#Creating a subset of the results where the p-value is less than 0.05
gonad_con_res_p05 <- gonad_con_res_noNA[gonad_con_res_noNA$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_MGonad_trin_genes <- cbind(rownames(head(gonad_con_res_p05[order(
  gonad_con_res_p05$log2FoldChange, decreasing = TRUE),], n=50)))

#Run once to create the file and then commented out
#write.table(top50_MGonad_trin_genes,
 #        'SA_maleGon_top50TRgenes.txt', 
  #        sep = "", 
   #       quote=FALSE, 
    #      row.names = FALSE, 
     #     col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FGonad_trin_genes <- cbind(rownames(head(gonad_con_res_p05[order(
  gonad_con_res_p05$log2FoldChange, decreasing = FALSE),], n=50)))

#Run once to create the file and then commented out
#write.table(top50_FGonad_trin_genes,
 #           'SA_femGon_top50TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```