---
title: "Differential Expression Analysis in _Stigmatopora argus_"
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: yes
    toc: true
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
```

This .Rmd is the analysis that follows the reassignment of identity to _S. argus_ samples after initial visualisation and inspection of previous sample identities was thought to be incorrect.

``` {r library, include = TRUE, message = FALSE, warning = FALSE}
setwd("~/Documents/GitHub/pipefish_sbge/stigmatopora/S_argus")
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
library(ggpubr)
library(cowplot)
library(UpSetR)
library(ggplot2)
library(kableExtra)
library(tidyverse)
library(knitr)
library(magick)
library(multcompView)
library(spfTools)
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.argus <- readRDS("data/txi.salmon_SA.RDS")

#The samples file generated for tximport
samples <- read.table("data/SA_sample_ID.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)

```

The package `DESeq2` was used for the differential expression analysis outlined below.

# Single factor analysis - Comparing Males v Females across all organs
To analyse your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions. The model for this single factor analysis was run as counts ~ Sex.

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_SA <- DESeqDataSetFromTximport(txi.argus, 
                                   colData = samples,
                                   design = ~ sex)

```

The data is then pre-filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

The cutoff here was to remove rows that had counts fewer than 10 across all samples.

```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_SA)) >= 10
dds_SA <- dds_SA[keep, ]

```

After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.

```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_SA_exp <- DESeq(dds_SA)

#Compile the results
res <- results(dds_SA_exp)
res
```

Once that has finished we can now start exploring some of the single-factor analysis results
```{r investigate-DESeq-results-SA}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#How many ADJUSTED p-values were less than 0.1?
sum(res$padj < 0.1, na.rm = TRUE)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_SA_exp, alpha = 0.05)
summary(res05)
head(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```

## Visualizing the results
### MA-plot - MvF Differntial expression
Generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be coloured if the adjusted p-value is less that 0.1.

```{r MA-plot, echo=FALSE, fig.cap="LogFC versus the mean normalized count for all of the genes. Points that are blue have a p-value less than 0.1. The two plots are the same with the only difference coming from an adjusted y-limit. Points that are triangles represent genes with a fold change higher/lower than the y-limit."}

par(mfrow=c(1,2))
plotMA(res, ylim = c(-2,2))
plotMA(res)

```

### Heatmap - overall expression
We can also generate a heat map to look at overall expression levels across our samples. Note, this is not differentially expressed genes.

```{r heatmap, echo=FALSE, fig.cap= "Heatmap showing the expression level across all organs for the top 20 genes with the highest expression levels."}
df <- as.data.frame(samples[,c("sex", "organ")])
rownames(df) <- samples$ID
df$organ <- as.character(df$organ)

organ_cols<-c("gill" = "#20B2AA",
              "gonad" = "#EEB422", 
              "liver" = "#EE8262")

sex_cols<-c("F" = "#7fc97f", "M" = "#beaed4")

ann_colors = list(sex=sex_cols,
                  organ=organ_cols)

col_order <- c(rownames(df[df$sex=="F" & df$organ=="gill",]),
               rownames(df[df$sex=="M" & df$organ=="gill",]),
               rownames(df[df$sex=="F" & df$organ=="gonad",]),
               rownames(df[df$sex=="M" & df$organ=="gonad",]),
               rownames(df[df$sex=="F" & df$organ=="liver",]),
               rownames(df[df$sex=="M" & df$organ=="liver",]))


#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(dds_SA_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]


#Transform the data
vsd <- vst(dds_SA_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select, col_order], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df, 
         annotation_colors = ann_colors)
```

From the heatmap we can see that for all of these top 20 expressed genes the highest expression in found in the gonads, which is different from the other two _Stigmatopora_ spp. that had the highest expression in the liver. There are also a few genes that appear to be highly expressed across all of the tissue types. 

With this heatmap we can also pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.

```{r heatmap_genes}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))
#Run once to create the file and then commented out
#write.table(heatmap_TG,
 #           'data/heatmap_trinitygenes.txt',
  #          sep = "",
   #         quote=FALSE,
    #        row.names = FALSE,
     #       col.names = FALSE)
```

### Sample-dist Heatmap

We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples. The darker the colour, the more similar the samples are. The diagonal line of very dark blue represents the comparison within the same samples.

```{r sample-dist, echo=FALSE, fig.cap="Sample-to-sample distances heatmap showing the similarities and differences between all of the samples. The darker the colour, the more similar the samples are. The diagonal line of very dark blue represents the comparisons between the same samples.", warning=FALSE}
#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$sex, vsd$organ, sep = "-")
colnames(sampleDistsMatrix) <- vsd$ID

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)
```

We can see that the highest similarities (aside from within samples comparisons) is among samples from the same organ. After that, the next highest similarities can be seen between the male and female gonads. S39 (male liver) is clustering with gills - might need to remove it later. We'll see what it looks like in the PCAs. 

### PCA plots

```{r pca-pairs-plot, echo=FALSE, fig.cap="Principal components analysis pairsplot showing PCA axis 1 - 4."}
#Generate the PCA data to create the pairs plot, using the PCAtools package
p <- pca(assay(vsd), 
         metadata = colData(dds_SA))

#Create the pairs plot
pairsplot(p,
          components = getComponents(p, c(1:4)),
          colby = "organ",
          triangle = FALSE,
          hline = 0, vline = 0,
          colkey = c("gill" = "#6E8B3D75", 
                     "gonad" = "#EEB42275", 
                     "liver" = "#EE826275"),
          shape = "sex",
          shapekey = c(16, 17),
          pointSize = 3,
          margingaps = unit(c(0.1, 0.1, 0.1, 0.1), 'cm'),
          legendPosition = "left", legendIconSize = 3, legendLabSize = 10)

```

From the pairs plot we can see that PC1 appears to be explaining differences between the gonad and the other tissues, PC2 appears to be explaining differences between gonadal and somatic tissue. PC3 appears to be explaining differences between the sexes, especially in the gonads. S39 is still clustering away from the other organs, so I suspect we will remove it. 

Next we can look at PCA plots for our samples to see how our different samples are grouping.
```{r pca-plot, echo=FALSE, message=FALSE, warning=FALSE}

#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$organ <- samples$organ
pca_plotting_data$sex <- samples$sex
pca_plotting_data$ID <- samples$ID

#Calculate the percent variation attributed to each axis 
percentVar <- round(p$variance, digits = 2)


my_colors <- c("gill" = "#6E8B3D75", "gonad" = "#EEB42275", "liver" = "#EE826275")
my_shapes <- c(16, 17)

ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed') + 
  geom_text(aes(label = ID), vjust = -1, hjust = 0.5)
```
S39 appears to be grouping away from the other organs. It also had a low mapping rate compared to the other samples (~60% for S39 compared to ~98% for all the others).

### Remove outlier
Based on the heatmaps and PCA plots, I am going to remove S39.

```{r rm-39-DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_SA <- DESeqDataSetFromTximport(txi.argus, 
                                   colData = samples,
                                   design = ~ sex)

```


```{r rm-39-pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_SA)) >= 10
dds_SA <- dds_SA[keep, ]

#remove S39
dds_SA <- dds_SA[, !(dds_SA$ID %in% c("S39"))]

```

After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.

```{r rm-39-diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_SA_exp <- DESeq(dds_SA)

#Compile the results
res <- results(dds_SA_exp)
res
```

```{r rm-39-investigate-DESeq-results-SA}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#How many ADJUSTED p-values were less than 0.1?
sum(res$padj < 0.1, na.rm = TRUE)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_SA_exp, alpha = 0.05)
summary(res05)
head(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```

### Heatmap - overall expression
Let's visualise again - generate a heat map to look at overall expression levels across our samples. Note, this is not differentially expressed genes.

```{r rm-39-heatmap, echo=FALSE, fig.cap= "Heatmap showing the expression level across all organs for the top 20 genes with the highest expression levels."}
samples <- samples[!(samples$ID %in% c("S39")),]
df <- as.data.frame(samples[,c("sex", "organ")])
rownames(df) <- samples$ID
df$organ <- as.character(df$organ)

organ_cols<-c("gill" = "#20B2AA",
              "gonad" = "#EEB422", 
              "liver" = "#EE8262")

sex_cols<-c("F" = "#7fc97f", "M" = "#beaed4")

ann_colors = list(sex=sex_cols,
                  organ=organ_cols)

col_order <- c(rownames(df[df$sex=="F" & df$organ=="gill",]),
               rownames(df[df$sex=="M" & df$organ=="gill",]),
               rownames(df[df$sex=="F" & df$organ=="gonad",]),
               rownames(df[df$sex=="M" & df$organ=="gonad",]),
               rownames(df[df$sex=="F" & df$organ=="liver",]),
               rownames(df[df$sex=="M" & df$organ=="liver",]))


#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(dds_SA_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]


#Transform the data
vsd <- vst(dds_SA_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select, col_order], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df, 
         annotation_colors = ann_colors)
```

From the new heatmap, we can see that for all of these top 20 expressed genes the highest expression in found in the gonads. There are also a few genes that appear to be highly expressed across all of the tissue types. 

With this heatmap we can also pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.

```{r rm-39-heatmap_genes}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG_rm39 <- cbind(row.names(assay(vsd)[select,]))
#Run once to create the file and then commented out
#write.table(heatmap_TG_rm39,
 #           'data/heatmap_trinitygenes_rm39.txt',
  #          sep = "",
   #         quote=FALSE,
    #        row.names = FALSE,
     #       col.names = FALSE)
```

### PCA plots

```{r rm-39-pca-pairs-plot, echo=FALSE, fig.cap="Principal components analysis pairsplot showing PCA axis 1 - 4."}
#Generate the PCA data to create the pairs plot, using the PCAtools package
p <- pca(assay(vsd), 
         metadata = colData(dds_SA))

#Create the pairs plot
pairsplot(p,
          components = getComponents(p, c(1:4)),
          colby = "organ",
          triangle = FALSE,
          hline = 0, vline = 0,
          colkey = c("gill" = "#6E8B3D75", 
                     "gonad" = "#EEB42275", 
                     "liver" = "#EE826275"),
          shape = "sex",
          shapekey = c(16, 17),
          pointSize = 3,
          margingaps = unit(c(0.1, 0.1, 0.1, 0.1), 'cm'),
          legendPosition = "left", legendIconSize = 3, legendLabSize = 10)

```

From the pairs plot we can see that PC1 appears to be explaining differences between the gonads and the other tissues, PC2 appears to be explaining differences between gonadal and somatic tissues, PC3 seems to account for differences between the sexes, particularly in the gonad samples, and in PC4 we can see three gill samples sticking out from all the rest. I am going to plot all of PC2 - 4 against PC1 below for a clearer picture.

Next we can look at PCA plots for our samples to see how our different samples are grouping.
```{r rm-39-pca-plot, echo=FALSE, message=FALSE, warning=FALSE}

#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$organ <- samples$organ
pca_plotting_data$sex <- samples$sex
pca_plotting_data$ID <- samples$ID

#Calculate the percent variation attributed to each axis 
percentVar <- round(p$variance, digits = 2)


my_colors <- c("gill" = "#6E8B3D75", "gonad" = "#EEB42275", "liver" = "#EE826275")
my_shapes <- c(16, 17)

ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed') + 
  geom_text(aes(label = ID), vjust = -1, hjust = 0.5)
```

```{r pca1v3, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC3. This third axis explains 9.08% of the variation, largely attributed to sex differences between the gonads samples.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC3, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC3: ",percentVar[3],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed') #+ 
  #geom_text(aes(label = ID), vjust = -1, hjust = 0.5)
```

```{r pca1v4, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC4. This fourth axis explains 6.63% of the variation, largely attributed to three of the gill samples.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC4, color = organ, shape = sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC4: ",percentVar[4],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed') #+ 
  #geom_text(aes(label = ID), vjust = -1, hjust = 0.5)
```

# Multifactor design - Comparing M v F across the different tissue types 
If we investigate the column of our DESeq dataset, we can see that each sample has both a sex and an organ type attached to it - we will be using these two factors in our multifactor analysis.  The sample "S39" appeared to a bit of an outlier in the above analysis, particularly clear in the sample-dist heatmap and the PCA, so it will be removed from the analysis going forward.

```{r multi-factor-analysis, message=FALSE, warning=FALSE}
#import samples.txt 
samples <- read.table("data/SA_sample_ID.txt", header = TRUE)

#Create an additional column the has the two conditions combined(sex and organ type)
samples$group <- factor(paste0(samples$sex, samples$organ))

##Create a copy of the DESeq dataset
ddsMF_SA <- DESeqDataSetFromTximport(txi.argus,
                                     colData = samples,
                                     design = ~ group)

##Remove S19 and S39 from the dataset
ddsMF_SA <- ddsMF_SA[, ddsMF_SA$ID != "S19","S39"]

##Filter the dataset, only keeping rows that have at least 10 reads total, but less than 1,000,000
keep <- rowSums(counts(ddsMF_SA)) >= 10 & rowSums(counts(ddsMF_SA)) < 1e6
ddsMF_SA <- ddsMF_SA[keep, ]

#Run the differential expression analysis
ddsMF_SA_exp <- DESeq(ddsMF_SA)
resultsNames(ddsMF_SA_exp)

```

## Invesitgate the results of the differential expression analysis
Thanks to the multi-factor analysis we can now explore differential expression between all of the different combinations:   

  - Male Liver v. Female Liver 
  - Male Gill v. Female Gill  
  - Male Gonad v. Female Gonad 
  - All of the within sex tissue comparisons (e.g. Male Liver v. Male Gill, etc.)

### M-F liver comparison
```{r m-f-liver-results}
##Pulling out the liver M-F results with an alpha of 0.05
liver_con_res <- results(ddsMF_SA_exp, contrast = c("group", "Mliver", "Fliver"),
                         alpha = 0.05)
liver_con_res$trin_geneid <- rownames(liver_con_res)
head(liver_con_res)

summary(liver_con_res)
```

There are many criteria that have been employed previously to denote sex bias. For this study we are classifying sex-biased genes as **genes with a p-value < 0.05 AND a logFC $\ge$ |2|**. With that criteria in the liver there are `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(liver_con_res$padj <= 0.05 & liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. 

I have then pulled out all of the male-biased, female-biased, and non-biased genes based on the criteria outlined above. I determined non-biased genes as a p-vaule > 0.05.

```{r liver-sex-biased}
#Removing the rows where padj. is NA in results
liver_con_res_noNA <- liver_con_res[!is.na(liver_con_res$padj),]
summary(liver_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#write.csv(as.data.frame(liver_con_res_noNA), "data/liver_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the liver
liver_mal_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange >= 2
                                             & liver_con_res_noNA$padj <= 0.05),]
liver_fem_biased <- liver_con_res_noNA[which(liver_con_res_noNA$log2FoldChange <= -2 
                                             & liver_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the liver
liver_non_biased <- liver_con_res_noNA[which(liver_con_res_noNA$padj > 0.05),]

```

I will be generating a table that outlines the genes that are the most male or female biased in each organ type. To do this I need to pull all differentially expressed genes in males or females, get the Trinity gene IDs and then BLAST the corresponding sequences. Here I am pulling out the Trinity gene IDs for those genes.


```{r all-liver}
#Creating a subset of the results where the p-value is less than 0.05
liver_con_res_p05 <- liver_con_res_noNA[liver_con_res_noNA$padj <= 0.05, ]


#Pulling the all diff expressed genes in Males
MLiver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(
  liver_con_res_p05$log2FoldChange, decreasing = TRUE),], n=354)))
#Run once to create the file and then commented out
#write.table(MLiver_trin_genes,
 #           'SA_maleL_TRgenes.txt', 
  #           sep = "", 
   #          quote=FALSE, 
    #         row.names = FALSE, 
     #        col.names = FALSE)

#Pulling all diff expressed genes in Females
Fliver_trin_genes <- cbind(rownames(head(liver_con_res_p05[order(
  liver_con_res_p05$log2FoldChange, decreasing = FALSE),], n=267)))

#Run once to create the file and then commented out
#write.table(Fliver_trin_genes,
 #           'SA_femL_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### M-F Gill Comparisson
```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res <- results(ddsMF_SA_exp, contrast = c("group", "Mgill", "Fgill"),
                        alpha = 0.05)
gill_con_res$trin_geneid <- rownames(gill_con_res)
head(gill_con_res)

summary(gill_con_res)
```

In the gills there were `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res$padj <= 0.05 & gill_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

I have then pulled out all of the male-biased, female-biased, and non-biased genes to save them into their own objects.

```{r gill-sex-biased}
#Removing the rows where padj. is NA in results
gill_con_res_noNA <- gill_con_res[!is.na(gill_con_res$padj), ]
summary(gill_con_res_noNA) #We can now see that there are no outliers or 
                           #low counts since the NAs have been removed
#write.csv(as.data.frame(gill_con_res_noNA), "data/gill_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the gills
gill_mal_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange >= 2 
                                           & gill_con_res_noNA$padj <= 0.05),]
gill_fem_biased <- gill_con_res_noNA[which(gill_con_res_noNA$log2FoldChange <= -2 
                                           & gill_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gills, p>0.05
gill_non_biased <- gill_con_res_noNA[which(gill_con_res_noNA$padj > 0.05),]
```

Here I am getting the trinity gene IDs that correspond to all differentially expressed genes for males and females. Because there are only 22 female-biased and male-biased genes I will only use those.

```{r all-gill}
#Creating a subset of results where p-value is less than 0.05
gill_con_res_p05 <- gill_con_res_noNA[gill_con_res_noNA$padj <= 0.05,]

#Pulling all diff expressed genes in Males
MGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(
  gill_con_res_p05$log2FoldChange, decreasing = TRUE),], n=13)))

#Run once to create the file and then commented out
#write.table(MGill_trin_genes,
 #           'SA_maleG_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)

#Pulling all diff expressed genes in Females 
FGill_trin_genes <- cbind(rownames(head(gill_con_res_p05[order(
  gill_con_res_p05$log2FoldChange, decreasing = FALSE),], n=182)))

#Run once to create the file and then commented out
#write.table(FGill_trin_genes,
 #           'SA_femG_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### M-F Gonad Comparisson
```{r m-f-gonads}
##Pulling out the gonad M-F results
gonad_con_res <- results(ddsMF_SA_exp, 
                         contrast = c("group", "Mgonad", "Fgonad"),
                         alpha = 0.05)
gonad_con_res$trin_geneid <- rownames(gonad_con_res)
head(gonad_con_res)

summary(gonad_con_res)
```

Between the ovaries and testes (gonads) there were `r sum(gonad_con_res$padj <= 0.05 & gonad_con_res$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gonad_con_res$padj <= 0.05 & gonad_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

From the gonad M-F results I have then pulled out all of the male-biased, female-biased, and non-biased genes and saved them to their own objects.

```{r gonad-sex-biased}
#Removing the rows where padj. is NA in results
gonad_con_res_noNA <- gonad_con_res[!is.na(gonad_con_res$padj), ]
summary(gonad_con_res_noNA) #We can now see that there are no outliers or 
                            #low counts since the NAs have been removed
#write.csv(as.data.frame(gonad_con_res_noNA), "data/gonad_res.csv", row.names = TRUE)

#Creating an object that contains all of the male-biased and female-biased genes in the gonads
gonad_mal_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$log2FoldChange >= 2 
                                             & gonad_con_res_noNA$padj <= 0.05),]
gonad_fem_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$log2FoldChange <= -2 
                                             & gonad_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gonads, p>0.05
gonad_non_biased <- gonad_con_res_noNA[which(gonad_con_res_noNA$padj > 0.05),]
```

I then pulled out the trinity gene IDs that corresponded to all differentially expressed genes in males and females so that I could BLAST the sequences.


```{r all-gonad}
#Creating a subset of the results where the p-value is less than 0.05
gonad_con_res_p05 <- gonad_con_res_noNA[gonad_con_res_noNA$padj <= 0.05,]

#Pulling all diff expressed genes in Males
MGonad_trin_genes <- cbind(rownames(head(gonad_con_res_p05[order(
  gonad_con_res_p05$log2FoldChange, decreasing = TRUE),], n=4992)))

#Run once to create the file and then commented out
#write.table(MGonad_trin_genes,
 #        'SA_maleGon_TRgenes.txt', 
  #        sep = "", 
   #       quote=FALSE, 
    #      row.names = FALSE, 
     #     col.names = FALSE)

#Pulling all diff expressed genes in Females
FGonad_trin_genes <- cbind(rownames(head(gonad_con_res_p05[order(
  gonad_con_res_p05$log2FoldChange, decreasing = FALSE),], n=1163)))

#Run once to create the file and then commented out
#write.table(FGonad_trin_genes,
 #           'SA_femGon_TRgenes.txt', 
  #          sep = "", 
   #         quote=FALSE, 
    #        row.names = FALSE, 
     #       col.names = FALSE)
```

### MA Plots
```{r MA-plot-MF, echo=FALSE, fig.height= 15, fig.width=8, fig.cap="MA-plots generated for each organ type that compares logFC to the mean expression. Female-biased and male-biased genes are represented by colour (female: green; male: purple) and are determined with a fold change cutoff of 2 and a p-value cutoff of 0.05"}
liver_MA <- ggmaplot(liver_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Liver" %->% "Female Liver"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gill_MA <- ggmaplot(gill_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Gill" %->% "Female Gill"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gonad_MA <- ggmaplot(gonad_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Testes" %->% "Ovaries"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "gray27"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

#Merge all the plots into one
plot_grid(liver_MA, gill_MA, gonad_MA,
          ncol = 1,
          nrow = 3,
          labels = c('A', 'B', 'C'),
          label_fontfamily = 'sans',
          label_fontface = 'bold',
          label_size = 18,
          rel_heights = c(4.10, 4.10, 4.10))

```

From the MA plots we can see the highest number of biased genes are found in the gonads (Fig. \@ref(fig:MA-plot-MF)) and then the liver and lastly the gills. In both the liver and the gonads there are more female-biased genes than male-biased genes.

## Creating an Upset Plot
```{r upset-plot-mal, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in males."}
#Male-biased genes
listInputMB <- list("Testes" = rownames(gonad_mal_biased),
                  "Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased)
                  )
upset(fromList(listInputMB),
      mainbar.y.label = "Number of Shared Male-Biased Genes",
      sets.x.label = "Total Number of Male-Biased Genes",
      point.size = 3)
```

```{r upset-plot-fem, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in females."}
#Female-biased genes
listInputFB <- list("Ovaries" = rownames(gonad_fem_biased),
                  "Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased)
                  )
upset(fromList(listInputFB),
      mainbar.y.label = "Number of Shared Female-Biased Genes",
      sets.x.label = "Total Number of Female-Biased Genes",
      point.size = 3)
```

```{r upset-plot-all, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs across both sexes."}
#Exporting data used to make the Upset plot (run once and them commented out)
#write.table(as.data.frame(gonad_mal_biased), "data/gonad_mal_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(gonad_fem_biased), "data/gonad_fem_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(gill_mal_biased), "data/gill_mal_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(gill_fem_biased), "data/gill_fem_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(liver_mal_biased), "data/liver_mal_biased.txt", 
 #           row.names = TRUE)
#write.table(as.data.frame(liver_fem_biased), "data/liver_fem_biased.txt", 
 #           row.names = TRUE)

#Checking there is no overlap between the two sexes
listInputall <- list("Testes" = rownames(gonad_mal_biased),
                  "Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased),
                  "Ovaries" = rownames(gonad_fem_biased),
                  "Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased))

upset(fromList(listInputall),
      mainbar.y.label = "Number of Shared Sex-Biased Genes",
      sets.x.label = "Total Number of Sex-Biased Genes",
      point.size = 3,
      nsets = 6,
      nintersects = NA)

```

In males, the highest number of shared sex-biased genes are found between the gonads and the liver, and in females the highest number of shared sex-biased genes are found between the gills and liver. There are no genes that are shared across the same organ between males and females (e.g., biased in males and females in the liver) which is good. 

## Variation in FC across sex-bias and tissue type
I want to create a plot the highlights the variation we see in fold-change both across the different biases groups (male-biased, female-biased, and non-biased) within one tissue type and across all of the tissue types. To do this I first need to create a "long" dataset in the following format:

| Tissue_type|     Bias      |   Fold-Change   |
|:----------:|:-------------:|:---------------:|
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   fem_bias    |      XXXXX      |
|    Gill    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   fem_bias    |      XXXXX      |
|    Liver   |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   fem_bias    |      XXXXX      |
|   Gonad    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |

That dataset is created here, with the Trinity gene IDs included as well:

```{r logFC-long-dataset}
logFC_long<- data.frame(
  tissue=c(rep("Gill",nrow(gill_fem_biased)),
           rep("Gill", nrow(gill_mal_biased)),
           rep("Gill", nrow(gill_non_biased)),
           rep("Gonad", nrow(gonad_fem_biased)),
           rep("Gonad", nrow(gonad_mal_biased)),
           rep("Gonad", nrow(gonad_non_biased)),
           rep("Liver", nrow(liver_fem_biased)),
           rep("Liver", nrow(liver_mal_biased)),
           rep("Liver", nrow(liver_non_biased))
         ),
  bias=c(rep("FB",nrow(gill_fem_biased)),
         rep("MB",nrow(gill_mal_biased)),
         rep("UB", nrow(gill_non_biased)),
         rep("FB", nrow(gonad_fem_biased)),
         rep("MB", nrow(gonad_mal_biased)),
         rep("UB", nrow(gonad_non_biased)),
         rep("FB", nrow(liver_fem_biased)),
         rep("MB", nrow(liver_mal_biased)),
         rep("UB", nrow(liver_non_biased))
         ),
  logFC=c(gill_fem_biased$log2FoldChange,
          gill_mal_biased$log2FoldChange,
          gill_non_biased$log2FoldChange,
          gonad_fem_biased$log2FoldChange,
          gonad_mal_biased$log2FoldChange,
          gonad_non_biased$log2FoldChange,
          liver_fem_biased$log2FoldChange,
          liver_mal_biased$log2FoldChange,
          liver_non_biased$log2FoldChange
          ),
  geneID=c(rownames(gill_fem_biased),
           rownames(gill_mal_biased),
           rownames(gill_non_biased),
           rownames(gonad_fem_biased),
           rownames(gonad_mal_biased),
           rownames(gonad_non_biased),
           rownames(liver_fem_biased),
           rownames(liver_mal_biased),
           rownames(liver_non_biased)
           )
  
)

#Exporting this data, run once and then commented out
#write.csv(logFC_long, "data/logFC_long_sexbias.csv", row.names = FALSE)
```

With the dataset in the proper format now, we can generate the plot

```{r FC-var-plot, echo=FALSE, fig.cap="Absolute value of the logFC for genes that are female-biased, male-biased, and unbiased across each tissue type. Raw fold-change values are added on top of the boxplot as jitters."}
my_colors <- c("FB" = "#7fc97f", "MB" = "#beaed4", "UB" = "darkgray")

ggplot(logFC_long,
       aes(x = bias, y = abs(logFC), fill = bias)) +
  geom_boxplot() +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_point(data=logFC_long, aes(x=bias, y=abs(logFC)),
             size=0.4, 
             position = position_jitter(width = 0.15)) + 
  labs(x="Bias Level", y="|logFC|") +
  guides(fill = guide_legend(title = "Bias Level", order = 3))

```

There appears to be differences in the logFC across both the organs and the sex-bias groups (male-biased, fem-biased, and non-biased) (Fig. \@ref(fig:FC-var-plot)). Let's explore some analyses to see if we can put some statistical power behind those differences.

```{r FC-var-stats, warning=FALSE, eval=FALSE}
#Looking at some summary statistics for logFC between the different groups
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), mean)
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), sd)

par(mfrow=c(1,1))
interaction.plot(logFC_long$tissue, logFC_long$bias, logFC_long$logFC)

logFC_long$tissue <- as.factor(logFC_long$tissue)
logFC_long$bias <- as.factor(logFC_long$bias)

fc_var_aov <- aov(abs(logFC_long$logFC) ~ logFC_long$bias*logFC_long$tissue)
anova(fc_var_aov)

par(mfrow = c(2,2))
plot(fc_var_aov)

library(car)
leveneTest(fc_var_aov)
par(mfrow=c(1,1))
hist(resid(fc_var_aov))
#shapiro.test(resid(fc_var_aov)) 

plot(TukeyHSD(fc_var_aov))
library(multcompView)
multcompLetters4(fc_var_aov, TukeyHSD(fc_var_aov))

##Fitting a negative binomial model
library(MASS)
model <- glm.nb(abs(logFC_long$logFC) ~ logFC_long$tissue * logFC_long$bias)

linear_model <- lm(abs(logFC_long$logFC)~ logFC_long$tissue * logFC_long$bias)
```

## Categorising sex-specific genes
Now that we have investigated what the sex-bias is like across the different tissue types I want to dive further into genes with **sex-specific expression**. These are genes expressed only in one sex or the other. I am classifying a sex-specific genes within each tissue as ones where the expression of that gene is less than 10 for all of one sex and there is a median of $\ge 20$ in the other sex. 

Prior to running the for loop, any "outliers" that were determined by DESeq were removed from the pool of genes. Additionally, normalized counts were used for the classification of sex-specific genes as DESeq2 uses the moralized counts for all of the logFC/sexbias calculations.

```{r sex-specific-for-loop}
#Pulling out the geneIDs for genes that were categorised as "outliers" by DESeq2
#Calculating the Cooks threshold that would have been used
np <- length(resultsNames(ddsMF_SA_exp))
nsamp <- ncol(ddsMF_SA_exp)
cooks_thresh <- qf(0.99, df1 = np, df2 = nsamp-np)

out_ids <- names(mcols(ddsMF_SA_exp)$maxCooks[mcols(ddsMF_SA_exp)$maxCooks >
                                                cooks_thresh])

#Checking to make sure we have correct number of outliers
length(out_ids)

#Filtering out the dds dataset to remove the outliers determined by DESeq
ddsMF_SA_exp_filtered <- ddsMF_SA_exp[!(rownames(ddsMF_SA_exp) %in% out_ids), ]

#Create a vector with the different organ types
organs <- levels(colData(ddsMF_SA_exp_filtered)$organ)

#Create an empty list to store my datasets in
SA_sex_specific_genes <- list()

#Generate the for loop to identify MSpecific and FSpecific genes in each 
#organ based on Medians
for(organ in organs){
  browser()
  #Male-Specific Genes
  ##Pull out all of the rows where fem count <=10 in every female sample
  fem0_organ_names <- which(rowSums(t(apply(counts(ddsMF_SA_exp_filtered, 
                                                   normalized = TRUE)[, ddsMF_SA_exp_filtered$sex == "F" 
                                                                      & 
                                                                        ddsMF_SA_exp_filtered$organ == organ], 
                                            1, 
                                            function(x) x <= 10)
                                      )
                                ) == ncol(counts(ddsMF_SA_exp_filtered, 
                                                 normalized = TRUE)[, ddsMF_SA_exp_filtered$sex == "F" &
                                                                      ddsMF_SA_exp_filtered$organ == organ])
                      )
  
  fem0_organ <- counts(ddsMF_SA_exp_filtered, 
                       normalized = TRUE)[rownames(counts(ddsMF_SA_exp_filtered, 
                                                          normalized = TRUE)) %in% 
                                            names(fem0_organ_names), 
                                          ddsMF_SA_exp_filtered$organ == organ]
                                                         
  
  ##Pull out rows where median of male count >=20 for that organ
  mal10_organ <- apply(counts(ddsMF_SA_exp_filtered, 
                              normalized = TRUE)[, ddsMF_SA_exp_filtered$sex == "M" 
                                                 & ddsMF_SA_exp_filtered$organ == organ], 
                       1,
                       function(x) median(x) >=20)
  
  ##Keep only the rows where all fem samples <=10 and the Male median>=20
  fem0_mal10_organ <- fem0_organ[rownames(fem0_organ) %in% 
                                   names(mal10_organ[mal10_organ == TRUE]),
                                 ]
  
  ##Create a new object with a name based on the organ type
  organ_malsp <- sub("$", "_male_specific", organ)
  SA_sex_specific_genes[[organ_malsp]] <- fem0_mal10_organ
  
  
  
  #Female-Specific Genes
  ##Pull out all of the rows where male count <=10 in every male sample
  mal0_organ_names <- which(rowSums(t(apply(counts(ddsMF_SA_exp_filtered, 
                                                   normalized = TRUE)[, ddsMF_SA_exp_filtered$sex == "M" 
                                                                      & 
                                                                        ddsMF_SA_exp_filtered$organ == organ],
                                            1, 
                                            function(x) x <= 10)
                                      )
                                    ) == ncol(counts(ddsMF_SA_exp_filtered, 
                                                     normalized = TRUE)[, ddsMF_SA_exp_filtered$sex == "M" & 
                                                                          ddsMF_SA_exp_filtered$organ == organ])
                            )
  
  mal0_organ <- counts(ddsMF_SA_exp_filtered, 
                       normalized = TRUE)[rownames(counts(ddsMF_SA_exp_filtered, 
                                                          normalized = TRUE)) %in% 
                                            names(mal0_organ_names), 
                                          ddsMF_SA_exp_filtered$organ == organ]
  
  ##Pull out rows where median of female count >=10 for that organ
  fem10_organ <- apply(counts(ddsMF_SA_exp_filtered, 
                              normalized = TRUE)[, ddsMF_SA_exp_filtered$sex == "F" &
                                                   ddsMF_SA_exp_filtered$organ == organ],
                       1,
                       function(x) median(x) >=20)
  
  #Keep only the rows where male=0 and the fem median>=10
  mal0_fem10_organ <- mal0_organ[rownames(mal0_organ) %in% 
                                   names(fem10_organ[fem10_organ == TRUE]),
                                 ]
  
  # Create a new object with a name based on the organ type
  organ_femsp <- sub("$", "_female_specific", organ)
  SA_sex_specific_genes[[organ_femsp]] <- mal0_fem10_organ
  
} 
```
