---
title: "Weighted Gene Co-expression Network Analysis (WGCNA) in ANZ and Aus _Stigmatopora nigra_"
output:
  bookdown::pdf_document2:
    fig_caption: yes
    keep_tex: yes
    number_sections: yes
    toc: true
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
knitr::opts_chunk$set(tidy.opts = list(width.cutoff = 60), tidy = TRUE)
```

``` {r library, include = TRUE, message = FALSE, warning = FALSE}
setwd("~/Documents/GitHub/pipefish_sbge/stigmatopora/wgcna")
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(flashClust)
library(tidyverse)
library(pheatmap)
library(dplyr)
library(readr)
library(WGCNA)
library(foreach)
library(doParallel)
```

## Using OrthoFinder in the RCC

Because the transcriptomes were assembled de novo and have no gene annotation, we'll need to use OrthoFinder to be able to meaningfully compare the two populations.

To run OrthoFinder, you need protein sequences. To achieve this, use TransDecoder:

```{bash, transdecoder, eval = FALSE}

# For each transcriptome FASTA
TransDecoder.LongOrfs -t species1_transcripts.fasta 
TransDecoder.Predict -t species1_transcripts.fasta
```

Running `./TransDecoder.LongOrfs` doesn't take long - no more than 10 minutes on the RCC.
For NZ nigra this was run as: `./TransDecoder.LongOrfs -t ../emily_files/S_nigra_2024/nigra_supertranscript.fasta`
For Aus nigra this was run as: `./TransDecoder.LongOrfs -t ../emily_files/australia_nigra_raw_reads/supertranscriptome.fasta`

and

For NZ nigra this was run as: `./TransDecoder.Predict -t ../emily_files/S_nigra_2024/nigra_supertranscript.fasta`
     redundancy-minimized set includes 4870 / 5000 = 97.40%
     
For Aus nigra this was run as: `./TransDecoder.Predict -t ../emily_files/australia_nigra_raw_reads/supertranscriptome.fasta`
     redundancy-minimized set includes 4978 / 5000 = 99.56%

Running `./TransDecoder.Predict` takes much longer - depending on the size of your transcriptome, it could take several hours. Run with nohup or in a screen session.  
Running this code gives you a `.pep` file for each - your protein sequences.


Then you have to prepare the input for OrthoFinder by creating a directory like this:

```{bash, prepare-input, eval = FALSE}
orthofinder_input/
  nz_nigra.fa     # <- from TransDecoder
  aus_nigra.fa     # <- from TransDecoder
```

Then you can run OrthoFinder

```{bash, run-orthofinder, eval = FALSE}

orthofinder -f orthofinder_input/ -t 8

```

This was run as `./orthofinder -f ../TransDecoder-TransDecoder-v5.7.1/orthofinder_input/`

Running this command will: 1) Run DIAMOND to align proteins, 2) cluster transcripts into orthogroups, 3) Infer orthologs and gene trees. 

Then you can use the orthogroups.tsv to merge with expression data. This will allow you to summarise expression values by orthogroup and build a shared expression matrix for WGCNA. 

Note - CITATION:
 When publishing work that uses OrthoFinder please cite:
 Emms D.M. & Kelly S. (2019), Genome Biology 20:238

 If you use the species tree in your work then please also cite:
 Emms D.M. & Kelly S. (2017), MBE 34(12): 3267-3278
 Emms D.M. & Kelly S. (2018), bioRxiv https://doi.org/10.1101/267914

## Get expression data for NZ _S. nigra_

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.nigra <- readRDS("data/txi.salmon_SN.RDS")

#The samples file generated for tximport
samples <- read.table("data/SN_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)

```

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_SN <- DESeqDataSetFromTximport(txi.nigra, 
                                   colData = samples,
                                   design = ~ sex)

```

The data is then pre-filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

The cutoff here was to remove rows that had counts fewer than 10 across all samples.

```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at least 10 reads total
keep <- rowSums(counts(dds_SN)) >= 10
dds_SN <- dds_SN[keep, ]

dds_SN <- dds_SN[, !(dds_SN$ID %in% c("S34", "S41"))]
samples <- samples[!(samples$ID %in% c ("S34", "S41")),]
```

After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.

```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_SN_exp <- DESeq(dds_SN)

#Compile the results
res <- results(dds_SN_exp)
res

```

## Get expression data for Aus _S. nigra_

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.aus <- readRDS("data/txi.salmon_aus.RDS")

#The samples file generated for tximport
samples <- read.table("data/aus_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
samples$sex <- as.factor(samples$sex)
samples$organ <- as.factor(samples$organ)

```

The package `DESeq2` was used for the differential expression analysis outlined below.

# Single factor analysis - Comparing Males v Females across all organs
To analyze your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions. The model for this single factor analysis was run as counts ~ Sex.

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_aus <- DESeqDataSetFromTximport(txi.aus, 
                                   colData = samples,
                                   design = ~ sex)

dds_aus <- dds_aus[, !(dds_aus$ID %in% c("S3","S10","S13","S16"))]

samples <- samples[!(dds_aus$ID %in% c("S3","S10","S13","S16")), ]
```

The data is then pre-filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

The cutoff here was to remove rows that had counts fewer than 10 across all samples.

```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at least 10 reads total
keep <- rowSums(counts(dds_aus)) >= 10
dds_aus <- dds_aus[keep, ]

```

After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.

```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_aus_exp <- DESeq(dds_aus)

#Compile the results
res <- results(dds_aus_exp)
res

```

################################### work through this

```{r}
## Load and prepare orthology data
orthogroups <- read.delim("data/Orthogroups.tsv", header = TRUE, stringsAsFactors = FALSE, fill = TRUE)

## Function to filter expression data based on orthogroups
filter_by_orthogroup <- function(expr_data, orthologs, species_column) {
  # Get unique genes for each species
  ortholog_pairs <- split(orthologs[, c(species_column, "GeneID")], 
                         orthologs[[species_column]])
  
  # Filter expression data to include only genes with orthologs
  filtered_genes <- unlist(lapply(names(ortholog_pairs), 
                                function(sp) ortholog_pairs[[sp]][["GeneID"]]))
  
  # Return filtered expression matrix
  expr_data[, rownames(expr_data) %in% filtered_genes]
}

## Apply filtering after VST transformation
datExpr_SN_filtered <- filter_by_orthogroup(datExpr_SN, orthologs, "nz_nigra")
datExpr_AUS_filtered <- filter_by_orthogroup(datExpr_AUS, orthologs, "aus_nigra")

## Verify dimensions after filtering
dim(datExpr_SN_filtered)
dim(datExpr_AUS_filtered)

## Continue with WGCNA analysis using filtered data
gsg_SN_filtered <- goodSamplesGenes(datExpr_SN_filtered, verbose = 3)
gsg_AUS_filtered <- goodSamplesGenes(datExpr_AUS_filtered, verbose = 3)

# Remove outliers if present
if (!gsg_SN_filtered$allOK) {
  datExpr_SN_filtered <- datExpr_SN_filtered[gsg_SN_filtered$goodSamples, 
                                            gsg_SN_filtered$goodGenes]
}
if (!gsg_AUS_filtered$allOK) {
  datExpr_AUS_filtered <- datExpr_AUS_filtered[gsg_AUS_filtered$goodSamples,
                                               gsg_AUS_filtered$goodGenes]
}

## Re-run soft threshold selection with filtered data
powers <- c(1:20)
sft_SN_filtered <- pickSoftThreshold(datExpr_SN_filtered, 
                                    powerVector = powers, 
                                    verbose = 5)
softPower_SN_filtered <- sft_SN_filtered$powerEstimate

sft_AUS_filtered <- pickSoftThreshold(datExpr_AUS_filtered, 
                                     powerVector = powers, 
                                     verbose = 5)
softPower_AUS_filtered <- sft_AUS_filtered$powerEstimate
```


##################################



## Format and filter data for WGCNA 

```{r}
## Format, filter, and normalize data for WGCNA

# Remove missing values
dds_SN_exp <- na.omit(dds_SN_exp)
dds_aus_exp <- na.omit(dds_aus_exp)

# Function to filter low-variance genes and apply VST
prepare_wgcna_data <- function(dds_obj) {
  # Get raw counts
  counts_matrix <- counts(dds_obj)
  
  # Filter out genes with zero variance
  variances <- apply(counts_matrix, 1, var)
  filtered_counts <- counts_matrix[variances > 0, ]
  
  # Subset DESeqDataSet
  keep <- rownames(filtered_counts) %in% rownames(dds_obj)
  dds_filtered <- dds_obj[keep, ]
  
  # Apply variance stabilizing transformation
  vsd <- vst(dds_filtered, blind = FALSE)
  
  # Extract and transpose expression matrix
  datExpr <- t(assay(vsd))
  
  return(datExpr)
}

# Apply to both datasets
datExpr_SN <- prepare_wgcna_data(dds_SN_exp)
datExpr_AUS <- prepare_wgcna_data(dds_aus_exp)

# Check dimensions
dim(datExpr_SN)  # Should be (samples × genes)
dim(datExpr_AUS)
```


```{r check-outliers}
# Check for outliers
gsg_SN <- goodSamplesGenes(datExpr_SN, verbose = 3)
gsg_AUS <- goodSamplesGenes(datExpr_AUS, verbose = 3)

# If outliers are detected, remove them
if (!gsg_SN$allOK) {
  datExpr_SN <- datExpr_SN[gsg_SN$goodSamples, gsg_SN$goodGenes]
}
if (!gsg_AUS$allOK) {
  datExpr_AUS <- datExpr_AUS[gsg_AUS$goodSamples, gsg_AUS$goodGenes]
}
```

## Construct co-expression networks 

```{r}
powers <- c(1:20) # Try a range of soft-threshold powers

# Choose soft threshold for NZ population
sft_SN <- pickSoftThreshold(datExpr_SN, powerVector = powers, verbose = 5)
softPower_SN <- sft_SN$powerEstimate

# Choose soft threshold for AUS population
sft_AUS <- pickSoftThreshold(datExpr_AUS, powerVector = powers, verbose = 5)
softPower_AUS <- sft_AUS$powerEstimate

# Plot the scale-free topology fit index
par(mfrow = c(1,2))
plot(sft_SN$fitIndices[,1], -sign(sft_SN$fitIndices[,3])*sft_SN$fitIndices[,2], type="b",
     main="Soft Thresholding NZ", xlab="Soft Threshold", ylab="Scale-Free Topology Fit")
plot(sft_AUS$fitIndices[,1], -sign(sft_AUS$fitIndices[,3])*sft_AUS$fitIndices[,2], type="b",
     main="Soft Thresholding AUS", xlab="Soft Threshold", ylab="Scale-Free Topology Fit")

```

 
## Construct the network and identify gene modules 

```{r matrix-and-TOM}

# Process one dataset at a time
adjacency_SN <- adjacency(datExpr_SN, power = softPower_SN)
TOM_SN <- TOMsimilarity(adjacency_SN)
dissTOM_SN <- 1 - TOM_SN

# Save results and clear
save(dissTOM_SN, file="dissTOM_SN.RData")
rm(adjacency_SN, TOM_SN, dissTOM_SN)

# Repeat for second dataset
adjacency_AUS <- adjacency(datExpr_AUS, power = softPower_AUS)
TOM_AUS <- TOMsimilarity(adjacency_AUS)
dissTOM_AUS <- 1 - TOM_AUS
save(dissTOM_AUS, file="dissTOM_AUS.RData")

```

## Identify gene modules using hierarchical clustering

```{r module-clustering}
# Cluster genes and define modules
geneTree_SN <- hclust(as.dist(dissTOM_SN), method = "average")
dynamicMods_SN <- cutreeDynamic(dendro = geneTree_SN, distM = dissTOM_SN, deepSplit = 2, pamRespectsDendro = FALSE, minClusterSize = 30)

geneTree_AUS <- hclust(as.dist(dissTOM_AUS), method = "average")
dynamicMods_AUS <- cutreeDynamic(dendro = geneTree_AUS, distM = dissTOM_AUS, deepSplit = 2, pamRespectsDendro = FALSE, minClusterSize = 30)

# Convert module labels to colors
moduleColors_SN <- labels2colors(dynamicMods_SN)
moduleColors_AUS <- labels2colors(dynamicMods_AUS)

# Plot dendrogram with module colors
par(mfrow = c(1,2))
plotDendroAndColors(geneTree_SN, moduleColors_SN, "Modules NZ", dendroLabels = FALSE, hang = 0.03)
plotDendroAndColors(geneTree_AUS, moduleColors_AUS, "Modules AUS", dendroLabels = FALSE, hang = 0.03)

```

The goal is to pick a β value where R² ≥ 0.8, ensuring a scale-free topology.

In my plots:
-	NZ population: The fit index reaches R² ≈ 0.8 around β = 9–10.
-	AUS population: The fit index reaches R² ≈ 0.8 around β = 12–13.

This suggests I should select β = 9 or 10 for NZ and β = 12 or 13 for AUS for optimal network construction

## Relate gene modules to traits

```{r trait-modules}
# Load metadata (ensure it includes sample IDs, sex, tissue type, etc.)
metadata <- read.csv("metadata.csv", row.names = 1)

# Relate modules to traits
MEs_SN <- moduleEigengenes(datExpr_SN, colors = moduleColors_SN)$eigengenes
MEs_AUS <- moduleEigengenes(datExpr_AUS, colors = moduleColors_AUS)$eigengenes

moduleTraitCor_SN <- cor(MEs_SN, metadata, use = "p")
moduleTraitPval_SN <- corPvalueStudent(moduleTraitCor_SN, nrow(datExpr_SN))

moduleTraitCor_AUS <- cor(MEs_AUS, metadata, use = "p")
moduleTraitPval_AUS <- corPvalueStudent(moduleTraitCor_AUS, nrow(datExpr_AUS))

# Heatmap of correlations
par(mfrow = c(1,2))
labeledHeatmap(Matrix = moduleTraitCor_SN, xLabels = colnames(metadata), yLabels = names(MEs_SN), 
               colorLabels = FALSE, main = "Module-Trait Correlation (NZ)")
labeledHeatmap(Matrix = moduleTraitCor_AUS, xLabels = colnames(metadata), yLabels = names(MEs_AUS), 
               colorLabels = FALSE, main = "Module-Trait Correlation (AUS)")

```

## Identify hub genes

```{r hub-genes}

hubGenes_SN <- chooseTopHubInEachModule(datExpr_SN, moduleColors_SN, power = softPower_SN)
hubGenes_AUS <- chooseTopHubInEachModule(datExpr_AUS, moduleColors_AUS, power = softPower_AUS)

```

## Comparing module preservation between populations in WGCNA

To assess whether gene co-expression modules are conserved between the two _S. nigra_ populations, we will use module preservation statistics in WGCNA. This  analysis will help determine whether the same sets of genes remain co-expressed across populations or whether there are population-specific differences in co-expression patterns. First, we have to format the data for module preservation.

```{r format-mod-preservation}

# Define data sets
multiExpr <- list(
  NZ = list(data = datExpr_SN),   # NZ expression matrix
  AUS = list(data = datExpr_AUS)  # AUS expression matrix
)

# Define module colors (same ones from previous WGCNA analysis)
multiColor <- list(
  NZ = moduleColors_SN,   # NZ module assignments
  AUS = moduleColors_AUS  # AUS module assignments
)
```

```{r run-mod-preservation}
# Set up parameters
set.seed(123)
mp_results <- modulePreservation(
  multiExpr,          # Expression data from both populations
  multiColor,         # Module colors from both populations
  referenceNetworks = 1,  # Use NZ as the reference
  testNetworks = 2,       # Compare against AUS
  nPermutations = 200,    # Number of permutations (increase for stability)
  randomSeed = 42,
  verbose = 3
)

```

## Interpreting module preservation results 

The function `modulePreservation` generates multiple preservation statistics, including:

Zsummary: A key metric where:

* Z > 10 → Strongly preserved

* Z 5–10 → Moderately preserved

* Z < 5 → Not preserved (population-specific)

MedianRank: A lower rank suggests stronger preservation.

We will extract and visualise the results 
```{r stats}
# Extract key statistics
preservationStats <- mp_results$preservation$Z[[2]]  # AUS relative to NZ

# Convert to a dataframe for easier visualisation
preservationDF <- data.frame(
  Module = rownames(preservationStats),
  Zsummary = preservationStats$Zsummary.pres,
  MedianRank = preservationStats$medianRank.pres
)

# Sort modules by preservation strength
preservationDF <- preservationDF[order(-preservationDF$Zsummary), ]

# Print top preserved modules
print(head(preservationDF))

# Visualisation: Zsummary vs. MedianRank
ggplot(preservationDF, aes(x = MedianRank, y = Zsummary, label = Module)) +
  geom_point(aes(color = Zsummary > 10), size = 3) +
  geom_text_repel() +
  theme_minimal() +
  labs(title = "Module Preservation Analysis", x = "Median Rank", y = "Zsummary")

```

## Identifying sex-biased and tissue-specific modules in WGCNA

To identify sex-biased and tissue-specific gene co-expression modules, we will relate module eigengenes to sex and tissue metadata. This will help pinpoint which modules are associated with sex-biased gene expression and which ones show tissue specificity. 

```{r create-metadata}
# Read NZ sample metadata
sn_meta <- read_tsv("SN_samples.txt", col_names = TRUE)

# Read AUS sample metadata
aus_meta <- read_tsv("aus_samples.txt", col_names = TRUE)

# Before merging, make sure they are structured correctly
head(sn_meta)
head(aus_meta)

# Add population labels
sn_meta <- sn_meta %>% mutate(Population = "NZ")
aus_meta <- aus_meta %>% mutate(Population = "AUS")

# Combine NZ and AUS metadata
metaData <- bind_rows(sn_meta, aus_meta)

# Ensure sample IDs match expression data
rownames(metaData) <- metaData$Sample_ID

# Check the structure
head(metaData)

# Save and then comment out
write_tsv(metaData, "metadata_combined.tsv")


```

## Calculate module eigengenes
Each module has a module eigengene (ME), which is the first principal component summarising gene expression within that module. We calculate these eigengenes and correlate them sex and tissue.

```{r calculate-ME}
# Compute module eigengenes
MEs_NZ <- moduleEigengenes(datExpr_SN, moduleColors_SN)$eigengenes
MEs_AUS <- moduleEigengenes(datExpr_AUS, moduleColors_AUS)$eigengenes

# Add metadata to eigengenes
MEs_NZ$Sex <- metaData$Sex
MEs_NZ$Tissue <- metaData$Tissue
MEs_AUS$Sex <- metaData$Sex
MEs_AUS$Tissue <- metaData$Tissue

```

## Identifying sex-biased modules
We test for differences in module eigengene expression between males and females using a linear model.

```{r}
# Linear model for sex differences in NZ population
sex_assoc_NZ <- apply(MEs_NZ[, -c(1:2)], 2, function(x) {
  summary(lm(x ~ metaData$Sex))$coefficients[2, 4]  # Extract p-value for Sex effect
})

# Linear model for sex differences in AUS population
sex_assoc_AUS <- apply(MEs_AUS[, -c(1:2)], 2, function(x) {
  summary(lm(x ~ metaData$Sex))$coefficients[2, 4]
})

# Adjust for multiple testing (FDR correction)
sex_assoc_NZ_adj <- p.adjust(sex_assoc_NZ, method = "fdr")
sex_assoc_AUS_adj <- p.adjust(sex_assoc_AUS, method = "fdr")

# Identify significantly sex-biased modules
sex_biased_modules_NZ <- names(sex_assoc_NZ_adj)[sex_assoc_NZ_adj < 0.05]
sex_biased_modules_AUS <- names(sex_assoc_AUS_adj)[sex_assoc_AUS_adj < 0.05]

print("Sex-biased modules in NZ:")
print(sex_biased_modules_NZ)

print("Sex-biased modules in AUS:")
print(sex_biased_modules_AUS)

```

## Identifying tissue-specific modules
### Calculate tau

```{r tau}
# Function to calculate tau
##################################ADD MY OTHER TAU GENERATION FUNCTION####################################
calculate_tau <- function(expression_matrix) {
  max_expression <- apply(expression_matrix, 1, max)  # Get max expression per gene
  tau_values <- rowSums(1 - expression_matrix / max_expression, na.rm = TRUE) / (ncol(expression_matrix) - 1)
  return(tau_values)
}

# Split expression data by population
tissue_expression_NZ <- t(datExpr_SN)  # Make sure samples are in columns, genes in rows
tissue_expression_AUS <- t(datExpr_AUS)

# Compute tau for each population
tau_NZ <- calculate_tau(tissue_expression_NZ)
tau_AUS <- calculate_tau(tissue_expression_AUS)

# Add to data frames
tau_df_NZ <- data.frame(Gene = rownames(tissue_expression_NZ), Tau = tau_NZ)
tau_df_AUS <- data.frame(Gene = rownames(tissue_expression_AUS), Tau = tau_AUS)

# Save results
write_tsv(tau_df_NZ, "tau_NZ.tsv")
write_tsv(tau_df_AUS, "tau_AUS.tsv")

```

Compare tau with WGCNA tissue-specific modules. 
Now that we have tau values, check if genes in tissue-associated modules have higher tau values than expected.

```{r}
# Add module assignments to tau data
tau_df_NZ$Module <- moduleColors_SN[match(tau_df_NZ$Gene, names(moduleColors_SN))]
tau_df_AUS$Module <- moduleColors_AUS[match(tau_df_AUS$Gene, names(moduleColors_AUS))]

# Compare τ across modules
boxplot(Tau ~ Module, data = tau_df_NZ, main = "Tau Distribution Across Modules (NZ)", las = 2)
boxplot(Tau ~ Module, data = tau_df_AUS, main = "Tau Distribution Across Modules (AUS)", las = 2)

```

To test if certain modules have significantly higher tau values, we will use Kruskal-Wallis tests.

High τ values in a module → Genes in that module are likely tissue-specific.

Low τ values in a module → Genes are broadly expressed across tissues.

Comparing WGCNA-derived tissue modules with τ values helps confirm true tissue-specific expression patterns.

```{r KW-test}
kruskal.test(Tau ~ Module, data = tau_df_NZ)
kruskal.test(Tau ~ Module, data = tau_df_AUS)

#If significant, follow up with pairwise comparisons
pairwise.wilcox.test(tau_df_NZ$Tau, tau_df_NZ$Module, p.adjust.method = "fdr")
pairwise.wilcox.test(tau_df_AUS$Tau, tau_df_AUS$Module, p.adjust.method = "fdr")

```

```{r ME-generate}
# Ensure metadata is properly formatted
metaData$Tissue <- as.factor(metaData$Tissue)

# Get module eigengenes
MEs_SN <- moduleEigengenes(datExpr_SN, moduleColors_SN)$eigengenes
MEs_AUS <- moduleEigengenes(datExpr_AUS, moduleColors_AUS)$eigengenes

# Convert tissue to numeric for correlation
tissue_trait_SN <- model.matrix(~ 0 + metaData$Tissue[metaData$Population == "NZ"])
tissue_trait_AUS <- model.matrix(~ 0 + metaData$Tissue[metaData$Population == "AUS"])

# Compute module-trait correlation
moduleTraitCor_SN <- cor(MEs_SN, tissue_trait_SN, use = "p")
moduleTraitCor_AUS <- cor(MEs_AUS, tissue_trait_AUS, use = "p")

# Compute p-values
moduleTraitPvalue_SN <- corPvalueStudent(moduleTraitCor_SN, nrow(datExpr_SN))
moduleTraitPvalue_AUS <- corPvalueStudent(moduleTraitCor_AUS, nrow(datExpr_AUS))
```

Now we overlay tau values onto the heatmap by calculating the median tau value for each module

```{r heatmap-module}
# Merge tau values with module assignments
tau_df_SN$Module <- moduleColors_SN[match(tau_df_SN$Gene, names(moduleColors_SN))]
tau_df_AUS$Module <- moduleColors_AUS[match(tau_df_AUS$Gene, names(moduleColors_AUS))]

# Compute median τ per module
median_tau_SN <- aggregate(tau_df_SN$Tau, by = list(Module = tau_df_SN$Module), FUN = median)
median_tau_AUS <- aggregate(tau_df_AUS$Tau, by = list(Module = tau_df_AUS$Module), FUN = median)

# Rename columns
colnames(median_tau_SN) <- c("Module", "Median_Tau")
colnames(median_tau_AUS) <- c("Module", "Median_Tau")

# Define a custom function to add tau values to labels
add_tau_labels <- function(corMatrix, tauValues) {
  module_labels <- rownames(corMatrix)
  tau_map <- setNames(tauValues$Median_Tau, tauValues$Module)
  tau_labels <- sapply(module_labels, function(m) {
    tau <- tau_map[m]
    if (!is.na(tau)) {
      return(paste0(m, " (τ=", round(tau, 2), ")"))
    } else {
      return(m)
    }
  })
  return(tau_labels)
}

# Generate heatmap with tau labels
pheatmap(moduleTraitCor_SN,
         color = colorRampPalette(c("blue", "white", "red"))(100),
         main = "Module-Tissue Correlation (NZ)",
         labels_row = add_tau_labels(moduleTraitCor_SN, median_tau_SN),
         display_numbers = round(moduleTraitPvalue_SN, 3))

pheatmap(moduleTraitCor_AUS,
         color = colorRampPalette(c("blue", "white", "red"))(100),
         main = "Module-Tissue Correlation (AUS)",
         labels_row = add_tau_labels(moduleTraitCor_AUS, median_tau_AUS),
         display_numbers = round(moduleTraitPvalue_AUS, 3))

```

Interpretation
Red cells → Strong positive correlation between a module and a tissue.

Blue cells → Strong negative correlation.

τ values (τ=0 to τ=1) added to module labels indicate whether modules highly correlated with a tissue also contain highly tissue-specific genes.

Modules with both high correlation and high τ → Strongly tissue-specific.