---
title: "Tissue Specificity in _Syngnathus fuscus_"
author: "Coley Tosto"
date: "`r Sys.Date()`"
output:
  github_document: 
    toc: true 
  #bookdown::pdf_document2:
    #fig_caption: yes
    #keep_tex: yes
    #number_sections: no
    #toc: false
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',
                     fig_path="../imgs/")
```

``` {r library, message = FALSE, warning = FALSE}
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(spfTools)
```

``` {r read-data, message = FALSE, warning = FALSE}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.salmon <- readRDS("data/txi.salmon_FU.RDS")

#The samples file generated for tximport
samples <- read.table("FU_samples.txt", header = TRUE)

samples$group <- factor(paste0(samples$Sex, samples$Organ))

#Changing "Gonad" to be more specific to testis or ovaries
samples$Organ <- ifelse(samples$Sex == "F" & samples$Organ =="Gonad",
                        paste0("Ovaries"),
                        ifelse(samples$Sex == "M" & samples$Organ == "Gonad",
                               paste0("Testis"),
                               paste0(samples$Organ))
                        )

#Make sure the conditions are in the samples file as a factor
samples$Sex <- as.factor(samples$Sex)
samples$Organ <- as.factor(samples$Organ)

#Format colData to be used in the tau function
colData <- as.data.frame(samples)
rownames(colData) <- samples$ID

```

# Differential Expression Analysis
The main goal of this analysis is to relate measure of pleiotropic constraints (tissue specificity) to sex bias. In order to do that I need to combine my sex-bias data that I generated in a separate .Rmd (`fuscus_diff_expr_analysis.Rmd`) with the tau values I will be generating in this document.

Here I am reading in the sex-bias datasets that correspond to each of the organ's single-factor analysis. These datasets include all of the genes where the p-value was not equal to "NA".

```{r read-in-sb-datasets}
liver_res <- read.csv("data/liver_res.csv", 
                      row.names = 1, 
                      header = TRUE)
gill_res <- read.csv("data/gill_res.csv", 
                      row.names = 1, 
                      header = TRUE)
gonad_res <- read.csv("data/gonad_res.csv", 
                      row.names = 1, 
                      header = TRUE)
```

# Calculating Tissue Specificity
To estimate tissue specificity, the TPM estimates are needed which is the number of transcripts from a particular gene normalized first by gene length, and then by sequencing depth (in millions) in the sample. The output quant.sf files from salmon have the following format:

```
Name | Length | EffectiveLength | TPM | NumReads

```

## Pulling out TPM values
From the salmon outputs I pulled out the TPM values for each sample. All of the necessary .sf files are stored in a folder called `fuscus_expression_files`. 

The function below is pulling out the TPMs for each of the samples into a column and then binding them together into one dataset. I then remove the three samples that were not included in the differential expression analysis.

```{r getTPMs}
#Get the list of file names/paths for all of the quant.sf files
files <- list.files(pattern = ".sf", path = "data/fuscus_expression_files", 
                    full.names = TRUE)

#For each quant.sf file pull out the TPM column
tpms <- do.call(cbind, lapply(files, function(file){

  dat <- read.delim(file, row.names = "Name")
  tpm <- dat["TPM"]
  colnames(tpm) <- gsub("data/fuscus_expression_files/(.*)_quant.sf","\\1",file)
  
  return(tpm)
}))

#Remove the samples that were removed in the diff. expression analysis
tpms <- tpms[, !(colnames(tpms) %in% c("FUT11M4",
                                       "FUG11M4",
                                       "FUL11M4"))]
```

## Filtering the TPM dataset
I know want to filter the TPM dataset to remove a few things. These include similar things that I filtered out of the DESeq2 datasets prior to performing the differential expression analysis. This includes:

  1. Keeping only the rows that weren't filtered out in the DESeq2 dataset due to low counts (rowSum $\le$ 10).
  
  2. Removing the rows that corresponded to genes that were "outliers" in the DESeq2 analysis.
  
  3. Removing the samples related to individual 11M4 that were removed in the differential expression analysis (see above).

This will be imperfect filtering in this case because I did an individual single-factor analysis within each organ type. To try and make life easier I am going to run the multi-factor analysis on the DESeq dataset and then apply the filtering from that onto the TPM dataset. 

It should also be noted that the DESeq2 filtering was done based off of the count data and not the TPMs, but as TPMs are just normalized counts they should be correlated and something that was had low gene counts should also have a low TPM and anything that was considered an outlier in the counts could also be an outlier in terms of TPM, but it may not be perfect.

```{r filterTPMs}
#Running the MF diff. expr. analysis
#Create the DESeq dataset
dds_FU <- DESeqDataSetFromTximport(txi.salmon,
                                   colData = samples,
                                   design = ~ group)

##Remove all 11M4 organs from the dataset
dds_FU <- dds_FU[, !(dds_FU$ID %in% c("FUT11M4", "FUG11M4", "FUL11M4"))]

##Filter the dataset, only keeping rows that have at least 10 reads total
keep <- rowSums(counts(dds_FU)) >= 10 
dds_FU <- dds_FU[keep, ]

#Generate the expression values
dds_FU_exp <- DESeq(dds_FU)

#Filtering the tpm dataset
#Only keeping the rows that weren't filtered out due to low counts
tpms <- tpms[rownames(tpms) %in% rownames(dds_FU), ]

#Pulling out the geneIDs for genes that were categorized as "outliers" by DESeq2
#Calculating the Cooks threshold that would have been used
np <- length(resultsNames(dds_FU_exp))
nsamp <- ncol(dds_FU_exp)
cooks_thresh <- qf(0.99, df1 = np, df2 = nsamp-np)

out_ids <- names(mcols(dds_FU_exp)$maxCooks[mcols(dds_FU_exp)$maxCooks > cooks_thresh])

#Removing the rows in the tpm dataset that were deemed "outliers" by DESeq2
tpms <- tpms[!(rownames(tpms) %in% out_ids), ]

```


## Generating $\tau$
These filtered TPM estimates can then be used to estimate $\tau$, a tissue specificity estimator that can range from 0 to 1. $\tau$ is calculated for each tissue, $i$, as follows:

$$
\tau=\frac{\sum_i{[1-ln(TPM_i)/ln(TPM_{max})]}}{N-1}
$$

where $TPM_{max}$ is the maximum **average** TPM for a given tissue type, and $TPM_i$ is the **average** TPM for tissue $i$. If $\tau = 0$, that gene is evenly expressed across tissues; if $\tau=1$, the gene is expressed in an entirely tissue-specific fashion. Because TPM values approach 0 are impacted by sampling stochasticity, any genes that had an expression approaching 0 were set to $TPM=2$. 

```{r est_tauFXN}
#Function for estimating tau given the TPM matrix and metadata file
est_tau <- function(geneDat,colDat){
  
  #For each row in the TPM matrix cbind it with the metadata file,
  #this attaches organ type information to the TPM values
  tissue_dat<-data.frame(cbind(colDat,
                               geneDat))
  
  #For the TPM values approaching 0, set them to 2
  tissue_dat$geneDat[tissue_dat$geneDat < 2] <- 2
  
  
  #Get the average TPM for each tissue type (TPMi)
  tissue_avgs <- tapply(tissue_dat$geneDat,tissue_dat$Organ,mean)
  
  #Get the maximum value from the average TPMS (TPMmax)
  tpmMax <- max(tissue_avgs, na.rm=TRUE)
    
  #IF running tau on JUST males of JUST females, this accounts for the
  #fact that ovary or testis will return an NA in the averaging
  if(length(unique(tissue_dat$Organ)) == 3){
    
    #calcualte tau
    tau <- sum(1-(log(tissue_avgs[unique(tissue_dat$Organ)])/log(tpmMax)))/
      (length(unique(tissue_dat$Organ))-1)
    
    #pull out appropriate organ
    if(tau == 0){
    
    organ <- paste0(NA)
    
    }else{
    
    organ <- names(tissue_avgs[tissue_avgs == tpmMax])
    
    }
  
  #Combine tau and the organ information together
  tau_organ <- data.frame(tau = tau, organ = organ)
  
  return(tau_organ)
    
  }
  
  #IF using the WHOLE dataset, calculate tau
  tau <- sum(1-(log(tissue_avgs)/log(tpmMax)))/(length(unique(tissue_dat$Organ))-1)
  
  #Pull out the information about WHICH tissue the expression is specific in
  ##If tau == 0 (not tissue specific) then return NA, but if above zero
  ##pull out the name of the organ with the highest expression
  if(tau == 0){
    
    organ <- paste0(NA)
    
  }else{
    
    organ <- names(tissue_avgs[tissue_avgs == tpmMax])
    
  }
  
  #Combine tau and the organ information together
  tau_organ <- data.frame(tau = tau, organ = organ)
  
  return(tau_organ)
}

```

I then applied that function across each row in the TPMs matrix with my metadata stored in the object `colData`. The metadata file includes the sample ID, Sex, and Organ type for every column present in the TPM matrix.

I ran the $\tau$ function with the whole dataset, and then for just males and just females.

```{r calcTau}
colData <- colData[!(colData$ID %in% c("FUT11M4", "FUG11M4", "FUL11M4")),]

tau <- do.call(rbind, apply(tpms, 1, est_tau, colDat=colData))


tau_fem <- do.call(rbind, apply(tpms[,which(colData$Sex=="F")], 1, est_tau,
                                colData[which(colData$Sex=="F"),]))

tau_mal <- do.call(rbind, apply(tpms[,which(colData$Sex=="M")], 1, est_tau,
                                colData[which(colData$Sex=="M"),]))

```

```{r tauHist,fig.cap="Distribution of tissue specificity estimates for only the male samples (left) versus only the female samples (middle) versus the male samples and the female samples (right).", echo=FALSE}
par(mfrow=c(1,3))
hist(tau_mal$tau,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="Male-only")

hist(tau_fem$tau,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="Female-only")

hist(tau$tau,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="All samples")
```

We can see that the distribution of $\tau$ doesn't vary drastically when looking at only male/female samples versus when we included all of the samples in the calculation. Because of this, I can be confident in using the $\tau$ with all the samples moving forward.

## Validating the $\tau$ calculations
To make sure $\tau$ is being calculated in a way that makes sense, I am checking some of the TPM values and plotting the corresponding counts for the genes with a high tissue specificity index and a low tissue specificity index.

I also want to make sure that the organ classification I implemented in the tissue specificity function is accurate.

```{r ts-plot-counts-high, echo=FALSE, fig.height=20, fig.width=15, fig.cap="Plots of the counts from the DESeq2 Dataset for genes with the highest tau", message=FALSE}
#Getting list of genes with the greatest Tau
top_tau <- head(tau[order(tau$tau, decreasing = TRUE), ], 
                n=9)

top_tau

#Looking at the TPMS
tpms[rownames(tpms)== rownames(top_tau[1,]),] #Gill-specific
tpms[rownames(tpms)== rownames(top_tau[2,]),] #Ovary-specific
tpms[rownames(tpms)== rownames(top_tau[3,]),] #Gill-specific
tpms[rownames(tpms)== rownames(top_tau[4,]),] #Ovary-specific
tpms[rownames(tpms)== rownames(top_tau[5,]),] #Gill-specific
tpms[rownames(tpms)== rownames(top_tau[6,]),] #Ovary-specific
tpms[rownames(tpms)== rownames(top_tau[7,]),] #Ovary-specific
tpms[rownames(tpms)== rownames(top_tau[8,]),] #Ovary-specific
tpms[rownames(tpms)== rownames(top_tau[9,]),] #Liver-specific

#Plotting the gene counts from DESeq2 for the toptau genes
par(mfrow=c(3, 3))
plotCounts(dds_FU, gene= rownames(top_tau[1,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(top_tau[2,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(top_tau[3,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(top_tau[4,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(top_tau[5,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(top_tau[6,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(top_tau[7,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(top_tau[8,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(top_tau[9,]), intgroup="Organ")
```

```{r ts-plot-counts-low, echo=FALSE, fig.height=20, fig.width=15, fig.cap="Plots of the counts from the DESeq2 Dataset for genes with the lowest tau", message=FALSE}
#Getting list of genes with the lowest Tau
bot_tau <- head(tau[order(tau$tau, decreasing = FALSE), ], 
                n=9)

bot_tau

#Looking at the TPMS
tpms[rownames(tpms)== rownames(bot_tau[1,]),]
tpms[rownames(tpms)== rownames(bot_tau[2,]),]
tpms[rownames(tpms)== rownames(bot_tau[3,]),]
tpms[rownames(tpms)== rownames(bot_tau[4,]),]
tpms[rownames(tpms)== rownames(bot_tau[5,]),]
tpms[rownames(tpms)== rownames(bot_tau[6,]),]
tpms[rownames(tpms)== rownames(bot_tau[7,]),]
tpms[rownames(tpms)== rownames(bot_tau[8,]),]
tpms[rownames(tpms)== rownames(bot_tau[9,]),]

#Plotting the gene counts from DESeq2 for the toptau genes
par(mfrow=c(3, 3))
plotCounts(dds_FU, gene= rownames(bot_tau[1,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(bot_tau[2,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(bot_tau[3,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(bot_tau[4,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(bot_tau[5,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(bot_tau[6,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(bot_tau[7,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(bot_tau[8,]), intgroup="Organ")
plotCounts(dds_FU, gene= rownames(bot_tau[9,]), intgroup="Organ")

```

We can see that for the genes that have high tau values the plots of the normalized counts show higher expression in just that tissue, whereas for the plots of the lower tau values no clear patterns are showing up in the plots of the counts. This is what we want to see, increasing confidence that the the calculations of tau are accurate. We can also see that the assignment of the tissues from the est_tau function match the outputs from looking at the raw counts.

# Sex Bias and Tissue Specificity
Now that I have calculated $\tau$ I want to relate it back to information about sex bias to see what, if any, relationship exists. I am going to do this separately for the different organs to see if the relationship changes based on organ type.

## $\tau$ and logFC in sex-biased genes for each organ
A gene was determined to be sex-biased if it had a log-fold change $\ge |2|$ AND an adjusted p-value of $< 0.05$. Earlier I read in four datasets that contain information for all of the significantly expressed genes in each organ (i.e. the genes that were not given a p-value of NA from DESeq). 

From these datasets I can pull out the female- and male-biased genes to plot logFC against $\tau$ within the genes that are sex-biased.

```{r tau-v-bias, echo=FALSE, fig.cap="Sex bias (in terms of the square root of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are sex biased in the gonads, liver, skin, and brain."}

#Pull out the sex-biased genes (based on FC of >= |2| and padj of < 0.05) for each organ
liver_biased <- liver_res[which(abs(liver_res$log2FoldChange) >= 2 & liver_res$padj <= 0.05),]
gonad_biased <- gonad_res[which(abs(gonad_res$log2FoldChange) >= 2 & gonad_res$padj <= 0.05),]
gill_biased <- gill_res[which(abs(gill_res$log2FoldChange) >= 2 & gill_res$padj <= 0.05),]

#Merge the sex bias information with the tau values based on row names for each organ
liver_bias_tau <- merge(as.data.frame(liver_biased), tau, by = "row.names")
gonad_bias_tau <- merge(as.data.frame(gonad_biased), tau, by = "row.names")
gill_bias_tau <- merge(as.data.frame(gill_biased), tau, by = "row.names")

#Plot the relationship between FC (bias) and tau for each organ
par(mar=c(5.1, 4.1, 4.1, 8.1), xpd=FALSE)
plot(sqrt(abs(gonad_bias_tau$log2FoldChange)),
     gonad_bias_tau$tau,
     pch=19,
     col = "#EEB42275",
     xlab="Sex bias (|log2 fold change|)",
     ylab=expression(tau["TPM"]))
points(sqrt(abs(liver_bias_tau$log2FoldChange)),
     liver_bias_tau$tau,
     pch=19,
     col="#EE826275")
points(sqrt(abs(gill_bias_tau$log2FoldChange)),
     gill_bias_tau$tau,
     pch=19,
     col = "#20B2AA75")


#Add the best fit lines
abline(lm(gonad_bias_tau$tau ~ sqrt(abs(gonad_bias_tau$log2FoldChange))), col="#EEB422", lwd = 3)
abline(lm(liver_bias_tau$tau ~ sqrt(abs(liver_bias_tau$log2FoldChange))), col = "#EE8262", lwd = 3, lty = 2)
abline(lm(gill_bias_tau$tau ~ sqrt(abs(gill_bias_tau$log2FoldChange))), col = "#20B2AA", lwd = 3, lty = 3)

#Add a legend
outer_legend("right",
             c("Gonad", "Liver", "Gill"),
             col = c("#EEB422", "#EE8262", "#20B2AA"),
             pch = 19,
             lwd = 3,
             bty = 'n',
             cex = 1.5,
             lty = 1:3,
             ncol = 1)


```

The absolute value of the log2 fold change was square root transformed. There appears to be a relationship between $\tau$ and sex bias for the gonads, but potentially not for the liver and the gills. This may be due to a low number of sex-biased genes. We can look further into this relationship with a Spearman's rank correlation test. This was done between $\tau$ and the $|log2FoldChange|$ for each organ separately.

```{r tau-v-bias-cor, echo=FALSE, message=FALSE, warning=FALSE}
cor.test(gonad_bias_tau$tau, sqrt(abs(gonad_bias_tau$log2FoldChange)), method = "spearman")
cor.test(liver_bias_tau$tau, sqrt(abs(liver_bias_tau$log2FoldChange)), method = "spearman")
cor.test(gill_bias_tau$tau, sqrt(abs(gill_bias_tau$log2FoldChange)), method = "spearman")

```

The liver and gonad both show a significant correlation. This isn't super obvious from just looking at the plot.

## $\tau$ and logFC for all the signficantly expressed genes in each organ
Only looking at the logFC for the sex-biased genes may be affecting the relationship that I am seeing between sex bias and $\tau$, especially in the organs where the samples size is very small if I am only including sex-biased genes (like the liver and gills). 

I used the same steps as above to see how the relationship changes when I add in ALL of the genes that are significantly expressed for each organ. In this datasets I have removed all rows where the adjusted p-value is NA (outliers and low-counts).

```{r tau-v-FC, echo=FALSE, fig.cap="Sex bias (in terms of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are signifcantly expressed in the gonads, liver, and gills."}
#Merge the datasets that include all of the sex-biased and non-sex-biased genes with the tau datasets
liver_bias_tau_all <- merge(liver_res, tau, by = "row.names")
#write.csv(liver_bias_tau_all, "data/liver_bias_tau.csv", row.names = FALSE)

gonad_bias_tau_all <- merge(gonad_res, tau, by = "row.names")
#write.csv(gonad_bias_tau_all, "gonad_bias_tau.csv", row.names = FALSE)

gill_bias_tau_all <- merge(gill_res, tau, by = "row.names")
#write.csv(brain_bias_tau_all, "data/brain_bias_tau.csv", row.names = FALSE)

#Plot logFC against Tau for all of the organs
par(mar=c(5.1, 4.1, 4.1, 8.1), xpd=FALSE)
plot(sqrt(abs(gonad_bias_tau_all$log2FoldChange)),
     gonad_bias_tau_all$tau,
     pch=19,
     col = "#EEB42275",
     xlab="sex bias (|log2 fold change|)",
     ylab=expression(tau["TPM"]),
     xlim = c(0, 5.5),
     ylim = c(0, 1))
points(sqrt(abs(liver_bias_tau_all$log2FoldChange)),
     liver_bias_tau_all$tau,
     pch=19,
     col="#EE826275")
points(sqrt(abs(gill_bias_tau_all$log2FoldChange)),
     gill_bias_tau_all$tau,
     pch=19,
     col = "#20B2AA75")


#Add the best fit lines
abline(lm(gonad_bias_tau_all$tau ~ sqrt(abs(gonad_bias_tau_all$log2FoldChange))), col="#EEB422", lwd = 3)
abline(lm(liver_bias_tau_all$tau ~ sqrt(abs(liver_bias_tau_all$log2FoldChange))), col = "#EE8262", lwd = 3, lty = 2)
abline(lm(gill_bias_tau_all$tau ~ sqrt(abs(gill_bias_tau_all$log2FoldChange))), col = "#6E8B3D", lwd = 3, lty = 3)

outer_legend("right",
             c("Gonad", "Liver", "Gill"),
             col = c("#EEB422", "#EE8262", "#6E8B3D"),
             #pch = 19,
             lwd = 3,
             bty = 'n',
             cex = 1.5,
             lty = 1:3,
             ncol = 1)

outer_legend("right",
             c("              ", "             ", "            "),
             col = c("#EEB422", "#EE8262", "#20B2AA"),
             pch = 21,
             pt.bg = c("#EEB42275", "#EE826275", "#20B2AA75"),
             #lwd = 3,
             bty = 'n',
             cex = 1.5,
             #lty = 1:3,
             ncol = 1)


```

The absolute value of the log2 fold change was square root transformed. We can see similar patterns as when we only plotted the sex biased genes with the strongest relationship showing up in the gonads. When we add in all of the points, however, it seems that the best fit line for the liver and gills now points downward. Let's see how the Spearman's rank correlation test have changed. This was done between $\tau$ and the $|log2FoldChange|$ for each organ separately.

```{r tau-v-FC-cor, echo=FALSE, warning=FALSE}
cor.test(gonad_bias_tau_all$tau, abs(gonad_bias_tau_all$log2FoldChange), method = "spearman")
cor.test(liver_bias_tau_all$tau, abs(liver_bias_tau_all$log2FoldChange), method = "spearman")
cor.test(gill_bias_tau_all$tau, abs(gill_bias_tau_all$log2FoldChange), method = "spearman")

```

All correlations are now significant, however, for the gills and the liver there is a significantly negative correlation between $\tau$ and the degree of sex bias. 

## Investigating genes with high tau and high logFC`

Because we have information about which tissue each gene is specific in, I was to see if genes that are very sex-biased in a tissue are also tissue-specific in that same tissue.

### Gonads

```{r gonad-tau-logFC, echo=FALSE, fig.cap="_Sex bias (in terms of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are sex-biased in the gonads. Color represents the tissue that it is specific in which may be either the testis or ovaries (yellow), the liver (pink), the gills (blue), or none (black)._"}

plot(sqrt(abs(gonad_bias_tau$log2FoldChange)),
     gonad_bias_tau$tau,
     pch=19,
     col = ifelse(gonad_bias_tau$organ %in% c("Ovaries", "Testis"),
                  "#EEB42275",
                  ifelse(gonad_bias_tau$organ == "Liver",
                         "#EE826275",
                         ifelse(gonad_bias_tau$organ == "Gill",
                                "#20B2AA75",
                                "#00000075")
                         )
                  ),
     xlab="sex bias (|log2 fold change|)",
     ylab=expression(tau["TPM"]),
     ylim = c(0, 1))

```

In the gonads it looks like a lot of the genes that are extremely sex-biased are also tissue specific in the gonads (whether that is tissue-specific in the testis or in the ovaries).

### Liver

```{r liver-tau-logFC, echo=FALSE, fig.cap="_Sex bias (in terms of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are sex-biased in the liver. Color represents the tissue that it is specific in which may be either the testis or ovaries (yellow), the liver (pink), the gills (blue), or none (black)._"}

plot(sqrt(abs(liver_bias_tau$log2FoldChange)),
     liver_bias_tau$tau,
     pch=19,
     col = ifelse(liver_bias_tau$organ %in% c("Ovaries", "Testis"),
                  "#EEB42275",
                  ifelse(liver_bias_tau$organ == "Liver",
                         "#EE826275",
                         ifelse(liver_bias_tau$organ == "Gill",
                                "#20B2AA75",
                                "#00000075")
                         )
                  ),
     xlab="sex bias (|log2 fold change|)",
     ylab=expression(tau["TPM"]),
     ylim = c(0, 1))

```

Similar to the gonads, most of the extremely biased genes in the liver are also tissue specific in the liver, expect for a few cases. It looks like some of the  most biased genes in the liver are showing more expression in the gonads and there are two cases where tau is 0. Other than that, many of the more intermediately sex-biased genes show the highest expression in gonads.

### Gills

```{r gills-tau-logFC, echo=FALSE, fig.cap="_Sex bias (in terms of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are sex-biased in the gills. Color represents the tissue that it is specific in which may be either the testis or ovaries (yellow), the liver (pink), the gills (blue), or none (black)._"}

plot(sqrt(abs(gill_bias_tau$log2FoldChange)),
     gill_bias_tau$tau,
     pch=19,
     col = ifelse(gill_bias_tau$organ %in% c("Ovaries", "Testis"),
                  "#EEB42275",
                  ifelse(gill_bias_tau$organ == "Liver",
                         "#EE826275",
                         ifelse(gill_bias_tau$organ == "Gill",
                                "#20B2AA75",
                                "#00000075")
                         )
                  ),
     xlab="sex bias (|log2 fold change|)",
     ylab=expression(tau["TPM"]),
     ylim = c(0, 1))

```

There aren't many biased genes in the gills, however, of the ones that are there is quite a mix in terms of which tissue is showing the highest overall expression.