---
title: "Differential Expression Analysis in _Syngnathus fuscus_"
author: "Coley Tosto"
date: "`r Sys.Date()`"
output:
  github_document: 
    toc: true 
  #bookdown::pdf_document2:
    #fig_caption: yes
    #keep_tex: yes
    #number_sections: yes
    #toc: true
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
```

``` {r library, include = FALSE}
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
library(spfTools)
library(magick)
library(patchwork)
library(ggpubr)
library(cowplot)
library(UpSetR)
library(ggplot2)
library(kableExtra)
library(tidyverse)
library(knitr)
library(webshot)
library(htmltools)
library(dplyr)
library(pdftools)

```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
FU_txi.salmon <- readRDS("data/txi.salmon_FU.RDS")

#The samples file generated for tximport
FU_samples <- read.table("FU_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
FU_samples$Sex <- as.factor(FU_samples$Sex)
FU_samples$Organ <- as.factor(FU_samples$Organ)

```

The package `DESeq2` was used for the differential expression analysis outlined below.

# Single factor analysis - Comparing Males v Females
I am starting with a single factor analysis to look at overall differences between male and female expression across the organs and to do some exploratory analysis of the data. 

To analyze my data with DESeq2 I must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` which is stored in the object "FU_txi.salmon" and was generated at the end of the `Analyzing_fuscus_RNAseq_data_from_MSU.Rmd` document and a `samples` file which is stored in the object "FU_samples" that lays out all of the conditions.

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_FU <- DESeqDataSetFromTximport(FU_txi.salmon, 
                                   colData = FU_samples,
                                   design = ~ Sex)

```

I am then going to pre-filter the data to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

The cutoff used here was to remove reads that had a **total** count of less than 10 in each row.

```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_FU)) >= 10
dds_FU <- dds_FU[keep, ]

```

After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.

```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_FU_exp <- DESeq(dds_FU)

#Compile the results
res <- results(dds_FU_exp)
res

```

Once that has finished we can now start exploring some of the single-factor analysis results.

```{r investigate-DESeq-results-SF}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_FU_exp, alpha = 0.05)
summary(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```

Looking at the summary of our results we can see that there are 2,626 male-biased genes and 2,488 female-biased genes when all of the organs are organized together. While this is informative, I am more interested in a multi-factor approach that takes into account both **Sex** and **Tissue** type. Before the multi-factor analysis, we can visualize some results from the single-factor analysis and use it for exploratory data analysis.

Interestingly, the summary is saying that there are 0 outliers, which I have a hard time believing. 

## Visualizing the results
### MA-plot - MvF Diff.
Generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be colored if the adjusted p-value is less that 0.05.

```{r MA-plot, echo=FALSE, fig.cap="LogFC versus the mean normalized count for all of the genes. Points that are blue have a p-value less than 0.05. The two plots are the same with the only difference coming from an adjusted y-limit. Points that are triangles represent genes with a fold change higher/lower than the y-limit."}

par(mfrow=c(1,2))
plotMA(res05, ylim = c(-2,2))
plotMA(res05)

```

### Heatmap - Looking at patterns of overall expression
We can also generate a heat map to look ar overall expression levels across our samples. Note, this is not differentially expressed genes.

```{r heatmap, echo=FALSE, fig.cap= "Heatmap showing the expression level across all organs for the top 20 genes with the highest expression levels."}

#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(dds_FU_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]
df <- as.data.frame(colData(dds_FU)[,c("Sex", "Organ")])

#Transform the data
vsd <- vst(dds_FU_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df)

```

From the heatmap we can see that for a lot of the top expressed genes the highest expression are found in the liver. There appears to be a few genes that are highly expressed across all tissues and a few genes that are highly expressed in the gill tissues relative to the other tissues. With this heatmap we can also pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.

```{r heatmap_genes, eval = FALSE}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))

#Export the gene names
write.table(heatmap_TG,
            'FU_heatmap_trinitygenes.txt',
            sep = "",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)

```

### Sample-dist Heatmap
We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples.

```{r sample-dist, echo=FALSE, fig.cap="Sample-to-sample distances heatmap showing the similarities and differences between all of the samples. The darker the color, the more similar the samples are. The diagonal line of very dark blue represents the comparissons between the same samples.", warning=FALSE}

#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$Sex, vsd$Organ, sep = "-")
colnames(sampleDistsMatrix) <- paste(vsd$ID)

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)

```

We can see that the highest similarities (aside from same samples comparisons) is between samples from the same organ. After that we can see high similarities between the male gonads and the female gonads. There are also a couple samples that may be an issue for use as they are not similar to any of the other samples from the same group. This includes FUG11M4 and FUT11M4. I may end up removing all organs related to 11M4.

### PCA plots
Next we can look at PCA plots for our samples to see how our different samples are grouping.

```{r pca-pairs-plot, echo=FALSE, fig.cap="Principal components analysis pairsplot showing PCA axis 1 - 4."}
#Generate the PCA data to create the pairs plot, using the PCAtools package
p <- pca(assay(vsd), 
         metadata = colData(dds_FU))

#Create the pairs plot
pairsplot(p,
          components = getComponents(p, c(1:4)),
          colby = "Organ",
          triangle = FALSE,
          hline = 0, vline = 0,
          colkey = c("Gill" = "#6E8B3D75", 
                     "Gonad" = "#EEB42275", 
                     "Liver" = "#EE826275"),
          shape = "Sex",
          shapekey = c(16, 17),
          pointSize = 3,
          margingaps = unit(c(0.1, 0.1, 0.1, 0.1), 'cm'),
          legendPosition = "left", legendIconSize = 3, legendLabSize = 10)

```

From the pairs plot we can see that PC1 appears to be explaining differences between the liver and the other tissues, PC2 appears to be explaining differences between gonadal and somatic tissues, PC3 seems to account for differences between male and female samples, especially in the gonads and in PC4 we can see the one problematic gill sample from above sticking out from all the rest. I am going to plot all of PC2 - 4 against PC1 below for a clearer picture.

```{r pca-plot, echo=FALSE, message=FALSE, fig.cap="Principal components analysis reflects that most variation corresponds to differences in expression between organs (green: gill; yellow: gonad; pink: liver), rather than variation due to sex (circle = female; triangle = male). The first axis explains 31% of variation in the dataset, and largely explains differences between the liver and the other tissues. The second axis explains 20% of the variation in gene expression, with somatic tissues on the opposite side of the axis from the gonads.", warning=FALSE}

#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$Organ <- FU_samples$Organ
pca_plotting_data$Sex <- FU_samples$Sex
pca_plotting_data$ID <- FU_samples$ID

#Calculate the percent variation attributed to each axis 
percentVar <- round(p$variance, digits = 2)


my_colors <- c("Gill" = "#6E8B3D75", "Gonad" = "#EEB42275", "Liver" = "#EE826275")
my_shapes <- c(16, 17)

ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = Organ, shape = Sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

```{r pca1v3, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC3. This third axis explains 10% of the variation, largely attributed to differences between the sexes, particularily in the gonads.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC3, color = Organ, shape = Sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC3: ",percentVar[3],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

```{r pca1v4, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC4. This fourth axis explains 6% of the variation, largely attributed to one of the gill samples.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC4, color = Organ, shape = Sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC4: ",percentVar[4],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

#### Saving the PCA plots to use in figure creation
I will be including the PCAs describing the relationship between PC1 - 3 in the main text, as they account for the majority of the differences between our samples. The plot with PC1 and PC4 I still want to export, and will likely include as a supplemental to highlight why samples related to 11M4 will be removed from the subsequent analyses.

```{r setParams-PCAs}
#Setting the shapes I want to show up for the different organs
organ_shapes <- c(Gill = 16, Liver = 17, Gonad = 15)

#Setting the colors I want for each sex
sex_cols <- c("F" = "#7fc97f", "M" = "#beaed4")

```

```{r figPCA1v2}
#Create the blank pdf to store the plot in
pdf("docs/figs/Fig_PCA1v2.pdf",height = 6,width=6)

#Set the plotting parameters
par(mar=c(4,5,4,1), oma=c(2,2,2,2))

#Create the plot for PC1 v PC2
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC2,
     col = paste0(sex_cols[pca_plotting_data$Sex],"75"),
     pch = organ_shapes[pca_plotting_data$Organ],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC2: ",percentVar[2], "% variance"),
     bty = 'l',
     xpd = TRUE)

#Add a legend describing the sex
outer_legend("top",
             c("Female","Male"),
             pch = 18,
             bty = 'n',
             col=paste0(sex_cols,"75"),
             cex = 1.75,
             ncol = 2,
             pt.cex = 3)

dev.off()
```

```{r figPCA1v3}
#Create the blank pdf to store the plot in
pdf("docs/figs/Fig_PCA1v3.pdf",height = 6,width=6)

#Set the plotting parameters
par(mar=c(4,5,4,1), oma=c(2,2,2,2))

#Create the plot for PC1 v PC3
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC3,
     col = paste0(sex_cols[pca_plotting_data$Sex],"75"),
     pch = organ_shapes[pca_plotting_data$Organ],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC3: ",percentVar[3], "% variance"),
     bty = 'l',
     xpd = TRUE)

#Add  legend describing the organs
outer_legend("top",
             c("Gill","Gonad","Liver"),
             pch = c(16,15,17),
             bty = 'n',
             col = 'darkgrey',
             cex = 1.75,
             ncol = 3,
             pt.cex = 3)

dev.off()
```

```{r figPCA1v4}
#Create the blank pdf to store the plot in
pdf("docs/figs/Fig_PCA1v4.pdf",height = 6,width=8)

#Set the plotting parameters
par(mar=c(4,5,1,5),oma=c(2,2,2,2))

#Create the plot for PC1 v PC4
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC4,
     col = paste0(sex_cols[pca_plotting_data$Sex],"75"),
     pch = organ_shapes[pca_plotting_data$Organ],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC4: ",percentVar[4], "% variance"),
     bty = 'l',
     xpd = TRUE)

text(pca_plotting_data$PC1[pca_plotting_data$ID == "FUG11M4"] + 50,
     pca_plotting_data$PC4[pca_plotting_data$ID == "FUG11M4"] - 25,
     labels = pca_plotting_data$ID[pca_plotting_data$ID == "FUG11M4"])

#Add  legend describing the organs and sex
outer_legend("right",
             c("Gill","Gonad","Liver"),
             pch = c(16,15,17),
             bty = 'n',
             col = 'darkgrey',
             cex = 1.5,
             ncol = 1,
             pt.cex = 2.25)

outer_legend("topright",
             c("Female","Male"),
             pch = 18,
             bty = 'n',
             col=paste0(sex_cols,"75"),
             cex = 1.5,
             ncol = 1,
             pt.cex = 2.25)
dev.off()
```

#### Creating heatmaps based on the PCA axes
I now want to generate heatmaps based on the genes that have the highest loadings for each of the PCA axes. I will be using a cut off of 0.02 to determine which genes will be includd in the heatmaps. Once again I am doing this for PCA1-4, but will likely only inlcude PCA1-3 in a figure for the maintext and PC4 will go in the supplemental.

```{r setParams-heatmaps}

df <- as.data.frame(FU_samples[,c("Sex", "Organ")])
rownames(df) <- FU_samples$ID
df$Organ <- as.character(df$Organ)

organ_cols<-c("Gill" = "#20B2AA",
              "Gonad" = "#EEB422", 
              "Liver" = "#EE8262")

ann_colors = list(Sex=sex_cols,
                  Organ=organ_cols)

col_order <- c(rownames(df[df$Sex=="F" & df$Organ=="Gill",]),
               rownames(df[df$Sex=="M" & df$Organ=="Gill",]),
               rownames(df[df$Sex=="F" & df$Organ=="Gonad",]),
               rownames(df[df$Sex=="M" & df$Organ=="Gonad",]),
               rownames(df[df$Sex=="F" & df$Organ=="Liver",]),
               rownames(df[df$Sex=="M" & df$Organ=="Liver",]))

pca_rotation <- pca$rotation[, 1:4]

```

```{r savePheatmapFxn}

#Function to use to export the heatmaps to pdfs
# from https://stackoverflow.com/questions/43051525/how-to-draw-pheatmap-plot-to-screen-and-also-save-to-file
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

```

```{r figheatmap-PC1}

pc1 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,1]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors = ann_colors,
                cellwidth = 9,
                fontsize = 16,
                annotation_legend = FALSE,
                main = "Top loading genes on PC1")

save_pheatmap_pdf(pc1, "docs/figs/Fig_pc1_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC2}

pc2 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,2]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                annotation_legend = FALSE,
                main = "Top loading genes on PC2")

save_pheatmap_pdf(pc2, "docs/figs/Fig_pc2_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC3}

pc3 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,3]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                border_color = NA,
                main = "Top loading genes on PC3")

save_pheatmap_pdf(pc3, "docs/figs/Fig_pc3_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC4}

pc4 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,4]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                border_color = NA,
                main = "Top loading genes on PC4")

save_pheatmap_pdf(pc4, "docs/figs/Fig_pc4_heatmap.pdf",
                  width=6,
                  height=6)

```

#### Making the combined figure

```{r assembleFigPCA, message=FALSE}
figPCAa <- image_ggplot(image_read_pdf('docs/figs/Fig_PCA1v2.pdf'),
                        interpolate = TRUE)
figPCAb <- image_ggplot(image_read_pdf('docs/figs/Fig_PCA1v3.pdf'),
                        interpolate = TRUE)
figPCAc <- image_ggplot(image_read_pdf('docs/figs/Fig_pc1_heatmap.pdf'),
                        interpolate = TRUE)
figPCAd <- image_ggplot(image_read_pdf('docs/figs/Fig_pc2_heatmap.pdf'),
                        interpolate = TRUE)
figPCAe <- image_ggplot(image_read_pdf('docs/figs/Fig_pc3_heatmap.pdf'),
                        interpolate = TRUE)

# make two patchworks
pcas <- figPCAa + figPCAb
hms <- figPCAc + figPCAd + figPCAe

# put the patchworks together
figPCA <-  wrap_plots(pcas,
                      hms,
                      ncol=1) + 
  plot_annotation(tag_levels = 'A') 

ggsave("docs/figs/FigPCA.pdf", figPCA, height=4, width=5)
ggsave("docs/figs/FigPCA.png", figPCA, height=4, width=5) # also save as a png
```

```{r assembleFigPCA-supp, message=FALSE}
figPCA4 <- image_ggplot(image_read_pdf('docs/figs/Fig_PCA1v4.pdf'),
                        interpolate = TRUE)
figPCA4_hm <- image_ggplot(image_read_pdf('docs/figs/Fig_pc4_heatmap.pdf'),
                           interpolate = TRUE)

# put the patchworks together
figPCA_supp <-  wrap_plots(figPCA4, figPCA4_hm, ncol=2) + 
  plot_annotation(tag_levels = 'A') 

ggsave("docs/figs/FigPCA_supp.pdf", figPCA_supp, height=4, width=8)
ggsave("docs/figs/FigPCA_supp.png", figPCA_supp, height=4, width=8) # also save as a png
```

```{r showAssembledFigPCA, eval=TRUE}
figPCA
```


# Multifactor design - Comparing M v F across the diff tissue types
If we investigate the column data of our DESeq dataset we can see that each sample has both a sex and organ type attached to it, we will be using these two factors in our multi-factor analysis. Prior to running the multi-factor analysis I am going to remove the samples that correspond to the individual 11M4.

```{r multi-factor-analysis, message=FALSE, warning=FALSE}

#Create an additional column the has the two conditions combined(sex and organ type)
FU_samples$group <- factor(paste0(FU_samples$Sex, FU_samples$Organ))

##Create a copy of the DESeq dataset
ddsMF_FU <- DESeqDataSetFromTximport(FU_txi.salmon,
                                     colData = FU_samples,
                                     design = ~ group)

##Remove all 11M4 organs from the dataset
ddsMF_FU <- ddsMF_FU[, !(ddsMF_FU$ID %in% c("FUT11M4", "FUG11M4", "FUL11M4"))]

##Filter the dataset, only keeping rows that have at least 10 reads total
keep <- rowSums(counts(ddsMF_FU)) >= 10
ddsMF_FU <- ddsMF_FU[keep, ]

#Run the differential expression analysis
ddsMF_FU_exp <- DESeq(ddsMF_FU)
resultsNames(ddsMF_FU_exp)

#Looking at the Normalized read counts for each sample
cbind(FU_samples[!(FU_samples$ID %in% 
                     c("FUT11M4", "FUG11M4", "FUL11M4")),],
      normalizedReads=colSums(counts(ddsMF_FU_exp, 
                                     normalized=TRUE)
                              )
      )

```

If we look at the summary of normalized reads across the different tissue types, we can see that there are considerably more in the liver relative to the other tissue types. Because during a multi-factor analysis normalization occurs across all of the tissues, I am opting to also remove some of the extremely expressed genes to try and offset the differences.

```{r refilter-mf-analysis, message=FALSE, warning=FALSE}

##Filter the dataset, only keeping rows that have at least 10 reads total AND less than one million
keep <- rowSums(counts(ddsMF_FU)) >= 10 & rowSums(counts(ddsMF_FU)) < 1e6
ddsMF_FU <- ddsMF_FU[keep, ]

#Re-run the differential expression analysis
ddsMF_FU_exp <- DESeq(ddsMF_FU)

#Re investigate the Normalized read counts for each sample
cbind(FU_samples[!(FU_samples$ID %in% 
                     c("FUT11M4", "FUG11M4", "FUL11M4")),],
      normalizedReads=colSums(counts(ddsMF_FU_exp, 
                                     normalized=TRUE)
                              )
      )

```

With this change we can now see that there is more even levels of expression across the different tissue types. 

## Invesitgate the results of the differential expression
Thanks to the multi-factor analysis we can now explore differential expression between all of the different combinations of tissues and sexes: 

  - Male Liver v. Female Liver 
  - Male Gill v. Female Gill  
  - Male Gonad v. Female Gonad  
  - All of the within sex tissue comparisons (e.g. Male Liver v. Male Gill, etc.)  

### M-F Liver Comparisson

```{r m-f-liver-results}
##Pulling out the liver M-F results with an alpha of 0.05
FU_liver_con_res <- results(ddsMF_FU_exp, 
                            contrast = c("group", "MLiver", "FLiver"), 
                            alpha = 0.05)
FU_liver_con_res$trin_geneid <- rownames(FU_liver_con_res)
head(FU_liver_con_res)

summary(FU_liver_con_res)
```

I will be classifying sex-biased genes the same way that I did for _Syngnathus floridae_ as **genes with a p-value < 0.05 AND a logFC $\ge$ |2|**. With that criteria in the liver there are `r sum(FU_liver_con_res$padj <= 0.05 & FU_liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(FU_liver_con_res$padj <= 0.05 & FU_liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. 

I have then pulled out all of the male-biased, female-biased, and non-biased genes based on the criteria outlined above. I determined non-biased genes as a p-vaule > 0.05.

```{r liver-sex-biased}
#Removing the rows where padj. is NA in results
FU_liver_con_res_noNA <- FU_liver_con_res[!is.na(FU_liver_con_res$padj),]
summary(FU_liver_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating a vector that contains all of the male-biased and female-biased genes in the liver
FU_liver_mal_biased <- FU_liver_con_res_noNA[which(FU_liver_con_res_noNA$log2FoldChange >= 2 &
                                                     FU_liver_con_res_noNA$padj <= 0.05),]

FU_liver_fem_biased <- FU_liver_con_res_noNA[which(FU_liver_con_res_noNA$log2FoldChange <= -2 &
                                                     FU_liver_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the liver
FU_liver_non_biased <- FU_liver_con_res_noNA[which(FU_liver_con_res_noNA$padj > 0.05),]

```

I will be generating a table that outlines the genes that are the most male or female biased in each organ type. To do this I need to pull the top 50 differentially expressed genes in males or females, get the Trinity gene IDs and then BLAST the corresponding sequences. Here I am pulling out the Trinity gene IDs for those genes.

```{r top50-liver, eval = FALSE}
#Creating a subset of the results where the p-value is less than 0.05
FU_liver_con_res_p05 <- FU_liver_con_res_noNA[FU_liver_con_res_noNA$padj <= 0.05, ]


#Pulling the top 50 diff expressed genes in Males
top50_FU_MLiver <- head(FU_liver_con_res_p05[order(FU_liver_con_res_p05$log2FoldChange, 
                                                   decreasing = TRUE), ], 
                        n = 50)

#Run once to generate the file and then commented out
write.table(cbind(rownames(top50_FU_MLiver)),
            'FU_maleL_top50TRgenes.txt', 
             sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)


#Pulling the top 50 diff expressed genes in Females
top50_FU_Fliver <- head(FU_liver_con_res_p05[order(FU_liver_con_res_p05$log2FoldChange, 
                                                   decreasing = FALSE), ], 
                        n = 50)

#Run once to generate the file and then commented out
write.table(cbind(rownames(top50_FU_Fliver)),
            'FU_femL_top50TRgenes.txt', 
             sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```


### M-F Gill Comparisson

```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res_FU <- results(ddsMF_FU_exp, 
                           contrast = c("group", "MGill", "FGill"), 
                           alpha = 0.5)
gill_con_res_FU$trin_geneid <- rownames(gill_con_res_FU)
head(gill_con_res_FU)

summary(gill_con_res_FU)
```

In the gills there were `r sum(gill_con_res_FU$padj <= 0.05 & gill_con_res_FU$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res_FU$padj <= 0.05 & gill_con_res_FU$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

I have then pulled out all of the male-biased, female-biased, and non-biased genes to save them into their own objects.

```{r gill-sex-biased}
#Removing the rows where padj. is NA in results
FU_gill_con_res_noNA <- gill_con_res_FU[!is.na(gill_con_res_FU$padj), ]
summary(FU_gill_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating a vector that contains all of the male-biased and female-biased genes in the gills
FU_gill_mal_biased <- FU_gill_con_res_noNA[which(FU_gill_con_res_noNA$log2FoldChange >= 2 
                                                 & FU_gill_con_res_noNA$padj <= 0.05),]
FU_gill_fem_biased <- FU_gill_con_res_noNA[which(FU_gill_con_res_noNA$log2FoldChange <= -2 
                                                 & FU_gill_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gills, p>0.05
FU_gill_non_biased <- FU_gill_con_res_noNA[which(FU_gill_con_res_noNA$padj > 0.05),]
```

Here I am getting the trinity gene IDs that correspond to the top 50 differentially expressed genes for males and females.

```{r top50-gill, eval = FALSE}
#Creating a subset of results where p-value is less than 0.05
FU_gill_con_res_p05 <- FU_gill_con_res_noNA[FU_gill_con_res_noNA$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_FU_MGill <- head(FU_gill_con_res_p05[order(FU_gill_con_res_p05$log2FoldChange, 
                                                 decreasing = TRUE),], 
                       n=50)

#Run once to generate the file and then commented out
write.table(cbind(rownames(top50_FU_MGill)),
            'FU_maleG_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FU_FGill <- head(FU_gill_con_res_p05[order(FU_gill_con_res_p05$log2FoldChange, 
                                                 decreasing = FALSE),], 
                       n=50)

#Run once to generate the file and then commented out
write.table(cbind(rownames(top50_FU_FGill)),
            'FU_femG_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

```

### M-F Gonad Comparisson

```{r m-f-gonads}
##Pulling out the gonad M-F results
gonad_con_res_FU <- results(ddsMF_FU_exp, 
                            contrast = c("group", "MGonad", "FGonad"), 
                            alpha = 0.05)
gonad_con_res_FU$trin_geneid <- rownames(gonad_con_res_FU)
head(gonad_con_res_FU)

summary(gonad_con_res_FU)
```

Between the ovaries and testis (gonads) there were `r sum(gonad_con_res_FU$padj <= 0.05 & gonad_con_res_FU$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gonad_con_res_FU$padj <= 0.05 & gonad_con_res_FU$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

From the gonad M-F results I have then pulled out all of the male-biased, female-biased, and non-biased genes and saved them to their own objects.

```{r gonad-sex-biased}
#Removing the rows where padj. is NA in results
FU_gonad_con_res_noNA <- gonad_con_res_FU[!is.na(gonad_con_res_FU$padj), ]
summary(FU_gonad_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating an object that contains all of the male-biased and female-biased genes in the gonads
FU_gonad_mal_biased <- FU_gonad_con_res_noNA[which(FU_gonad_con_res_noNA$log2FoldChange >= 2 
                                                   & FU_gonad_con_res_noNA$padj <= 0.05),]
FU_gonad_fem_biased <- FU_gonad_con_res_noNA[which(FU_gonad_con_res_noNA$log2FoldChange <= -2 
                                                   & FU_gonad_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gonads, p>0.05
FU_gonad_non_biased <- FU_gonad_con_res_noNA[which(FU_gonad_con_res_noNA$padj > 0.05),]
```

I then pulled out the trinity gene IDs that corresponded to the top 50 differentially expressed genes in males and females so that I could BLAST the sequences.

```{r top50-gonad, eval=FALSE}
#Creating a subset of the results where the p-value is less than 0.05
FU_gonad_con_res_p05 <- FU_gonad_con_res_noNA[FU_gonad_con_res_noNA$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_FU_MGonad <- head(FU_gonad_con_res_p05[order(FU_gonad_con_res_p05$log2FoldChange, 
                                                   decreasing = TRUE),], 
                        n=50)

#Run once to generate the file and then commented out
write.table(cbind(rownames(top50_FU_MGonad)),
            'FU_maleGon_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FU_FGonad <- head(FU_gonad_con_res_p05[order(FU_gonad_con_res_p05$log2FoldChange, 
                                                   decreasing = FALSE),], 
                        n=50)

#Run once to generate the file and then commented out
write.table(cbind(head(top50_FU_FGonad)),
            'FU_femGon_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```


### MA Plots

```{r MA-plot-MF, echo=FALSE, fig.height= 15, fig.width=8, fig.cap="MA-plots generated for each organ type that compares logFC to the mean expression. Female-biased and Male-biased genes are represented by color (green and purple correspondingly) and are determing with a fc cutoff of 2 and a p-value cutoff of 0.05."}
liver_MA <- ggmaplot(FU_liver_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Liver" %->% "Female Liver"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gill_MA <- ggmaplot(gill_con_res_FU,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Gill" %->% "Female Gill"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gonad_MA <- ggmaplot(gonad_con_res_FU,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Testis" %->% "Ovaries"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

#Merge all the plots into one
plot_grid(liver_MA, gill_MA, gonad_MA,
          ncol = 1,
          nrow = 3,
          labels = c('A', 'B', 'C'),
          label_fontfamily = 'sans',
          label_fontface = 'bold',
          label_size = 18,
          rel_heights = c(4.10, 4.10, 4.10))

```

## Variation in FC across sex-bias and tissue type
I want to create a plot the highlights the variation we see in fold-change both across the different biases groups (male-biased, female-biased, and non-biased) within one tissue type and across all of the tissue types. To do this I first need to create a "long" dataset in the following format:

| Tissue_type|     Bias      |   Fold-Change   |
|:----------:|:-------------:|:---------------:|
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   fem_bias    |      XXXXX      |
|    Gill    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   fem_bias    |      XXXXX      |
|    Liver   |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   fem_bias    |      XXXXX      |
|   Gonad    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |

That dataset is created here, with the Trinity gene IDs included as well:

```{r logFC-long-dataset}
FU_logFC_long <- data.frame(
  tissue=c(rep("Gill",nrow(FU_gill_fem_biased)),
           rep("Gill", nrow(FU_gill_mal_biased)),
           rep("Gill", nrow(FU_gill_non_biased)),
           rep("Gonad", nrow(FU_gonad_fem_biased)),
           rep("Gonad", nrow(FU_gonad_mal_biased)),
           rep("Gonad", nrow(FU_gonad_non_biased)),
           rep("Liver", nrow(FU_liver_fem_biased)),
           rep("Liver", nrow(FU_liver_mal_biased)),
           rep("Liver", nrow(FU_liver_non_biased))
         ),
  bias=c(rep("FB",nrow(FU_gill_fem_biased)),
         rep("MB",nrow(FU_gill_mal_biased)),
         rep("NB", nrow(FU_gill_non_biased)),
         rep("FB", nrow(FU_gonad_fem_biased)),
         rep("MB", nrow(FU_gonad_mal_biased)),
         rep("NB", nrow(FU_gonad_non_biased)),
         rep("FB", nrow(FU_liver_fem_biased)),
         rep("MB", nrow(FU_liver_mal_biased)),
         rep("NB", nrow(FU_liver_non_biased))
         ),
  logFC=c(FU_gill_fem_biased$log2FoldChange,
          FU_gill_mal_biased$log2FoldChange,
          FU_gill_non_biased$log2FoldChange,
          FU_gonad_fem_biased$log2FoldChange,
          FU_gonad_mal_biased$log2FoldChange,
          FU_gonad_non_biased$log2FoldChange,
          FU_liver_fem_biased$log2FoldChange,
          FU_liver_mal_biased$log2FoldChange,
          FU_liver_non_biased$log2FoldChange
          ),
  geneID=c(rownames(FU_gill_fem_biased),
           rownames(FU_gill_mal_biased),
           rownames(FU_gill_non_biased),
           rownames(FU_gonad_fem_biased),
           rownames(FU_gonad_mal_biased),
           rownames(FU_gonad_non_biased),
           rownames(FU_liver_fem_biased),
           rownames(FU_liver_mal_biased),
           rownames(FU_liver_non_biased)
           )
  
)
```

With the dataset in the proper format now, we can generate the plot

```{r FC-var-plot, echo=FALSE, fig.cap="Absolute value of the logFC for genes that are female-biased, male-biased, and unbiased across each tissue type. Raw fold-change values are added ontop of the boxplot as jitters."}

pdf("docs/figs/FigMF_logFC_boxplots.pdf", width = 10, height=4)

ymax <- max(abs(FU_logFC_long$logFC),
            na.rm = TRUE) + 5
sex_bias_colors <- c("FB" = "#7fc97f", 
                     "MB" = "#beaed4", 
                     "UB" = "darkgray")
organs <- levels(as.factor(FU_logFC_long$tissue))

par(mfrow=c(1,3), 
    oma=c(4,5,2,8), 
    mar=c(1,1,1,0))

for(organ in organs){
   
   # add jittered points
   plot(abs(FU_logFC_long$logFC[FU_logFC_long$tissue==organ]) ~ 
            jitter(as.numeric(as.factor(FU_logFC_long$bias[FU_logFC_long$tissue==organ]))),
        col="#00000075", 
        axes=FALSE,
        cex.main=2,
        xlim=c(0,4),
        ylim=c(0,ymax)
        )
   
   # make the boxplot
   boxplot(abs(FU_logFC_long$logFC[FU_logFC_long$tissue==organ]) ~  
             FU_logFC_long$bias[FU_logFC_long$tissue==organ],
           col=scales::alpha(sex_bias_colors,0.75),
           add = TRUE,
           yaxt='n',
           las=1,
           cex.axis=1.75,
           frame=FALSE,
           lwd=1.5,
           outline=FALSE# do not plot outliers
           )
   
   # make the axis lines longer and thicker
   axis(1, labels=NA,
        at=-1:4,
        lwd=2, 
        lwd.ticks=0, 
        las = 3, 
        cex.axis = 1.5)
   axis(2, labels=seq(-5,ymax,10),
        line=NA, 
        at=seq(-5,ymax,10), 
        lwd=2,
        ylim=c(0,ymax),
        cex.axis=2,
        las=2)
   
   mtext(organ, cex=1.5,outer=FALSE, line=-1)
   
}

# add x axis label
mtext("Bias level",side = 1,cex=1.5, outer=TRUE, line=2)

# add y axis label
mtext(expression("|log"[2]*"FC|"),side = 2,cex=1.5, outer=TRUE, line=2.25)

# add legend
spfTools::outer_legend("right",
                       legend=c("Female\nbiased",
                                "Male\nbiased",
                                "Unbiased"),
                       pt.bg=sex_bias_colors,
                       pch=22,
                       bty='n',
                       ncol=1,
                       cex=2,
                       y.intersp = 1.5,
                       pt.cex=2)

dev.off()

```

There are some weird patterns popping up in this analysis. The first one is the magnitude of differences in male- and female-biased genes in the liver. This weirdness is then confirmed in the above plot, we can see this clumping pattern showing um in the male-biased genes in the liver. There is a cluster of genes with a logFC between 5 -15 and then another clump at 25 and above, and very little in between. This may be a result of normalizing across all of the organs. I am going to try re-running the DE analysis within each organ and see if that helps anything.

# Running a single factor analysis between males and females WITHIN each organ
## Liver M-F differential expression

```{r liver-sf-analysis, message=FALSE}
#Create a subset of the DESeq dataset that only includes the liver samples
dds_FU_liver <- ddsMF_FU[, ddsMF_FU$Organ == "Liver"]

#Filtering to remove any rows that have less than 10 reads total
keep <- rowSums(counts(dds_FU_liver)) >= 10
dds_FU_liver <- dds_FU_liver[keep, ]

design(dds_FU_liver) <- ~ Sex

#Run the differential expression analysis
dds_FU_liver_exp <- DESeq(dds_FU_liver)

#Compiling the results
res_liver <- results(dds_FU_liver_exp, alpha = 0.05)

summary(res_liver)

```

The single factor analysis from the liver showed `r sum(res_liver$padj <= 0.05 & res_liver$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(res_liver$padj <= 0.05 & res_liver$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. I am going to pull out all of the female- and male-biased genes to store them in their own objects as well as genes that are non-biased. Non-biased genes will be categorized as genes with a p-value > 0.05, regardless of what the logFC may be.

```{r liver-sf-sexbiased}
#Removing the rows where padj. is NA in results
liver_res_noNA <- res_liver[!is.na(res_liver$padj),]
summary(liver_res_noNA) #We can now see that there are no outliers or 
                            #low counts since the NAs have been removed
#write.csv(as.data.frame(liver_res_noNA), "data/liver_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the brain
liver_mal_biased <- liver_res_noNA[which(liver_res_noNA$log2FoldChange >= 2
                                             & liver_res_noNA$padj <= 0.05),]
liver_fem_biased <- liver_res_noNA[which(liver_res_noNA$log2FoldChange <= -2 
                                             & liver_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the brain
liver_non_biased <- liver_res_noNA[which(liver_res_noNA$padj > 0.05),]
```

## Gonad M-F differential expression

```{r gonad-sf-analysis, message=FALSE}
#Create a subset of the DESeq dataset that only includes the gonad samples
dds_FU_gonad <- ddsMF_FU[, ddsMF_FU$Organ == "Gonad"]

#Filtering to remove any rows that have less than 10 reads total
keep <- rowSums(counts(dds_FU_gonad)) >= 10
dds_FU_gonad <- dds_FU_gonad[keep, ]

design(dds_FU_gonad) <- ~ Sex

#Run the differential expression analysis
dds_FU_gonad_exp <- DESeq(dds_FU_gonad)

#Compiling the results
res_gonad <- results(dds_FU_gonad_exp, alpha = 0.05)

summary(res_gonad)

```

The single factor analysis from the gonad showed `r sum(res_gonad$padj <= 0.05 & res_gonad$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(res_gonad$padj <= 0.05 & res_gonad$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. I am going to pull out all of the female- and male-biased genes to store them in their own objects as well as genes that are non-biased. Non-biased genes will be categorized as genes with a p-value > 0.05, regardless of what the logFC may be.

```{r gonad-sf-sexbiased}
#Removing the rows where padj. is NA in results
gonad_res_noNA <- res_gonad[!is.na(res_gonad$padj),]
summary(gonad_res_noNA) #We can now see that there are no outliers or 
                            #low counts since the NAs have been removed
#write.csv(as.data.frame(gonad_res_noNA), "data/gonad_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the brain
gonad_mal_biased <- gonad_res_noNA[which(gonad_res_noNA$log2FoldChange >= 2
                                             & gonad_res_noNA$padj <= 0.05),]
gonad_fem_biased <- gonad_res_noNA[which(gonad_res_noNA$log2FoldChange <= -2 
                                             & gonad_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the brain
gonad_non_biased <- gonad_res_noNA[which(gonad_res_noNA$padj > 0.05),]
```

## Gill M-F differential expression

```{r gill-sf-analysis, message=FALSE}
#Create a subset of the DESeq dataset that only includes the gill samples
dds_FU_gill <- ddsMF_FU[, ddsMF_FU$Organ == "Gill"]

#Filtering to remove any rows that have less than 10 reads total
keep <- rowSums(counts(dds_FU_gill)) >= 10
dds_FU_gill <- dds_FU_gill[keep, ]

design(dds_FU_gill) <- ~ Sex

#Run the differential expression analysis
dds_FU_gill_exp <- DESeq(dds_FU_gill)

#Compiling the results
res_gill <- results(dds_FU_gill_exp, alpha = 0.05)

summary(res_gill)

```

The single factor analysis from the gill showed `r sum(res_gill$padj <= 0.05 & res_gill$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(res_gill$padj <= 0.05 & res_gill$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. I am going to pull out all of the female- and male-biased genes to store them in their own objects as well as genes that are non-biased. Non-biased genes will be categorized as genes with a p-value > 0.05, regardless of what the logFC may be.

```{r gill-sf-sexbiased}
#Removing the rows where padj. is NA in results
gill_res_noNA <- res_gill[!is.na(res_gill$padj),]
summary(gill_res_noNA) #We can now see that there are no outliers or 
                            #low counts since the NAs have been removed
#write.csv(as.data.frame(gill_res_noNA), "data/gill_res.csv", row.names = TRUE)

#Creating a vector that contains all of the male-biased and female-biased genes in the brain
gill_mal_biased <- gill_res_noNA[which(gill_res_noNA$log2FoldChange >= 2
                                             & gill_res_noNA$padj <= 0.05),]
gill_fem_biased <- gill_res_noNA[which(gill_res_noNA$log2FoldChange <= -2 
                                             & gill_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the brain
gill_non_biased <- gill_res_noNA[which(gill_res_noNA$padj > 0.05),]
```

## Re-looking at variation in logFC across both sex-bias categories and tissue types
I want to now re-create the plot the highlights the variation we see in fold-change both across the different biases groups (male-biased, female-biased, and non-biased) within one tissue type and across all of the tissue types to see if the weird patterns are still there. To do this I first need to re-generate the "long" style dataset with the new information:

```{r logFC-long-dataset-SFanalysis}
logFC_long <- data.frame(
  tissue=c(rep("Gill",nrow(gill_fem_biased)),
           rep("Gill", nrow(gill_mal_biased)),
           rep("Gill", nrow(gill_non_biased)),
           rep("Gonad", nrow(gonad_fem_biased)),
           rep("Gonad", nrow(gonad_mal_biased)),
           rep("Gonad", nrow(gonad_non_biased)),
           rep("Liver", nrow(liver_fem_biased)),
           rep("Liver", nrow(liver_mal_biased)),
           rep("Liver", nrow(liver_non_biased))
         ),
  bias=c(rep("FB",nrow(gill_fem_biased)),
         rep("MB",nrow(gill_mal_biased)),
         rep("NB", nrow(gill_non_biased)),
         rep("FB", nrow(gonad_fem_biased)),
         rep("MB", nrow(gonad_mal_biased)),
         rep("NB", nrow(gonad_non_biased)),
         rep("FB", nrow(liver_fem_biased)),
         rep("MB", nrow(liver_mal_biased)),
         rep("NB", nrow(liver_non_biased))
         ),
  logFC=c(gill_fem_biased$log2FoldChange,
          gill_mal_biased$log2FoldChange,
          gill_non_biased$log2FoldChange,
          gonad_fem_biased$log2FoldChange,
          gonad_mal_biased$log2FoldChange,
          gonad_non_biased$log2FoldChange,
          liver_fem_biased$log2FoldChange,
          liver_mal_biased$log2FoldChange,
          liver_non_biased$log2FoldChange
          ),
  geneID=c(rownames(gill_fem_biased),
           rownames(gill_mal_biased),
           rownames(gill_non_biased),
           rownames(gonad_fem_biased),
           rownames(gonad_mal_biased),
           rownames(gonad_non_biased),
           rownames(liver_fem_biased),
           rownames(liver_mal_biased),
           rownames(liver_non_biased)
           )
  
)

#Exporting this data, run once and then commented out
#write.csv(logFC_long, "data/logFC_long_sexbias.csv", row.names = FALSE)
```

After re-creating the long version of the dataset I can re-generate the plot.

```{r FC-var-plot-SFanalysis, echo=FALSE, fig.cap="Absolute value of the logFC for genes that are female-biased, male-biased, and unbiased across each tissue type. Raw fold-change values are added ontop of the boxplot as jitters."}

pdf("docs/figs/FigSF_logFC_boxplots.pdf", width = 10, height=4)

ymax <- max(abs(logFC_long$logFC),
            na.rm = TRUE) + 5
organs <- levels(as.factor(logFC_long$tissue))

par(mfrow=c(1,3), 
    oma=c(4,5,2,8), 
    mar=c(1,1,1,0))

for(organ in organs){
   
   # add jittered points
   plot(abs(logFC_long$logFC[logFC_long$tissue==organ]) ~ 
            jitter(as.numeric(as.factor(logFC_long$bias[logFC_long$tissue==organ]))),
        col="#00000075", 
        axes=FALSE,
        cex.main=2,
        xlim=c(0,4),
        ylim=c(0,ymax)
        )
   
   # make the boxplot
   boxplot(abs(logFC_long$logFC[logFC_long$tissue==organ]) ~  
             logFC_long$bias[logFC_long$tissue==organ],
           col=scales::alpha(sex_bias_colors,0.75),
           add = TRUE,
           yaxt='n',
           las=1,
           cex.axis=1.75,
           frame=FALSE,
           lwd=1.5,
           outline=FALSE# do not plot outliers
           )
   
   # make the axis lines longer and thicker
   axis(1, labels=NA,
        at=-1:4,
        lwd=2, 
        lwd.ticks=0, 
        las = 3, 
        cex.axis = 1.5)
   axis(2, labels=seq(-5,ymax,10),
        line=NA, 
        at=seq(-5,ymax,10), 
        lwd=2,
        ylim=c(0,ymax),
        cex.axis=2,
        las=2)
   
   mtext(organ, cex=1.5,outer=FALSE, line=-1)
   
}

# add x axis label
mtext("Bias level",side = 1,cex=1.5, outer=TRUE, line=2)

# add y axis label
mtext(expression("|log"[2]*"FC|"),side = 2,cex=1.5, outer=TRUE, line=2.25)

# add legend
spfTools::outer_legend("right",
                       legend=c("Female\nbiased",
                                "Male\nbiased",
                                "Unbiased"),
                       pt.bg=sex_bias_colors,
                       pch=22,
                       bty='n',
                       ncol=1,
                       cex=2,
                       y.intersp = 1.5,
                       pt.cex=2)

dev.off()

```

The weird pattern does appear to still be there for the male-biased genes in the liver. I am going to look more in detail to the expression level of the male- and female-biased genes within each organ across all of the tissues using heatmaps. What does change, however, is the degree of the logFC. When doing the multi-facotr analysis, there were several genes that had a logFC of 35. Now, the highest logFC we see is around 25. 

## Investigating expression across samples in the sex-biased genes
For each organ type I am going to pull out either all of the sex-biased genes, or just the top 200 if there are more than that, and look at the expression levels of those genes across all of the samples.

```{r heatmap_gonads, echo=FALSE}

#Set the colors used for the heatmap labeling
sex_cols <- c("F" = "#7fc97f", "M" = "#beaed4" )
ann_colors = list(
  Sex=sex_cols)

#Transform the data
vsd_gonad <- vst(dds_FU_gonad_exp, blind=FALSE)

#Set the information for how you want the columns to show up
df_gonad <- as.data.frame(colData(dds_FU_gonad_exp))
col_order_gonad <- c(
  rownames(df_gonad[df_gonad$Sex=="F",]),
  rownames(df_gonad[df_gonad$Sex=="M",])
)

#Pull out the top 200 male-biased genes
select_mal_gonad <- rownames(head(gonad_mal_biased[order(gonad_mal_biased$log2FoldChange,
                                                         decreasing = TRUE),], 
                                  n=200))


gonadM_HM <- pheatmap(assay(vsd_gonad)[select_mal_gonad, col_order_gonad], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df_gonad[2],
              annotation_colors=ann_colors,
              fontsize = 15,
              annotation_legend = FALSE,
              main = "Top 200 Male-biased genes in the Testis")

#Pull out the top 200 female-biased genes
select_fem_gonad <- rownames(head(gonad_fem_biased[order(gonad_fem_biased$log2FoldChange,
                                                         decreasing = FALSE),], 
                                  n=200))

gonadF_HM <- pheatmap(assay(vsd_gonad)[select_fem_gonad, col_order_gonad], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df_gonad[2],
              annotation_colors=ann_colors,
              fontsize = 15,
              annotation_legend = FALSE,
              main = "Top 200 Female-biased genes in the Ovaries")

```


```{r heatmap_liver, echo=FALSE}

#Transform the data
vsd_liver <- vst(dds_FU_liver_exp, blind=FALSE)

#Set the information for how you want the columns to show up
df_liver <- as.data.frame(colData(dds_FU_liver_exp))
col_order_liver <- c(
  rownames(df_liver[df_liver$Sex=="F",]),
  rownames(df_liver[df_liver$Sex=="M",])
)

#Pull out the male-biased genes
select_mal_liver <- rownames(liver_mal_biased[order(liver_mal_biased$log2FoldChange,
                                                         decreasing = TRUE),])


liverM_HM <- pheatmap(assay(vsd_liver)[select_mal_liver, col_order_liver], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              border_color=NA,
              annotation_col = df_liver[2],
              annotation_colors=ann_colors,
              fontsize = 15,
              annotation_legend = FALSE,
              main = "Male-biased genes in the Liver")

#Pull out the female-biased genes
select_fem_liver <- rownames(liver_fem_biased[order(liver_fem_biased$log2FoldChange,
                                                         decreasing = FALSE),])

liverF_HM <- pheatmap(assay(vsd_liver)[select_fem_liver, col_order_liver], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df_liver[2],
              annotation_colors=ann_colors,
              border_color=NA,
              fontsize = 15,
              annotation_legend = FALSE,
              main = "Female-biased genes in the Liver")

```


```{r heatmap_gill, echo=FALSE}
#Transform the data
vsd_gill <- vst(dds_FU_gill_exp, blind=FALSE)

#Set the information for how you want the columns to show up
df_gill <- as.data.frame(colData(dds_FU_gill_exp))
col_order_gill <- c(
  rownames(df_gill[df_gill$Sex=="F",]),
  rownames(df_gill[df_gill$Sex=="M",])
)

#Pull out the male-biased genes
select_mal_gill <- rownames(gill_mal_biased[order(gill_mal_biased$log2FoldChange,
                                                         decreasing = TRUE),])


gillM_HM <- pheatmap(assay(vsd_gill)[select_mal_gill, col_order_gill], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              border_color=NA,
              annotation_col = df_gill[2],
              annotation_colors=ann_colors,
              fontsize = 15,
              annotation_legend = FALSE,
              main = "Male-biased genes in the gill")

#Pull out the female-biased genes
select_fem_gill <- rownames(gill_fem_biased[order(gill_fem_biased$log2FoldChange,
                                                         decreasing = FALSE),])

gillF_HM <- pheatmap(assay(vsd_gill)[select_fem_gill, col_order_gill], 
              cluster_rows = FALSE, 
              show_rownames = FALSE, 
              cluster_cols = FALSE, 
              show_colnames = FALSE,
              annotation_col = df_gill[2],
              annotation_colors=ann_colors,
              border_color=NA,
              fontsize = 15,
              annotation_legend = FALSE,
              main = "Female-biased genes in the gill")
```

In the liver, for both males and females, expression in the top biased genes appears to be driven almost exclusively by two individuals. I am going to continue on with the SF analysis over the MF analysis for a more conservative approach. I am also choosing to leave those individuals in the analysis as they may represent true variation in expression of genes across the samples.

Looking back at the logFC plot made from the single-factor analysis, there appears to be differences in the logFC across both the organs and the sex-bias groups (male-biased, fem-biased, and non-biased). In the gill and liver, it looks like the female-biased genes are more biased on average than the male-biased genes, but this pattern reverses in the gonads. In every case the non-biased genes have the lowest median logFC even though there is some variation.

I am going to fit a linear model to the data and explore some of the summary statistics to see what is happening

```{r FC-var-stats, warning=FALSE}
#Looking at some summary statistics for logFC between the different groups
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), median)
tapply(abs(logFC_long$logFC), list(logFC_long$bias, logFC_long$tissue), sd)

model <- lm(abs(logFC_long$logFC)~ logFC_long$tissue * logFC_long$bias)
summary(model)
```

The summary statistics confirm the assumptions made from the boxplots, female-biased genes in the liver and gill have a higher logFC compared to the male-biased genes, but male-biased genes in the gonads have a higher logFC compared to the female-biased. In all three organs, unbiased genes possess the lowest logFC. Interestingly, the highest logFC across tissues can be found in the liver, followed by the gill, and lastly the gonads.

## Looking at shared sex-biased genes with an Upset plot
I now want to see if there are genes that are female/male-biased in multiple organs or if there are genes that are female-biased in one organ and then male-biased in a different organ, etc. In order to do this I will be creating upset plots first for the sexes individually and then across all biased genes.

```{r upset-plot-mal, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in males."}

#Male-biased genes
listInputMB <- list("Testis" = rownames(gonad_mal_biased),
                  "Male Gill"=rownames(gill_mal_biased), 
                  "Male Liver"=rownames(liver_mal_biased)
                  )

upset(fromList(listInputMB),
      mainbar.y.label = "Number of Shared Male-Biased Genes",
      sets.x.label = "Total Number of Male-Biased Genes",
      point.size = 3)
```


```{r upset-plot-fem, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs in females."}
#Female
listInputFB <- list("Ovaries" = rownames(gonad_fem_biased),
                  "Female Gill"=rownames(gill_fem_biased), 
                  "Female Liver"=rownames(liver_fem_biased)
                  )
upset(fromList(listInputFB),
      mainbar.y.label = "Number of Shared Female-Biased Genes",
      sets.x.label = "Total Number of Female-Biased Genes",
      point.size = 3)
```

```{r upset-plot-all, echo=FALSE, fig.cap="Upset plots to show the number of shared sex-biased genes across the organs across both sexes."}

#Checking there is no overlap between the two sexes
listInputall <- list("MB Gonad" = rownames(gonad_mal_biased),
                     "MB Gill" = rownames(gill_mal_biased), 
                     "MB Liver" = rownames(liver_mal_biased),
                     "FB Gonad" = rownames(gonad_fem_biased),
                     "FB Gill" = rownames(gill_fem_biased), 
                     "FB Liver" = rownames(liver_fem_biased))

#Plot exported and color added via illustrator
pdf("docs/figs/FigSB_sexbias_upset.pdf",width = 10, height=7)
upset(fromList(listInputall),
      mainbar.y.label = "# Shared Sex-Biased Genes",
      sets.x.label = "# Sex-Biased Genes",
      point.size = 3,
      nsets = 6,
      nintersects = NA,
      text.scale = c(2, 2, 2, 1.5, 2, 1.5)
)

dev.off()

```

In females, the highest number of shared sex-biased genes is found between the gonads and the liver, but in males, the highest number of shared sex-biased genes is found between the gills and the gonads. There are no genes that are shared across the same organ between males and females (i.e. biased in males and females in the liver) which is good. There are 3 and 2 genes that are biased in all three organs for males and females respectively. Overall, however, there is not a high amount of genes that are sex-biased in more than one organ type.

## Categorizing sex-specific genes
Now that we have investigated what the sex-bias is like across the different tissue types I want to dive further into genes with **sex-specific expression**. These are genes expressed only in one sex or the other. I am classifying a sex-specific genes within each tissue as ones where the expression of that gene is less than 10 for all of one sex and there is a median of $\ge 20$ in the other sex. 

Prior to running the for loop, any "outliers" that were determined by DESeq were removed from the pool of genes. Additionally, normalized counts were used for the classification of sex-specific genes as DESeq2 uses the moralized counts for all of the logFC/sexbias calculations.

```{r sex-specific-for-loop}
#Store all of the single factor analysis outputs in a list
DE_models <- list(dds_FU_gill_exp, 
                  dds_FU_gonad_exp, 
                  dds_FU_liver_exp)

#Create an empty list to store my sex-specific datasets in
FU_sex_specific_genes <- list()

#For each DESeq dataset in DE_models pull out the Mspecific and Fspecific
#genes based on medians
for (dataset in DE_models) {
  
  #Pull out the geneIDs for genes that were categorized as "outliers" by DESeq2
  ##Calculate the Cooks threshold that would have been used
  np <- length(resultsNames(dataset))
  nsamp <- ncol(dataset)
  cooks_thresh <- qf(0.99, df1 = np, df2 = nsamp - np)
  
  ##Apply threshold calculated above to the cooks values in the dataset
  out_ids <- names(mcols(dataset)$maxCooks[mcols(dataset)$maxCooks >
                                                cooks_thresh])
  
  #Filtering the dds dataset to remove the outliers identified by DESeq
  dataset_filtered <- dataset[!(rownames(dataset) %in% out_ids), ]
  
  
  #Male-specific Genes
  ##Pull out all of the rows where fem count <=10 in every female sample
  fem10_organ_names <- which(rowSums(t(apply(counts(dataset_filtered,
                                                    normalized = TRUE)[, dataset_filtered$Sex == "F"],
                                             1,
                                             function(x) x <= 10)
                                       )
                                     ) == ncol(counts(dataset_filtered, 
                                                      normalized = TRUE)[, dataset_filtered$Sex == "F"])
                             )
  
  fem10_organ <- counts(dataset_filtered,
                        normalized = TRUE)[rownames(counts(dataset_filtered,
                                                          normalized = TRUE)) %in%
                                            names(fem10_organ_names),]
  
  ##Pull out the rows where median of male count >=50  
  mal50_organ <- apply(counts(dataset_filtered,
                              normalized = TRUE)[, dataset_filtered$Sex == "M"],
                       1,
                       function(x) median(x) >= 50)
  
  ##Keep only rows where all female samples <=10 and the male median >= 50
  fem10_mal50_organ <- fem10_organ[rownames(fem10_organ) %in%
                                     names(mal50_organ[mal50_organ == TRUE]),
                                   ]
  
  ##Create a new object with a name based on the organ type
  organ_malsp <- sub("$", "_male_specific", dataset$Organ[1])
  FU_sex_specific_genes[[organ_malsp]] <- fem10_mal50_organ
  
  
  #Female-specific Genes
  ##Pull out ll the rows where male count is <=10 in every male sample
  mal10_organ_names <- which(rowSums(t(apply(counts(dataset_filtered,
                                                    normalized = TRUE)[, dataset_filtered$Sex == "M"],
                                             1,
                                             function(x) x <= 10)
                                       )
                                     ) == ncol(counts(dataset_filtered,
                                                      normalized = TRUE)[, dataset_filtered$Sex == "M"])
                             )
  
  mal10_organ <- counts(dataset_filtered,
                        normalized = TRUE)[rownames(counts(dataset_filtered,
                                                           normalized = TRUE)) %in%
                                             names(mal10_organ_names), ]
  
  ##Pull out rows where median of female count is >=50 
  fem50_organ <- apply(counts(dataset_filtered,
                              normalized = TRUE)[, dataset_filtered$Sex == "F"],
                       1,
                       function(x) median(x) >= 50)
  
  ##keep only the rows where male <=10 and fem median >= 50
  mal10_fem50_organ <- mal10_organ[rownames(mal10_organ) %in%
                                     names(fem50_organ[fem50_organ == TRUE]),
                                   ]
  
  ##Create a new object with a name based on the organ type
  organ_femsp <- sub("$", "_female_specific", dataset$Organ[1])
  FU_sex_specific_genes[[organ_femsp]] <- mal10_fem50_organ
  
}
```

### Investigating the results of the sex-specific subsetting
Let's now take a look at how many sex-specific genes we have in each tissue type:

```{r sex-specific-counts, echo = FALSE, eval = FALSE}
FU_sex_specific_counts <- data.frame(tissue = rep(c("Gill", "Gonad", "Liver"), 2),
                                  sex = c(rep("M", 3),
                                          rep("F", 3)),
                                  num_genes = c(nrow(FU_sex_specific_genes$Gill_male_specific),
                                                nrow(FU_sex_specific_genes$Gonad_male_specific),
                                                nrow(FU_sex_specific_genes$Liver_male_specific),
                                                nrow(FU_sex_specific_genes$Gill_female_specific),
                                                nrow(FU_sex_specific_genes$Gonad_female_specific),
                                                nrow(FU_sex_specific_genes$Liver_female_specific))
                                  )

output_table <- FU_sex_specific_counts %>% 
  group_by(tissue, sex) %>% 
  summarise("Number of Sex-specific Genes" = sum(num_genes)) %>%
  kable(caption = "Sex-specific Genes", format = "html",
        col.names = c("Tissue Type", "Sex", "Number of Sex-specific Genes")) %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F, position = "left")

# Save the table to a temporary HTML file
html_file <- tempfile(fileext = ".html")
save_kable(output_table, file = html_file)

# Ensure you have PhantomJS installed (required by webshot)
#webshot::install_phantomjs()

# Capture the HTML file as a PNG
webshot(html_file, file = "sex_specific_genes_table.png")

```

```{r show-ss-table, eval = FALSE}

ss_genes <- image_ggplot(image_read('docs/figs/sex_specific_genes_table.png'),interpolate = TRUE)

ss_genes

```

To get a better idea what is going on between the sex-specific genes vs the sex-biased genes I first determined how many overlaps there were between our sex-biased and sex-specific genes and then used plotCounts to plot the counts for each gene separately. 

In the **Male Gills** there are `r nrow(FU_gill_mal_biased[rownames(FU_gill_mal_biased) %in% rownames(FU_sex_specific_genes$Gill_male_specific),])` genes that overlap between sex-biased and sex-specific, there are `r nrow(FU_sex_specific_genes$Gonad_male_specific[rownames(FU_sex_specific_genes$Gonad_male_specific) %in% rownames(FU_gonad_mal_biased),])` overlapping in the **gonads** and `r nrow(FU_sex_specific_genes$Liver_male_specific[rownames(FU_sex_specific_genes$Liver_male_specific) %in% rownames(FU_liver_mal_biased),])` overlapping in the **liver**. For females we have `r nrow(FU_gill_fem_biased[rownames(FU_gill_fem_biased) %in% rownames(FU_sex_specific_genes$Gill_female_specific),])` overlapping sex-specific and sex-biased genes in the **gills**, `r nrow(FU_sex_specific_genes$Gonad_female_specific[rownames(FU_sex_specific_genes$Gonad_female_specific) %in% rownames(FU_gonad_fem_biased),])` genes in the **gonads**, and `r nrow(FU_sex_specific_genes$Liver_female_specific[rownames(FU_sex_specific_genes$Liver_female_specific) %in% rownames(FU_liver_fem_biased),])` genes in the **liver**.

### Pulling out the gene IDs for all of the sex-specific genes
```{r SS-geneIDs, eval=FALSE}
#Run once to create the file and then commented out
write.table(cbind(rownames(FU_sex_specific_genes$Gill_male_specific)),
            'data/FU_malG_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(FU_sex_specific_genes$Gill_female_specific)),
            'data/FU_femG_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(FU_sex_specific_genes$Gonad_male_specific)),
            'data/FU_malGon_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(FU_sex_specific_genes$Gonad_female_specific)),
            'data/FU_femGon_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(FU_sex_specific_genes$Liver_male_specific)),
            'data/FU_malL_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(FU_sex_specific_genes$Liver_female_specific)),
            'data/FU_femL_specific_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

```

## Creating categories and binning the sex-biased genes based on degree of logFC
This binned was only done on the sex-biased genes, will not have a category for the unbiased genes. The cutoffs for the different groups are as follows:

    1. Low - biased = LFC 2 - 3
    2. Medium - biased = LFC 3 - 5
    3. High - biased = LFC 5 - 10
    4. Extreme sex bias = LFC > 10

Counts of the sex-specific genes will be added on separately as they were not classified based on fold-change but rather a presence in one sex and an absence in the other sex. A for loop will then be used to make sure genes are not counted twice (i.e. as both extremely biased and sex-specific).

```{r bin-sex-bias}
#Make a vector that contains all of the groupings
biased_bins <- c("Unbiased", "Low", "Med", "High", "Extreme", "Sex-specific")

#Create a new column in the dataset and use ifelse statements to set the category limits
#abs(logFC) was used to account for the fem-biased genes possessing negative values
logFC_long$bias_cat <- ifelse(logFC_long$bias == "NB",
                              biased_bins[1],
                              ifelse(abs(logFC_long$logFC) >= 2 & abs(logFC_long$logFC) < 3,
                                     biased_bins[2],
                                     ifelse(abs(logFC_long$logFC) >= 3 & abs(logFC_long$logFC) < 5,
                                            biased_bins[3],
                                            ifelse(abs(logFC_long$logFC) >= 5 & abs(logFC_long$logFC) < 10,
                                                   biased_bins[4],
                                                   biased_bins[5])
                                            )
                                     )
                              )

#Making sure the genes we categorized as sex-specific are labeled as sex-specific for their 
#bias cat. in the dataset
organs <- c("Gill", "Gill", "Gonad", "Gonad", "Liver", "Liver")
bias <- c("MB", "FB", "MB", "FB", "MB", "FB")

for(i in 1:length(FU_sex_specific_genes)){

  tmp <- FU_sex_specific_genes[[i]]
  tmp <- as.data.frame(tmp)
  tmp$geneID <- rownames(tmp)
  
  for(j in 1:nrow(tmp)){
    
    one_gene <- tmp[j, ]
    
    if(one_gene[["geneID"]] %in%
       logFC_long[logFC_long$tissue == organs[[i]] &
                  logFC_long$bias == bias[[i]],"geneID"]){
       
      logFC_long[logFC_long$geneID == one_gene[["geneID"]] &
                   logFC_long$tissue == organs[[i]] &
                   logFC_long$bias == bias[[i]],
                 "bias_cat"] <- "Sex-specific"
    }else{
      
      one_gene_dat <- data.frame(matrix(ncol= ncol(logFC_long),
                                        nrow=1))
      colnames(one_gene_dat) <- colnames(logFC_long)
      
      one_gene_dat$tissue <- organs[[i]]
      one_gene_dat$geneID <- one_gene[["geneID"]]
      one_gene_dat$bias <- bias[[i]]
      one_gene_dat$bias_cat <- "Sex-specific"
      rownames(one_gene_dat) <- NULL
      
      logFC_long <- rbind(one_gene_dat, logFC_long)
      
      rownames(logFC_long) <- NULL
    }
  }
 }

#Create  subset of our long dataset that does not include the non-biased genes
logFC_long_noNB <- logFC_long[logFC_long$bias_cat != "Unbiased",]

#Make a table to count the number of genes in each category for each organ
bias_cat_table <- table(logFC_long_noNB$bias, logFC_long_noNB$bias_cat, logFC_long_noNB$tissue)

#Seperate out the different organs
bias_cat_gill <- bias_cat_table[,, "Gill"]
#write.table(bias_cat_gill, "data/bias_cat_gills.txt", row.names = TRUE)

bias_cat_gonad <- bias_cat_table[,, "Gonad"]
#write.table(bias_cat_gonad, "data/bias_cat_gonad.txt", row.names = TRUE)

bias_cat_liver <- bias_cat_table[,, "Liver"]
write.table(bias_cat_liver, "data/bias_cat_liver.txt", row.names = TRUE)

#Plot the counts for each tissue type
sex_cols <- c("F" = "#7fc97f", "M" = "#beaed4" )
labs <- c("Low", "Med", "High", "Extreme", "Specific")

pdf("docs/figs/FigSB_biasCat_counts.pdf",width = 10, height=4)

par(mfrow=c(1, 3), oma=c(6,4,2,8), mar=c(1,2.5,1,0), 
    cex.main=2,
    cex.axis=2)

bp <- barplot(bias_cat_gill[,biased_bins[biased_bins!="Unbiased"]], 
        beside = TRUE,
        xaxt='n',
        ylim = c(0, max(bias_cat_gill)+5), 
        col = sex_cols,
        main = "")
mtext("Gill",3,outer = FALSE,cex=1.5,line=-1)
axis(2,lwd=2)
text(cex=2, x=colMeans(bp), y=-0.5, labs, xpd=NA, srt=35, adj = 1)

barplot(bias_cat_gonad[,biased_bins[biased_bins!="Unbiased"]], 
        beside = TRUE, 
        ylim = c(0, max(bias_cat_gonad)+1000), 
        xaxt='n',
        col = sex_cols,
        main = "",
        cex.main=2,
        cex.axis=2)
mtext("Gonad",3,outer = FALSE,cex=1.5,line=-1)
axis(2,lwd=2,labels = NA)
text(cex=2, x=colMeans(bp), y=-100, labs, xpd=NA, srt=35, adj=1)

barplot(bias_cat_liver[,biased_bins[biased_bins!="Unbiased"]], 
        beside = TRUE, 
        ylim = c(0, max(bias_cat_liver)+20), 
        xaxt='n',
        col = sex_cols,
        main = "",
        cex.main=2,
        cex.axis=2)
axis(2,lwd=2, labels = NA)
text(cex=2, x=colMeans(bp), y=-2, labs, xpd=NA, srt=35,adj=1)
mtext("Liver",3,outer = FALSE,cex=1.5,line=-1)

mtext("Number of Genes",2,outer=TRUE, cex=1.5, line=2.25)
mtext("Bias category",1, outer=TRUE, cex=1.5, line=4)

outer_legend("right", 
       legend = c("Female\nbiased", "Male\nbiased"), 
       pt.bg = sex_cols,
       pch=22,
       bty='n',
       ncol=1,
       cex=2,
       y.intersp = 1.5,
       pt.cex=2)

dev.off()
```

There is a weird pattern showing up in the liver where most of the male biased genes are showing an "Extreme" level of sex bias. With the heatmaps we created above, we likely guess that these extreme biased genes are the ones driven by those two male individuals.

## Create a combined figure
```{r figSB, message=FALSE}

figSBa <- image_ggplot(image_read_pdf('docs/figs/FigSF_logFC_boxplots.pdf'),interpolate = TRUE)
figSBb <- image_ggplot(image_read_pdf('docs/figs/FigSB_biasCat_counts.pdf'),interpolate = TRUE)

figSB <- wrap_plots(
  figSBa,
  figSBb,
  nrow = 2)

figSB <- figSB + plot_annotation(tag_levels = 'A')
figSB

ggsave("docs/figs/FigSB.pdf", figSB, height=8, width=10)
ggsave("docs/figs/FigSB.png", figSB, height=8, width=10)

```

```{r figSB-HM, message=FALSE}

figHMa <- image_ggplot(image_read_pdf('docs/figs/liverMB_heatmap.pdf'),interpolate = TRUE)
figHMb <- image_ggplot(image_read_pdf('docs/figs/liverFB_heatmap.pdf'),interpolate = TRUE)

figSBHM <- wrap_plots(
  figSBa,
  figHMa,
  figSBb,
  figHMb,
  ncol = 2,
  nrow = 2)

figSBHM <- figSBHM + plot_annotation(tag_levels = 'A')
figSBHM

ggsave("docs/figs/FigSBHM.pdf", figSBHM, height=8, width=10)
ggsave("docs/figs/FigSBHM.png", figSBHM, height=8, width=10)

```
# Gene Ontologogy Analysis and BLASTing the different genes
There are several things that I want to accomplish with the gene ontology analysis, to start with I want to classify the sex-biased genes and then the sex-specific genes.

## BLASTING against the zebrafish proteome
The first step to the GO analysis is to BLAST the trinity sequences against one of the PANTHER organisms. The zebrafish, _Danio reiro_, was used for this as it was the only fish in the list.

A BLAST database was generated with the _D. reiro_ proteome as:
`makeblastdb -in ../ncbi_dataset/data/GCF_000002035.6/protein.faa -out d_rerio_prot -dbtyp prot`

The trinity gene IDs that correspond to ALL of the female- or male-biased genes were pulled out and saved to `.txt` files.

```{r biased-trin-ids, eval=FALSE}
write.table(cbind(rownames(gill_fem_biased)),
            'FU_femG_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(gill_mal_biased)),
            'FU_malG_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(gonad_fem_biased)),
            'FU_femGon_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(gonad_mal_biased)),
            'FU_malGon_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(liver_fem_biased)),
            'FU_femL_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
write.table(cbind(rownames(liver_mal_biased)),
            'FU_malL_biased_TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

```

The `.txt` files were imported into the RCC and then BLASTed against the zebrafish proteome using the following script:

```{bash blast_pipeline, file = 'bash/blast_pipeline.sh', eval = FALSE}


```

This script was run as `bash bash_scripts/blast_pipeline.sh fuscus_high_exp_genes/high_expTG_names other_scripts/subset_fasta_file trinity_supertran_fuscus.fasta blastx ../genomes/d_rerio_prot/d_reiro_prot _reiro_blast.txt fuscus_high_exp_genes/high_exp_results/`.

The trinity gene IDs that correspond to all of the sex-specific genes were pulled out above. Those were BLASTed with the script above using **blastn** against the _Syngnathus scovelli_ genome as: 
`bash bash_scripts/blast_pipeline.sh fuscus_high_exp_genes/high_expTG_names other_scripts/subset_fasta_file trinity_supertran_fuscus.fasta blastn ../genomes/s_scov/s_scov_genome _blast.txt fuscus_high_exp_genes/high_exp_results/`.

## Read in and filter the BLAST results
I then exported all of the blast results as .txt files from the RCC and need to first read them into R so that I can filter the results and make sure only one hit is kept per gene.

```{r readin_blast_results_dreiro, echo=FALSE}
#Specify the directory where BLAST results are located
blast_path <- "data/fuscus_BLAST/"

#Create a list of the files I want
FU_blast_files <- list.files(blast_path, pattern = "reiro")

#Create an empty list to store my datasets in
FU_blast_list <- list()

#Create a loop to read in all of the blast results
for(file_name in FU_blast_files){
  
  #Read the file and store it in a new object
  file_data <- read.delim(file.path(blast_path, file_name), stringsAsFactors = FALSE, quote = "", sep = "\t", header = FALSE)

  #Add column names to the dataset
  colnames(file_data) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "reiro_prot_info", "prot_id", "reiro_prot_start", "reiro_prot_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  #Create a new object with a name based on the file name
  blast_name <- sub(".txt$", "", file_name) #Removes the file extension
  FU_blast_list[[blast_name]] <- file_data
}

```

```{r readin_blast_results_SSgenes, echo=FALSE}
#Create a list of the files I want
FU_blast_files_sex_specific <- list.files(blast_path, pattern = "specific")

#Create an empty list to store my datasets in
FU_blast_list_sex_specific <- list()

#Create a loop to read in all of the blast results
for(file_name in FU_blast_files_sex_specific){
  
  #Read the file and store it in a new object
  file_data <- read.delim(file.path(blast_path, file_name), 
                          stringsAsFactors = FALSE, quote = "", 
                          sep = "\t", header = FALSE)

  #Add column names to the dataset
  colnames(file_data) <- c("trin_geneid", 
                           "trin_gene_start", "trin_gene_end", 
                           "scov_gene_info", "scov_prot_info",
                           "scov_gene_start", "scov_gene_end", 
                           "evalue", "bit_score", "length", "pident", 
                           "gaps")
  
  #Create a new object with a name based on the file name
  blast_name <- sub(".txt$", "", file_name) #Removes the file extension
  FU_blast_list_sex_specific[[blast_name]] <- file_data
}

```

Not that I have all of the BLAST files for SS genes and SB genes in R I am going to filter them:

```{r filter-BLAST-dreiro}
#Use lapply to apply the function to each dataset stored in the list created above
blast_output_filtered <- lapply(FU_blast_list, function(data_frame){
 
  #Pull out the Unique Trinity gene IDs
  uniqueID <- unique(data_frame[1])
  
  #Create an empty dataframe to store intermediate results in
  output <- data.frame(matrix(data = NA, ncol = ncol(data_frame), nrow = 0))
  colnames(output) <- c("trin_geneid", "trin_gene_start", "trin_gene_end", "reiro_prot_info", "prot_id", "reiro_prot_start", "reiro_prot_end", "evalue", "bit_score", "length", "pident", "gaps")
  
  #Generate a for loop that pulls out the lowest e-value + highest % identity for each gene
  for(gene in uniqueID$trin_geneid){
    
    #Subset the dataset for each Trinity gene
    this_trin_gene<- subset(data_frame, trin_geneid==gene)
    #Pull out gene with smallest e-value
    uni_gene_subset_lowe <- this_trin_gene[which(this_trin_gene$evalue == min(this_trin_gene$evalue)),]
    #In case mult. genes have same e-value, pull out gene with highest % identity
    uni_gene_subset_lowe_highpid <- uni_gene_subset_lowe[which(uni_gene_subset_lowe$pident == max(uni_gene_subset_lowe$pident)),]
    
    # keep only one of multiple identical rows
    uni_gene_subset_lowe_highpid <- unique(uni_gene_subset_lowe_highpid)

    # check that there is a single gene ID in scovelli, if not, only first row is kept
    if(length(gsub("^.*(GeneID:\\d+)\\].*$","\\1",uni_gene_subset_lowe_highpid$reiro_prot_info))>1){
    
    
      uni_gene_subset_lowe_highpid<-uni_gene_subset_lowe_highpid[1,]
    }
    
    #Add the final gene into the empty dataframe from above
    output <- rbind(output, uni_gene_subset_lowe_highpid)
  }
  
  return(output)
})

```

```{r filter-BLAST-SS}
#Use lapply to apply the function to each dataset stored in the list created above
blast_output_filtered_SS <- lapply(FU_blast_list_sex_specific, function(data_frame){
 
  #Pull out the Unique Trinity gene IDs
  uniqueID <- unique(data_frame[1])
  
  #Create an empty dataframe to store intermediate results in
  output <- data.frame(matrix(data = NA, ncol = ncol(data_frame), nrow = 0))
  colnames(output) <- c("trin_geneid", "trin_gene_start", "trin_gene_end",
                        "scov_gene_info", "scov_prot_info",
                        "scov_gene_start", "scov_gene_end", 
                        "evalue", "bit_score", "length", "pident", "gaps")
  
  #Generate a for loop that pulls out the lowest e-value + highest % identity for each gene
  for(gene in uniqueID$trin_geneid){
    
    #Subset the dataset for each Trinity gene
    this_trin_gene<- subset(data_frame, trin_geneid==gene)
    #Pull out gene with smallest e-value
    uni_gene_subset_lowe <- this_trin_gene[which(this_trin_gene$evalue ==
                                                   min(this_trin_gene$evalue)),]
    #In case mult. genes have same e-value, pull out gene with highest % identity
    uni_gene_subset_lowe_highpid <- 
      uni_gene_subset_lowe[which(uni_gene_subset_lowe$pident ==
                                   max(uni_gene_subset_lowe$pident)),]
    
    # keep only one of multiple identical rows
    uni_gene_subset_lowe_highpid <- unique(uni_gene_subset_lowe_highpid)

    # check that there is a single gene ID in scovelli, if not, only first row is kept
    if(length(gsub("^.*(GeneID:\\d+)\\].*$",
                   "\\1",
                   uni_gene_subset_lowe_highpid$scov_gene_info))>1){
      
      uni_gene_subset_lowe_highpid <- uni_gene_subset_lowe_highpid[1,]
    }
    
    #Add the final gene into the empty dataframe from above
    output <- rbind(output, uni_gene_subset_lowe_highpid)
  }
  
  return(output)
})

```

## Collect gene names corresponding to each protein -sex-biased genes
In order to run PANTHER we need IDs that follow one of their supported formats. From NCBI, the gene name works (e.g. rpl27). That information is not automatically supplied with the BLAST output, but we can use the .gff file from NCBI to pull out the gene names that correspond to the proteinID (e.g. NP_956018.1).

```{r gff-gene-names}
#Read in the GFF file for zebrafish that has info about the geneID
reiro_gff <- read.delim("data/d_reiro_genomic.gff",
                        header = FALSE,
                        comment.char = "#")

#Keep only the columns I want and rename them
reiro_gff <- reiro_gff[,c(1:5,9)]
colnames(reiro_gff) <- c("seqname", "source", "feature", "start", "end", "gene_info")

#Subset the dataset to only include the rows we're interested in
genes <- reiro_gff[reiro_gff$feature == "CDS",]

#Pull out the gene name that corresponds to each protein ID
genes$gene_name <- gsub("^(.*;)(gene=)(\\w+\\d*);(.*$)", "\\3", genes$gene_info)
genes$prot_id <- gsub("^(.*;)(protein_id=)(.*)$", "\\3", genes$gene_info)

#Merge the gene names pulled out above with the rest of the BLAST datasets based on the proteinID
blast_output_merged <- lapply(blast_output_filtered, function(dataframe){
  
   output <- merge(dataframe, unique(genes[,7:8]), by = "prot_id")
  
   return(output)
})

```

I now need to save only the gene_id for all of the BLASTed genes to then upload to PANTHER.

```{r export-danio-geneID, eval=FALSE}

write.table(c(blast_output_merged$FU_femG_biased_TRgenes_reiro_blast$gene_name, 
              blast_output_merged$FU_femGon_biased_TRgenes_reiro_blast$gene_name,
              blast_output_merged$FU_femL_biased_TRgenes_reiro_blast$gene_name, 
              blast_output_merged$FU_malG_biased_TRgenes_reiro_blast$gene_name, 
              blast_output_merged$FU_malGon_biased_TRgenes_reiro_blast$gene_name, 
              blast_output_merged$FU_malL_biased_TRgenes_reiro_blast$gene_name),
           'FU_GOnames_SBG.txt', 
            sep = " ", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

write.table(c(blast_output_merged$FU_femL_biased_TRgenes_reiro_blast$gene_name, 
              blast_output_merged$FU_malL_biased_TRgenes_reiro_blast$gene_name),
           'FUL_GOnames_SBG.txt', 
            sep = " ", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

```

## Pull out the gene names and protein ID for sex-specific BLAST
Once all the BLAST results for the sex-specific genes have been filtered I want to pull out specifically the name of the gene and the protein ID from the  `scov_gene_info` column and save them as their own column. 

```{r scov-geneIDs}
for (i in 1:length(blast_output_filtered_SS)) {
  
  #Pulling out the gene name and adding it to a new column
  blast_output_filtered_SS[[i]]$gene <- gsub("^(.*)(gene=)(\\w+\\d*)(.*$)",
                                             "\\3",
                                             blast_output_filtered_SS[[i]]$scov_gene_info)
  
  #Pulling out the protein name and adding it to a new column
  blast_output_filtered_SS[[i]]$protein <- gsub("^(.*)(protein=)(.*)([[:graph:]]\\s[[:graph:]])(protein_id=)(.*$)",
                                                "\\3",
                                                blast_output_filtered_SS[[i]]$scov_gene_info)
  
  
}

```

Now that I have isolated the gene names and protein IDS I want to reform the BLAST results so that they can all be compiled into one single dataset.

```{r long_BLASTSS}
#Create a long format dataset that has the tissue information, BLAST gene and protein, 
#geneID, and sex 
blast_long_SS <- data.frame(
  tissue=c(rep("Gill", 
               sum(nrow(blast_output_filtered_SS$FU_femG_specific_TRgenes_blast),
                   nrow(blast_output_filtered_SS$FU_malG_specific_TRgenes_blast))),
           rep("Gonad", 
               sum(nrow(blast_output_filtered_SS$FU_femGon_specific_TRgenes_blast),
                   nrow(blast_output_filtered_SS$FU_malGon_specific_TRgenes_blast))),
           rep("Liver",
               sum(nrow(blast_output_filtered_SS$FU_femL_specific_TRgenes_blast),
                   nrow(blast_output_filtered_SS$FU_malL_specific_TRgenes_blast)))
         ),
  sex=c(rep("Female",
            nrow(blast_output_filtered_SS$FU_femG_specific_TRgenes_blast)),
        rep("Male",
            nrow(blast_output_filtered_SS$FU_malG_specific_TRgenes_blast)),
        rep("Female",
            nrow(blast_output_filtered_SS$FU_femGon_specific_TRgenes_blast)),
        rep("Male",
            nrow(blast_output_filtered_SS$FU_malGon_specific_TRgenes_blast)),
        rep("Female",
            nrow(blast_output_filtered_SS$FU_femL_specific_TRgenes_blast)),
        rep("Male",
            nrow(blast_output_filtered_SS$FU_malL_specific_TRgenes_blast))
        ),
  gene=c(blast_output_filtered_SS$FU_femG_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$FU_malG_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$FU_femGon_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$FU_malGon_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$FU_femL_specific_TRgenes_blast$gene,
         blast_output_filtered_SS$FU_malL_specific_TRgenes_blast$gene
         ),
  protein=c(blast_output_filtered_SS$FU_femG_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$FU_malG_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$FU_femGon_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$FU_malGon_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$FU_femL_specific_TRgenes_blast$protein,
            blast_output_filtered_SS$FU_malL_specific_TRgenes_blast$protein
            ),
  trinityID=c(blast_output_filtered_SS$FU_femG_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$FU_malG_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$FU_femGon_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$FU_malGon_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$FU_femL_specific_TRgenes_blast$trin_geneid,
              blast_output_filtered_SS$FU_malL_specific_TRgenes_blast$trin_geneid
              )
  )

#write.csv(blast_long_SS, "FU_BLAST_results_sex_specific.csv", row.names = FALSE)
```

## Reading in the PANTHER results - Sex-biased Genes
In PANTHER I uploaded the `FU_GOnames_SBG.txt` file, selected _Danio rerio_ as the organism, and then chose "Functional classification viewed in gene list" as the analysis type. 

I exported those results to a .txt file and will now be reading that into R and visualizing the results.

```{r panther-results-SBG}
#Read in the .txt file containing PANTHER results
panther <- read.delim("data/pantherGeneList_SBG.txt", 
                      header = FALSE)

#Add the approporiate column names for the PANTHER data
colnames(panther) <- c("GeneID", "MappedID", 
                       "GeneName", "pantherFAM", 
                       "panther_prot_class", "species")

#Add the identified PANTHER protein classes to the BLAST datasets
blast_output_merged$FU_femG_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$FU_femG_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$FU_femGon_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$FU_femGon_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$FU_femL_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$FU_femL_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$FU_malG_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$FU_malG_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$FU_malGon_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$FU_malGon_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)
blast_output_merged$FU_malL_biased_TRgenes_reiro_blast <-
  merge(blast_output_merged$FU_malL_biased_TRgenes_reiro_blast, 
        panther[, c(2,5)], 
        by.x = "gene_name", 
        by.y = "MappedID", 
        all.x = TRUE)


```

Now that we have the PANTHER protein classes attached to all of the successfully BLASTed sex-biased genes within each organ, we can plot the proportions of genes falling into each category to see what is most important for each of the organ types.

```{r plot-panther-SBG}
#Count the number of genes that are falling into each protein class
go_tables <- lapply(blast_output_merged, function(dat){
  
  #Create a table that counts the number of genes falling within each PANTHER prot. class
  panther_table <- data.frame(table(dat$panther_prot_class))
  
  #Make sure the protein names are in there as a character
  panther_table$Var1 <- as.character(panther_table$Var1)
  
  #Change all of the null results to say Unclassified
  panther_table$Var1[panther_table$Var1==""] <- "Unclassified"

  return(panther_table)
  })

#Merge all of the different organ data into one dataset
#Create an empty dataframe to store all of the values in
all_go_dat <- data.frame(matrix(ncol=4,nrow=0))

#Add column names to that empty dataset
colnames(all_go_dat) <- c(
  colnames(go_tables$FU_femG_biased_TRgenes_reiro_blast),
  "bias_cat", "tissue")

#Create a vector of you tissues, in the same order that they appear in the go_table list
tissues <- c("Gill", "Gonad", "Liver", 
             "Gill", "Gonad", "Liver")

#Loop through all of the go_tables and combine into one
for(i in 1:length(go_tables)){
  #browser()
  tmp <- go_tables[[i]]
  if (nrow(tmp) == 0) {
    print(tmp)
  }else{
  
  tmp$bias_cat <- names(go_tables)[i]
  tmp$tissue <- tissues[[i]]
  all_go_dat <- rbind(all_go_dat, tmp)}
  
}

#Calculate the frequency at which each protein class occurs within the different tissue types
##Create an empty dataset to store the values in
all_go_sums <- data.frame(matrix(ncol = 4,
                                 nrow = 0))

##Add appropriate column names
colnames(all_go_sums) <- c("Freq", "prot_class", 
                           "tissue", "prop")

##Loop through each organ type to calculate frequencies
for(organ in unique(all_go_dat$tissue)){
  
  #subset the dataset based on the organ
  tmp <- all_go_dat[all_go_dat$tissue == organ, ]
  
  #Sum up all of the genes in each category
  go_sums <- as.data.frame(tapply(tmp$Freq, tmp$Var1, sum))
  
  #Add the protein class as a column rather than row names
  go_sums$prot_class <- rownames(go_sums)
  
  #Change the column name to "Frequency"
  colnames(go_sums)[1] <- "Freq"
  
  #Add the organ in a separate column and remove rownames
  go_sums$tissue <- organ
  rownames(go_sums) <- NULL
  
  #Calculate the respective proportion for each protein class
  go_sums$prop <- go_sums$Freq/sum(go_sums$Freq)
  
  #Export the data to the previously created dataframe
  all_go_sums <- rbind(all_go_sums, go_sums)
}

#Change all of the protein classes that have a frequency of less than 0.01 to "Other"
for(class in unique(all_go_sums$prot_class)){
  
  #Pull out the rows that contain that protein class
  match_rows <- which(all_go_sums$prot_class==class)
  
  #If any of the proportions associated with those rows is less than 0.02, classify as other
  if(any(all_go_sums$prop[match_rows]>0.01)==FALSE){
    
    all_go_sums$prot_class[match_rows] <- " other"
    
  }
}

#Change format of "Unclassified so it will come first"
all_go_sums$prot_class[all_go_sums$prot_class=="Unclassified"] <-  " unclassified"

#write.csv(all_go_sums, "data/GO_freq_SBG.csv", row.names = FALSE)
#Generate the GO plot
organ_cols <- c("Gill" = "#20B2AA", "Gonad" = "#EEB422",
                "Liver" = "#EE8262")

pdf("docs/figs/FigGO_SBGbarplot2.pdf",height=14, width=18)

all_go_sums$prot_class <- str_wrap(all_go_sums$prot_class, 
                                       width = 60)
all_go_sums$prot_class[all_go_sums$prot_class=="unclassified"]<- " unclassified"
all_go_sums$prot_class[all_go_sums$prot_class=="other"]<- " other"

ggplot(all_go_sums, 
       aes(fct_rev(prot_class), prop, fill = tissue)) +   
  geom_bar(position = "dodge", stat="identity") +
  scale_fill_manual(values = organ_cols) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1,size=20),
        axis.text.y = element_text(size=15, hjust = 0.5, 
                                   margin = margin(b = 20),
                                   vjust = 0.75),
        text=element_text(size=20),
        legend.position = "bottom",
        axis.ticks.length.y = unit(.25, "cm")) +
  coord_flip() + 
  labs(y="Proportion of sex-biased genes", x="")

dev.off()

```

## Looking at GO groups for the male- and female-biased genes
```{r plot-panther-SBG-liver}
#Calculate the frequency at which each protein class occurs within the different tissue types
##Create an empty dataset to store the values in
all_go_sums_liver <- data.frame(matrix(ncol = 5,
                                      nrow = 0))

##Add appropriate column names
colnames(all_go_sums_liver) <- c(colnames(all_go_dat),
                                "prop")

##Loop through each bias type to calculate frequencies
for (bias in c("FU_femL_biased_TRgenes_reiro_blast",
               "FU_malL_biased_TRgenes_reiro_blast")) {
  
  tmp <- all_go_dat[all_go_dat$bias_cat == bias, ]
  
  #Calculate the respective proportion for each protein class
  tmp$prop <- tmp$Freq/sum(tmp$Freq)
  
  #Export the data to the previously created dataframe
  all_go_sums_liver <- rbind(tmp, all_go_sums_liver)
  
}


#Change all of the protein classes that have a frequency of less than 0.01 to "Other"
for(class in unique(all_go_sums_liver$Var1)){
  
  #Pull out the rows that contain that protein class
  match_rows <- which(all_go_sums_liver$Var1==class)
  
  #If any of the proportions associated with those rows is less than 0.01, classify as other
  if(any(all_go_sums_liver$prop[match_rows]>0.01)==FALSE){
    
    all_go_sums_liver$Var1[match_rows] <- " other"
    
  }
}

#Change format of "Unclassified so it will come first"
all_go_sums_liver$Var1[all_go_sums_liver$Var1=="Unclassified"] <-  " unclassified"

#Generate the GO plot
bias_col <- c("FU_femL_biased_TRgenes_reiro_blast" = "#7fc97f", 
               "FU_malL_biased_TRgenes_reiro_blast" = "#beaed4")


pdf("docs/figs/FigGO_SBGliver.pdf",height=14, width=12)

ggplot(all_go_sums_liver, 
       aes(fct_rev(Var1), prop, fill = bias_cat)) +   
  geom_bar(position = "dodge", stat="identity") +
  scale_fill_manual(values = bias_col) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text.x = element_text(angle = 90, 
                                   vjust = 0.5, 
                                   hjust=1,size=20),
        axis.text.y = element_text(size=15, hjust = 1,
                                   vjust = 0.3),
        text=element_text(size=20),
        legend.position = "bottom",
        axis.ticks.length.y = unit(.25, "cm")) +
  coord_flip() + 
  labs(y="Proportion of sex-biased genes", x="")

dev.off()

```

```{r plot-panther-SBG-gonads}
#Calculate the frequency at which each protein class occurs within the different tissue types
##Create an empty dataset to store the values in
all_go_sums_gonad <- data.frame(matrix(ncol = 5,
                                      nrow = 0))

##Add appropriate column names
colnames(all_go_sums_gonad) <- c(colnames(all_go_dat),
                                "prop")

##Loop through each bias type to calculate frequencies
for (bias in c("FU_femGon_biased_TRgenes_reiro_blast",
               "FU_malGon_biased_TRgenes_reiro_blast")) {
  
  tmp <- all_go_dat[all_go_dat$bias_cat == bias, ]
  
  #Calculate the respective proportion for each protein class
  tmp$prop <- tmp$Freq/sum(tmp$Freq)
  
  #Export the data to the previously created dataframe
  all_go_sums_gonad <- rbind(tmp, all_go_sums_gonad)
  
}


#Change all of the protein classes that have a frequency of less than 0.01 to "Other"
for(class in unique(all_go_sums_gonad$Var1)){
  
  #Pull out the rows that contain that protein class
  match_rows <- which(all_go_sums_gonad$Var1==class)
  
  #If any of the proportions associated with those rows is less than 0.01, classify as other
  if(any(all_go_sums_gonad$prop[match_rows]>0.01)==FALSE){
    
    all_go_sums_gonad$Var1[match_rows] <- " other"
    
  }
}

#Change format of "Unclassified so it will come first"
all_go_sums_gonad$Var1[all_go_sums_gonad$Var1=="Unclassified"] <-  " unclassified"

#Generate the GO plot
bias_col <- c("FU_femGon_biased_TRgenes_reiro_blast" = "#7fc97f", 
               "FU_malGon_biased_TRgenes_reiro_blast" = "#beaed4")


pdf("docs/figs/FigGO_SBGgonad.pdf",height=14, width=12)

ggplot(all_go_sums_gonad, 
       aes(fct_rev(Var1), prop, fill = bias_cat)) +   
  geom_bar(position = "dodge", stat="identity") +
  scale_fill_manual(values = bias_col) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.text.x = element_text(angle = 90, 
                                   vjust = 0.5, 
                                   hjust=1,size=20),
        axis.text.y = element_text(size=15, hjust = 1,
                                   vjust = 0.3),
        text=element_text(size=20),
        legend.position = "bottom",
        axis.ticks.length.y = unit(.25, "cm")) +
  coord_flip() + 
  labs(y="Proportion of sex-biased genes", x="")

dev.off()

```

```{r}
#Combine the GO figures
figGOa <- image_ggplot(image_read_pdf('docs/figs/FigGO_SBGliver.pdf'),
                        interpolate = TRUE)
figGOb <- image_ggplot(image_read_pdf('docs/figs/FigGO_SBGgonad.pdf'),
                           interpolate = TRUE)

# put the patchworks together
figGO <-  wrap_plots(figGOa, figGOb, ncol=2) + 
  plot_annotation(tag_levels = 'A') 

ggsave("docs/figs/FigGO_indv.pdf", figGO, height=5, width=8)
ggsave("docs/figs/FigGO_indv.png", figGO, height=5, width=8)
```


```{r export-danio-geneID-INDV, eval=FALSE}

write.table(blast_output_merged$FU_femGon_biased_TRgenes_reiro_blast$gene_name,
            'FU_GOnames_SBG_ovary.txt',
            sep = " ",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)

write.table(blast_output_merged$FU_malGon_biased_TRgenes_reiro_blast$gene_name,
            'FU_GOnames_SBG_testes.txt',
            sep = " ",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)

write.table(blast_output_merged$FU_femL_biased_TRgenes_reiro_blast$gene_name,
            'FU_GOnames_SBG_fLiver.txt',
            sep = " ",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)

write.table(blast_output_merged$FU_malL_biased_TRgenes_reiro_blast$gene_name,
            'FU_GOnames_SBG_mLiver.txt',
            sep = " ",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)
```