---
title: "Differential Expression Analysis in _Syngnathus fuscus_"
author: "Coley Tosto"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    code_folding: show
    toc: yes
    toc_float: yes
    number_sections: yes
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
```

``` {r library, include = FALSE}
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
library(ggpubr)
library(cowplot)
library(UpSetR)
library(ggplot2)
library(kableExtra)
library(tidyverse)
library(knitr)
```

``` {r functions}

```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
FU_txi.salmon <- readRDS("data/txi.salmon_FU.RDS")

#The samples file generated for tximport
FU_samples <- read.table("FU_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
FU_samples$Sex <- as.factor(FU_samples$Sex)
FU_samples$Organ <- as.factor(FU_samples$Organ)

```

The package `DESeq2` was used for the differential expression analysis outlined below.

# Single factor analysis - Comparing Males v Females
To analyze your data with DESeq2 you must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` and a `samples` file that lays out all of the conditions.
```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_FU <- DESeqDataSetFromTximport(FU_txi.salmon, 
                                   colData = FU_samples,
                                   design = ~ Sex)

```


The data is then pre-filtered to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.
```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_FU)) >= 10
dds_FU <- dds_FU[keep, ]

```


After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.
```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_FU_exp <- DESeq(dds_FU)

#Compile the results
res <- results(dds_FU_exp)
res
```

Once that has finished we can now start exploring some of the single-factor analysis results
```{r investigate-DESeq-results-SF}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#How many ADJUSTED p-values were less than 0.1?
sum(res$padj < 0.1, na.rm = TRUE)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_FU_exp, alpha = 0.05)
summary(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```
Looking at the summary of our results we can see that there are 2,626 male-biased genes and 2,488 female-biased genes when all of the organs are organized together. While this is informative, I am more interested in a multi-factor approach that takes into account both Sex and Tissue type. Before the multi-factor analysis, we can visualize some results from the single-factor analysis and use it for exploratory data analysis.

## Visualizing the results {.tabset}
### MA-plot - MvF Diff. {-}
Generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be colored if the adjusted p-value is less that 0.05.
```{r MA-plot, echo=FALSE}
par(mfrow=c(1,2))
plotMA(res05, ylim = c(-2,2))
plotMA(res05)

```

### Heatmap - overall expression {-}
We can also generate a heat map to look ar overall expression levels across our samples. Note, this is not differentially expressed genes.
```{r heatmap, echo=FALSE}
#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(dds_FU_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]
df <- as.data.frame(colData(dds_FU)[,c("Sex", "Organ")])

#Transform the data
vsd <- vst(dds_FU_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df)
```

From the heatmap we can pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.
```{r heatmap_genes}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))

#Only run once to create the file and then commented out
#write.table(heatmap_TG,
            #'FU_heatmap_trinitygenes.txt',
            #sep = "",
            #quote=FALSE,
            #row.names = FALSE,
            #col.names = FALSE)
```

### Sample-dist Heatmap {-}
We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples
```{r sample-dist, echo=FALSE}
#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$Sex, vsd$Organ, sep = "-")
colnames(sampleDistsMatrix) <- paste(vsd$Sex, vsd$Organ, sep = "-")

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)
```

### PCA plots {-}
```{r pca-plot, echo=FALSE, message=FALSE}
colData <- as.data.frame(FU_samples)
rownames(colData) <- FU_samples$ID

#Prepare the PCA using the transformed data from above
pcaFU <- pca(assay(vsd), metadata = colData)

#Plot the PCA
biplot(pcaFU,
       showLoadings = FALSE,
       colby = 'Organ', colkey = c('Liver'='forestgreen', 'Gill'='purple', 'Gonad'='gold'),
       colLegendTitle = "Organ Type",
       hline = 0, vline = 0,
       vlineType = 'dashed',
       pointSize = 3,
       legendPosition = 'left', legendLabSize = 10, legendIconSize = 6.0, legendTitleSize = 10,
       shape = 'Sex', shapekey = c('M'=17, 'F'=8),
       shapeLegendTitle = 'Sex',
       drawConnectors = FALSE,
       title = 'PCA bi-plot for Syngnathus fuscus')

pairsplot(pcaFU,
          components = getComponents(pcaFU, c(1, 2, 3)),
          triangle = FALSE,
          hline = 0, vline = 0,
          pointSize = 3,
          gridlines.major = FALSE, gridlines.minor = FALSE,
          colby = 'Organ', colkey = c('Liver'='forestgreen', 'Gill'='purple', 'Gonad'='gold'),
          shape = 'Sex', shapekey = c('M'=17, 'F'=8),
          title = 'Pairs plot', titleLabSize = 22,
          legendPosition = 'left',
          axisLabSize = 10, plotaxes = TRUE,
          margingaps = unit(c(0.1, 0.1, 0.1, 0.1), 'cm'))
```

# Multifactor design - Comparing M v F across the diff tissue types
If we investigate the column data of our DESeq dataset we can see that each sample has both a sex and organ type attached to it, we will be using these two factors in our multi-factor analysis

```{r multi-factor-analysis, message=FALSE, warning=FALSE}
##Create a copy of the DESeq dataset
ddsMF_FU <- dds_FU

#Create an additional column the has the two conditions combined(sex and organ type)
ddsMF_FU$group <- factor(paste0(ddsMF_FU$Sex, ddsMF_FU$Organ))

#Reapply that new column as the design for this analysis
design(ddsMF_FU) <- ~ group

#Run the differential expression analysis
ddsMF_FU_exp <- DESeq(ddsMF_FU)
resultsNames(ddsMF_FU_exp)

```

## Invesitgate the results of the differential expression {.tabset}
Thanks to the multi-factor analysis we can now explore differential expression between all of the different combinations of tissues and sexes: \  

  - Male Liver v. Female Liver \  
  - Male Gill v. Female Gill \  
  - Male Gonad v. Female Gonad \  
  - All of the within sex tissue comparisons (e.g. Male Liver v. Male Gill, etc.) \  

### M-F Liver Comparisson {-}
```{r m-f-liver-results}
##Pulling out the liver M-F results with an alpha of 0.05
FU_liver_con_res <- results(ddsMF_FU_exp, contrast = c("group", "MLiver", "FLiver"), alpha = 0.05)
FU_liver_con_res$trin_geneid <- rownames(FU_liver_con_res)
head(FU_liver_con_res)

summary(FU_liver_con_res)
```
I will be classifying sex-biased genes the same way that I did for _Syngnathus floridae_ as **genes with a p-value < 0.05 AND a logFC $\ge$ |2|**. With that criteria in the liver there are `r sum(FU_liver_con_res$padj <= 0.05 & FU_liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(FU_liver_con_res$padj <= 0.05 & FU_liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. 

I have then pulled out all of the male-biased, female-biased, and non-biased genes based on the criteria outlined above. I determined non-biased genes as a p-vaule > 0.05.

```{r liver-sex-biased}
#Removing the rows where padj. is NA in results
FU_liver_con_res_noNA <- FU_liver_con_res[!is.na(FU_liver_con_res$padj),]
summary(FU_liver_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating a vector that contains all of the male-biased and female-biased genes in the liver
FU_liver_mal_biased <- FU_liver_con_res_noNA[which(FU_liver_con_res_noNA$log2FoldChange >= 2 & FU_liver_con_res_noNA$padj <= 0.05),]
FU_liver_fem_biased <- FU_liver_con_res_noNA[which(FU_liver_con_res_noNA$log2FoldChange <= -2 & FU_liver_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the liver
FU_liver_non_biased <- FU_liver_con_res_noNA[which(FU_liver_con_res_noNA$padj > 0.05),]

```

I will be generating a table that outlines the genes that are the most male or female biased in each organ type. To do this I need to pull the top 50 differentially expressed genes in males or females, get the Trinity gene IDs and then BLAST the corresponding sequences. Here I am pulling out the Trinity gene IDs for those genes.

```{r top50-liver}
#Creating a subset of the results where the p-value is less than 0.05
FU_liver_con_res_p05 <- FU_liver_con_res_noNA[FU_liver_con_res_noNA$padj <= 0.05, ]


#Pulling the top 50 diff expressed genes in Males
top50_FU_MLiver <- head(FU_liver_con_res_p05[order(FU_liver_con_res_p05$log2FoldChange, decreasing = TRUE), ], n = 50)
#Run once to generate the file and then commented out
#write.table(cbind(rownames(top50_FU_MLiver)),
 #           'FU_maleL_top50TRgenes.txt', 
 #            sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)


#Pulling the top 50 diff expressed genes in Females
top50_FU_Fliver <- head(FU_liver_con_res_p05[order(FU_liver_con_res_p05$log2FoldChange, decreasing = FALSE), ], n = 50)
#Run once to generate the file and then commented out
#write.table(cbind(rownames(top50_FU_Fliver)),
 #           'FU_femL_top50TRgenes.txt', 
 #            sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)
```


### M-F Gill Comparisson {-}
```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res_FU <- results(ddsMF_FU_exp, contrast = c("group", "MGill", "FGill"), alpha = 0.5)
gill_con_res_FU$trin_geneid <- rownames(gill_con_res_FU)
head(gill_con_res_FU)

summary(gill_con_res_FU)
```

In the gills there were `r sum(gill_con_res_FU$padj <= 0.05 & gill_con_res_FU$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res_FU$padj <= 0.05 & gill_con_res_FU$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

I have then pulled out all of the male-biased, female-biased, and non-biased genes to save them into their own objects.

```{r gill-sex-biased}
#Removing the rows where padj. is NA in results
FU_gill_con_res_noNA <- gill_con_res_FU[!is.na(gill_con_res_FU$padj), ]
summary(FU_gill_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating a vector that contains all of the male-biased and female-biased genes in the gills
FU_gill_mal_biased <- FU_gill_con_res_noNA[which(FU_gill_con_res_noNA$log2FoldChange >= 2 & FU_gill_con_res_noNA$padj <= 0.05),]
FU_gill_fem_biased <- FU_gill_con_res_noNA[which(FU_gill_con_res_noNA$log2FoldChange <= -2 & FU_gill_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gills, p>0.05
FU_gill_non_biased <- FU_gill_con_res_noNA[which(FU_gill_con_res_noNA$padj > 0.05),]
```


Here I am getting the trinity gene IDs that correspond to the top 50 differentially expressed genes for males and females.

```{r top50-gill}
#Creating a subset of results where p-value is less than 0.05
FU_gill_con_res_p05 <- FU_gill_con_res_noNA[FU_gill_con_res_noNA$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_FU_MGill <- head(FU_gill_con_res_p05[order(FU_gill_con_res_p05$log2FoldChange, decreasing = TRUE),], n=50)
#Run once to generate the file and then commented out
#write.table(cbind(rownames(top50_FU_MGill)),
 #           'FU_maleG_top50TRgenes.txt', 
 #           sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FU_FGill <- head(FU_gill_con_res_p05[order(FU_gill_con_res_p05$log2FoldChange, decreasing = FALSE),], n=50)
#Run once to generate the file and then commented out
#write.table(cbind(rownames(top50_FU_FGill)),
 #           'FU_femG_top50TRgenes.txt', 
 #           sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)
```

### M-F Gonad Comparisson {-}
```{r m-f-gonads}
##Pulling out the gonad M-F results
gonad_con_res_FU <- results(ddsMF_FU_exp, contrast = c("group", "MGonad", "FGonad"), alpha = 0.05)
gonad_con_res_FU$trin_geneid <- rownames(gonad_con_res_FU)
head(gonad_con_res_FU)

summary(gonad_con_res_FU)
```

Between the ovaries and testis (gonads) there were `r sum(gonad_con_res_FU$padj <= 0.05 & gonad_con_res_FU$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gonad_con_res_FU$padj <= 0.05 & gonad_con_res_FU$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

From the gonad M-F results I have then pulled out all of the male-biased, female-biased, and non-biased genes and saved them to their own objects.

```{r gonad-sex-biased}
#Removing the rows where padj. is NA in results
FU_gonad_con_res_noNA <- gonad_con_res_FU[!is.na(gonad_con_res_FU$padj), ]
summary(FU_gonad_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating an object that contains all of the male-biased and female-biased genes in the gonads
FU_gonad_mal_biased <- FU_gonad_con_res_noNA[which(FU_gonad_con_res_noNA$log2FoldChange >= 2 & FU_gonad_con_res_noNA$padj <= 0.05),]
FU_gonad_fem_biased <- FU_gonad_con_res_noNA[which(FU_gonad_con_res_noNA$log2FoldChange <= -2 & FU_gonad_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gonads, p>0.05
FU_gonad_non_biased <- FU_gonad_con_res_noNA[which(FU_gonad_con_res_noNA$padj > 0.05),]
```

I then pulled out the trinity gene IDs that corresponded to the top 50 differentially expressed genes in males and females so that I could BLAST the sequences.
```{r top50-gonad}
#Creating a subset of the results where the p-value is less than 0.05
FU_gonad_con_res_p05 <- FU_gonad_con_res_noNA[FU_gonad_con_res_noNA$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_FU_MGonad <- head(FU_gonad_con_res_p05[order(FU_gonad_con_res_p05$log2FoldChange, decreasing = TRUE),], n=50)
#Run once to generate the file and then commented out
#write.table(cbind(rownames(top50_FU_MGonad)),
 #           'FU_maleGon_top50TRgenes.txt', 
 #           sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FU_FGonad <- head(FU_gonad_con_res_p05[order(FU_gonad_con_res_p05$log2FoldChange, decreasing = FALSE),], n=50)
#Run once to generate the file and then commented out
#write.table(cbind(head(top50_FU_FGonad)),
 #           'FU_femGon_top50TRgenes.txt', 
 #           sep = "", 
 #           quote=FALSE, 
 #           row.names = FALSE, 
 #           col.names = FALSE)
```

### MA Plots {-}
```{r MA-plot-MF, echo=FALSE, fig.height= 20}
liver_MA <- ggmaplot(FU_liver_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Liver" %->% "Female Liver"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("orange2", "yellowgreen", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gill_MA <- ggmaplot(gill_con_res_FU,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Gill" %->% "Female Gill"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("orange2", "yellowgreen", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gonad_MA <- ggmaplot(gonad_con_res_FU,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Testis" %->% "Ovaries"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("orange2", "yellowgreen", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

#Merge all the plots into one
plot_grid(liver_MA, gill_MA, gonad_MA,
          ncol = 1,
          nrow = 3,
          labels = c('A', 'B', 'C'),
          label_fontfamily = 'sans',
          label_fontface = 'bold',
          label_size = 18,
          rel_heights = c(4.10, 4.10, 4.10))

```

## Creating an Upset Plot
```{r upset-plot, echo=FALSE}
par(mfrow=c(1,2))
#Male-biased genes
listInputMB <- list("Testis" = rownames(FU_gonad_mal_biased),
                  "Male Gill"=rownames(FU_gill_mal_biased), 
                  "Male Liver"=rownames(FU_liver_mal_biased)
                  )
upset(fromList(listInputMB),
      mainbar.y.label = "Number of Shared Male-Biased Genes",
      sets.x.label = "Total Number of Male-Biased Genes",
      point.size = 3)

#Female
listInputFB <- list("Ovaries" = rownames(FU_gonad_fem_biased),
                  "Female Gill"=rownames(FU_gill_fem_biased), 
                  "Female Liver"=rownames(FU_liver_fem_biased)
                  )
upset(fromList(listInputFB),
      mainbar.y.label = "Number of Shared Female-Biased Genes",
      sets.x.label = "Total Number of Female-Biased Genes",
      point.size = 3)
```

## Variation in FC across sex-bias and tissue type
I want to create a plot the highlights the variation we see in fold-change both across the different biases groups (male-biased, female-biased, and non-biased) within one tissue type and across all of the tissue types. To do this I first need to create a "long" dataset in the following format:

| Tissue_type|     Bias      |   Fold-Change   |
|:----------:|:-------------:|:---------------:|
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   fem_bias    |      XXXXX      |
|    Gill    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   fem_bias    |      XXXXX      |
|    Liver   |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   fem_bias    |      XXXXX      |
|   Gonad    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |

That dataset is created here, with the Trinity gene IDs included as well:

```{r logFC-long-dataset}
FU_logFC_long<- data.frame(
  tissue=c(rep("Gill",nrow(FU_gill_fem_biased)),
           rep("Gill", nrow(FU_gill_mal_biased)),
           rep("Gill", nrow(FU_gill_non_biased)),
           rep("Gonad", nrow(FU_gonad_fem_biased)),
           rep("Gonad", nrow(FU_gonad_mal_biased)),
           rep("Gonad", nrow(FU_gonad_non_biased)),
           rep("Liver", nrow(FU_liver_fem_biased)),
           rep("Liver", nrow(FU_liver_mal_biased)),
           rep("Liver", nrow(FU_liver_non_biased))
         ),
  bias=c(rep("FB",nrow(FU_gill_fem_biased)),
         rep("MB",nrow(FU_gill_mal_biased)),
         rep("NB", nrow(FU_gill_non_biased)),
         rep("FB", nrow(FU_gonad_fem_biased)),
         rep("MB", nrow(FU_gonad_mal_biased)),
         rep("NB", nrow(FU_gonad_non_biased)),
         rep("FB", nrow(FU_liver_fem_biased)),
         rep("MB", nrow(FU_liver_mal_biased)),
         rep("NB", nrow(FU_liver_non_biased))
         ),
  logFC=c(FU_gill_fem_biased$log2FoldChange,
          FU_gill_mal_biased$log2FoldChange,
          FU_gill_non_biased$log2FoldChange,
          FU_gonad_fem_biased$log2FoldChange,
          FU_gonad_mal_biased$log2FoldChange,
          FU_gonad_non_biased$log2FoldChange,
          FU_liver_fem_biased$log2FoldChange,
          FU_liver_mal_biased$log2FoldChange,
          FU_liver_non_biased$log2FoldChange
          ),
  geneID=c(rownames(FU_gill_fem_biased),
           rownames(FU_gill_mal_biased),
           rownames(FU_gill_non_biased),
           rownames(FU_gonad_fem_biased),
           rownames(FU_gonad_mal_biased),
           rownames(FU_gonad_non_biased),
           rownames(FU_liver_fem_biased),
           rownames(FU_liver_mal_biased),
           rownames(FU_liver_non_biased)
           )
  
)
```

With the dataset in the proper format now, we can generate the plot
```{r FC-var-plot, echo=FALSE}
#ggplot(logFC_long,
       #aes(x = tissue, y = abs(logFC), fill = bias)) +
  #geom_boxplot() +
  #geom_jitter(color="black", size=0.4, alpha=0.9) +
  #theme_minimal()

my_colors <- c("FB" = "darkolivegreen4", "MB" = "goldenrod2", "NB" = "salmon2")

ggplot(FU_logFC_long,
       aes(x = bias, y = abs(logFC), fill = bias)) +
  geom_boxplot() +
  scale_fill_manual(values = my_colors) +
  facet_grid(. ~ tissue) +
  theme_minimal() +
  geom_point(data=FU_logFC_long, aes(x=bias, y=abs(logFC)),
             size=0.4, 
             position = position_jitter(width = 0.15)) + 
  labs(x="Bias Level", y="|logFC|") +
  guides(fill = guide_legend(title = "Bias Level", order = 3))
  #scale_fill_discrete(labels=c('Female-Biased', 'Male-Biased', 'Non-Biased'))
  #geom_jitter(color="black", size=0.4, alpha=0.9)
```

## Categorizing sex-specific genes
Now that we have investigated what the sex-bias is like across the different tissue types I want to dive further into genes with **sex-specific expression**. These are genes expressed only in one sex or the other. I am classifying a sex-specific genes within each tissue as ones where the expression of that gene is 0 for all of one sex and there is a median of $\ge 10$ in the other sex.

```{r sex-specific-for-loop}
#Create a vector with the different organ types
organs <- levels(colData(dds_FU)$Organ)

#Create an empty list to store my datasets in
FU_sex_specific_genes <- list()

#Generate the for loop to identify MSpecific and FSpecific genes in each organ based on Medians
for(organ in organs){
  
  #Male-Specific Genes
  ##Pull out all of the rows where fem count = 0
  fem0_organ <- assay(dds_FU)[rowSums(assay(dds_FU)[, dds_FU$Sex == "F" & dds_FU$Organ==organ]) < 1, dds_FU$Organ == organ]
  
  ##Pull out rows where median of male count >=10 for that organ
  mal10_organ <- apply(assay(dds_FU)[, dds_FU$Sex == "M" & dds_FU$Organ == organ],
                       1,
                       function(x) median(x) >=10)
  
  ##Keep only the rows where fem=0 and the Male median>=10
  fem0_mal10_organ <- fem0_organ[rownames(fem0_organ) %in% names(mal10_organ[mal10_organ == TRUE]),]
  
  ##Create a new object with a name based on the organ type
  organ_malsp <- sub("$", "_male_specific", organ)
  FU_sex_specific_genes[[organ_malsp]] <- fem0_mal10_organ
  
  
  
  #Female-Specific Genes
  ##Pull out all of the rows where male count = 0
  mal0_organ <- assay(dds_FU)[rowSums(assay(dds_FU)[, dds_FU$Sex == "M" & dds_FU$Organ==organ]) < 1, dds_FU$Organ == organ]
  
  ##Pull out rows where median of female count >=10 for that organ
  fem10_organ <- apply(assay(dds_FU)[, dds_FU$Sex == "F" & dds_FU$Organ == organ],
                       1,
                       function(x) median(x) >=10)
  
  #Keep only the rows where male=0 and the fem median>=10
  mal0_fem10_organ <- mal0_organ[rownames(mal0_organ) %in% names(fem10_organ[fem10_organ == TRUE]),]
  
  # Create a new object with a name based on the organ type
  organ_femsp <- sub("$", "_female_specific", organ)
  FU_sex_specific_genes[[organ_femsp]] <- mal0_fem10_organ
} 

```

### Investigating the results of the sex-specific subsetting
Let's now take a look at how many sex-specific genes we have in each tissue type:
```{r sex-specific-counts, echo=FALSE, message=FALSE, warning=FALSE}
FU_sex_specific_counts <- data.frame(tissue = rep(c("Gill", "Gonad", "Liver"), 2),
                                  sex = c(rep("M", 3),
                                          rep("F", 3)),
                                  num_genes = c(nrow(FU_sex_specific_genes$Gill_male_specific),
                                                nrow(FU_sex_specific_genes$Gonad_male_specific),
                                                nrow(FU_sex_specific_genes$Liver_male_specific),
                                                nrow(FU_sex_specific_genes$Gill_female_specific),
                                                nrow(FU_sex_specific_genes$Gonad_female_specific),
                                                nrow(FU_sex_specific_genes$Liver_female_specific))
                                  )

FU_sex_specific_counts %>% 
  group_by(tissue, sex) %>% 
  summarise("Number of Sex-specific Genes" = sum(num_genes)) %>%
  kable(caption = "Sex-specific Genes", format = "html",
        col.names = c("Tissue Type", "Sex", "Number of Sex-specific Genes")) %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F, position = "left")
```

To get a better idea what is going on between the sex-specific genes vs the sex-biased genes I first determined how many overlaps there were between our sex-biased and sex-specific genes and then used plotCounts to plot the counts for each gene separately. 

In the **Male Gills** there are `r nrow(FU_gill_mal_biased[rownames(FU_gill_mal_biased) %in% rownames(FU_sex_specific_genes$Gill_male_specific),])` genes that overlap between sex-biased and sex-specific, there are `r nrow(FU_sex_specific_genes$Gonad_male_specific[rownames(FU_sex_specific_genes$Gonad_male_specific) %in% rownames(FU_gonad_mal_biased),])` overlapping in the **gonads** and `r nrow(FU_sex_specific_genes$Liver_male_specific[rownames(FU_sex_specific_genes$Liver_male_specific) %in% rownames(FU_liver_mal_biased),])` overlapping in the **liver**. For females we have `r nrow(FU_gill_fem_biased[rownames(FU_gill_fem_biased) %in% rownames(FU_sex_specific_genes$Gill_female_specific),])` overlapping sex-specific and sex-biased genes in the **gills**, `r nrow(FU_sex_specific_genes$Gonad_female_specific[rownames(FU_sex_specific_genes$Gonad_female_specific) %in% rownames(FU_gonad_fem_biased),])` genes in the **gonads**, and `r nrow(FU_sex_specific_genes$Liver_female_specific[rownames(FU_sex_specific_genes$Liver_female_specific) %in% rownames(FU_liver_fem_biased),])` genes in the **liver**.

```{r gills-sex-specific, eval = FALSE}
## Female sex-specific
fem_sp_geneIDs <- rownames(FU_sex_specific_genes$Gill_female_specific)

##Set the plotting window
par(mfrow=c(1,1))

for(gene in fem_sp_geneIDs){
  
  plotCounts(ddsMF_FU[,ddsMF_FU$Organ == 'Gill'], gene=gene, intgroup="Sex")
  
}

## Female sex-biased
gill_fem_biased_geneIDS <- rownames(FU_gill_fem_biased)

##Set the plotting window
par(mfrow=c(1,1))

for(gene in gill_fem_biased_geneIDS){
  
  plotCounts(ddsMF_FU[,ddsMF_FU$Organ == 'Gill'], gene=gene, intgroup="Sex", col= 'salmon', pch = 19)
  
}
```
