---
title: "Differential Expression Analysis in _Syngnathus fuscus_"
author: "Coley Tosto"
date: "`r Sys.Date()`"
output:
  github_document: 
    toc: true 
  #bookdown::pdf_document2:
    #fig_caption: yes
    #keep_tex: yes
    #number_sections: yes
    #toc: true
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
always_allow_html: true
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',fig_path="../imgs/")
```

``` {r library, include = FALSE}
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(PCAtools)
library(spfTools)
library(magick)
library(patchwork)
library(ggpubr)
library(cowplot)
library(UpSetR)
library(ggplot2)
library(kableExtra)
library(tidyverse)
library(knitr)
```

``` {r read-data}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
FU_txi.salmon <- readRDS("data/txi.salmon_FU.RDS")

#The samples file generated for tximport
FU_samples <- read.table("FU_samples.txt", header = TRUE)

#Make sure the conditions are in the samples file as a factor
FU_samples$Sex <- as.factor(FU_samples$Sex)
FU_samples$Organ <- as.factor(FU_samples$Organ)

```

The package `DESeq2` was used for the differential expression analysis outlined below.

# Single factor analysis - Comparing Males v Females
I am starting with a single factor analysis to look at overall differences between male and female expression across the organs and to do some exploratory analysis of the data. 

To analyze my data with DESeq2 I must first generate the DESeqDataSet. In order to do this we need the abundance matrix generated with `tximport` which is stored in the object "FU_txi.salmon" and was generated at the end of the `Analyzing_fuscus_RNAseq_data_from_MSU.Rmd` document and a `samples` file which is stored in the object "FU_samples" that lays out all of the conditions.

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_FU <- DESeqDataSetFromTximport(FU_txi.salmon, 
                                   colData = FU_samples,
                                   design = ~ Sex)

```

I am then going to pre-filter the data to remove low gene counts before running further DESeq2 functions. By doing this we remove rows in which there are very few reads thus reducing the memory size of the `dds` object and increasing the speed at which we can use the transformation and testing functions in DESeq2.

The cutoff used here was to remove reads that had a **total** count of less than 10 in each row.

```{r pre-filtering, message=FALSE, warning=FALSE}
#only keeping rows that have at lead 10 reads total
keep <- rowSums(counts(dds_FU)) >= 10
dds_FU <- dds_FU[keep, ]

```

After filtering we can now perform the standard differential expression analysis that is wrapped into DESeq2.

```{r diff-exp, message=FALSE, warning=FALSE}
#Generate the expression values
dds_FU_exp <- DESeq(dds_FU)

#Compile the results
res <- results(dds_FU_exp)
res

```

Once that has finished we can now start exploring some of the single-factor analysis results.

```{r investigate-DESeq-results-SF}
##Ordering our results based on p-value
resOrdered <- res[order(res$pvalue),]
resOrdered

summary(res)

#Looking at  an alpha=0.05, the default is 0.1
res05 <- results(dds_FU_exp, alpha = 0.05)
summary(res05)
sum(res05$padj < 0.05, na.rm = TRUE)

```

Looking at the summary of our results we can see that there are 2,626 male-biased genes and 2,488 female-biased genes when all of the organs are organized together. While this is informative, I am more interested in a multi-factor approach that takes into account both **Sex** and **Tissue** type. Before the multi-factor analysis, we can visualize some results from the single-factor analysis and use it for exploratory data analysis.

Interestingly, the summary is saying that there are 0 outliers, which I have a hard time believing. 

## Visualizing the results
### MA-plot - MvF Diff.
Generate an MA-plot to show the log2 fold changes attributable to sex over the mean of normalized counts for all of the samples in the `dds`. Points will be colored if the adjusted p-value is less that 0.05.

```{r MA-plot, echo=FALSE, fig.cap="LogFC versus the mean normalized count for all of the genes. Points that are blue have a p-value less than 0.05. The two plots are the same with the only difference coming from an adjusted y-limit. Points that are triangles represent genes with a fold change higher/lower than the y-limit."}

par(mfrow=c(1,2))
plotMA(res05, ylim = c(-2,2))
plotMA(res05)

```

### Heatmap - Looking at patterns of overall expression
We can also generate a heat map to look ar overall expression levels across our samples. Note, this is not differentially expressed genes.

```{r heatmap, echo=FALSE, fig.cap= "Heatmap showing the expression level across all organs for the top 20 genes with the highest expression levels."}

#Pull out the top 20 rows with the most expression
select <- order(rowMeans(counts(dds_FU_exp, normalized = TRUE)),
                decreasing = TRUE)[1:20]
df <- as.data.frame(colData(dds_FU)[,c("Sex", "Organ")])

#Transform the data
vsd <- vst(dds_FU_exp, blind=FALSE)

#Run the heat map with the function pheatmap
pheatmap(assay(vsd)[select,], 
         cluster_rows = FALSE, 
         show_rownames = TRUE, 
         cluster_cols = FALSE, 
         annotation_col = df)

```

From the heatmap we can see that for a lot of the top expressed genes the highest expression are found in the liver. There appears to be a few genes that are highly expressed across all tissues and a few genes that are highly expressed in the gill tissues relative to the other tissues. With this heatmap we can also pull out the names of the Trinity genes that are showing this high expression and BLAST the corresponding sequences to see what the genes are.

```{r heatmap_genes, eval = FALSE}
#Pull out the corresponding trinity_geneIDS that are plotted in the heatmap
heatmap_TG <- cbind(row.names(assay(vsd)[select,]))

#Export the gene names
write.table(heatmap_TG,
            'FU_heatmap_trinitygenes.txt',
            sep = "",
            quote=FALSE,
            row.names = FALSE,
            col.names = FALSE)

```

### Sample-dist Heatmap
We can then generate a sample-to-sample distances heatmap that will give an overview of the similarities and differences between our samples.

```{r sample-dist, echo=FALSE, fig.cap="Sample-to-sample distances heatmap showing the similarities and differences between all of the samples. The darker the color, the more similar the samples are. The diagonal line of very dark blue represents the comparissons between the same samples.", warning=FALSE}

#Calculate the sample-to-sample distances from the transformed count matrix
sampleDists <- dist(t(assay(vsd)))

#Transform into a matrix
sampleDistsMatrix <- as.matrix(sampleDists)
rownames(sampleDistsMatrix) <- paste(vsd$Sex, vsd$Organ, sep = "-")
colnames(sampleDistsMatrix) <- paste(vsd$ID)

#Run the heatmap
colors <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
pheatmap(sampleDistsMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists, 
         col = colors)

```

We can see that the highest similarities (aside from same samples comparisons) is between samples from the same organ. After that we can see high similarities between the male gonads and the female gonads. There are also a couple samples that may be an issue for use as they are not similar to any of the other samples from the same group. This includes FUG11M4 and FUT11M4. I may end up removing all organs related to 11M4.

### PCA plots
Next we can look at PCA plots for our samples to see how our different samples are grouping.

```{r pca-pairs-plot, echo=FALSE, fig.cap="Principal components analysis pairsplot showing PCA axis 1 - 4."}
#Generate the PCA data to create the pairs plot, using the PCAtools package
p <- pca(assay(vsd), 
         metadata = colData(dds_FU))

#Create the pairs plot
pairsplot(p,
          components = getComponents(p, c(1:4)),
          colby = "Organ",
          triangle = FALSE,
          hline = 0, vline = 0,
          colkey = c("Gill" = "#6E8B3D75", 
                     "Gonad" = "#EEB42275", 
                     "Liver" = "#EE826275"),
          shape = "Sex",
          shapekey = c(16, 17),
          pointSize = 3,
          margingaps = unit(c(0.1, 0.1, 0.1, 0.1), 'cm'),
          legendPosition = "left", legendIconSize = 3, legendLabSize = 10)

```

From the pairs plot we can see that PC1 appears to be explaining differences between the liver and the other tissues, PC2 appears to be explaining differences between gonadal and somatic tissues, PC3 seems to account for differences between male and female samples, especially in the gonads and in PC4 we can see the one problematic gill sample from above sticking out from all the rest. I am going to plot all of PC2 - 4 against PC1 below for a clearer picture.

```{r pca-plot, echo=FALSE, message=FALSE, fig.cap="Principal components analysis reflects that most variation corresponds to differences in expression between organs (green: gill; yellow: gonad; pink: liver), rather than variation due to sex (circle = female; triangle = male). The first axis explains 31% of variation in the dataset, and largely explains differences between the liver and the other tissues. The second axis explains 20% of the variation in gene expression, with somatic tissues on the opposite side of the axis from the gonads.", warning=FALSE}

#Generate the PCA dataframe
pca <- prcomp(t(assay(vsd)))
pca_plotting_data<-as.data.frame(pca$x)

#Add the metadata to the PCA dataset
pca_plotting_data$Organ <- FU_samples$Organ
pca_plotting_data$Sex <- FU_samples$Sex
pca_plotting_data$ID <- FU_samples$ID

#Calculate the percent variation attributed to each axis 
percentVar <- round(p$variance, digits = 2)


my_colors <- c("Gill" = "#6E8B3D75", "Gonad" = "#EEB42275", "Liver" = "#EE826275")
my_shapes <- c(16, 17)

ggplot(pca_plotting_data, aes(x=PC1, y=PC2, color = Organ, shape = Sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

```{r pca1v3, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC3. This third axis explains 10% of the variation, largely attributed to differences between the sexes, particularily in the gonads.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC3, color = Organ, shape = Sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC3: ",percentVar[3],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

```{r pca1v4, echo=FALSE, fig.cap="The same Principal Components Analysis as above but with PC4. This fourth axis explains 6% of the variation, largely attributed to one of the gill samples.", warning=FALSE}

ggplot(pca_plotting_data, aes(x=PC1, y=PC4, color = Organ, shape = Sex)) +
  scale_color_manual(values = my_colors) +
  scale_shape_manual(values = my_shapes) +
  geom_point(size = 5) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC4: ",percentVar[4],"% variance")) + 
  coord_fixed() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  geom_hline(yintercept = 0,
             linetype = 'dashed') +
  geom_vline(xintercept = 0,
             linetype = 'dashed')
```

#### Saving the PCA plots to use in figure creation
I will be including the PCAs describing the relationship between PC1 - 3 in the main text, as they account for the majority of the differences between our samples. The plot with PC1 and PC4 I still want to export, and will likely include as a supplemental to highlight why samples related to 11M4 will be removed from the subsequent analyses.

```{r setParams-PCAs}
#Setting the shapes I want to show up for the different organs
organ_shapes <- c(Gill = 16, Liver = 17, Gonad = 15)

#Setting the colors I want for each sex
sex_cols <- c("F" = "#7fc97f", "M" = "#beaed4")

```

```{r figPCA1v2, eval = FALSE}
#Create the blank pdf to store the plot in
pdf("docs/figs/Fig_PCA1v2.pdf",height = 6,width=6)

#Set the plotting parameters
par(mar=c(4,5,4,1), oma=c(2,2,2,2))

#Create the plot for PC1 v PC2
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC2,
     col = paste0(sex_cols[pca_plotting_data$Sex],"75"),
     pch = organ_shapes[pca_plotting_data$Organ],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC2: ",percentVar[2], "% variance"),
     bty = 'l',
     xpd = TRUE)

#Add a legend describing the sex
outer_legend("top",
             c("Female","Male"),
             pch = 18,
             bty = 'n',
             col=paste0(sex_cols,"75"),
             cex = 1.75,
             ncol = 2,
             pt.cex = 3)

dev.off()
```

```{r figPCA1v3, eval = FALSE}
#Create the blank pdf to store the plot in
pdf("docs/figs/Fig_PCA1v3.pdf",height = 6,width=6)

#Set the plotting parameters
par(mar=c(4,5,4,1), oma=c(2,2,2,2))

#Create the plot for PC1 v PC3
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC3,
     col = paste0(sex_cols[pca_plotting_data$Sex],"75"),
     pch = organ_shapes[pca_plotting_data$Organ],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC3: ",percentVar[3], "% variance"),
     bty = 'l',
     xpd = TRUE)

#Add  legend describing the organs
outer_legend("top",
             c("Gill","Gonad","Liver"),
             pch = c(16,15,17),
             bty = 'n',
             col = 'darkgrey',
             cex = 1.75,
             ncol = 3,
             pt.cex = 3)

dev.off()
```

```{r figPCA1v4, eval = FALSE}
#Create the blank pdf to store the plot in
pdf("docs/figs/Fig_PCA1v4.pdf",height = 6,width=8)

#Set the plotting parameters
par(mar=c(4,5,1,5),oma=c(2,2,2,2))

#Create the plot for PC1 v PC4
plot(pca_plotting_data$PC1,
     pca_plotting_data$PC4,
     col = paste0(sex_cols[pca_plotting_data$Sex],"75"),
     pch = organ_shapes[pca_plotting_data$Organ],
     cex = 2,
     cex.lab = 2,
     cex.axis = 1.75,
     xlab = paste0("PC1: ",percentVar[1], "% variance"),
     ylab = paste0("PC4: ",percentVar[4], "% variance"),
     bty = 'l',
     xpd = TRUE)

text(pca_plotting_data$PC1[pca_plotting_data$ID == "FUG11M4"] + 50,
     pca_plotting_data$PC4[pca_plotting_data$ID == "FUG11M4"] - 25,
     labels = pca_plotting_data$ID[pca_plotting_data$ID == "FUG11M4"])

#Add  legend describing the organs and sex
outer_legend("right",
             c("Gill","Gonad","Liver"),
             pch = c(16,15,17),
             bty = 'n',
             col = 'darkgrey',
             cex = 1.5,
             ncol = 1,
             pt.cex = 2.25)

outer_legend("topright",
             c("Female","Male"),
             pch = 18,
             bty = 'n',
             col=paste0(sex_cols,"75"),
             cex = 1.5,
             ncol = 1,
             pt.cex = 2.25)
dev.off()
```

#### Creating heatmaps based on the PCA axes
I now want to generate heatmaps based on the genes that have the highest loadings for each of the PCA axes. I will be using a cut off of 0.02 to determine which genes will be includd in the heatmaps. Once again I am doing this for PCA1-4, but will likely only inlcude PCA1-3 in a figure for the maintext and PC4 will go in the supplemental.

```{r setParams-heatmaps}

df <- as.data.frame(FU_samples[,c("Sex", "Organ")])
rownames(df) <- FU_samples$ID
df$Organ <- as.character(df$Organ)

organ_cols<-c("Gill" = "#20B2AA",
              "Gonad" = "#EEB422", 
              "Liver" = "#EE8262")

ann_colors = list(Sex=sex_cols,
                  Organ=organ_cols)

col_order <- c(rownames(df[df$Sex=="F" & df$Organ=="Gill",]),
               rownames(df[df$Sex=="M" & df$Organ=="Gill",]),
               rownames(df[df$Sex=="F" & df$Organ=="Gonad",]),
               rownames(df[df$Sex=="M" & df$Organ=="Gonad",]),
               rownames(df[df$Sex=="F" & df$Organ=="Liver",]),
               rownames(df[df$Sex=="M" & df$Organ=="Liver",]))

pca_rotation <- pca$rotation[, 1:4]

```

```{r savePheatmapFxn}

#Function to use to export the heatmaps to pdfs
# from https://stackoverflow.com/questions/43051525/how-to-draw-pheatmap-plot-to-screen-and-also-save-to-file
save_pheatmap_pdf <- function(x, filename, width=7, height=7) {
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  pdf(filename, width=width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}

```

```{r figheatmap-PC1, eval = FALSE}

pc1 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,1]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors = ann_colors,
                cellwidth = 9,
                fontsize = 16,
                annotation_legend = FALSE,
                main = "Top loading genes on PC1")

save_pheatmap_pdf(pc1, "docs/figs/Fig_pc1_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC2, eval=FALSE}

pc2 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,2]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                annotation_legend = FALSE,
                main = "Top loading genes on PC2")

save_pheatmap_pdf(pc2, "docs/figs/Fig_pc2_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC3, eval=FALSE}

pc3 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,3]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                border_color = NA,
                main = "Top loading genes on PC3")

save_pheatmap_pdf(pc3, "docs/figs/Fig_pc3_heatmap.pdf",
                  width=6,
                  height=6)

```

```{r figheatmap-PC4, eval=FALSE}

pc4 <- pheatmap(assay(vsd)[which(abs(pca_rotation[,4]) >= 0.02), col_order], 
                cluster_rows = FALSE, 
                show_rownames = FALSE, 
                cluster_cols = FALSE, 
                show_colnames = FALSE,
                annotation_col = df,
                annotation_colors=ann_colors,
                cellwidth = 9,
                fontsize = 16,
                border_color = NA,
                main = "Top loading genes on PC4")

save_pheatmap_pdf(pc4, "docs/figs/Fig_pc4_heatmap.pdf",
                  width=6,
                  height=6)

```

#### Making the combined figure

```{r assembleFigPCA, message=FALSE}
figPCAa <- image_ggplot(image_read_pdf('docs/figs/Fig_PCA1v2.pdf'),
                        interpolate = TRUE)
figPCAb <- image_ggplot(image_read_pdf('docs/figs/Fig_PCA1v3.pdf'),
                        interpolate = TRUE)
figPCAc <- image_ggplot(image_read_pdf('docs/figs/Fig_pc1_heatmap.pdf'),
                        interpolate = TRUE)
figPCAd <- image_ggplot(image_read_pdf('docs/figs/Fig_pc2_heatmap.pdf'),
                        interpolate = TRUE)
figPCAe <- image_ggplot(image_read_pdf('docs/figs/Fig_pc3_heatmap.pdf'),
                        interpolate = TRUE)

# make two patchworks
pcas <- figPCAa + figPCAb
hms <- figPCAc + figPCAd + figPCAe

# put the patchworks together
figPCA <-  wrap_plots(pcas,
                      hms,
                      ncol=1) + 
  plot_annotation(tag_levels = 'A') 

ggsave("docs/figs/FigPCA.pdf", figPCA, height=4, width=5)
ggsave("docs/figs/FigPCA.png", figPCA, height=4, width=5) # also save as a png
```

```{r assembleFigPCA-supp, message=FALSE}
figPCA4 <- image_ggplot(image_read_pdf('docs/figs/Fig_PCA1v4.pdf'),
                        interpolate = TRUE)
figPCA4_hm <- image_ggplot(image_read_pdf('docs/figs/Fig_pc4_heatmap.pdf'),
                           interpolate = TRUE)

# put the patchworks together
figPCA_supp <-  wrap_plots(figPCA4, figPCA4_hm, ncol=2) + 
  plot_annotation(tag_levels = 'A') 

ggsave("docs/figs/FigPCA_supp.pdf", figPCA_supp, height=4, width=8)
ggsave("docs/figs/FigPCA_supp.png", figPCA_supp, height=4, width=8) # also save as a png
```

```{r showAssembledFigPCA, eval=TRUE}
figPCA
```


# Multifactor design - Comparing M v F across the diff tissue types
If we investigate the column data of our DESeq dataset we can see that each sample has both a sex and organ type attached to it, we will be using these two factors in our multi-factor analysis. Prior to running the multi-factor analysis I am going to remove the samples that correspond to the individual 11M4.

```{r multi-factor-analysis, message=FALSE, warning=FALSE}

#Create an additional column the has the two conditions combined(sex and organ type)
FU_samples$group <- factor(paste0(FU_samples$Sex, FU_samples$Organ))

##Create a copy of the DESeq dataset
ddsMF_FU <- DESeqDataSetFromTximport(FU_txi.salmon,
                                     colData = FU_samples,
                                     design = ~ group)

##Remove all 11M4 organs from the dataset
ddsMF_FU <- ddsMF_FU[, !(ddsMF_FU$ID %in% c("FUT11M4", "FUG11M4", "FUL11M4"))]

##Filter the dataset, only keeping rows that have at least 10 reads total
keep <- rowSums(counts(ddsMF_FU)) >= 10
ddsMF_FU <- ddsMF_FU[keep, ]

#Run the differential expression analysis
ddsMF_FU_exp <- DESeq(ddsMF_FU)
resultsNames(ddsMF_FU_exp)

#Looking at the Normalized read counts for each sample
cbind(FU_samples[!(FU_samples$ID %in% 
                     c("FUT11M4", "FUG11M4", "FUL11M4")),],
      normalizedReads=colSums(counts(ddsMF_FU_exp, 
                                     normalized=TRUE)
                              )
      )

```

If we look at the summary of normalized reads across the different tissue types, we can see that there are considerably more in the liver relative to the other tissue types. Because during a multi-factor analysis normalization occurs across all of the tissues, I am opting to also remove some of the extremely expressed genes to try and offset the differences.

```{r refilter-mf-analysis, message=FALSE, warning=FALSE}

##Filter the dataset, only keeping rows that have at least 10 reads total AND less than one million
keep <- rowSums(counts(ddsMF_FU)) >= 10 & rowSums(counts(ddsMF_FU)) < 1e6
ddsMF_FU <- ddsMF_FU[keep, ]

#Re-run the differential expression analysis
ddsMF_FU_exp <- DESeq(ddsMF_FU)

#Re investigate the Normalized read counts for each sample
cbind(FU_samples[!(FU_samples$ID %in% 
                     c("FUT11M4", "FUG11M4", "FUL11M4")),],
      normalizedReads=colSums(counts(ddsMF_FU_exp, 
                                     normalized=TRUE)
                              )
      )

```

With this change we can now see that there is more even levels of expression across the different tissue types. 

## Invesitgate the results of the differential expression
Thanks to the multi-factor analysis we can now explore differential expression between all of the different combinations of tissues and sexes: 

  - Male Liver v. Female Liver 
  - Male Gill v. Female Gill  
  - Male Gonad v. Female Gonad  
  - All of the within sex tissue comparisons (e.g. Male Liver v. Male Gill, etc.)  

### M-F Liver Comparisson

```{r m-f-liver-results}
##Pulling out the liver M-F results with an alpha of 0.05
FU_liver_con_res <- results(ddsMF_FU_exp, 
                            contrast = c("group", "MLiver", "FLiver"), 
                            alpha = 0.05)
FU_liver_con_res$trin_geneid <- rownames(FU_liver_con_res)
head(FU_liver_con_res)

summary(FU_liver_con_res)
```

I will be classifying sex-biased genes the same way that I did for _Syngnathus floridae_ as **genes with a p-value < 0.05 AND a logFC $\ge$ |2|**. With that criteria in the liver there are `r sum(FU_liver_con_res$padj <= 0.05 & FU_liver_con_res$log2FoldChange >= 2, na.rm = TRUE)` male-biased genes and `r sum(FU_liver_con_res$padj <= 0.05 & FU_liver_con_res$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes. 

I have then pulled out all of the male-biased, female-biased, and non-biased genes based on the criteria outlined above. I determined non-biased genes as a p-vaule > 0.05.

```{r liver-sex-biased}
#Removing the rows where padj. is NA in results
FU_liver_con_res_noNA <- FU_liver_con_res[!is.na(FU_liver_con_res$padj),]
summary(FU_liver_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating a vector that contains all of the male-biased and female-biased genes in the liver
FU_liver_mal_biased <- FU_liver_con_res_noNA[which(FU_liver_con_res_noNA$log2FoldChange >= 2 &
                                                     FU_liver_con_res_noNA$padj <= 0.05),]

FU_liver_fem_biased <- FU_liver_con_res_noNA[which(FU_liver_con_res_noNA$log2FoldChange <= -2 &
                                                     FU_liver_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the liver
FU_liver_non_biased <- FU_liver_con_res_noNA[which(FU_liver_con_res_noNA$padj > 0.05),]

```

I will be generating a table that outlines the genes that are the most male or female biased in each organ type. To do this I need to pull the top 50 differentially expressed genes in males or females, get the Trinity gene IDs and then BLAST the corresponding sequences. Here I am pulling out the Trinity gene IDs for those genes.

```{r top50-liver, eval = FALSE}
#Creating a subset of the results where the p-value is less than 0.05
FU_liver_con_res_p05 <- FU_liver_con_res_noNA[FU_liver_con_res_noNA$padj <= 0.05, ]


#Pulling the top 50 diff expressed genes in Males
top50_FU_MLiver <- head(FU_liver_con_res_p05[order(FU_liver_con_res_p05$log2FoldChange, 
                                                   decreasing = TRUE), ], 
                        n = 50)

#Run once to generate the file and then commented out
write.table(cbind(rownames(top50_FU_MLiver)),
            'FU_maleL_top50TRgenes.txt', 
             sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)


#Pulling the top 50 diff expressed genes in Females
top50_FU_Fliver <- head(FU_liver_con_res_p05[order(FU_liver_con_res_p05$log2FoldChange, 
                                                   decreasing = FALSE), ], 
                        n = 50)

#Run once to generate the file and then commented out
write.table(cbind(rownames(top50_FU_Fliver)),
            'FU_femL_top50TRgenes.txt', 
             sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```


### M-F Gill Comparisson

```{r m-f-gill}
##Pulling out the gill M-F results
gill_con_res_FU <- results(ddsMF_FU_exp, 
                           contrast = c("group", "MGill", "FGill"), 
                           alpha = 0.5)
gill_con_res_FU$trin_geneid <- rownames(gill_con_res_FU)
head(gill_con_res_FU)

summary(gill_con_res_FU)
```

In the gills there were `r sum(gill_con_res_FU$padj <= 0.05 & gill_con_res_FU$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gill_con_res_FU$padj <= 0.05 & gill_con_res_FU$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

I have then pulled out all of the male-biased, female-biased, and non-biased genes to save them into their own objects.

```{r gill-sex-biased}
#Removing the rows where padj. is NA in results
FU_gill_con_res_noNA <- gill_con_res_FU[!is.na(gill_con_res_FU$padj), ]
summary(FU_gill_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating a vector that contains all of the male-biased and female-biased genes in the gills
FU_gill_mal_biased <- FU_gill_con_res_noNA[which(FU_gill_con_res_noNA$log2FoldChange >= 2 
                                                 & FU_gill_con_res_noNA$padj <= 0.05),]
FU_gill_fem_biased <- FU_gill_con_res_noNA[which(FU_gill_con_res_noNA$log2FoldChange <= -2 
                                                 & FU_gill_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gills, p>0.05
FU_gill_non_biased <- FU_gill_con_res_noNA[which(FU_gill_con_res_noNA$padj > 0.05),]
```

Here I am getting the trinity gene IDs that correspond to the top 50 differentially expressed genes for males and females.

```{r top50-gill, eval = FALSE}
#Creating a subset of results where p-value is less than 0.05
FU_gill_con_res_p05 <- FU_gill_con_res_noNA[FU_gill_con_res_noNA$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_FU_MGill <- head(FU_gill_con_res_p05[order(FU_gill_con_res_p05$log2FoldChange, 
                                                 decreasing = TRUE),], 
                       n=50)

#Run once to generate the file and then commented out
write.table(cbind(rownames(top50_FU_MGill)),
            'FU_maleG_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FU_FGill <- head(FU_gill_con_res_p05[order(FU_gill_con_res_p05$log2FoldChange, 
                                                 decreasing = FALSE),], 
                       n=50)

#Run once to generate the file and then commented out
write.table(cbind(rownames(top50_FU_FGill)),
            'FU_femG_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

```

### M-F Gonad Comparisson

```{r m-f-gonads}
##Pulling out the gonad M-F results
gonad_con_res_FU <- results(ddsMF_FU_exp, 
                            contrast = c("group", "MGonad", "FGonad"), 
                            alpha = 0.05)
gonad_con_res_FU$trin_geneid <- rownames(gonad_con_res_FU)
head(gonad_con_res_FU)

summary(gonad_con_res_FU)
```

Between the ovaries and testis (gonads) there were `r sum(gonad_con_res_FU$padj <= 0.05 & gonad_con_res_FU$log2FoldChange >= 2, na.rm = TRUE)` genes that we can consider male-biased and `r sum(gonad_con_res_FU$padj <= 0.05 & gonad_con_res_FU$log2FoldChange <= -2, na.rm = TRUE)` female-biased genes based on our criteria for sex-bias.

From the gonad M-F results I have then pulled out all of the male-biased, female-biased, and non-biased genes and saved them to their own objects.

```{r gonad-sex-biased}
#Removing the rows where padj. is NA in results
FU_gonad_con_res_noNA <- gonad_con_res_FU[!is.na(gonad_con_res_FU$padj), ]
summary(FU_gonad_con_res_noNA) #We can now see that there are no outliers or low counts since the NAs have been removed

#Creating an object that contains all of the male-biased and female-biased genes in the gonads
FU_gonad_mal_biased <- FU_gonad_con_res_noNA[which(FU_gonad_con_res_noNA$log2FoldChange >= 2 
                                                   & FU_gonad_con_res_noNA$padj <= 0.05),]
FU_gonad_fem_biased <- FU_gonad_con_res_noNA[which(FU_gonad_con_res_noNA$log2FoldChange <= -2 
                                                   & FU_gonad_con_res_noNA$padj <= 0.05),]

#Creating an object that contains all of the non-biased genes in the gonads, p>0.05
FU_gonad_non_biased <- FU_gonad_con_res_noNA[which(FU_gonad_con_res_noNA$padj > 0.05),]
```

I then pulled out the trinity gene IDs that corresponded to the top 50 differentially expressed genes in males and females so that I could BLAST the sequences.

```{r top50-gonad, eval=FALSE}
#Creating a subset of the results where the p-value is less than 0.05
FU_gonad_con_res_p05 <- FU_gonad_con_res_noNA[FU_gonad_con_res_noNA$padj <= 0.05,]

#Pulling the top 50 diff expressed genes in Males
top50_FU_MGonad <- head(FU_gonad_con_res_p05[order(FU_gonad_con_res_p05$log2FoldChange, 
                                                   decreasing = TRUE),], 
                        n=50)

#Run once to generate the file and then commented out
write.table(cbind(rownames(top50_FU_MGonad)),
            'FU_maleGon_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)

#Pulling the top 50 diff expressed genes in Females
top50_FU_FGonad <- head(FU_gonad_con_res_p05[order(FU_gonad_con_res_p05$log2FoldChange, 
                                                   decreasing = FALSE),], 
                        n=50)

#Run once to generate the file and then commented out
write.table(cbind(head(top50_FU_FGonad)),
            'FU_femGon_top50TRgenes.txt', 
            sep = "", 
            quote=FALSE, 
            row.names = FALSE, 
            col.names = FALSE)
```


### MA Plots

```{r MA-plot-MF, echo=FALSE, fig.height= 15, fig.width=8, fig.cap="MA-plots generated for each organ type that compares logFC to the mean expression. Female-biased and Male-biased genes are represented by color (green and purple correspondingly) and are determing with a fc cutoff of 2 and a p-value cutoff of 0.05."}
liver_MA <- ggmaplot(FU_liver_con_res,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Liver" %->% "Female Liver"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gill_MA <- ggmaplot(gill_con_res_FU,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Male Gill" %->% "Female Gill"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

gonad_MA <- ggmaplot(gonad_con_res_FU,
         fdr = 0.05, fc = 2,
         size = 1.4,
         top = 0,
         main = expression("Testis" %->% "Ovaries"),
         font.label = c("bold", 11), label.rectangle = TRUE,
         palette = c("#beaed4", "#7fc97f", "darkgray"),
         select.top.method = c("padj", "fc"),
         ggtheme = ggplot2::theme_minimal()
         )

#Merge all the plots into one
plot_grid(liver_MA, gill_MA, gonad_MA,
          ncol = 1,
          nrow = 3,
          labels = c('A', 'B', 'C'),
          label_fontfamily = 'sans',
          label_fontface = 'bold',
          label_size = 18,
          rel_heights = c(4.10, 4.10, 4.10))

```

## Variation in FC across sex-bias and tissue type
I want to create a plot the highlights the variation we see in fold-change both across the different biases groups (male-biased, female-biased, and non-biased) within one tissue type and across all of the tissue types. To do this I first need to create a "long" dataset in the following format:

| Tissue_type|     Bias      |   Fold-Change   |
|:----------:|:-------------:|:---------------:|
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   male_bias   |      XXXXX      |
|    Gill    |   fem_bias    |      XXXXX      |
|    Gill    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   male_bias   |      XXXXX      |
|    Liver   |   fem_bias    |      XXXXX      |
|    Liver   |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   male_bias   |      XXXXX      |
|   Gonad    |   fem_bias    |      XXXXX      |
|   Gonad    |   no_bias     |      XXXXX      |
|    ...     |    ....       |       ...       |

That dataset is created here, with the Trinity gene IDs included as well:

```{r logFC-long-dataset}
FU_logFC_long <- data.frame(
  tissue=c(rep("Gill",nrow(FU_gill_fem_biased)),
           rep("Gill", nrow(FU_gill_mal_biased)),
           rep("Gill", nrow(FU_gill_non_biased)),
           rep("Gonad", nrow(FU_gonad_fem_biased)),
           rep("Gonad", nrow(FU_gonad_mal_biased)),
           rep("Gonad", nrow(FU_gonad_non_biased)),
           rep("Liver", nrow(FU_liver_fem_biased)),
           rep("Liver", nrow(FU_liver_mal_biased)),
           rep("Liver", nrow(FU_liver_non_biased))
         ),
  bias=c(rep("FB",nrow(FU_gill_fem_biased)),
         rep("MB",nrow(FU_gill_mal_biased)),
         rep("NB", nrow(FU_gill_non_biased)),
         rep("FB", nrow(FU_gonad_fem_biased)),
         rep("MB", nrow(FU_gonad_mal_biased)),
         rep("NB", nrow(FU_gonad_non_biased)),
         rep("FB", nrow(FU_liver_fem_biased)),
         rep("MB", nrow(FU_liver_mal_biased)),
         rep("NB", nrow(FU_liver_non_biased))
         ),
  logFC=c(FU_gill_fem_biased$log2FoldChange,
          FU_gill_mal_biased$log2FoldChange,
          FU_gill_non_biased$log2FoldChange,
          FU_gonad_fem_biased$log2FoldChange,
          FU_gonad_mal_biased$log2FoldChange,
          FU_gonad_non_biased$log2FoldChange,
          FU_liver_fem_biased$log2FoldChange,
          FU_liver_mal_biased$log2FoldChange,
          FU_liver_non_biased$log2FoldChange
          ),
  geneID=c(rownames(FU_gill_fem_biased),
           rownames(FU_gill_mal_biased),
           rownames(FU_gill_non_biased),
           rownames(FU_gonad_fem_biased),
           rownames(FU_gonad_mal_biased),
           rownames(FU_gonad_non_biased),
           rownames(FU_liver_fem_biased),
           rownames(FU_liver_mal_biased),
           rownames(FU_liver_non_biased)
           )
  
)
```

With the dataset in the proper format now, we can generate the plot

```{r FC-var-plot, echo=FALSE, fig.cap="Absolute value of the logFC for genes that are female-biased, male-biased, and unbiased across each tissue type. Raw fold-change values are added ontop of the boxplot as jitters."}

#pdf("docs/figs/FigMF_logFC_boxplots.pdf", width = 10, height=4)

ymax <- max(abs(FU_logFC_long$logFC),
            na.rm = TRUE) + 5
sex_bias_colors <- c("FB" = "#7fc97f", 
                     "MB" = "#beaed4", 
                     "UB" = "darkgray")
organs <- levels(as.factor(FU_logFC_long$tissue))

par(mfrow=c(1,3), 
    oma=c(4,5,2,8), 
    mar=c(1,1,1,0))

for(organ in organs){
   
   # add jittered points
   plot(abs(FU_logFC_long$logFC[FU_logFC_long$tissue==organ]) ~ 
            jitter(as.numeric(as.factor(FU_logFC_long$bias[FU_logFC_long$tissue==organ]))),
        col="#00000075", 
        axes=FALSE,
        cex.main=2,
        xlim=c(0,4),
        ylim=c(0,ymax)
        )
   
   # make the boxplot
   boxplot(abs(FU_logFC_long$logFC[FU_logFC_long$tissue==organ]) ~  
             FU_logFC_long$bias[FU_logFC_long$tissue==organ],
           col=scales::alpha(sex_bias_colors,0.75),
           add = TRUE,
           yaxt='n',
           las=1,
           cex.axis=1.75,
           frame=FALSE,
           lwd=1.5,
           outline=FALSE# do not plot outliers
           )
   
   # make the axis lines longer and thicker
   axis(1, labels=NA,
        at=-1:4,
        lwd=2, 
        lwd.ticks=0, 
        las = 3, 
        cex.axis = 1.5)
   axis(2, labels=seq(-5,ymax,10),
        line=NA, 
        at=seq(-5,ymax,10), 
        lwd=2,
        ylim=c(0,ymax),
        cex.axis=2,
        las=2)
   
   mtext(organ, cex=1.5,outer=FALSE, line=-1)
   
}

# add x axis label
mtext("Bias level",side = 1,cex=1.5, outer=TRUE, line=2)

# add y axis label
mtext(expression("|log"[2]*"FC|"),side = 2,cex=1.5, outer=TRUE, line=2.25)

# add legend
spfTools::outer_legend("right",
                       legend=c("Female\nbiased",
                                "Male\nbiased",
                                "Unbiased"),
                       pt.bg=sex_bias_colors,
                       pch=22,
                       bty='n',
                       ncol=1,
                       cex=2,
                       y.intersp = 1.5,
                       pt.cex=2)

#dev.off()

```

There are some weird patterns popping up in this analysis. The first one is the magnitude of differences in male- and female-biased genes in the liver. This weirdness is then confirmed in the above plot, we can see this clumping pattern showing um in the male-biased genes in the liver. There is a cluster of genes with a logFC between 5 -15 and then another clump at 25 and above, and very little in between. This may be a result of normalizing across all of the organs. I am going to try re-running the DE analysis within each organ and see if that helps anything.

