---
title: "fuscus_orthologs"
author: "Sarah P. Flanagan"
date: "`r Sys.Date()`"
output:
  github_document: 
    toc: true 
  #bookdown::pdf_document2:
    #fig_caption: yes
    #keep_tex: yes
    #number_sections: no
    #toc: false
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
---



```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',
                     fig_path="../imgs/")
```

```{r loadlibs, include=FALSE}
library(knitr)
library(kableExtra)
library(spfTools)
```


The goal is to better understand the sex specific genes -- what's the evolutionary rate? 

## Identifying orthologs using genome sequences

So I will first identify orthologs. I've downloaded protein sequences from GenBank:

```{r genomeTable, include=FALSE}
genomes<- data.frame(
  species = c(
    "Syngnathus scovelli",
    "Syngnathus floridae",
    "Syngnathus typhle",
    "Syngnathus acus",
    "Hippocampus zosterae",
    "Hippocampus comes"
  ),
  assembly_id = c(
    "GCF_024217435.2",
    "GCA_010014945.2",
    "GCF_033458585.1",
    "GCF_901709675.1",
    "GCF_025434085.1",
    "GCF_001891065.2"
  ),
  type = c(
    "proteins",
    "fasta and annotations",
    rep("proteins",4)
  )
)
kable(genomes,"latex")
```

Based on the OrthoFinder tutorial on best practices, I'm using the proteomes where available. I have named these S_scovelli.protein.zip etc. Unfortunately Syngnathus floridae did not have a protein file available.

I've downloaded two Hippocampus protein files because in case we are really interested in changes at the Syngnathus fuscus branch (as opposed to comparing across clades) -- according to the [OrthoFinder tutorial](https://davidemms.github.io/orthofinder_tutorials/orthofinder-best-practices.html):

```
If you’re interested in what happened on a particular branch of the species tree, then you should likewise ensure good species sampling—ideally at least two species below the branch, at least two species on the closest branch above and two or more species in the outgroup.
```

I unzipped the ncbi downloads and pulled all the fasta files out into one folder. I'm skipping S. floridae for now. 

I used the orthofinder tool to identify primary transcripts (not sure it did anything, since these are not ensembl files)

```{bash, eval=FALSE}
for f in *faa ; do python ~/Programs/OrthoFinder/tools/primary_transcript.py $f ; done

```

This resulted in a folder called primary_transcripts containing the protein fasta files from the genomes. Note that I have not yet included the S. fuscus transcriptome here. I then identified orthologs:

```{bash, eval=FALSE}
/home/sarah/Programs/OrthoFinder/orthofinder -f primary_transcripts/

```

The overall statistics report that of 112330 genes, 111101 were placed in orthogroups (`r 111101/112330*100` %), which is good. The total number of single-copy orthogroups is 10,504. 

The species tree looks to correctly reflect the known phylogenetic tree structure:

```{r}
library(ape)
ortho_tree <- ape::read.tree(text="((H_comes.protein:0.0356618,H_zosterae.protein:0.0383317)0.987331:0.062338,(S_scovelli.protein:0.0242953,(S_typhle.protein:0.0260101,S_acus.protein:0.016738)0.563686:0.0118948)0.987331:0.062338);")
plot(ortho_tree)
```

## Align S. fuscus transcriptome to S. scovelli genome

The question is what type of transcripts to use, and how to use them. The easiest/current form is SuperTranscripts, which I don't think will work particularly well for this type of analysis. For sure I want to focus on sex-specific genes only. The question is, what *form* of the genes? Here are my ideas so far:

- align supertranscripts to S. scovelli and try to sort out 'correct' orientation of isoforms (eek this will be hard)
- use the longest isoform for this analysis and just use those directly in orthofinder
- put all isoforms into orthofinder, but this will probably result in 0 single copy ones so that's not ideal
- something else?? 

I think I will try (1) and see what the alignments look like. I'll use STAR to align to the S. scovelli genome (according to Stiller et al. 2022, they are most closely related of the available Syngnathus genomes). 

So first, I need to create a genome. I downloaded the data from NCBI using

```{bash, eval=FALSE}
# trinity-em, shared/coley_files/
# get the tools
curl -o datasets 'https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/v2/linux-amd64/datasets'
curl -o dataformat 'https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/v2/linux-amd64/dataformat'
chmod +x datasets dataformat
# get the data
./datasets download genome accession GCF_024217435.2 --include gff3,rna,cds,protein,genome,seq-report
unzip ncbi_download.zip
mv ncbi_dataset/data/GCF_024217435.2/ ./
mkdir scovelli_genome
```

and install STAR:
```{bash, eval=FALSE}
wget https://github.com/alexdobin/STAR/archive/2.7.11b.tar.gz
tar -xzf 2.7.11b.tar.gz
cd STAR-2.7.11b
cd source
make
cp STAR /usr/local/bin
cp ../bin/STARlong /usr/local/bin
```

Finally, need to retrieve the sex-specific transcript sequences, which are on the FlanaganLab server.

```{r}
"smb://file/Research/FlanaganLab/intermediate_data/SFU_DE_pipeline_analysis_2023/06_diff_expr_analysis/02_high_exp_trinity_gene_seqs"
```



```{bash, eval=FALSE}
# on trinity-em, in shared/coley_files/
# generate genome reference
STAR --runThreadN 6 \
     --runMode genomeGenerate \
     --genomeDir scovelli_genome \
     --genomeFastaFiles GCF_024217435.2/GCF_024217435.2_RoL_Ssco_1.2_genomic.fna \
     --sjdbGTFfile GCF_024217435.2/genomic.gff \
     --genomeSAindexNbases 13
```

Because I'm mapping assembled transcripts to a genome, I need to use the 'long reads' version of STAR (otherwise it does not work properly).

```{bash, eval=FALSE}

for faa in $(ls sex_specific_transcripts/*fasta); do
  name=$(basename $faa .fasta)
  STARlong --genomeDir scovelli_genome/ \
       --runThreadN 6 \
       --readFilesIn $faa \
       --outFileNamePrefix star_fuscus_to_scovelli/${name}_ \
       --outSAMtype BAM SortedByCoordinate \
       --outSAMunmapped Within \
       --outSAMattributes Standard \
       --outFilterMismatchNoverLmax 0.04 \
       --outReadsUnmapped Fastx
done &
```

From these, what I want to know are how the alignments map to the scovelli genome (which genes). So I want to then extract just the matches and locations of the matches:

```{bash, eval=FALSE}
for bam in $(ls *bam); do
  echo $bam
  samtools view -F 4 $bam | sed -E 's/^(\w+\t\w+\t\w+\.?\w+?\t\w+\t\w+).*$/\1/g' >> alignments.txt # use -F 4 to remove unmatched reads
done
```

## Identifying orthologs in aligned transcriptome

Now I can try to figure out orthologs of these sex-specific genes... but first I need to find the gene names in the scovelli genome.

```{r}
# using files in the FlanaganLab server 
# setwd to FlanaganLab/intermediate_data/SFU_DE_pipeline_analysis_2023/08_sex_specific_evolution
gff <- read.delim("genomic.gff",
                  comment.char = "#", 
                  header = FALSE)
colnames(gff) <- c("chromosome","method","type","start","stop","x","strand","y","description")
alignments <- read.delim("alignments.txt", header = FALSE)
colnames(alignments) <- c("fuscus_trinity_gene","alginment_key","chromosome","start","quality")

```

```{r}
alignment_gff <- do.call(rbind,apply(alignments, 1, function(this_alignment){
  this_chr <- this_alignment["chromosome"]
  this_pos <- as.numeric(this_alignment["start"])
  pot_gff <- gff[which(gff$chromosome == this_chr & gff$start <= this_pos & gff$stop > this_pos),]
  if(nrow(pot_gff)>1){
    pot_gff <- pot_gff[which.min(abs(pot_gff$start - this_pos)),]
  
  }
  pot_gff$fuscus_match <- this_alignment["fuscus_trinity_gene"]
  pot_gff$fuscus_start <- this_pos
  return(pot_gff)
}))

alignment_gff$ID <- gsub(";Parent.*$","",
                         gsub("\\w+\\-(\\w+.*)","\\1",
  gsub("ID=(\\w+.*);D.*$","\\1",alignment_gff$description)
))
```

```{r}
files<-list.files(path="../06_diff_expr_analysis/01_high_exp_trinity_gene_names",
           pattern="specific",full.names = TRUE)
specific_genes<-do.call(rbind,lapply(files,function(file){
  dat<-read.delim(file, header=FALSE)
  colnames(dat)<-"Trinity_gene"
  dat$sex <- gsub("\\.\\.\\/.*FU_(\\w{3})(\\wo?).*$","\\1",file)
  dat$organ <- gsub("\\.\\.\\/.*FU_(\\w{3})(\\wo?).*$","\\2",file)
  return(dat)
}))


alignment_info <- unique(merge(specific_genes, alignment_gff, by.x="Trinity_gene",by.y="fuscus_match",all=TRUE))
```


Most of the genes do not seem to have aligned successfully to the S. scovelli genome, but I will move forward with those that have.

```{r}
aligned<- alignment_info[!is.na(alignment_info$chromosome),]

aligned_coding <- aligned[!aligned$type %in% c("region","lnc_RNA"),]
aligned_coding$gene <- gsub("^.*gene=(\\w+.*);.*$",
                 "\\1",
                 aligned_coding$description)
aligned_coding$gene <- gsub(";.*$","",aligned_coding$gene)
```

Ok and also narrowing down the overall gff file to only include relevant things

```{r}
gff$gene <- gsub("^.*gene=(\\w+.*);.*$",
                 "\\1",
                 gff$description)
gff$gene <- gsub(";.*$", "",gff$gene)
gff_coding <- gff[gff$type %in% aligned_coding$type,]
gff_coding$Name <- gsub("^.*Name=(\\w+.*);.*$",
                 "\\1",
                 gff_coding$description)
gff_coding$Name <- gsub(";.*$","",gff_coding$Name)
```


Ok so the trick now is that the aligned data have IDs associated with genes, not with proteins. So I need to take all the scovelli orthologs and find the matching gene info (chromosome, location, name, etc). I will focus on single copy groups for simplicity.

```{r}
# these should be intermediate_data/ but are also on Dropbox

single_copy_groups <- read.table("Results_Feb11/Orthogroups/Orthogroups.tsv", sep='\t', header=TRUE)

single_copy_list <- read.table("Results_Feb11/single_copy_seq.txt")
single_copy_list <- gsub("Single_Copy_Orthologue_Sequences\\/(.*)\\.fa","\\1",single_copy_list$V1)
single_copy_groups <- single_copy_groups[single_copy_groups$Orthogroup %in% single_copy_list,]
```


```{r}
single_copy_gff<-merge(single_copy_groups, 
                       gff_coding,
                       by.x="S_scovelli.protein",
                       by.y="Name",
                       all.x=TRUE) # this has multiple rows per gene

single_copy_aligned <- merge(single_copy_gff, 
                             aligned_coding,
                             by="gene", all.y=TRUE)
# remove extra rows
single_copy_aligned <- single_copy_aligned[!duplicated(paste0(single_copy_aligned$Trinity_gene, single_copy_aligned$S_scovelli.protein, single_copy_aligned$sex,single_copy_aligned$organ)),]

# remove ones that didn't match
aligned_not_single <-single_copy_aligned[is.na(single_copy_aligned$Orthogroup),]
single_copy_aligned <- single_copy_aligned[!is.na(single_copy_aligned$Orthogroup),]

```

This results in 49 sex-biased genes (47 unique trinity genes).

```{r}
aligned_sc_simple <- single_copy_aligned[,c("Trinity_gene",
                                            "sex",
                                            "organ",
                                            "gene",
                                            "S_scovelli.protein",
                                            "Orthogroup")]
write.csv(aligned_sc_simple,
          "fuscus_to_scovelli_single_copy_genes.csv",
          row.names = FALSE)
```

Now I want to do the following for each row:
- extract the trinity gene sequence from the transcriptome file
- append that sequence to the orthogroup file (or align?)

I think I could write a script to do this per gene methodically, but I think the easiest first step is to extract the transcripts. I can use seqtk subseq to do this if I have a list of the gene names, but it puts them all in one file. So I'll fix my old code to do this job.

```{r}
write.table(unique(aligned_sc_simple$Trinity_gene),
           "fuscus_transcripts_to_extract.txt",
           row.names=FALSE,
           col.names=FALSE,
           quote=FALSE)
```



```{bash, eval=FALSE}
# put all gonad/sex-specific fastas into one
cat sex_specific_transcripts/*fasta > all_sex_specific_genes.fa

# make a new directory for these gene sequences
mkdir all_sex_specific_genes/
cd all_specific_genes/

# extract the genes so we have one file per transcript
 ./extract_genes_fasta -f ../all_sex_specific_genes.fa -l fuscus_transcripts_to_extract.txt

```


### Reversing the alignments that were reverse complements

Many of the alignments to the S. scovelli genome were reverse complements (containing the flag 16). So I want to reverse complement those ones and then use the reverse complements for the rest of the analysis. The flags are in the alignments file/object.

I will therefore write a script using those files that will then run seqtk on the relevant files.

```{r}
# create header and file
write("#!/bin/bash","reverse_complement.sh")
command_string<-"seqtk -r "
tmp<-apply(alignments[alignments$alginment_key==16 & alignments$fuscus_trinity_gene %in% aligned_sc_simple$Trinity_gene,],1,function(x){
  write(paste0(command_string,
               x["fuscus_trinity_gene"],".fasta",
               " > ", x["fuscus_trinity_gene"],"_reverse.fasta"),
        "reverse_complement.sh",
        append = TRUE)
})
```

I also will change the contents of the file with the matching gene info
```{r}
aligned_sc_simple$Trinity_gene[aligned_sc_simple$Trinity_gene %in% alignments$fuscus_trinity_gene[alignments$alginment_key==16]]<-paste0(aligned_sc_simple$Trinity_gene[aligned_sc_simple$Trinity_gene %in% alignments$fuscus_trinity_gene[alignments$alginment_key==16]],"_reverse")
write.csv(aligned_sc_simple,
          "fuscus_to_scovelli_single_copy_genes.csv",
          row.names = FALSE)
```
```{r, eval=TRUE}
aligned_sc_simple<-read.csv("fuscus_to_scovelli_single_copy_genes.csv")
```



## Preparing sex-biased orthogroups for selection analysis


so now I just need to match them to the ortholog info. To run Hyphy, I will need the nucleotide alignments. The orthogroups are amino acids, so I need to get the original sequences from the CDS files available from NCBI.

```{r}
aligned_sc_simple <- read.csv("Results_Feb11/fuscus_to_scovelli_single_copy_genes.csv")
all_gene_names<-merge(aligned_sc_simple, single_copy_groups, by="Orthogroup")
write.csv(all_gene_names, "Results_Feb11/sc_orthogroups_plus_trinity.csv",row.names = FALSE)
```

```{r, eval=TRUE}
all_gene_names<-read.csv("Results_Feb11/sc_orthogroups_plus_trinity.csv")
```

```{r}
write.table(all_gene_names$S_scovelli.protein.x, "generating_alignments/S_scovelli_protein_names.txt",quote=FALSE,row.names = FALSE, col.names = FALSE, eol='\n')
write.table(all_gene_names$H_comes.protein, "generating_alignments/H_comes_protein_names.txt",quote=FALSE,row.names = FALSE, col.names = FALSE, eol='\n')
write.table(all_gene_names$H_zosterae.protein, "generating_alignments/H_zosterae_protein_names.txt",quote=FALSE,row.names = FALSE, col.names = FALSE, eol='\n')
write.table(all_gene_names$S_acus.protein, "generating_alignments/S_acus_protein_names.txt",quote=FALSE,row.names = FALSE, col.names = FALSE, eol='\n')
write.table(all_gene_names$S_typhle.protein, "generating_alignments/S_typhle_protein_names.txt",quote=FALSE,row.names = FALSE, col.names = FALSE, eol='\n')
```

```{bash, file='generating_alignments/get_headers.sh', eval=FALSE}

```

Now I can extract them using my custom extract_genes_fasta program. For it to work, I had to edit the program to allow the names to be simplified (remove anything after the first space in the header).

```{bash, eval=FALSE}
# in generating_alignments
for fa in *fna; do spp=$(echo $fa | perl -pe 's/([A-Z]\_[a-z]+)\_cds_from_genomic.fna/\1/g'); ../extract_genes_fasta -f $fa -l ${spp}_fna_headers.txt -o ${spp}_ -s; done
```

Now, I need a set of files with all the fasta names per orthogroup, so that I can then use those to grep across all samples. **Note: this needs reversed/not reversed**

```{r}
apply(all_gene_names, 1, function(this_row){
  write(this_row[c("Trinity_gene","S_scovelli.protein.x","H_comes.protein","H_zosterae.protein","S_acus.protein","S_typhle.protein")],
        file=paste0("generating_alignments/",this_row["Orthogroup"],".txt"))
})
```

I'll move the trinity gene seqs into this folder and then use a cheeky grep + cat + sed to combine all six fasta nucleotide sequences per orthogroup.

```{bash, eval=FALSE}
ls *fasta > all_fastas.txt
# remove windows endlines from the og lists
sed -i 's/\r//g' OG*.txt

# now use grep plus cat plus sed to merge files into one
for og in OG*.txt; do og_id=$(basename $og .txt); grep -f $og all_fastas.txt | xargs cat | sed 's/>/\n>/g' > $og_id.fasta; done 
 
```



25 May 2025 - I have taken one of the alignments that looked good and deleted sequences that extended beyond the Trinity gene sequence (in geneious). I have run that in hyphy successfully (no evidence of selection). 

So now I need to do all the alignments, clean them up, and run them. I installed clustalw on trinity-big, and then in /mnt/intermediate_data/SFU_DE_pipeline_analysis_2023/08_sex_specific_evolution/generating_alignments/ I ran:

```{bash, eval=FALSE}
for fa in OG*fasta; do clustalw $fa; done

```




## Extra: Translating the transcriptome

```{bash,eval=FALSE}
#transdecoder is installed on trinity-big (not on trinity-em)
# from shared/coley_files/all_sex_specific_genes/
for fa in *fasta; do ../TransDecoder-TransDecoder-v5.7.1/TransDecoder.LongOrfs -t $fa -m 50 --complete_orfs_only; done
```

Some of them have multiple sequences and they're all in sub-folders now, which is kind of annoying

The manual says it works better on a whole transcriptome than on single genes, so I'll run it on the transcriptome file, I guess. 

```{bash, eval=FALSE}
# now in shared/coley_files
../TransDecoder-TransDecoder-v5.7.1/TransDecoder.LongOrfs -t  trinity_supertran_fuscus.fasta -m 50 --complete_orfs_only
```

I had to lower the length threshold to include at one of the sex-specific genes (931), if not more (I only checked that one).