---
title: "fuscus_orthologs"
author: "Sarah P. Flanagan"
date: "`r Sys.Date()`"
output:
  github_document: 
    toc: true 
  #bookdown::pdf_document2:
    #fig_caption: yes
    #keep_tex: yes
    #number_sections: no
    #toc: false
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
---



```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',
                     fig_path="../imgs/")
knitr::opts_chunk(eval=FALSE)
```

```{r loadlibs, include=FALSE}
library(knitr)
library(kableExtra)
library(spfTools)
```
```{r}
parse_gff<-function(filename){
  gff <- read.delim(filename,
                  comment.char = "#", 
                  header = FALSE)
  colnames(gff) <- c("chromosome","method","type","start","stop","x","strand","y","description")
  return(gff)
}
```


The goal is to better understand the sex specific genes -- what's the evolutionary rate? 

## Translating the transcriptome

I will translate the transcriptome to be able to use it in the orthofinder analysis with genome sequences

```{bash,eval=FALSE}
#transdecoder is installed on trinity-big (not on trinity-em)
# from shared/coley_files/all_sex_specific_genes/
for fa in *fasta; do ../TransDecoder-TransDecoder-v5.7.1/TransDecoder.LongOrfs -t $fa -m 50 --complete_orfs_only; done
```

Some of them have multiple sequences and they're all in sub-folders now, which is kind of annoying

The manual says it works better on a whole transcriptome than on single genes, so I'll run it on the transcriptome file, I guess. 

```{bash, eval=FALSE}
# now in shared/coley_files
../TransDecoder-TransDecoder-v5.7.1/TransDecoder.LongOrfs -t  trinity_supertran_fuscus.fasta -m 50 --complete_orfs_only
```

I had to lower the length threshold to include at one of the sex-specific genes (931), if not more (I only checked that one).

## Identifying orthologs using genome sequences

So I will first identify orthologs. I've downloaded protein sequences from GenBank:

```{r genomeTable, include=FALSE, eval=TRUE}
genomes<- data.frame(
  species = c(
    "Syngnathus scovelli",
    "Syngnathus floridae",
    "Syngnathus typhle",
    "Syngnathus acus",
    "Hippocampus zosterae",
    "Hippocampus comes"
  ),
  assembly_id = c(
    "GCF_024217435.2",
    "GCA_010014945.2",
    "GCF_033458585.1",
    "GCF_901709675.1",
    "GCF_025434085.1",
    "GCF_001891065.2"
  ),
  type = c(
    "proteins",
    "fasta and annotations",
    rep("proteins",4)
  )
)
kable(genomes,"latex")
```

Based on the OrthoFinder tutorial on best practices, I'm using the proteomes where available. I have named these S_scovelli.protein.zip etc. Unfortunately Syngnathus floridae did not have a protein file available.

I've downloaded two Hippocampus protein files because in case we are really interested in changes at the Syngnathus fuscus branch (as opposed to comparing across clades) -- according to the [OrthoFinder tutorial](https://davidemms.github.io/orthofinder_tutorials/orthofinder-best-practices.html):

```
If you’re interested in what happened on a particular branch of the species tree, then you should likewise ensure good species sampling—ideally at least two species below the branch, at least two species on the closest branch above and two or more species in the outgroup.
```

I unzipped the ncbi downloads and pulled all the fasta files out into one folder. I'm skipping S. floridae for now. 

I used the orthofinder tool to identify primary transcripts (not sure it did anything, since these are not ensembl files)

```{bash, eval=FALSE}
for f in *faa ; do python ~/Programs/OrthoFinder/tools/primary_transcript.py $f ; done

```

This resulted in a folder called primary_transcripts containing the protein fasta files from the genomes. Note that I have not yet included the S. fuscus transcriptome here. I then identified orthologs:

```{bash, eval=FALSE}
/home/sarah/Programs/OrthoFinder/orthofinder -f primary_transcripts/

```

I also need to put the outputs from transdecoder into the folder with these primary transcripts 

Then, I need to run OrthoFinder:

```{bash, eval=FALSE}
# on C001KR -- can't use dropbox because the "(" breaks things
 ./orthofinder -f /mnt/BigData/genomes/orthofinder/
```


(genomes only) The overall statistics report that of 726348 genes, 319862 were placed in orthogroups (`r 319862/726348*100` %), which is fairly good considering the messiness of the transcriptome (there were ~112k genes without the transcriptome). The total number of single-copy orthogroups is 2278 (down from 10,504 without the transcriptome). 

The species tree looks to correctly reflect the known phylogenetic tree structure (albeit with a long branch to *S. fuscus*):

```{r orthoTree, eval=TRUE}
library(ape)
ortho_tree <- ape::read.tree(text="((H_comes.protein:0.0300949,H_zosterae.protein:0.0355681)0.947398:0.0607965,((S_scovelli.protein:0.0112549,longest_orfs:0.274165)0.438367:0.0217156,(S_typhle.protein:0.0191163,S_acus.protein:0.0120793)0.490145:0.0126776)0.947398:0.0607965);")
plot(ortho_tree)
```

### single-copy orthologs

```{bash, eval=FALSE}
cd OrthoFinder/Results_Aug15/
ls Single_Copy_Orthologue_Sequences/*fa > Single_Copy_Orthologue_Sequences.txt

```

```{r}
# these should be intermediate_data/ but are also on Dropbox

single_copy_groups <- read.table("OrthoFinder/Results_Aug15/Orthogroups/Orthogroups.tsv", sep='\t', header=TRUE)

single_copy_list <- read.table("OrthoFinder/Results_Aug15/Single_Copy_Orthologue_Sequences.txt")
single_copy_list <- gsub("Single_Copy_Orthologue_Sequences\\/(.*)\\.fa","\\1",single_copy_list$V1)
single_copy_groups <- single_copy_groups[single_copy_groups$Orthogroup %in% single_copy_list,]
```

Are any of the single copy orthologs the sex-biased or sex-specific genes?


```{r getBiasInfo, eval=FALSE}
# get the info about which genes were sex specific in which tissues
# now includes ones that did not align
# in FlanaganLab/intermediate_data/SFU_DE_pipeline_analysis_2023
# on linux: setwd("/run/user/1000/gvfs/smb-share:server=file,share=research/FlanaganLab/intermediate_data/SFU_DE_pipeline_analysis_2023/")
files<-c(list.files(path="06_diff_expr_analysis/01_high_exp_trinity_gene_names",
           pattern="specific",full.names = TRUE),
         list.files(path="06_diff_expr_analysis/01_high_exp_trinity_gene_names",
           pattern="biased",full.names = TRUE) )
         
degs<-do.call(rbind,lapply(files,function(file){
  dat<-read.delim(file, header=FALSE)
  colnames(dat)<-"Trinity_gene"
  dat$sex <- gsub("^.*FU_(\\w{3})(\\w+).*$","\\1",file)
  dat$organ <- gsub("^.*FU_(\\w{3})(\\w+)_.*_.*$","\\2",file)
  dat$DEG_cat<-gsub("^.*FU_(\\w{3})(\\w+)_(\\w+)_.*$","\\3",file)
  return(dat)
}))

# prep the single copy groups for merging
single_copy_groups$S_fuscus_gene<-gsub("(g\\d+)\\.p\\d+","\\1",single_copy_groups$longest_orfs)

deg_sc_orthologs<-merge(single_copy_groups,
                        degs,
                                 by.x="S_fuscus_gene",
                                 by.y = "Trinity_gene",
                        all.x=TRUE) # keep all single copy orthogroups but not all DEGs
deg_sc_orthologs<-unique(deg_sc_orthologs)

# fill in gaps to avoid NAs
deg_sc_orthologs$DEG_cat[is.na(deg_sc_orthologs$DEG_cat)]<-"unbiased"
deg_sc_orthologs$sex[is.na(deg_sc_orthologs$sex)]<-"unbiased"
deg_sc_orthologs$organ[is.na(deg_sc_orthologs$organ)]<-"unbiased"

# save to file
write.csv(deg_sc_orthologs,"single_copy_DEG.csv",quote=FALSE, row.names = FALSE)
```


```{r getDEGorthogroups}
deg_sc_orthologs<-read.csv("single_copy_DEG.csv")
# factorise for easier investigation
deg_sc_orthologs$DEG_cat<-as.factor(deg_sc_orthologs$DEG_cat)
deg_sc_orthologs$sex<-as.factor(deg_sc_orthologs$sex)
deg_sc_orthologs$organ<-as.factor(deg_sc_orthologs$organ)
```

Now we can summarise the information

```{r}
summary(deg_sc_orthologs$DEG_cat)
table(deg_sc_orthologs$DEG_cat, deg_sc_orthologs$sex)
```


One interesting (?) development is that some of the trinity genes that seem to have the same name have different proteins in the single copy groups. For example:

```{r}
deg_sc_orthologs[deg_sc_orthologs$S_fuscus_gene=="TRINITY_DN0_c2_g1",]
```


## Aligning single-copy orthologs for selection analysis

To run Hyphy, I will need the nucleotide alignments. The orthogroups are amino acids, so I need to get the original sequences from the files available from NCBI. Once that is done, I will save all of the protein names to files to then process them. 

```{r writeProteinNames, eval=FALSE}
# dropbox
write.table(single_copy_groups$S_scovelli.protein, "generating_alignments/S_scovelli_protein_names.txt",quote=FALSE,row.names = FALSE, col.names = FALSE, eol='\n')
write.table(single_copy_groups$H_comes.protein, "generating_alignments/H_comes_protein_names.txt",quote=FALSE,row.names = FALSE, col.names = FALSE, eol='\n')
write.table(single_copy_groups$H_zosterae.protein, "generating_alignments/H_zosterae_protein_names.txt",quote=FALSE,row.names = FALSE, col.names = FALSE, eol='\n')
write.table(single_copy_groups$S_acus.protein, "generating_alignments/S_acus_protein_names.txt",quote=FALSE,row.names = FALSE, col.names = FALSE, eol='\n')
write.table(single_copy_groups$S_typhle.protein, "generating_alignments/S_typhle_protein_names.txt",quote=FALSE,row.names = FALSE, col.names = FALSE, eol='\n')
write.table(single_copy_groups$longest_orfs,
            "generating_alignments/S_fuscus_protein_names.txt", quote=FALSE, row.names = FALSE, col.names = FALSE, eol='\n')
```

Then I use this script to extract the headers and remove the ">" symbol at the start of the line:

```{bash, file='generating_alignments/get_headers.sh', eval=FALSE}

```


Now I can extract them using my custom extract_genes_fasta program. For it to work, I had to edit the program to allow the names to be simplified (remove anything after the first space in the header). *Note*: this will not find matches if the ">" is in the list of headers

```{bash, eval=FALSE}
# in generating_alignments
for fa in *genomic.fna; do spp=$(echo $fa | perl -pe 's/([A-Z]\_[a-z]+)\_cds_from_genomic.fna/\1/g'); ../extract_genes_fasta -f $fa -l ${spp}_fna_headers.txt -o ${spp}_ -s; done
```

I will also need to do this with the S. fuscus longest ORFs

```{bash,eval=FALSE}
# extract the header info
grep -f S_fuscus_protein_names.txt ../trinity_supertran_fuscus.fasta.transdecoder_dir/longest_orfs.cds | perl -pe 's/>//g' > S_fuscus_headers.txt
# extract the CDS files from the fuscus transcriptome
../extract_genes_fasta -f ../trinity_supertran_fuscus.fasta.transdecoder_dir/longest_orfs.cds -l S_fuscus_headers.txt -o S_fuscus_ -s
```


Now, I need a set of files with all the fasta names per orthogroup, so that I can then use those to grep across all samples.

```{r}
apply(single_copy_groups, 1, function(this_row){
  write(this_row[-1],
        file=paste0("generating_alignments/",this_row["Orthogroup"],".txt"))
})
```

I'll use a cheeky grep + cat + sed to combine all six fasta nucleotide sequences per orthogroup.

```{bash, eval=FALSE}
# in generating_alignments
ls *fasta > all_fastas.txt
# remove windows endlines from the og lists (if OGs were created on windows R)
sed -i 's/\r//g' OG*.txt

# now use grep plus cat plus sed to merge files into one
for og in OG*.txt; do og_id=$(basename $og .txt); grep -f $og all_fastas.txt | xargs cat | sed 's/>/\n>/g' > $og_id.fasta; done 
```

I'm going to move these fasta files into a new folder for better file organisation

```{bash, eval=FALSE}
mkdir ../OrthoFinder/Results_Aug15/Single_Copy_Orthologue_Seqs_NT
mv OG*fasta ../OrthoFinder/Results_Aug15/Single_Copy_Orthologue_Seqs_NT/
```



I installed prank v.250331 from source (from github) and copied the executable to /usr/bin on C001KR.

Now I will try aligning the single copy orthologues using PRANK. I will use the codon-aware option, but I am letting it infer the tree itself (mainly because the species tree created by orthofinder and the fasta files have different names).
I have to run this in a directory that does not have weird characters (so I'm running it in /mnt/BigData/genomes/orthofinder) and I need to use the -shortnames flag to avoid an allocation error. 

```{bash, eval=FALSE, file='../bash/prank.sh'}

```

For some of the files it does not seem to work -- throws a std::bad_alloc() error. But I'm running it anyway just to see what's going on. It might be that the ones it breaks on are poor alignments and/or ones where the CDS of the trinity genes is very short (anecdotal -- based on first few orthogroups). Ok it seems that 3186 threw the error and only 684 ran successfully. But 2278 files were created. 


I will likely need to check some of these alignments. 

## Selection analysis

I am installing hyphy v 2.5.78 on C001KR from source. My goal is to run it on all the single copy orthologs and then we can compare estimates of selection on the genes in different categories.

Before running the analysis program, I need to remove stop codons. hyphy has a tool I can use for that

```{bash, eval=FALSE}

for fa in OrthoFinder/Results_Aug15/Single_Copy_Orthologue_Seqs_NT/OG*.fasta
do
  og=$(basename $fa .fasta)
  hyphy ~/Programs/hyphy-2.5.78/res/TemplateBatchFiles/CleanStopCodons.bf Universal $fa "Disallow stops" $og_clean.fa
done
```

And finally, we can run aBSREL to estimate selection -- need trees though. The tree needs to match the header, I think. What order are the headers in? In the OG.txt, they are H. comes -> H. zosterae -> S. acus -> S. scovelli -> S. typhle -> S. fuscus. But that does not seem to be their order in OG.fasta, which seems to be in alphabetical order? Maybe based on the all_fastas.txt order. 


```{bash,eval=FALSE}
# 1 = OG; 2 = H_comes; 3=H_zos; 4=S_acus; 5=S_scov; 6=S_typhle; 7=S_fus
grep "OG0016975" single_copy_DEG.csv | perl -pe 's/^.*\,(OG\d+)\,(.*)\,(.*)\,(.*)\,(.*)\,(.*)\,(.*)\,(.*)\,(.*),(.*)/\2/g'

```


I could loop through/grep the single_copy_DEG file to get the names of the genes

I think the tree(s) will also need to be annotated with the FG branch associated with S. fuscus. 

```{bash, eval=FALSE}
/path/to/hyphy/hyphy absrel --alignment <alignment_file> --tree <tree_file> --code <genetic_code> --branches FG --output <output_file>
```


## archive

## Align S. fuscus transcriptome to S. scovelli genome

The question is what type of transcripts to use, and how to use them. The easiest/current form is SuperTranscripts, which I don't think will work particularly well for this type of analysis. For sure I want to focus on sex-specific genes only. The question is, what *form* of the genes? Here are my ideas so far:

- align supertranscripts to S. scovelli and try to sort out 'correct' orientation of isoforms (eek this will be hard)
- use the longest isoform for this analysis and just use those directly in orthofinder
- put all isoforms into orthofinder, but this will probably result in 0 single copy ones so that's not ideal
- something else?? 

I think I will try (1) and see what the alignments look like. I'll use STAR to align to the S. scovelli genome (according to Stiller et al. 2022, they are most closely related of the available Syngnathus genomes). 

So first, I need to create a genome. I downloaded the data from NCBI using

```{bash, eval=FALSE}
# trinity-em, shared/coley_files/
# get the tools
curl -o datasets 'https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/v2/linux-amd64/datasets'
curl -o dataformat 'https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/v2/linux-amd64/dataformat'
chmod +x datasets dataformat
# get the data
./datasets download genome accession GCF_024217435.2 --include gff3,rna,cds,protein,genome,seq-report
unzip ncbi_download.zip
mv ncbi_dataset/data/GCF_024217435.2/ ./
mkdir scovelli_genome
```

and install STAR:
```{bash, eval=FALSE}
wget https://github.com/alexdobin/STAR/archive/2.7.11b.tar.gz
tar -xzf 2.7.11b.tar.gz
cd STAR-2.7.11b
cd source
make
cp STAR /usr/local/bin
cp ../bin/STARlong /usr/local/bin
```

Finally, need to retrieve the sex-specific transcript sequences, which are on the FlanaganLab server.

```{bash, eval=FALSE}
"smb://file/Research/FlanaganLab/intermediate_data/SFU_DE_pipeline_analysis_2023/06_diff_expr_analysis/02_high_exp_trinity_gene_seqs"
```



```{bash, eval=FALSE}
# on trinity-em, in shared/coley_files/
# generate genome reference
STAR --runThreadN 6 \
     --runMode genomeGenerate \
     --genomeDir scovelli_genome \
     --genomeFastaFiles GCF_024217435.2/GCF_024217435.2_RoL_Ssco_1.2_genomic.fna \
     --sjdbGTFfile GCF_024217435.2/genomic.gff \
     --genomeSAindexNbases 13
```

Because I'm mapping assembled transcripts to a genome, I need to use the 'long reads' version of STAR (otherwise it does not work properly).

```{bash, eval=FALSE}

for faa in $(ls sex_specific_transcripts/*fasta); do
  name=$(basename $faa .fasta)
  STARlong --genomeDir scovelli_genome/ \
       --runThreadN 6 \
       --readFilesIn $faa \
       --outFileNamePrefix star_fuscus_to_scovelli/${name}_ \
       --outSAMtype BAM SortedByCoordinate \
       --outSAMunmapped Within \
       --outSAMattributes Standard \
       --outFilterMismatchNoverLmax 0.04 \
       --outReadsUnmapped Fastx
done &
```

From these, what I want to know are how the alignments map to the scovelli genome (which genes). So I want to then extract just the matches and locations of the matches:

```{bash, eval=FALSE}
for bam in $(ls *bam); do
  echo $bam
  samtools view -F 4 $bam | sed -E 's/^(\w+\t\w+\t\w+\.?\w+?\t\w+\t\w+).*$/\1/g' >> alignments.txt # use -F 4 to remove unmatched reads
done
```

## Identifying orthologs in aligned transcriptome

Now I can try to figure out orthologs of these sex-specific genes... but first I need to find the gene names in the scovelli genome.

```{r}
# using files in the FlanaganLab server 
# setwd to FlanaganLab/intermediate_data/SFU_DE_pipeline_analysis_2023/08_sex_specific_evolution
gff <- parse_gff("genomic.gff")
```


```{r, eval=TRUE}
# these are 142 transcripts that aligned to the S. scovelli genome
alignments <- read.delim("alignments.txt", header = FALSE)
colnames(alignments) <- c("fuscus_trinity_gene","alginment_key","chromosome","start","quality")

```

```{r}
alignment_gff <- do.call(rbind,apply(alignments, 1, function(this_alignment){
  this_chr <- this_alignment["chromosome"]
  this_pos <- as.numeric(this_alignment["start"])
  pot_gff <- gff[which(gff$chromosome == this_chr & gff$start <= this_pos & gff$stop > this_pos),]
  if(nrow(pot_gff)>1){
    pot_gff <- pot_gff[which.min(abs(pot_gff$start - this_pos)),]
  
  }
  pot_gff$fuscus_match <- this_alignment["fuscus_trinity_gene"]
  pot_gff$fuscus_start <- this_pos
  return(pot_gff)
}))

alignment_gff$ID <- gsub(";Parent.*$","",
                         gsub("\\w+\\-(\\w+.*)","\\1",
  gsub("ID=(\\w+.*);D.*$","\\1",alignment_gff$description)
))
```

```{r}
# get the info about which genes were sex specific in which tissues
# now includes ones that did not align
files<-list.files(path="../06_diff_expr_analysis/01_high_exp_trinity_gene_names",
           pattern="specific",full.names = TRUE)
specific_genes<-do.call(rbind,lapply(files,function(file){
  dat<-read.delim(file, header=FALSE)
  colnames(dat)<-"Trinity_gene"
  dat$sex <- gsub("\\.\\.\\/.*FU_(\\w{3})(\\wo?).*$","\\1",file)
  dat$organ <- gsub("\\.\\.\\/.*FU_(\\w{3})(\\wo?).*$","\\2",file)
  return(dat)
}))


alignment_info <- unique(merge(specific_genes, alignment_gff, by.x="Trinity_gene",by.y="fuscus_match",all=TRUE))
```


Most of the 1114 Trinity genes do not seem to have aligned successfully to the S. scovelli genome, but I will move forward with those that have (142 trinity genes)

```{r}
aligned<- alignment_info[!is.na(alignment_info$chromosome),]

# remove any that are not in any type of coding region
aligned_coding <- aligned[!aligned$type %in% c("region","lnc_RNA"),]
aligned_coding$gene <- gsub("^.*gene=(\\w+.*);.*$",
                 "\\1",
                 aligned_coding$description)
aligned_coding$gene <- gsub(";.*$","",aligned_coding$gene)
```

Ok and also narrowing down the overall gff file to only include relevant things

```{r}
gff$gene <- gsub("^.*gene=(\\w+.*);.*$",
                 "\\1",
                 gff$description)
gff$gene <- gsub(";.*$", "",gff$gene)
gff_coding <- gff[gff$type %in% aligned_coding$type,]
gff_coding$Name <- gsub("^.*Name=(\\w+.*);.*$",
                 "\\1",
                 gff_coding$description)
gff_coding$Name <- gsub(";.*$","",gff_coding$Name)
```


Ok so the trick now is that the aligned data have IDs associated with genes, not with proteins. So I need to take all the scovelli orthologs and find the matching gene info (chromosome, location, name, etc). I will focus on single copy groups for simplicity.

```{r}
# these should be intermediate_data/ but are also on Dropbox

single_copy_groups <- read.table("Results_Feb11/Orthogroups/Orthogroups.tsv", sep='\t', header=TRUE)

single_copy_list <- read.table("Results_Feb11/single_copy_seq.txt")
single_copy_list <- gsub("Single_Copy_Orthologue_Sequences\\/(.*)\\.fa","\\1",single_copy_list$V1)
single_copy_groups <- single_copy_groups[single_copy_groups$Orthogroup %in% single_copy_list,]
```


```{r}
single_copy_gff<-merge(single_copy_groups, 
                       gff_coding,
                       by.x="S_scovelli.protein",
                       by.y="Name",
                       all.x=TRUE) # this has multiple rows per gene

single_copy_aligned <- merge(single_copy_gff, 
                             aligned_coding,
                             by="gene", all.y=TRUE)
# add alignment key
single_copy_aligned <- merge(single_copy_aligned, alignments[,1:2],
           by.x = "Trinity_gene", by.y="fuscus_trinity_gene")
# remove ones that didn't match
aligned_not_single <-single_copy_aligned[is.na(single_copy_aligned$Orthogroup),]
single_copy_aligned <- single_copy_aligned[!is.na(single_copy_aligned$Orthogroup),]

# remove extra columns related to the protein matching
extras<-c("chromosome.x", "method.x", "type.x", "start.x","stop.x", "x.x", "strand.x", "y.x","description.x")
single_copy_aligned <- single_copy_aligned[,!colnames(single_copy_aligned) %in% extras]


# remove extra rows containing multiple cds per protein
single_copy_aligned <- single_copy_aligned[!duplicated(paste0(single_copy_aligned$Trinity_gene, single_copy_aligned$S_scovelli.protein, single_copy_aligned$sex,single_copy_aligned$organ)),]



write.csv(single_copy_aligned,"single_copy_aligned_all_info.csv",row.names = FALSE)

```
```{r}
# in intermediate_data/SFU_DE_pipeline_analysis_2023/08_sex_specific_evolution
single_copy_aligned<-read.csv("single_copy_aligned_all_info.csv")
```


This results in 49 sex-biased genes (47 unique trinity genes).

```{r}
aligned_sc_simple <- single_copy_aligned[,c("Trinity_gene",
                                            "sex",
                                            "organ",
                                            "gene",
                                            "S_scovelli.protein",
                                            "Orthogroup")]
write.csv(aligned_sc_simple,
          "fuscus_to_scovelli_single_copy_genes.csv",
          row.names = FALSE)
```
```{r, eval=TRUE}
aligned_sc_simple<-read.csv("fuscus_to_scovelli_single_copy_genes.csv")
```

### Extracting each transcript to its own file

Ok, so the next step is to get these trinity genes so I can append them to the orthogroup sequences. I will use my extract_genes_fasta code to do this, so I first need a list of the transcripts to extract. 

```{r}
write.table(unique(aligned_sc_simple$Trinity_gene),
           "fuscus_transcripts_to_extract.txt",
           row.names=FALSE,
           col.names=FALSE,
           quote=FALSE)
```

```{bash, eval=FALSE}
# put all gonad/sex-specific fastas into one
cat sex_specific_transcripts/*fasta > all_sex_specific_genes.fa

# make a new directory for these gene sequences
mkdir all_sex_specific_genes/
cd all_specific_genes/

# extract the genes so we have one file per transcript
 ./extract_genes_fasta -f ../all_sex_specific_genes.fa -l fuscus_transcripts_to_extract.txt

```


### Reversing the alignments that were reverse complements

Many of the alignments to the S. scovelli genome were reverse complements (containing the flag 16). So I want to reverse complement those ones and then use the reverse complements for the rest of the analysis. The flags are in the alignments file/object.

I will therefore write a script using those files that will then run seqtk on the relevant files.

```{r}
# create header and file
write("#!/bin/bash","generating_alignments/reverse_complement.sh")
command_string<-"seqtk seq -r "
tmp<-apply(alignments[alignments$alginment_key==16 & alignments$fuscus_trinity_gene %in% aligned_sc_simple$Trinity_gene,],1,function(x){
  write(paste0(command_string,
               x["fuscus_trinity_gene"],".fasta",
               " > ", x["fuscus_trinity_gene"],"_reverse.fasta"),
        "generating_alignments/reverse_complement.sh",
        append = TRUE)
})
```

I also will change the contents of the file with the matching gene info
```{r}
aligned_sc_simple$Trinity_gene[aligned_sc_simple$Trinity_gene %in% alignments$fuscus_trinity_gene[alignments$alginment_key==16]]<-paste0(aligned_sc_simple$Trinity_gene[aligned_sc_simple$Trinity_gene %in% alignments$fuscus_trinity_gene[alignments$alginment_key==16]],"_reverse")
write.csv(aligned_sc_simple,
          "fuscus_to_scovelli_single_copy_genes.csv",
          row.names = FALSE)
```
```{r, eval=TRUE}
aligned_sc_simple<-read.csv("fuscus_to_scovelli_single_copy_genes.csv")
```




## Preparing sex-biased orthogroups for selection analysis


so now I just need to match them to the ortholog info. To run Hyphy, I will need the nucleotide alignments. The orthogroups are amino acids, so I need to get the original sequences from the files available from NCBI.

```{r}
all_gene_names<-merge(aligned_sc_simple, single_copy_groups, by="Orthogroup")
write.csv(all_gene_names, "Results_Feb11/sc_orthogroups_plus_trinity.csv",row.names = FALSE)
```

```{r, eval=TRUE}
all_gene_names<-read.csv("Results_Feb11/sc_orthogroups_plus_trinity.csv")
```




### identifying transcripts from gffs

Some of the genes have aligned to scovelli mRNA rather than CDS, so I'll extract transcriptome information. However, I might also need to check if it is reversed.. 
```{r}
# includes S. scovelli
species<-list.files(pattern="_genomic.gff")
all_gene_names <- all_gene_names[,-6] # remove duplicate S. scov

transcripts<-lapply(gsub("_genomic.gff","",species), function(this_spp){
  gff<-parse_gff(paste0(this_spp,"_genomic.gff"))
  proteins<-all_gene_names[,grep(this_spp, colnames(all_gene_names))]
  txpts<-do.call(rbind,lapply(proteins, function(prot, gff){
    prot_gff<-gff[grep(prot,gff$description),]
    rnas<-unique(gsub("^.*Parent=rna-(XM_\\d+\\.\\d+);.*$",
                      "\\1",
                      prot_gff$description))
    
    if(length(rnas)>1){
      rna_gff<-gff[grep(rnas,gff$description),]
      rna_gff<-rna_gff[rna_gff$type %in% c("mRNA","exon"),]
      browser()  
    } else{
      strand<-unique(gff$strand[grep(rnas, gff$description)])
      if(strand == "-"){
        return(data.frame(transcript = rnas, reverse=TRUE))
      }else{
        return(data.frame(transcript = rnas, reverse=FALSE))
      }
    }
  }, gff=gff))
  
  write.table(unique(txpts$transcript),
              paste0("transcripts/",this_spp,"_transcript_names.txt"),
              quote=FALSE,
              row.names = FALSE, 
              col.names = FALSE, 
              eol='\n')
  txpts$species <- this_spp
  txpts$Orthogroup <- all_gene_names$Orthogroup
  return(txpts)
})

```

```{bash, file='generating_alignments/get_transcript_fastas.sh'}

```

I also believe that I want to reverse any alignments that were on the - strand, so I'll write a script to do that.

```{r}
# create header and file
write("#!/bin/bash","transcripts/reverse_complement_transcripts.sh")
command_string<-"seqtk seq -r "
tmp<-lapply(transcripts,function(x){
  y<-x$transcript[x$reverse==TRUE]
  write(paste0(command_string,
               unique(x$species),"_",
               y,".fasta",
               " > ", unique(x$species),"_",y,"_reverse.fasta"),
        "transcripts/reverse_complement_transcripts.sh",
        append = TRUE)
})
```


And now let's combine them

Now, I need a set of files with all the fasta names per orthogroup, so that I can then use those to grep across all samples.

```{r}
# from the transcripts object, get all the combined info 
orthogroup_transcripts <- merge(all_gene_names, transcripts[[1]], by="Orthogroup")


orthogroup_transcripts <- orthogroup_transcripts[,c("Orthogroup","Trinity_gene","transcript","reverse","species")]
colnames(orthogroup_transcripts)[colnames(orthogroup_transcripts)=="transcript"]<-paste0(unique(orthogroup_transcripts$species),"_transcript")
orthogroup_transcripts$H_comes_transcript[orthogroup_transcripts$reverse==TRUE]<-paste0(orthogroup_transcripts$H_comes_transcript[orthogroup_transcripts$reverse==TRUE], "_reverse")
orthogroup_transcripts<-orthogroup_transcripts[,c("Orthogroup","Trinity_gene","H_comes_transcript")]

orthogroup_transcripts <- merge(orthogroup_transcripts, transcripts[[2]],by="Orthogroup")
colnames(orthogroup_transcripts)[colnames(orthogroup_transcripts)=="transcript"]<-paste0(unique(orthogroup_transcripts$species),"_transcript")
orthogroup_transcripts$H_zosterae_transcript[orthogroup_transcripts$reverse==TRUE]<-paste0(orthogroup_transcripts$H_zosterae[orthogroup_transcripts$reverse==TRUE], "_reverse")
orthogroup_transcripts<-orthogroup_transcripts[,c("Orthogroup","Trinity_gene","H_comes_transcript", "H_zosterae_transcript")]

orthogroup_transcripts <- merge(orthogroup_transcripts, transcripts[[3]],
                                by="Orthogroup")
colnames(orthogroup_transcripts)[colnames(orthogroup_transcripts)=="transcript"]<-paste0(unique(orthogroup_transcripts$species),"_transcript")
orthogroup_transcripts$S_acus_transcript[orthogroup_transcripts$reverse==TRUE]<-paste0(orthogroup_transcripts$S_acus_transcript[orthogroup_transcripts$reverse==TRUE], "_reverse")
orthogroup_transcripts<-orthogroup_transcripts[,c("Orthogroup","Trinity_gene","H_comes_transcript", "H_zosterae_transcript","S_acus_transcript")]

orthogroup_transcripts <- merge(orthogroup_transcripts, transcripts[[4]],
                                by="Orthogroup")
colnames(orthogroup_transcripts)[colnames(orthogroup_transcripts)=="transcript"]<-paste0(unique(orthogroup_transcripts$species),"_transcript")
orthogroup_transcripts$S_scovelli_transcript[orthogroup_transcripts$reverse==TRUE]<-paste0(orthogroup_transcripts$S_scovelli_transcript[orthogroup_transcripts$reverse==TRUE], "_reverse")
orthogroup_transcripts<-orthogroup_transcripts[,c("Orthogroup","Trinity_gene","H_comes_transcript", "H_zosterae_transcript","S_acus_transcript","S_scovelli_transcript")]


orthogroup_transcripts <- merge(orthogroup_transcripts, transcripts[[5]],
                                by="Orthogroup")
colnames(orthogroup_transcripts)[colnames(orthogroup_transcripts)=="transcript"]<-paste0(unique(orthogroup_transcripts$species),"_transcript")
orthogroup_transcripts$S_typhle_transcript[orthogroup_transcripts$reverse==TRUE]<-paste0(orthogroup_transcripts$S_typhle_transcript[orthogroup_transcripts$reverse==TRUE], "_reverse")
orthogroup_transcripts<-orthogroup_transcripts[,c("Orthogroup","Trinity_gene","H_comes_transcript", "H_zosterae_transcript","S_acus_transcript","S_scovelli_transcript","S_typhle_transcript")]

# specify the reversed fuscus ones if relevant


# then write to file
apply(orthogroup_transcripts, 1, function(this_row){
  write(this_row[-1],
        file=paste0("transcripts/",this_row["Orthogroup"],"_transcripts.txt"))
})
```

I'll move the trinity gene seqs into this folder and then use a cheeky grep + cat + sed to combine all six fasta nucleotide sequences per orthogroup.

```{bash, eval=FALSE}
cd transcripts/
ls *fasta > all_fastas.txt #need to also add TRINITY genes....
# remove windows endlines from the og lists
sed -i 's/\r//g' OG*_transcripts.txt

# now use grep plus cat plus sed to merge files into one
for og in OG*_transcripts.txt; do og_id=$(basename $og _transcripts.txt); grep -f $og all_fastas.txt | xargs cat | sed 's/>/\n>/g' > ${og_id}_transcript.fasta; done 
```



Once clustalw is installed, it's very easy to run this on every orthogroup's fasta.

```{bash, eval=FALSE}
for fa in OG*_transcript.fasta; do clustalw $fa; done
```

The final step is trimming the alignments. I did this manually in geneious and MEGA, which also allowed me to visually inspect the alignments. Most did not look like proper multiple sequence alignments. 

## Estimating selection with hyphy

So far I've used the online interface of HyPhy rather than using the command line -- with only a few good alignments, it was easier that way. 

```{r}
library(jsonlite)
absrel_files<-list.files(pattern="absrel") # in Dropbox generating_alignments

test_results<-do.call(rbind,lapply(absrel_files, function(file){
  absrel_output<-read_json(file)

  test_result <- absrel_output$`test results`$`positive test results`
  attributes<-absrel_output$`branch attributes`$`0`
  fuscus<-attributes[grep("TRINITY", names(attributes))]
  fuscus_omega <- fuscus[[1]]$`Baseline MG94xREV omega ratio`
  return(c(test_result, fuscus_omega))

}))
rownames(test_results)<-absrel_files
colnames(test_results)<-c("selection", "omega")
kable(test_results,"latex",booktabs=TRUE)
```

Let's try to plot the one sequence that I've got that looks nice and is under selection (OG0016693)

```{r, include=FALSE, echo=FALSE, eval=FALSE}
devtools::install_github("vragh/seqvisr", build_manual = TRUE, build_vignettes = TRUE)
library(seqvisr)
msavisr(mymsa="OG0016693_modified.aln",myref="lcl|NC_090856.1_cds_XP_0495882")
```


```{r, include=FALSE, echo=FALSE, eval=FALSE}
library(ggmsa) # try this for a visualisation tool?
nt_sequences <- readDNAStringSet("OG0016693_modified.aln") # from Biostrings
ggmsa(nt_sequences, 
      start = 271, 
      end = 500, 
      char_width = 0.5, 
      seq_name = T) + geom_seqlogo()# + geom_msaBar()
```

```{r}
library(ggtree)


og<-all_gene_names[all_gene_names$Orthogroup=="OG0016693",]
og<-og[,-6]

absrel_output<-read_json("OG0016693_absrel.json")
omegas<-as.data.frame(do.call(rbind,lapply(absrel_output$`branch attributes`$`0`, function(x){
  return(round(x$`Baseline MG94xREV omega ratio`,3))
} )))
omegas$species<-lapply(rownames(omegas), function(x, og){
  ptn<-gsub("^.*(X\\w_\\d+)$","\\1",x)
  match<-grep(ptn,og[1,],value=T)
  return(gsub("\\..*$","",names(match)))
},og=og)
omegas$species[grep("TRINITY",rownames(omegas))]<-"S_fuscus"
colnames(omegas)[1]<-"omega"


tree<-read.tree("OG0016693.dnd")
tree$tip.label<-substr(tree$tip.label,1,30)

new_labels<-lapply(tree$tip.label,function(x, og){
  ptn<-gsub("^.*(X\\w_\\d+)$","\\1",x)
  match<-grep(ptn,og[1,],value=T)
  return(gsub("\\..*$","",names(match)))
},og=og)
new_labels<-unlist(new_labels)
new_labels[new_labels=="Trinity_gene"]<-"S_fuscus"

d <- data.frame(label = tree$tip.label, 
                label2 = new_labels)
d<- merge(d,omegas,by.x="label2",by.y="species")
d<-d[,c("label","label2","omega")]
```


```{r sequenceTree, fig.width=1522, fig.height=350}
p<-ggtree(tree)
p<-p %<+% d + geom_tiplab(aes(label=label2)) 
p<- p %>% rotate(8) # move seahorses together

p<- p+ geom_text(aes(x=branch, label=omega), 
            vjust=-.5, size=3) 
msaplot(p=p,
        offset=0.05,
        width=3.5,
        height=0.95,
        fasta="OG0016693_modified.aln", window=c(271,500))


```

```{r}
cols<-c("#B5D6F5FF", "#681A15FF", "#043CB2FF","#DC5750FF", "white")

```

