---
title: "Tissue Specificity in _Syngnathus floridae_"
author: "Coley Tosto"
date: "`r Sys.Date()`"
output:
  github_document: 
    toc: true
  #bookdown::pdf_document2:
   # fig_caption: yes
   # keep_tex: yes
  #  number_sections: no
   # toc: false
header-includes: >
  \usepackage{lipsum}
  \usepackage{float}
  \floatplacement{figure}{H}
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir='../',
                     fig_path="../imgs/")
```

``` {r library, message = FALSE, warning = FALSE}
#This is a cohesive list of all the libraries used in this document
library(DESeq2)
library(ggplot2)
library(EnvStats)
library(dplyr)
library(spfTools)
library(magick)
library(patchwork)
library(tidyverse)
library(knitr)

```

``` {r read-data, message = FALSE, warning = FALSE}
#The abundance matrix generated via salmon and tximport to be used for the DE analysis
txi.salmon <- readRDS("data/txi.salmon.floride.RDS")

#The samples file generated for tximport
samples <- read.table("FL_samples.txt", header = TRUE)

samples$group <- factor(paste0(samples$Sex, samples$Organ))

#Changing "Gonad" to be more specific to testis or ovaries
samples$Organ <- ifelse(samples$Sex == "F" & samples$Organ =="Gonad",
                        paste0("Ovaries"),
                        ifelse(samples$Sex == "M" & samples$Organ == "Gonad",
                               paste0("Testis"),
                               paste0(samples$Organ))
                        )

#Make sure the conditions are in the samples file as a factor
samples$Sex <- as.factor(samples$Sex)
samples$Organ <- as.factor(samples$Organ)

#Format colData to be used in the tau function
colData <- as.data.frame(samples)
rownames(colData) <- samples$ID

#Removing FLG4M4 and FLT8M7 from the metadata file
colData <- colData[!(colData$ID %in% c("FLG4M4", "FLT8M7")), ]

```

In this document I am re-doing the analysis conducted in "floridae_tissue_specificity.Rmd" but without sample FLT8M7. We are unable to determine whether that individual was pregnant and therefore want to remove it and see if there are any differences in the results that we found. There will be less detail in this document as that is all present in "floridae_tissue_specificity.Rmd".

# Differential Expression Analysis

```{r DESeqDataSet, message=FALSE, warning=FALSE}
#Create the DESeq dataset
dds_FL <- DESeqDataSetFromTximport(txi.salmon, 
                                   colData = samples,
                                   design = ~ group)
##Remove FLG4M4 from the dataset
dds_FL <- dds_FL[, !(dds_FL$ID %in% c("FLT8M7", "FLG4M4"))]

##Filter the dataset, only keeping rows that have at least 10 reads total, but less than 1,000,000
keep <- rowSums(counts(dds_FL)) >= 10 & rowSums(counts(dds_FL)) < 1e6
dds_FL <- dds_FL[keep, ]

#Generate the expression values
dds_FL_exp <- DESeq(dds_FL)

```

# Calculating Tissue Specificity
## Pulling out TPM values
From the salmon outputs I pulled out the TPM values for each sample.

```{r getTPMs}
#Get the list of file names/paths for all of the quant.sf files
files <- list.files(pattern = ".sf", path = "data/floridae_expression_files", 
                    full.names = TRUE)

#For each quant.sf file pull out the TPM column
tpms <- do.call(cbind, lapply(files, function(file){

  dat <- read.delim(file, row.names = "Name")
  tpm <- dat["TPM"]
  colnames(tpm) <- gsub("data/floridae_expression_files/(.*)_quant.sf","\\1",file)
  
  return(tpm)
}))

#Remove FLG4M4 and FLT8M7 from the dataset
tpms <- tpms[, !(colnames(tpms) %in% c("FLG4M4", "FLT8M7"))]
```

## Filtering the TPM dataset

```{r filterTPMs}
#Only keeping the rows that weren't filtered out due to low counts
tpms <- tpms[rownames(tpms) %in% rownames(dds_FL), ]

#Pulling out the geneIDs for genes that were categorized as "outliers" by DESeq2
#Calculating the Cooks threshold that would have been used
np <- length(resultsNames(dds_FL_exp))
nsamp <- ncol(dds_FL_exp)
cooks_thresh <- qf(0.99, df1 = np, df2 = 29-np)

out_ids <- names(mcols(dds_FL_exp)$maxCooks[mcols(dds_FL_exp)$maxCooks > cooks_thresh])

#Removing the rows in the tpm dataset that were deemed "outliers" by DESeq2
tpms <- tpms[!(rownames(tpms) %in% out_ids), ]

```

## Generating $\tau$

```{r est_tauFXN}
#Function for estimating tau given the TPM matrix and metadata file
est_tau<-function(geneDat,colDat){
  
  #For each row in the TPM matrix cbind it with the metadata file,
  #this attaches organ type information to the TPM values
  tissue_dat<-data.frame(cbind(colDat,
                               geneDat))
  
  #For the TPM values approaching 0, set them to 2
  tissue_dat$geneDat[tissue_dat$geneDat < 1] <- 2
  
  
  #Get the average TPM for each tissue type (TPMi)
  tissue_avgs<-tapply(tissue_dat$geneDat,tissue_dat$Organ,mean)
  
  #Get the maximum value from the average TPMS (TPMmax)
  tpmMax <- max(tissue_avgs, na.rm=TRUE)
    
  #IF running tau on JUST males of JUST females, this accounts for the
  #fact that ovary or testis will return an NA in the averaging
  if(length(unique(tissue_dat$Organ)) == 3){
    tau <- sum(1-(log(tissue_avgs[unique(tissue_dat$Organ)])/log(tpmMax)))/
      (length(unique(tissue_dat$Organ))-1)
    
    return(tau)
  }
  
  #IF using the WHOLE dataset, calculate tau
  tau<-sum(1-(log(tissue_avgs)/log(tpmMax)))/(length(unique(tissue_dat$Organ))-1)
  
  return(tau)
}

```

```{r calcTau}
tau <- apply(tpms, 1, est_tau, colDat=colData)


tau_fem<-apply(tpms[,which(colData$Sex=="F")], 1, est_tau,
               colData[which(colData$Sex=="F"),])

tau_mal<-apply(tpms[,which(colData$Sex=="M")], 1, est_tau,
               colData[which(colData$Sex=="M"),])

```

```{r tauHist,fig.cap="Distribution of tissue specificity estimates for only the male samples (left) versus only the female samples (middle) versus the male samples and the female samples (right).", echo=FALSE}
par(mfrow=c(1,3))
hist(tau_mal,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="Male-only")

hist(tau_fem,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="Female-only")

hist(tau,
     col="grey",
     xlab=expression(tau),
     ylab="Number of genes",
     main="All samples")
```

## Validating the $\tau$ calculations
To make sure $\tau$ is being calculated in a way that makes sense, I checked some of the TPM values and plotted the counts for the genes with a high tissue specificity index and a low tissue specificity index.

```{r ts-plot-counts-high, echo=FALSE, fig.height=20, fig.width=15, fig.cap="Plots of the counts from the DESeq2 Dataset for genes with the highest tau", message=FALSE}
#Getting list of genes with the greatest Tau
top_tau <- head(sort(tau, decreasing = TRUE), n=9)

#Looking at the TPMS
tpms[rownames(tpms)== names(top_tau[1]),]
tpms[rownames(tpms)== names(top_tau[2]),]
tpms[rownames(tpms)== names(top_tau[3]),]
tpms[rownames(tpms)== names(top_tau[4]),]
tpms[rownames(tpms)== names(top_tau[5]),]
tpms[rownames(tpms)== names(top_tau[6]),]
tpms[rownames(tpms)== names(top_tau[7]),]
tpms[rownames(tpms)== names(top_tau[8]),]
tpms[rownames(tpms)== names(top_tau[9]),]

#Plotting the gene counts from DESeq2 for the toptau genes
par(mfrow=c(3, 3))
plotCounts(dds_FL, gene= names(top_tau[1]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[2]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[3]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[4]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[5]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[6]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[7]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[8]), intgroup="Organ")
plotCounts(dds_FL, gene= names(top_tau[9]), intgroup="Organ")
```

```{r ts-plot-counts-low, echo=FALSE, fig.height=20, fig.width=15, fig.cap="Plots of the counts from the DESeq2 Dataset for genes with the lowest tau", message=FALSE}
#Getting list of genes with the lowest Tau
bot_tau <- head(sort(tau, decreasing = FALSE), n=9)

#Looking at the TPMS
tpms[rownames(tpms)== names(bot_tau[1]),]
tpms[rownames(tpms)== names(bot_tau[2]),]
tpms[rownames(tpms)== names(bot_tau[3]),]
tpms[rownames(tpms)== names(bot_tau[4]),]
tpms[rownames(tpms)== names(bot_tau[5]),]
tpms[rownames(tpms)== names(bot_tau[6]),]
tpms[rownames(tpms)== names(bot_tau[7]),]
tpms[rownames(tpms)== names(bot_tau[8]),]
tpms[rownames(tpms)== names(bot_tau[9]),]

#Plotting the gene counts from DESeq2 for the toptau genes
par(mfrow=c(3, 3))
plotCounts(dds_FL, gene= names(bot_tau[1]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[2]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[3]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[4]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[5]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[6]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[7]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[8]), intgroup="Organ")
plotCounts(dds_FL, gene= names(bot_tau[9]), intgroup="Organ")

```


# Sex Bias and Tissue Specificity
From the DESeq2 analysis I started earlier I can pull out the FC information for all of the sex biased genes and plot that against the $\tau$ calculations.

## $\tau$ and logFC in sex-biased genes for each organ
A gene was determined to be sex-biased if it had a log-fold change $\ge |2|$ AND an adjusted p-value of $< 0.05$. 

```{r tau-v-bias, echo=FALSE, fig.cap="Sex bias (in terms of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are sex biased in the gonads, liver, and gills."}
##Pull out the results of the MF contrasts for each organ
liver_con_res <- results(dds_FL_exp, contrast = c("group", "MLiver", "FLiver"), alpha = 0.05)
gonad_con_res <- results(dds_FL_exp, contrast = c("group", "MGonad", "FGonad"), alpha = 0.05)
gill_con_res <- results(dds_FL_exp, contrast = c("group", "MGill", "FGill"), alpha = 0.05)

#Remove rows where the p-value is NA (low counts and outliers) for each organ
liver_con_res_noNA <- liver_con_res[!is.na(liver_con_res$padj),]
gonad_con_res_noNA <- gonad_con_res[!is.na(gonad_con_res$padj),]
gill_con_res_noNA <- gill_con_res[!is.na(gill_con_res$padj),]

#Pull out the sex-biased genes (based on FC of >= |2| and padj of < 0.05) for each organ
liver_biased <- liver_con_res_noNA[which(abs(liver_con_res_noNA$log2FoldChange) >= 2 & liver_con_res_noNA$padj <= 0.05),]
gonad_biased <- gonad_con_res_noNA[which(abs(gonad_con_res_noNA$log2FoldChange) >= 2 & gonad_con_res_noNA$padj <= 0.05),]
gill_biased <- gill_con_res_noNA[which(abs(gill_con_res_noNA$log2FoldChange) >= 2 & gill_con_res_noNA$padj <= 0.05),]

#Merge the sex bias information with the tau values based on row names for each organ
liver_bias_tau <- merge(as.data.frame(liver_biased), as.data.frame(tau), by = "row.names")
gonad_bias_tau <- merge(as.data.frame(gonad_biased), as.data.frame(tau), by = "row.names")
gill_bias_tau <- merge(as.data.frame(gill_biased), as.data.frame(tau), by = "row.names")

#Plot the relationship between FC (bias) and tau for each organ
par(mar=c(5.1, 4.1, 4.1, 8.1), xpd=FALSE)
plot(sqrt(abs(gonad_bias_tau$log2FoldChange)),
     gonad_bias_tau$tau,
     pch=19,
     col = "#EEB42275",
     xlab="Sex bias (|log2 fold change|)",
     ylab=expression(tau["TPM"]))
points(sqrt(abs(liver_bias_tau$log2FoldChange)),
     liver_bias_tau$tau,
     pch=19,
     col="#EE826275")
points(sqrt(abs(gill_bias_tau$log2FoldChange)),
     gill_bias_tau$tau,
     pch=19,
     col = "#6E8B3D75")

#Add the best fit lines
abline(lm(gonad_bias_tau$tau ~ sqrt(abs(gonad_bias_tau$log2FoldChange))), col="#EEB422", lwd = 3)
abline(lm(liver_bias_tau$tau ~ sqrt(abs(liver_bias_tau$log2FoldChange))), col = "#EE8262", lwd = 3, lty = 2)
abline(lm(gill_bias_tau$tau ~ sqrt(abs(gill_bias_tau$log2FoldChange))), col = "#6E8B3D", lwd = 3, lty = 3)

#Add a legend
par(xpd=TRUE)
legend("topright", inset = c(-0.4, 0),
       legend = c("Gonads","Liver",
                    "Gills"), 
       lty=c(1,2,3), 
       col=c("#EEB422","#EE8262","#6E8B3D"), 
       bty='o',
       lwd=3)
legend("topright", inset = c(-0.4, 0),
       c("","",""), 
       pch=c(21,21,21),
       pt.bg = c("#EEB42275","#EE826275","#6E8B3D75"),
       pt.lwd = 1,
       pt.cex = 1.5,
       col=c("#EEB422","#EE8262","#6E8B3D"), bty='n',
       text.width = 1.1,
       cex=1.0
       )

```


```{r tau-v-bias-cor, echo=FALSE, message=FALSE, warning=FALSE}
cor.test(gonad_bias_tau$tau, abs(gonad_bias_tau$log2FoldChange), method = "spearman")
cor.test(liver_bias_tau$tau, abs(liver_bias_tau$log2FoldChange), method = "spearman")
cor.test(gill_bias_tau$tau, abs(gill_bias_tau$log2FoldChange), method = "spearman")
```


## $\tau$ and logFC for all the signficantly expressed genes in each organ

```{r tau-v-FC, echo=FALSE, fig.cap="Sex bias (in terms of the absolute value of the log2FoldChange) versus tissue specificity (tau) for all genes that are signifcantly expressed in the gonads, liver, and gills."}
#Merge the noNA datasets that include all of the sex-biased and non-sex-biased genes with the tau datasets
liver_bias_tau_all <- merge(as.data.frame(liver_con_res_noNA), as.data.frame(tau), by = "row.names")
gonad_bias_tau_all <- merge(as.data.frame(gonad_con_res_noNA), as.data.frame(tau), by = "row.names")
gill_bias_tau_all <- merge(as.data.frame(gill_con_res_noNA), as.data.frame(tau), by = "row.names")

#Plot logFC against Tau for all of the organs
par(mar=c(5.1, 4.1, 4.1, 8.1), xpd=FALSE)
plot(sqrt(abs(gonad_bias_tau_all$log2FoldChange)),
     gonad_bias_tau_all$tau,
     pch=19,
     col = "#EEB42275",
     xlab="sex bias (|log2 fold change|)",
     ylab=expression(tau["TPM"]),
     xlim = c(0, 7),
     ylim = c(0, 1))
points(sqrt(abs(liver_bias_tau_all$log2FoldChange)),
     liver_bias_tau_all$tau,
     pch=19,
     col="#EE826275")
points(sqrt(abs(gill_bias_tau_all$log2FoldChange)),
     gill_bias_tau_all$tau,
     pch=19,
     col = "#6E8B3D75")

#Add the best fit lines
abline(lm(gonad_bias_tau_all$tau ~ sqrt(abs(gonad_bias_tau_all$log2FoldChange))), col="#EEB422", lwd = 3)
abline(lm(liver_bias_tau_all$tau ~ sqrt(abs(liver_bias_tau_all$log2FoldChange))), col = "#EE8262", lwd = 3, lty = 2)
abline(lm(gill_bias_tau_all$tau ~ sqrt(abs(gill_bias_tau_all$log2FoldChange))), col = "#6E8B3D", lwd = 3, lty = 3)

#Add a legend
par(xpd=TRUE)
legend("topright", inset = c(-0.4, 0),
       legend = c("Gonads","Liver",
                    "Gills"), 
       lty=c(1,2,3), 
       col=c("#EEB422","#EE8262","#6E8B3D"), 
       bty='o',
       lwd=3)
legend("topright", inset = c(-0.4, 0),
       c("","",""), 
       pch=c(21,21,21),
       pt.bg = c("#EEB42275","#EE826275","#6E8B3D75"),
       pt.lwd = 1,
       pt.cex = 1.5,
       col=c("#EEB422","#EE8262","#6E8B3D"), bty='n',
       text.width = 1.5,
       cex=1.0
       )
```

The absolute value of the log2 fold change was square root transformed. We can see similar patterns as when we only plotted the sex biased genes (Fig. \@ref(fig:tau-v-bias)) with the strongest relationship showing up in the gonads. When we add in all of the points, however, it seems that the best fit line for the gills now slightly points downward (Fig. \@ref(fig:tau-v-FC)). Let's see how the Spearman's rank correlation test have changed. This was done between $\tau$ and the $|log2FoldChange|$ for each organ separately.

```{r tau-v-FC-cor, echo=FALSE, warning=FALSE}
cor.test(gonad_bias_tau_all$tau, abs(gonad_bias_tau_all$log2FoldChange), method = "spearman")
cor.test(liver_bias_tau_all$tau, abs(liver_bias_tau_all$log2FoldChange), method = "spearman")
cor.test(gill_bias_tau_all$tau, abs(gill_bias_tau_all$log2FoldChange), method = "spearman")
```

We can see the rho values going down for both the gonads and the liver, but the correlations are still significant. We can also see that in the gills we now have a significant negative correlation between sex bias and $\tau$.


## Categories of sex bias vs $\tau$

```{r bias-bin-cats-all, echo=FALSE}
#Create a long format dataset that has the tissue information, logFC, tau, and geneID 
logFC_long_all <- data.frame(
  tissue=c(rep("Gill", nrow(gill_bias_tau_all)),
           rep("Gonad", nrow(gonad_bias_tau_all)),
           rep("Liver", nrow(liver_bias_tau_all))
         ),
  logFC=c(gill_bias_tau_all$log2FoldChange,
          gonad_bias_tau_all$log2FoldChange,
          liver_bias_tau_all$log2FoldChange
          ),
  geneID=c(gill_bias_tau_all$Row.names,
           gonad_bias_tau_all$Row.names,
           liver_bias_tau_all$Row.names
           ),
  tau=c(gill_bias_tau_all$tau,
        gonad_bias_tau_all$tau,
        liver_bias_tau_all$tau
        ),
  padj=c(gill_bias_tau_all$padj,
         gonad_bias_tau_all$padj,
         liver_bias_tau_all$padj)
  
)

#Add a column to denote female-biased or male biased
logFC_long_all$bias <- ifelse(logFC_long_all$logFC <= -2 & logFC_long_all$padj <= 0.05,
                              paste0("FB"),
                              ifelse(logFC_long_all$logFC >= 2 & logFC_long_all$padj <= 0.05,
                                     paste0("MB"),
                                     paste0("UB")
                                     )
                            )
  
#Categorize the degree of sex bias in each row
#Make a vector that contains all of the groupings
biased_bins <- c("Unbiased", "Low", "Med", "High", "Extreme", "Sex-specific")

#Create a new column in the dataset and use ifelse statements to set the category limits
#abs(logFC) was used to account for the fem-biased genes possessing negative values
logFC_long_all$bias_cat <- ifelse(logFC_long_all$bias == "UB", 
                              biased_bins[1],
                              ifelse(abs(logFC_long_all$logFC) >= 2 & abs(logFC_long_all$logFC) < 3,
                                     biased_bins[2],
                                     ifelse(abs(logFC_long_all$logFC) >= 3 & abs(logFC_long_all$logFC) < 5,
                                            biased_bins[3],
                                            ifelse(abs(logFC_long_all$logFC) >= 5 & abs(logFC_long_all$logFC) < 10,
                                                   biased_bins[4],
                                                   biased_bins[5])
                                            )
                                     )
                              )
  

#Making sure the ordering stays correct
logFC_long_all$bias_cat <- factor(logFC_long_all$bias_cat,
    levels = biased_bins, ordered = TRUE)

logFC_long_all$tissue <- factor(logFC_long_all$tissue, levels = c("Gonad", "Liver", "Gill"), ordered = TRUE)

#write.csv(logFC_long_all, "data/logFC_long_taubias.csv", row.names = FALSE)
```

```{r sex-specific-for-loop}
#Filtering out the dds dataset to remove the outliers determined by DESeq
dds_FL_exp_filtered <- dds_FL_exp[!(rownames(dds_FL_exp) %in% out_ids), ]

#Changing the organ from testis or ovaries back to gonad
dds_FL_exp_filtered$Organ <- ifelse(dds_FL_exp_filtered$Organ =="Ovaries",
                        paste0("Gonad"),
                        ifelse(dds_FL_exp_filtered$Organ == "Testis",
                               paste0("Gonad"),
                               paste0(dds_FL_exp_filtered$Organ))
                        )
dds_FL_exp_filtered$Organ <- as.factor(dds_FL_exp_filtered$Organ)

#Create a vector with the different organ types
organs <- levels(colData(dds_FL_exp_filtered)$Organ)

#Create an empty list to store my datasets in
FL_sex_specific_genes <- list()

#Generate the for loop to identify MSpecific and FSpecific genes in each organ based on Medians
for(organ in organs){
  
  #Male-Specific Genes
  ##Pull out all of the rows where fem count <=10 in every female sample
  fem0_organ_names <- which(rowSums(t(apply(counts(dds_FL_exp_filtered, 
                                                   normalized = TRUE)[, dds_FL_exp_filtered$Sex == "F" & 
                                                                        dds_FL_exp_filtered$Organ == organ],
                                            1,
                                            function(x) x <= 10)
                                      )
                                    ) == ncol(counts(dds_FL_exp_filtered,
                                                     normalized = TRUE)[, dds_FL_exp_filtered$Sex == "F" & 
                                                                          dds_FL_exp_filtered$Organ == organ])
                            )
  
  fem0_organ <- counts(dds_FL_exp_filtered, 
                       normalized = TRUE)[rownames(counts(dds_FL_exp_filtered,
                                                          normalized = TRUE)) %in% 
                                            names(fem0_organ_names), 
                                          dds_FL_exp_filtered$Organ == organ]
  
  ##Pull out rows where median of male count >=20 for that organ
  mal10_organ <- apply(counts(dds_FL_exp_filtered, 
                              normalized = TRUE)[, dds_FL_exp_filtered$Sex == "M" 
                                                 & dds_FL_exp_filtered$Organ == organ],
                       1,
                       function(x) median(x) >=20)
  
  ##Keep only the rows where all fem samples <=10 and the Male median>=20
  fem0_mal10_organ <- fem0_organ[rownames(fem0_organ) %in% 
                                   names(mal10_organ[mal10_organ == TRUE]),]
  
  ##Create a new object with a name based on the organ type
  organ_malsp <- sub("$", "_male_specific", organ)
  FL_sex_specific_genes[[organ_malsp]] <- fem0_mal10_organ
  
  
  
  #Female-Specific Genes
  ##Pull out all of the rows where male count <=10 in every male sample
  mal0_organ_names <- which(rowSums(t(apply(counts(dds_FL_exp_filtered, 
                                                   normalized = TRUE)[, dds_FL_exp_filtered$Sex == "M" & 
                                                                        dds_FL_exp_filtered$Organ == organ],
                                            1,
                                            function(x) x <= 10)
                                      )
                                    ) == ncol(counts(dds_FL_exp_filtered, 
                                                     normalized = TRUE)[, dds_FL_exp_filtered$Sex == "M" & 
                                                                          dds_FL_exp_filtered$Organ == organ])
                            )
  
  mal0_organ <- counts(dds_FL_exp_filtered, 
                       normalized = TRUE)[rownames(counts(dds_FL_exp_filtered, 
                                                          normalized = TRUE)) %in% 
                                            names(mal0_organ_names), 
                                          dds_FL_exp_filtered$Organ == organ]
  
  ##Pull out rows where median of female count >=10 for that organ
  fem10_organ <- apply(counts(dds_FL_exp_filtered, 
                              normalized = TRUE)[, dds_FL_exp_filtered$Sex == "F" &
                                                   dds_FL_exp_filtered$Organ == organ],
                       1,
                       function(x) median(x) >=20)
  
  #Keep only the rows where male=0 and the fem median>=10
  mal0_fem10_organ <- mal0_organ[rownames(mal0_organ) %in%
                                   names(fem10_organ[fem10_organ == TRUE]),]
  
  # Create a new object with a name based on the organ type
  organ_femsp <- sub("$", "_female_specific", organ)
  FL_sex_specific_genes[[organ_femsp]] <- mal0_fem10_organ
} 
```


```{r cat-sex-specific}
#Changing the bias category for the genes we just labeled as sex-specific
organs <- c("Gill", "Gill", "Gonad", "Gonad", "Liver", "Liver")
bias <- c("MB", "FB", "MB", "FB", "MB", "FB")

for(i in 1:length(FL_sex_specific_genes)){#

  tmp <- FL_sex_specific_genes[[i]]
  tmp <- as.data.frame(tmp)
  tmp$geneID <- rownames(tmp)
  
  for(j in 1:nrow(tmp)){
    
    one_gene <- tmp[j, ]
    
    if(one_gene[["geneID"]] %in%
       logFC_long_all[logFC_long_all$tissue == organs[[i]] &
                      logFC_long_all$bias == bias[[i]],"geneID"]){
       
       logFC_long_all[logFC_long_all$geneID == one_gene[["geneID"]] &
                      logFC_long_all$tissue == organs[[i]] &
                      logFC_long_all$bias == bias[[i]],"bias_cat"
                      ] <- "Sex-specific"
    }else{
      
      one_gene_dat <- data.frame(matrix(ncol= ncol(logFC_long_all),
                                        nrow=1))
      colnames(one_gene_dat) <- colnames(logFC_long_all)
      
      one_gene_dat$tissue <- organs[[i]]
      one_gene_dat$geneID <- one_gene[["geneID"]]
      one_gene_dat$tau <- tau[names(tau) == one_gene[["geneID"]]]
      one_gene_dat$bias <- bias[[i]]
      one_gene_dat$bias_cat <- "Sex-specific"
      rownames(one_gene_dat) <- NULL
      
      logFC_long_all <- rbind(one_gene_dat, logFC_long_all)
      
      rownames(logFC_long_all) <- NULL
    }
  }
}

#Making sure the ordering stays correct
logFC_long_all$bias_cat <- factor(logFC_long_all$bias_cat,
    levels = biased_bins, ordered = TRUE)

#write.csv(logFC_long_all, "data/logFC_long_taubias_SS.csv", row.names = FALSE)
```

## Generating the Figures

```{r fig_setup}
logFC_long_all$tissue <- factor(logFC_long_all$tissue, 
                                levels = c("Gonad", "Liver", "Gill"), 
                                ordered = TRUE)

organs <- levels(logFC_long_all$tissue)

organ_cols <- c("Gill" = "#20B2AA",
                "Gonad" = "#EEB422", 
                "Liver" = "#EE8262")

sex_bias_colors <- c("FB" = "#7fc97f", 
                     "MB" = "#beaed4", 
                     "UB" = "darkgray")


bias_labs <- c("U","L", "M", "H", "E", "S")
biased_bins <- c("Low", "Med", "High", "Extreme", "Sex-specific")
bias_bins <- c("Unbiased", biased_bins)

```


```{r fig4a, eval=FALSE}
pdf("imgs/Fig4A_tau_sexbias_v2.pdf", width = 8, height=5)

#Set the graphical parameters
par(oma=c(2,2,1,2),
    mar=c(2,2,1,2),
    xpd=FALSE)

#Generate the base plot to be added on to
plot(logFC_long_all$tau[logFC_long_all$tissue=="Gonad"]~
       sqrt(abs(logFC_long_all$logFC[logFC_long_all$tissue=="Gonad"])),
     xlim=c(0,8),
     ylim=c(0,1),
     xlab="",
     ylab="",
     bty="n",
     type='n',
     axes=FALSE)

axis(1,pos=0,lwd=2,cex=2, cex.axis=2, las=1)
axis(2,pos=0,lwd=2,cex=2, cex.axis=2, las=1)
clip(0,8,0,1)

#Loop through the organs to add the points and best-fit line
for(organ in organs){
  
  points(logFC_long_all$tau[logFC_long_all$tissue==organ]~
           sqrt(abs(logFC_long_all$logFC[logFC_long_all$tissue==organ])),
         col = paste0(organ_cols[organ],"50"),
         pch = 19)

  }

for(organ in organs){
  
  abline(lm(logFC_long_all$tau[logFC_long_all$tissue==organ]~
              sqrt(abs(logFC_long_all$logFC[logFC_long_all$tissue==organ]))),
         col = "black",
         lwd = 3,
         lty = 1,
         xpd = FALSE)
  
  abline(lm(logFC_long_all$tau[logFC_long_all$tissue==organ]~
              sqrt(abs(logFC_long_all$logFC[logFC_long_all$tissue==organ]))),
         col = organ_cols[organ],
         lwd = 3,
         lty = which(organs %in% organ),
         xpd = FALSE)

  }

#Add a legend
outer_legend("top",
             names(organ_cols[order(names(organ_cols))]),
             col = organ_cols[order(names(organ_cols))],
             pch = 19,
             lwd = 3,
             bty = 'n',
             cex = 2,
             lty = 1:3,
             ncol = 3)

#Add the axes labels
mtext("|log fold change|",1,cex=2, line=2)
mtext(expression(tau["TPM"]),2,cex=2, line=2.5)

dev.off()

```


```{r fig4b, eval=FALSE}

pdf("imgs/Fig4B_tau_biascat_violin_jitter2.pdf", width = 12, height=3.75)

# Preprocessing to create a new variable indicating sample size greater than 10
logFC_long_all <- logFC_long_all %>%
  group_by(bias_cat, tissue, bias) %>%
  mutate(sample_size = n()) %>%
  ungroup() %>%
  mutate(plot_type = ifelse(sample_size > 10, "box_violin", "jitter"))

# Plotting
ggplot(logFC_long_all, aes(x = bias_cat, y = tau, fill = bias)) +
  geom_violin(data = filter(logFC_long_all, plot_type == "box_violin"), 
              position = position_dodge(), draw_quantiles = c(0.5)) +
  geom_boxplot(data = filter(logFC_long_all, plot_type == "box_violin"), 
               width = 0.1, color = "black", position = position_dodge(width = 0.9)) +
  geom_point(data = filter(logFC_long_all[logFC_long_all$bias == "MB",],
                           plot_type == "jitter"),
             aes(x = as.numeric(bias_cat) + 0.185, y = tau),
             size=1.5, 
             position = position_jitter(width = 0.10),
             bg=paste0(sex_bias_colors["MB"],"75"),
             col=sex_bias_colors["MB"], pch = 21) +
  geom_point(data = filter(logFC_long_all[logFC_long_all$bias == "FB",],
                           plot_type == "jitter"), 
             aes(x=as.numeric(bias_cat)-0.185, y=tau),
             size=1.5, 
             position = position_jitter(width = 0.10),
             bg=paste0(sex_bias_colors["FB"], "75"),
             col=sex_bias_colors["FB"],pch=21) +
  scale_x_discrete(labels= bias_labs) +
  scale_fill_manual(values = sex_bias_colors) +
  facet_grid(. ~ tissue) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.background = element_blank(), 
        strip.background = element_blank(),
        axis.line = element_line(colour = "black"),
        text=element_text(size=16)) +
  labs(x = "Bias Category", y = expression(tau["TPM"]))  +
  guides(fill = guide_legend(title = "Bias", order = 3)) +
  guides(color = guide_legend(title = "Bias", order = 3)) +
  stat_n_text(data = logFC_long_all[logFC_long_all$bias == "FB",], 
              aes(x = bias_cat, y = tau),
              y.pos = -0.05,
              color = sex_bias_colors["FB"]
  ) +
  stat_n_text(data = logFC_long_all[logFC_long_all$bias == "MB",], 
              aes(x = bias_cat, y = tau),
              y.pos = 0.95,
              color = sex_bias_colors["MB"]
  ) +
  stat_n_text(data = logFC_long_all[logFC_long_all$bias == "UB",], 
              aes(x = bias_cat, y = tau),
              y.pos = -0.05,
              color = sex_bias_colors["UB"]
  )

dev.off()
```

### Fig. 4 Assembly

```{r fig4, message=FALSE}
fig4a <- image_ggplot(image_read_pdf('imgs/Fig4A_tau_sexbias_v2.pdf'),
                      interpolate = TRUE)
fig4b <- image_ggplot(image_read_pdf('imgs/Fig4B_tau_biascat_violin_jitter2.pdf'),
                      interpolate = TRUE)


fig4 <- wrap_plots(fig4a,
                   fig4b,
                   ncol=2)

fig4 <- fig4 + plot_annotation(tag_levels = 'A')

ggsave("imgs/Fig4.pdf",fig4,height = 4, width=16)
ggsave("imgs/Fig4.png",fig4,height = 4, width=16)

```

```{r showAssembledFig4, eval=TRUE}
fig4
```


